{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6474b362-094a-408d-837c-bd461e6237c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import cudf\n",
    "import dask_cudf\n",
    "import hydra\n",
    "import cupy as cp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33659ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_openai import OpenAIEmbeddings, ChatOpenAI\n",
    "\n",
    "os.environ[\"OPENAI_API_KEY\"] = \"xxx\"\n",
    "\n",
    "emb_model = OpenAIEmbeddings(model='text-embedding-ada-002')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e9bd74a-892a-43f9-b79e-e26449b9896d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'_target_': 'talk2knowledgegraphs.tools.multimodal_subgraph_extraction', 'ollama_embeddings': ['nomic-embed-text'], 'temperature': 0.1, 'streaming': False, 'topk': 5, 'topk_e': 5, 'cost_e': 0.5, 'c_const': 0.01, 'root': -1, 'num_clusters': 1, 'pruning': 'gw', 'verbosity_level': 0, 'node_id_column': 'node_id', 'node_attr_column': 'node_attr', 'edge_src_column': 'edge_src', 'edge_attr_column': 'edge_attr', 'edge_dst_column': 'edge_dst', 'node_colors_dict': {'gene/protein': '#6a79f7', 'molecular_function': '#82cafc', 'cellular_component': '#3f9b0b', 'biological_process': '#c5c9c7', 'drug': '#c4a661', 'disease': '#80013f'}, 'biobridge': {'source': '/mnt/blockstorage/biobridge_multimodal/', 'node_type': ['gene/protein', 'molecular_function', 'cellular_component', 'biological_process', 'drug', 'disease']}}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load hydra configuration\n",
    "with hydra.initialize(version_base=None, config_path=\"../../../aiagents4pharma/talk2knowledgegraphs/configs\"):\n",
    "    cfg = hydra.compose(\n",
    "        config_name=\"config\", overrides=[\"tools/multimodal_subgraph_extraction=default\"]\n",
    "    )\n",
    "    cfg = cfg.tools.multimodal_subgraph_extraction\n",
    "cfg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0ec56d2c-7651-4cf4-bf85-5082758b2bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cfg.biobridge.source = \"/mnt/blockstorage/biobridge_multimodal\"\n",
    "# cfg.biobridge.source = \"../../../../AIAgents4Pharma/aiagents4pharma/talk2knowledgegraphs/tests/files/biobridge_multimodal\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a6b6fa3f-7713-4e51-9ce7-3dbde2e34872",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nodes enrichment\n",
      "['/mnt/blockstorage/biobridge_multimodal/nodes/enrichment/drug.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/enrichment/cellular_component.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/enrichment/gene_protein.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/enrichment/disease.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/enrichment/biological_process.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/enrichment/molecular_function.parquet.gzip']\n",
      "nodes embedding\n",
      "['/mnt/blockstorage/biobridge_multimodal/nodes/embedding/drug.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/embedding/cellular_component.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/embedding/gene_protein.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/embedding/disease.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/embedding/biological_process.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/nodes/embedding/molecular_function.parquet.gzip']\n",
      "edges enrichment\n",
      "['/mnt/blockstorage/biobridge_multimodal/edges/enrichment/edges.parquet.gzip']\n",
      "edges embedding\n",
      "['/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3050000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2250000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1150000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_600000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1550000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1750000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3100000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1600000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2750000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2950000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3700000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_500000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_800000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1050000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_300000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_250000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3600000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_900000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2400000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1100000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3450000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_400000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1500000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_0.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1250000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3900000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1400000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_50000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1450000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1800000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1000000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1950000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2100000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2350000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3500000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3750000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2800000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1350000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_150000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2500000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_450000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_850000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1850000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3000000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_550000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_200000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3300000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2650000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3350000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_100000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3400000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3550000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_350000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3800000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3200000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2850000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2200000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3850000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3250000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2900000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_750000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1300000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2300000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2700000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3150000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1900000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2150000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_950000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2450000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1700000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_650000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_3650000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2550000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_700000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2000000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2600000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_2050000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1200000.parquet.gzip', '/mnt/blockstorage/biobridge_multimodal/edges/embedding/edges_1650000.parquet.gzip']\n"
     ]
    }
   ],
   "source": [
    "# Loop over nodes and edges\n",
    "graph_dict = {}\n",
    "for element in [\"nodes\", \"edges\"]:\n",
    "    # Make an empty dictionary for each folder\n",
    "    graph_dict[element] = {}\n",
    "    for stage in [\"enrichment\", \"embedding\"]:\n",
    "        print(element, stage)\n",
    "        # Create the file pattern for the current subfolder\n",
    "        file_list = glob.glob(os.path.join(cfg.biobridge.source,\n",
    "                                           element,\n",
    "                                           stage, '*.parquet.gzip'))\n",
    "        print(file_list)\n",
    "        # Read and concatenate all dataframes in the folder\n",
    "        # Except the edges embedding, which is too large to read in one go\n",
    "        # We are using a chunk size to read the edges embedding in smaller parts instead\n",
    "        if element == \"edges\" and stage == \"embedding\":\n",
    "            # For edges embedding, only read two columns: triplet_index and edge_emb\n",
    "            # graph_dict[element][stage] = cudf.concat([cudf.read_parquet(f, columns=[\"triplet_index\", \"edge_emb\"]) for f in file_list[:2]], ignore_index=True)\n",
    "            # Loop by chunks\n",
    "            # file_list = file_list[:2]\n",
    "            chunk_size = 5\n",
    "            graph_dict[element][stage] = []\n",
    "            for i in range(0, len(file_list), chunk_size):\n",
    "                chunk_files = file_list[i:i+chunk_size]\n",
    "                chunk_df = cudf.concat([cudf.read_parquet(f, columns=[\"triplet_index\", \"edge_emb\"]) for f in chunk_files], ignore_index=True)\n",
    "                graph_dict[element][stage].append(chunk_df)\n",
    "        else:\n",
    "            # For nodes and edges enrichment, read and concatenate all dataframes in the folder\n",
    "            # This includes the nodes embedding, which is small enough to read in one go\n",
    "            graph_dict[element][stage] = cudf.concat([cudf.read_parquet(f) for f in file_list], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bbcf094a",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = graph_dict\n",
    "\n",
    "# Get nodes enrichment and embedding dataframes\n",
    "nodes_enrichment_df = graph['nodes']['enrichment']\n",
    "nodes_embedding_df = graph['nodes']['embedding']\n",
    "# Get edges enrichment and embedding dataframes\n",
    "edges_enrichment_df = graph['edges']['enrichment']\n",
    "edges_embedding_df = graph['edges']['embedding'] # consisted of a list of dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1ac28ba5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>triplet_index</th>\n",
       "      <th>primekg_head_index</th>\n",
       "      <th>primekg_tail_index</th>\n",
       "      <th>head_id</th>\n",
       "      <th>tail_id</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>edge_type_str</th>\n",
       "      <th>head_index</th>\n",
       "      <th>tail_index</th>\n",
       "      <th>feat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8889</td>\n",
       "      <td>PHYHIP_(0)</td>\n",
       "      <td>KIF15_(8889)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>0</td>\n",
       "      <td>8816</td>\n",
       "      <td>PHYHIP (gene/protein) has a direct relationshi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2798</td>\n",
       "      <td>GPANK1_(1)</td>\n",
       "      <td>PNMA1_(2798)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>1</td>\n",
       "      <td>2787</td>\n",
       "      <td>GPANK1 (gene/protein) has a direct relationshi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>5646</td>\n",
       "      <td>ZRSR2_(2)</td>\n",
       "      <td>TTC33_(5646)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>2</td>\n",
       "      <td>5610</td>\n",
       "      <td>ZRSR2 (gene/protein) has a direct relationship...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>11592</td>\n",
       "      <td>NRF1_(3)</td>\n",
       "      <td>MAN1B1_(11592)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>3</td>\n",
       "      <td>11467</td>\n",
       "      <td>NRF1 (gene/protein) has a direct relationship ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2122</td>\n",
       "      <td>PI4KA_(4)</td>\n",
       "      <td>RGS20_(2122)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>4</td>\n",
       "      <td>2117</td>\n",
       "      <td>PI4KA (gene/protein) has a direct relationship...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904605</th>\n",
       "      <td>3904605</td>\n",
       "      <td>52855</td>\n",
       "      <td>34572</td>\n",
       "      <td>B cell receptor transport into membrane raft_(...</td>\n",
       "      <td>CD24_(34572)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>45323</td>\n",
       "      <td>27800</td>\n",
       "      <td>B cell receptor transport into membrane raft (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904606</th>\n",
       "      <td>3904606</td>\n",
       "      <td>113352</td>\n",
       "      <td>34572</td>\n",
       "      <td>chemokine receptor transport out of membrane r...</td>\n",
       "      <td>CD24_(34572)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>71241</td>\n",
       "      <td>27800</td>\n",
       "      <td>chemokine receptor transport out of membrane r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904607</th>\n",
       "      <td>3904607</td>\n",
       "      <td>42264</td>\n",
       "      <td>57675</td>\n",
       "      <td>negative regulation of cytoskeleton organizati...</td>\n",
       "      <td>IQCJ-SCHIP1_(57675)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>35166</td>\n",
       "      <td>49927</td>\n",
       "      <td>negative regulation of cytoskeleton organizati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904608</th>\n",
       "      <td>3904608</td>\n",
       "      <td>109904</td>\n",
       "      <td>58770</td>\n",
       "      <td>mesendoderm migration_(109904)</td>\n",
       "      <td>APELA_(58770)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>67928</td>\n",
       "      <td>50777</td>\n",
       "      <td>mesendoderm migration (biological_process) has...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904609</th>\n",
       "      <td>3904609</td>\n",
       "      <td>44810</td>\n",
       "      <td>57692</td>\n",
       "      <td>regulation of endoplasmic reticulum unfolded p...</td>\n",
       "      <td>PIGBOS1_(57692)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>37608</td>\n",
       "      <td>49939</td>\n",
       "      <td>regulation of endoplasmic reticulum unfolded p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3904610 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         triplet_index  primekg_head_index  primekg_tail_index  \\\n",
       "0                    0                   0                8889   \n",
       "1                    1                   1                2798   \n",
       "2                    2                   2                5646   \n",
       "3                    3                   3               11592   \n",
       "4                    4                   4                2122   \n",
       "...                ...                 ...                 ...   \n",
       "3904605        3904605               52855               34572   \n",
       "3904606        3904606              113352               34572   \n",
       "3904607        3904607               42264               57675   \n",
       "3904608        3904608              109904               58770   \n",
       "3904609        3904609               44810               57692   \n",
       "\n",
       "                                                   head_id  \\\n",
       "0                                               PHYHIP_(0)   \n",
       "1                                               GPANK1_(1)   \n",
       "2                                                ZRSR2_(2)   \n",
       "3                                                 NRF1_(3)   \n",
       "4                                                PI4KA_(4)   \n",
       "...                                                    ...   \n",
       "3904605  B cell receptor transport into membrane raft_(...   \n",
       "3904606  chemokine receptor transport out of membrane r...   \n",
       "3904607  negative regulation of cytoskeleton organizati...   \n",
       "3904608                     mesendoderm migration_(109904)   \n",
       "3904609  regulation of endoplasmic reticulum unfolded p...   \n",
       "\n",
       "                     tail_id display_relation  \\\n",
       "0               KIF15_(8889)              ppi   \n",
       "1               PNMA1_(2798)              ppi   \n",
       "2               TTC33_(5646)              ppi   \n",
       "3             MAN1B1_(11592)              ppi   \n",
       "4               RGS20_(2122)              ppi   \n",
       "...                      ...              ...   \n",
       "3904605         CD24_(34572)   interacts with   \n",
       "3904606         CD24_(34572)   interacts with   \n",
       "3904607  IQCJ-SCHIP1_(57675)   interacts with   \n",
       "3904608        APELA_(58770)   interacts with   \n",
       "3904609      PIGBOS1_(57692)   interacts with   \n",
       "\n",
       "                                                 edge_type  \\\n",
       "0                        [gene/protein, ppi, gene/protein]   \n",
       "1                        [gene/protein, ppi, gene/protein]   \n",
       "2                        [gene/protein, ppi, gene/protein]   \n",
       "3                        [gene/protein, ppi, gene/protein]   \n",
       "4                        [gene/protein, ppi, gene/protein]   \n",
       "...                                                    ...   \n",
       "3904605  [biological_process, interacts with, gene/prot...   \n",
       "3904606  [biological_process, interacts with, gene/prot...   \n",
       "3904607  [biological_process, interacts with, gene/prot...   \n",
       "3904608  [biological_process, interacts with, gene/prot...   \n",
       "3904609  [biological_process, interacts with, gene/prot...   \n",
       "\n",
       "                                          edge_type_str  head_index  \\\n",
       "0                         gene/protein|ppi|gene/protein           0   \n",
       "1                         gene/protein|ppi|gene/protein           1   \n",
       "2                         gene/protein|ppi|gene/protein           2   \n",
       "3                         gene/protein|ppi|gene/protein           3   \n",
       "4                         gene/protein|ppi|gene/protein           4   \n",
       "...                                                 ...         ...   \n",
       "3904605  biological_process|interacts with|gene/protein       45323   \n",
       "3904606  biological_process|interacts with|gene/protein       71241   \n",
       "3904607  biological_process|interacts with|gene/protein       35166   \n",
       "3904608  biological_process|interacts with|gene/protein       67928   \n",
       "3904609  biological_process|interacts with|gene/protein       37608   \n",
       "\n",
       "         tail_index                                               feat  \n",
       "0              8816  PHYHIP (gene/protein) has a direct relationshi...  \n",
       "1              2787  GPANK1 (gene/protein) has a direct relationshi...  \n",
       "2              5610  ZRSR2 (gene/protein) has a direct relationship...  \n",
       "3             11467  NRF1 (gene/protein) has a direct relationship ...  \n",
       "4              2117  PI4KA (gene/protein) has a direct relationship...  \n",
       "...             ...                                                ...  \n",
       "3904605       27800  B cell receptor transport into membrane raft (...  \n",
       "3904606       27800  chemokine receptor transport out of membrane r...  \n",
       "3904607       49927  negative regulation of cytoskeleton organizati...  \n",
       "3904608       50777  mesendoderm migration (biological_process) has...  \n",
       "3904609       49939  regulation of endoplasmic reticulum unfolded p...  \n",
       "\n",
       "[3904610 rows x 11 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_idxs = nodes_enrichment_df.node_id.to_pandas().tolist()\n",
    "edges_enrichment_df[(edges_enrichment_df.head_id.isin(node_idxs)) & (edges_enrichment_df.head_id.isin(node_idxs))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9698ddb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[13814,\n",
       " 13815,\n",
       " 13816,\n",
       " 13817,\n",
       " 13818,\n",
       " 13819,\n",
       " 13820,\n",
       " 13821,\n",
       " 13822,\n",
       " 13823,\n",
       " 13824,\n",
       " 13825,\n",
       " 13826,\n",
       " 13827,\n",
       " 13828,\n",
       " 13829,\n",
       " 13830,\n",
       " 13831,\n",
       " 13832,\n",
       " 13833,\n",
       " 13834,\n",
       " 13835,\n",
       " 13836,\n",
       " 13837,\n",
       " 13838,\n",
       " 13839,\n",
       " 13840,\n",
       " 13841,\n",
       " 13842,\n",
       " 13843,\n",
       " 13844,\n",
       " 13845,\n",
       " 13846,\n",
       " 13847,\n",
       " 13848,\n",
       " 13849,\n",
       " 13850,\n",
       " 13851,\n",
       " 13852,\n",
       " 13853,\n",
       " 13854,\n",
       " 13855,\n",
       " 13856,\n",
       " 13857,\n",
       " 13858,\n",
       " 13859,\n",
       " 13860,\n",
       " 13861,\n",
       " 13862,\n",
       " 13863,\n",
       " 13864,\n",
       " 13865,\n",
       " 13866,\n",
       " 13867,\n",
       " 13868,\n",
       " 13869,\n",
       " 13870,\n",
       " 13871,\n",
       " 13872,\n",
       " 13873,\n",
       " 13874,\n",
       " 13875,\n",
       " 13876,\n",
       " 13877,\n",
       " 13878,\n",
       " 13879,\n",
       " 13880,\n",
       " 13881,\n",
       " 13882,\n",
       " 13883,\n",
       " 13884,\n",
       " 13885,\n",
       " 13886,\n",
       " 13887,\n",
       " 13888,\n",
       " 13889,\n",
       " 13890,\n",
       " 13891,\n",
       " 13892,\n",
       " 13893,\n",
       " 13894,\n",
       " 13895,\n",
       " 13896,\n",
       " 13897,\n",
       " 13898,\n",
       " 13899,\n",
       " 13900,\n",
       " 13901,\n",
       " 13902,\n",
       " 13903,\n",
       " 13904,\n",
       " 13905,\n",
       " 13906,\n",
       " 13907,\n",
       " 13908,\n",
       " 13909,\n",
       " 13910,\n",
       " 13911,\n",
       " 13912,\n",
       " 13913,\n",
       " 13914,\n",
       " 13915,\n",
       " 13916,\n",
       " 13917,\n",
       " 13918,\n",
       " 13919,\n",
       " 13920,\n",
       " 13921,\n",
       " 13922,\n",
       " 13923,\n",
       " 13924,\n",
       " 13925,\n",
       " 13926,\n",
       " 13927,\n",
       " 13928,\n",
       " 13929,\n",
       " 13930,\n",
       " 13931,\n",
       " 13932,\n",
       " 13933,\n",
       " 13934,\n",
       " 13935,\n",
       " 13936,\n",
       " 13937,\n",
       " 13938,\n",
       " 13939,\n",
       " 13940,\n",
       " 13941,\n",
       " 13942,\n",
       " 13943,\n",
       " 13944,\n",
       " 13945,\n",
       " 13946,\n",
       " 13947,\n",
       " 13948,\n",
       " 13949,\n",
       " 13950,\n",
       " 13951,\n",
       " 13952,\n",
       " 13953,\n",
       " 13954,\n",
       " 13955,\n",
       " 13956,\n",
       " 13957,\n",
       " 13958,\n",
       " 13959,\n",
       " 13960,\n",
       " 13961,\n",
       " 13962,\n",
       " 13963,\n",
       " 13964,\n",
       " 13965,\n",
       " 13966,\n",
       " 13967,\n",
       " 13968,\n",
       " 13969,\n",
       " 13970,\n",
       " 13971,\n",
       " 13972,\n",
       " 13973,\n",
       " 13974,\n",
       " 13975,\n",
       " 13976,\n",
       " 13977,\n",
       " 13978,\n",
       " 13979,\n",
       " 13980,\n",
       " 13981,\n",
       " 13982,\n",
       " 13983,\n",
       " 13984,\n",
       " 13985,\n",
       " 13986,\n",
       " 13987,\n",
       " 13988,\n",
       " 13989,\n",
       " 13990,\n",
       " 13991,\n",
       " 13992,\n",
       " 13993,\n",
       " 13994,\n",
       " 13995,\n",
       " 13996,\n",
       " 13997,\n",
       " 13998,\n",
       " 13999,\n",
       " 14000,\n",
       " 14001,\n",
       " 14002,\n",
       " 14003,\n",
       " 14004,\n",
       " 14005,\n",
       " 14006,\n",
       " 14007,\n",
       " 14008,\n",
       " 14009,\n",
       " 14010,\n",
       " 14011,\n",
       " 14012,\n",
       " 14013,\n",
       " 14014,\n",
       " 14015,\n",
       " 14016,\n",
       " 14017,\n",
       " 14018,\n",
       " 14019,\n",
       " 14020,\n",
       " 14021,\n",
       " 14022,\n",
       " 14023,\n",
       " 14024,\n",
       " 14025,\n",
       " 14026,\n",
       " 14027,\n",
       " 14028,\n",
       " 14029,\n",
       " 14030,\n",
       " 14031,\n",
       " 14032,\n",
       " 14033,\n",
       " 14034,\n",
       " 14035,\n",
       " 14036,\n",
       " 14037,\n",
       " 14038,\n",
       " 14039,\n",
       " 14040,\n",
       " 14041,\n",
       " 14042,\n",
       " 14043,\n",
       " 14044,\n",
       " 14045,\n",
       " 14046,\n",
       " 14047,\n",
       " 14048,\n",
       " 14049,\n",
       " 14050,\n",
       " 14051,\n",
       " 14052,\n",
       " 14053,\n",
       " 14054,\n",
       " 14055,\n",
       " 14056,\n",
       " 14057,\n",
       " 14058,\n",
       " 14059,\n",
       " 14060,\n",
       " 14061,\n",
       " 14062,\n",
       " 14063,\n",
       " 14064,\n",
       " 14065,\n",
       " 14066,\n",
       " 14067,\n",
       " 14068,\n",
       " 14069,\n",
       " 14070,\n",
       " 14071,\n",
       " 14072,\n",
       " 14073,\n",
       " 14074,\n",
       " 14075,\n",
       " 14076,\n",
       " 14077,\n",
       " 14078,\n",
       " 14079,\n",
       " 14080,\n",
       " 14081,\n",
       " 14082,\n",
       " 14083,\n",
       " 14084,\n",
       " 14085,\n",
       " 14086,\n",
       " 14087,\n",
       " 14088,\n",
       " 14089,\n",
       " 14090,\n",
       " 14091,\n",
       " 14092,\n",
       " 14093,\n",
       " 14094,\n",
       " 14095,\n",
       " 14096,\n",
       " 14097,\n",
       " 14098,\n",
       " 14099,\n",
       " 14100,\n",
       " 14101,\n",
       " 14102,\n",
       " 14103,\n",
       " 14104,\n",
       " 14105,\n",
       " 14106,\n",
       " 14107,\n",
       " 14108,\n",
       " 14109,\n",
       " 14110,\n",
       " 14111,\n",
       " 14112,\n",
       " 14113,\n",
       " 14114,\n",
       " 14115,\n",
       " 14116,\n",
       " 14117,\n",
       " 14118,\n",
       " 14119,\n",
       " 14120,\n",
       " 14121,\n",
       " 14122,\n",
       " 14123,\n",
       " 14124,\n",
       " 14125,\n",
       " 14126,\n",
       " 14127,\n",
       " 14128,\n",
       " 14129,\n",
       " 14130,\n",
       " 14131,\n",
       " 14132,\n",
       " 14133,\n",
       " 14134,\n",
       " 14135,\n",
       " 14136,\n",
       " 14137,\n",
       " 14138,\n",
       " 14139,\n",
       " 14140,\n",
       " 14141,\n",
       " 14142,\n",
       " 14143,\n",
       " 14144,\n",
       " 14145,\n",
       " 14146,\n",
       " 14147,\n",
       " 14148,\n",
       " 14149,\n",
       " 14150,\n",
       " 14151,\n",
       " 14152,\n",
       " 14153,\n",
       " 14154,\n",
       " 14155,\n",
       " 14156,\n",
       " 14157,\n",
       " 14158,\n",
       " 14159,\n",
       " 14160,\n",
       " 14161,\n",
       " 14162,\n",
       " 14163,\n",
       " 14164,\n",
       " 14165,\n",
       " 14166,\n",
       " 14167,\n",
       " 14168,\n",
       " 14169,\n",
       " 14170,\n",
       " 14171,\n",
       " 14172,\n",
       " 14173,\n",
       " 14174,\n",
       " 14175,\n",
       " 14176,\n",
       " 14177,\n",
       " 14178,\n",
       " 14179,\n",
       " 14180,\n",
       " 14181,\n",
       " 14182,\n",
       " 14183,\n",
       " 14184,\n",
       " 14185,\n",
       " 14186,\n",
       " 14187,\n",
       " 14188,\n",
       " 14189,\n",
       " 14190,\n",
       " 14191,\n",
       " 14192,\n",
       " 14193,\n",
       " 14194,\n",
       " 14195,\n",
       " 14196,\n",
       " 14197,\n",
       " 14198,\n",
       " 14199,\n",
       " 14200,\n",
       " 14201,\n",
       " 14202,\n",
       " 14203,\n",
       " 14204,\n",
       " 14205,\n",
       " 14206,\n",
       " 14207,\n",
       " 14208,\n",
       " 14209,\n",
       " 14210,\n",
       " 14211,\n",
       " 14212,\n",
       " 14213,\n",
       " 14214,\n",
       " 14215,\n",
       " 14216,\n",
       " 14217,\n",
       " 14218,\n",
       " 14219,\n",
       " 14220,\n",
       " 14221,\n",
       " 14222,\n",
       " 14223,\n",
       " 14224,\n",
       " 14225,\n",
       " 14226,\n",
       " 14227,\n",
       " 14228,\n",
       " 14229,\n",
       " 14230,\n",
       " 14231,\n",
       " 14232,\n",
       " 14233,\n",
       " 14234,\n",
       " 14235,\n",
       " 14236,\n",
       " 14237,\n",
       " 14238,\n",
       " 14239,\n",
       " 14240,\n",
       " 14241,\n",
       " 14242,\n",
       " 14243,\n",
       " 14244,\n",
       " 14245,\n",
       " 14246,\n",
       " 14247,\n",
       " 14248,\n",
       " 14249,\n",
       " 14250,\n",
       " 14251,\n",
       " 14252,\n",
       " 14253,\n",
       " 14254,\n",
       " 14255,\n",
       " 14256,\n",
       " 14257,\n",
       " 14258,\n",
       " 14259,\n",
       " 14260,\n",
       " 14261,\n",
       " 14262,\n",
       " 14263,\n",
       " 14264,\n",
       " 14265,\n",
       " 14266,\n",
       " 14267,\n",
       " 14268,\n",
       " 14269,\n",
       " 14270,\n",
       " 14271,\n",
       " 14272,\n",
       " 14273,\n",
       " 14274,\n",
       " 14275,\n",
       " 14276,\n",
       " 14277,\n",
       " 14278,\n",
       " 14279,\n",
       " 14280,\n",
       " 14281,\n",
       " 14282,\n",
       " 14283,\n",
       " 14284,\n",
       " 14285,\n",
       " 14286,\n",
       " 14287,\n",
       " 14288,\n",
       " 14289,\n",
       " 14290,\n",
       " 14291,\n",
       " 14292,\n",
       " 14293,\n",
       " 14294,\n",
       " 14295,\n",
       " 14296,\n",
       " 14297,\n",
       " 14298,\n",
       " 14299,\n",
       " 14300,\n",
       " 14301,\n",
       " 14302,\n",
       " 14303,\n",
       " 14304,\n",
       " 14305,\n",
       " 14306,\n",
       " 14307,\n",
       " 14308,\n",
       " 14309,\n",
       " 14310,\n",
       " 14311,\n",
       " 14312,\n",
       " 14313,\n",
       " 14314,\n",
       " 14315,\n",
       " 14316,\n",
       " 14317,\n",
       " 14318,\n",
       " 14319,\n",
       " 14320,\n",
       " 14321,\n",
       " 14322,\n",
       " 14323,\n",
       " 14324,\n",
       " 14325,\n",
       " 14326,\n",
       " 14327,\n",
       " 14328,\n",
       " 14329,\n",
       " 14330,\n",
       " 14331,\n",
       " 14332,\n",
       " 14333,\n",
       " 14334,\n",
       " 14335,\n",
       " 14336,\n",
       " 14337,\n",
       " 14338,\n",
       " 14339,\n",
       " 14340,\n",
       " 14341,\n",
       " 14342,\n",
       " 14343,\n",
       " 14344,\n",
       " 14345,\n",
       " 14346,\n",
       " 14347,\n",
       " 14348,\n",
       " 14349,\n",
       " 14350,\n",
       " 14351,\n",
       " 14352,\n",
       " 14353,\n",
       " 14354,\n",
       " 14355,\n",
       " 14356,\n",
       " 14357,\n",
       " 14358,\n",
       " 14359,\n",
       " 14360,\n",
       " 14361,\n",
       " 14362,\n",
       " 14363,\n",
       " 14364,\n",
       " 14365,\n",
       " 14366,\n",
       " 14367,\n",
       " 14368,\n",
       " 14369,\n",
       " 14370,\n",
       " 14371,\n",
       " 14372,\n",
       " 14373,\n",
       " 14374,\n",
       " 14375,\n",
       " 14376,\n",
       " 14377,\n",
       " 14378,\n",
       " 14379,\n",
       " 14380,\n",
       " 14381,\n",
       " 14382,\n",
       " 14383,\n",
       " 14384,\n",
       " 14385,\n",
       " 14386,\n",
       " 14387,\n",
       " 14388,\n",
       " 14389,\n",
       " 14390,\n",
       " 14391,\n",
       " 14392,\n",
       " 14393,\n",
       " 14394,\n",
       " 14395,\n",
       " 14396,\n",
       " 14397,\n",
       " 14398,\n",
       " 14399,\n",
       " 14400,\n",
       " 14401,\n",
       " 14402,\n",
       " 14403,\n",
       " 14404,\n",
       " 14405,\n",
       " 14406,\n",
       " 14407,\n",
       " 14408,\n",
       " 14409,\n",
       " 14410,\n",
       " 14411,\n",
       " 14412,\n",
       " 14413,\n",
       " 14414,\n",
       " 14415,\n",
       " 14416,\n",
       " 14417,\n",
       " 14418,\n",
       " 14419,\n",
       " 14420,\n",
       " 14421,\n",
       " 14422,\n",
       " 14423,\n",
       " 14424,\n",
       " 14425,\n",
       " 14426,\n",
       " 14427,\n",
       " 14428,\n",
       " 14429,\n",
       " 14430,\n",
       " 14431,\n",
       " 14432,\n",
       " 14433,\n",
       " 14434,\n",
       " 14435,\n",
       " 14436,\n",
       " 14437,\n",
       " 14438,\n",
       " 14439,\n",
       " 14440,\n",
       " 14441,\n",
       " 14442,\n",
       " 14443,\n",
       " 14444,\n",
       " 14445,\n",
       " 14446,\n",
       " 14447,\n",
       " 14448,\n",
       " 14449,\n",
       " 14450,\n",
       " 14451,\n",
       " 14452,\n",
       " 14453,\n",
       " 14454,\n",
       " 14455,\n",
       " 14456,\n",
       " 14457,\n",
       " 14458,\n",
       " 14459,\n",
       " 14460,\n",
       " 14461,\n",
       " 14462,\n",
       " 14463,\n",
       " 14464,\n",
       " 14465,\n",
       " 14466,\n",
       " 14467,\n",
       " 14468,\n",
       " 14469,\n",
       " 14470,\n",
       " 14471,\n",
       " 14472,\n",
       " 14473,\n",
       " 14474,\n",
       " 14475,\n",
       " 14476,\n",
       " 14477,\n",
       " 14478,\n",
       " 14479,\n",
       " 14480,\n",
       " 14481,\n",
       " 14482,\n",
       " 14483,\n",
       " 14484,\n",
       " 14485,\n",
       " 14486,\n",
       " 14487,\n",
       " 14488,\n",
       " 14489,\n",
       " 14490,\n",
       " 14491,\n",
       " 14492,\n",
       " 14493,\n",
       " 14494,\n",
       " 14495,\n",
       " 14496,\n",
       " 14497,\n",
       " 14498,\n",
       " 14499,\n",
       " 14500,\n",
       " 14501,\n",
       " 14502,\n",
       " 14503,\n",
       " 14504,\n",
       " 14505,\n",
       " 14506,\n",
       " 14507,\n",
       " 14508,\n",
       " 14509,\n",
       " 14510,\n",
       " 14511,\n",
       " 14512,\n",
       " 14513,\n",
       " 14514,\n",
       " 14515,\n",
       " 14516,\n",
       " 14517,\n",
       " 14518,\n",
       " 14519,\n",
       " 14520,\n",
       " 14521,\n",
       " 14522,\n",
       " 14523,\n",
       " 14524,\n",
       " 14525,\n",
       " 14526,\n",
       " 14527,\n",
       " 14528,\n",
       " 14529,\n",
       " 14530,\n",
       " 14531,\n",
       " 14532,\n",
       " 14533,\n",
       " 14534,\n",
       " 14535,\n",
       " 14536,\n",
       " 14537,\n",
       " 14538,\n",
       " 14539,\n",
       " 14540,\n",
       " 14541,\n",
       " 14542,\n",
       " 14543,\n",
       " 14544,\n",
       " 14545,\n",
       " 14546,\n",
       " 14547,\n",
       " 14548,\n",
       " 14549,\n",
       " 14550,\n",
       " 14551,\n",
       " 14552,\n",
       " 14553,\n",
       " 14554,\n",
       " 14555,\n",
       " 14556,\n",
       " 14557,\n",
       " 14558,\n",
       " 14559,\n",
       " 14560,\n",
       " 14561,\n",
       " 14562,\n",
       " 14563,\n",
       " 14564,\n",
       " 14565,\n",
       " 14566,\n",
       " 14567,\n",
       " 14568,\n",
       " 14569,\n",
       " 14570,\n",
       " 14571,\n",
       " 14572,\n",
       " 14573,\n",
       " 14574,\n",
       " 14575,\n",
       " 14576,\n",
       " 14577,\n",
       " 14578,\n",
       " 14579,\n",
       " 14580,\n",
       " 14581,\n",
       " 14582,\n",
       " 14583,\n",
       " 14584,\n",
       " 14585,\n",
       " 14586,\n",
       " 14587,\n",
       " 14588,\n",
       " 14589,\n",
       " 14590,\n",
       " 14591,\n",
       " 14592,\n",
       " 14593,\n",
       " 14594,\n",
       " 14595,\n",
       " 14596,\n",
       " 14597,\n",
       " 14598,\n",
       " 14599,\n",
       " 14600,\n",
       " 14601,\n",
       " 14602,\n",
       " 14603,\n",
       " 14604,\n",
       " 14605,\n",
       " 14606,\n",
       " 14607,\n",
       " 14608,\n",
       " 14609,\n",
       " 14610,\n",
       " 14611,\n",
       " 14612,\n",
       " 14613,\n",
       " 14614,\n",
       " 14615,\n",
       " 14616,\n",
       " 14617,\n",
       " 14618,\n",
       " 14619,\n",
       " 14620,\n",
       " 14621,\n",
       " 14622,\n",
       " 14623,\n",
       " 14624,\n",
       " 14625,\n",
       " 14626,\n",
       " 14627,\n",
       " 14628,\n",
       " 14629,\n",
       " 14630,\n",
       " 14631,\n",
       " 14632,\n",
       " 14633,\n",
       " 14634,\n",
       " 14635,\n",
       " 14636,\n",
       " 14637,\n",
       " 14638,\n",
       " 14639,\n",
       " 14640,\n",
       " 14641,\n",
       " 14642,\n",
       " 14643,\n",
       " 14644,\n",
       " 14645,\n",
       " 14646,\n",
       " 14647,\n",
       " 14648,\n",
       " 14649,\n",
       " 14650,\n",
       " 14651,\n",
       " 14652,\n",
       " 14653,\n",
       " 14654,\n",
       " 14655,\n",
       " 14656,\n",
       " 14657,\n",
       " 14658,\n",
       " 14659,\n",
       " 14660,\n",
       " 14661,\n",
       " 14662,\n",
       " 14663,\n",
       " 14664,\n",
       " 14665,\n",
       " 14666,\n",
       " 14667,\n",
       " 14668,\n",
       " 14669,\n",
       " 14670,\n",
       " 14671,\n",
       " 14672,\n",
       " 14673,\n",
       " 14674,\n",
       " 14675,\n",
       " 14676,\n",
       " 14677,\n",
       " 14678,\n",
       " 14679,\n",
       " 14680,\n",
       " 14681,\n",
       " 14682,\n",
       " 14683,\n",
       " 14684,\n",
       " 14685,\n",
       " 14686,\n",
       " 14687,\n",
       " 14688,\n",
       " 14689,\n",
       " 14690,\n",
       " 14691,\n",
       " 14692,\n",
       " 14693,\n",
       " 14694,\n",
       " 14695,\n",
       " 14696,\n",
       " 14697,\n",
       " 14698,\n",
       " 14699,\n",
       " 14700,\n",
       " 14701,\n",
       " 14702,\n",
       " 14703,\n",
       " 14704,\n",
       " 14705,\n",
       " 14706,\n",
       " 14707,\n",
       " 14708,\n",
       " 14709,\n",
       " 14710,\n",
       " 14711,\n",
       " 14712,\n",
       " 14713,\n",
       " 14714,\n",
       " 14715,\n",
       " 14716,\n",
       " 14717,\n",
       " 14718,\n",
       " 14719,\n",
       " 14720,\n",
       " 14721,\n",
       " 14722,\n",
       " 14723,\n",
       " 14724,\n",
       " 14725,\n",
       " 14726,\n",
       " 14727,\n",
       " 14728,\n",
       " 14729,\n",
       " 14730,\n",
       " 14731,\n",
       " 14732,\n",
       " 14733,\n",
       " 14734,\n",
       " 14735,\n",
       " 14736,\n",
       " 14737,\n",
       " 14738,\n",
       " 14739,\n",
       " 14740,\n",
       " 14741,\n",
       " 14742,\n",
       " 14743,\n",
       " 14744,\n",
       " 14745,\n",
       " 14746,\n",
       " 14747,\n",
       " 14748,\n",
       " 14749,\n",
       " 14750,\n",
       " 14751,\n",
       " 14752,\n",
       " 14753,\n",
       " 14754,\n",
       " 14755,\n",
       " 14756,\n",
       " 14757,\n",
       " 14758,\n",
       " 14759,\n",
       " 14760,\n",
       " 14761,\n",
       " 14762,\n",
       " 14763,\n",
       " 14764,\n",
       " 14765,\n",
       " 14766,\n",
       " 14767,\n",
       " 14768,\n",
       " 14769,\n",
       " 14770,\n",
       " 14771,\n",
       " 14772,\n",
       " 14773,\n",
       " 14774,\n",
       " 14775,\n",
       " 14776,\n",
       " 14777,\n",
       " 14778,\n",
       " 14779,\n",
       " 14780,\n",
       " 14781,\n",
       " 14782,\n",
       " 14783,\n",
       " 14784,\n",
       " 14785,\n",
       " 14786,\n",
       " 14787,\n",
       " 14788,\n",
       " 14789,\n",
       " 14790,\n",
       " 14791,\n",
       " 14792,\n",
       " 14793,\n",
       " 14794,\n",
       " 14795,\n",
       " 14796,\n",
       " 14797,\n",
       " 14798,\n",
       " 14799,\n",
       " 14800,\n",
       " 14801,\n",
       " 14802,\n",
       " 14803,\n",
       " 14804,\n",
       " 14805,\n",
       " 14806,\n",
       " 14807,\n",
       " 14808,\n",
       " 14809,\n",
       " 14810,\n",
       " 14811,\n",
       " 14812,\n",
       " 14813,\n",
       " ...]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "07f5bad1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18797"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(graph['nodes']['embedding'][mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fb872356",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_index</th>\n",
       "      <th>primekg_node_index</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_type</th>\n",
       "      <th>desc</th>\n",
       "      <th>feat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10770</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PHYHIP_(0)</td>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PHYHIP belongs to gene/protein node. PHYHIP is...</td>\n",
       "      <td>MELLSTPHSIEINNITCDSFRISWAMEDSDLERVTHYFIDLNKKEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10771</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GPANK1_(1)</td>\n",
       "      <td>GPANK1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GPANK1 belongs to gene/protein node. GPANK1 is...</td>\n",
       "      <td>MSRPLLITFTPATDPSDLWKDGQQQPQPEKPESTLDGAAARAFYEA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10772</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>ZRSR2_(2)</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ZRSR2 belongs to gene/protein node. ZRSR2 is z...</td>\n",
       "      <td>MAAPEKMTFPEKPSHKKYRAALKKEKRKKRRQELARLRDSGLSQKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10773</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>NRF1_(3)</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NRF1 belongs to gene/protein node. NRF1 is nuc...</td>\n",
       "      <td>MEEHGVTQTEHMATIEAHAVAQQVQQVHVATYTEHSMLSADEDSPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10774</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>PI4KA_(4)</td>\n",
       "      <td>PI4KA</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PI4KA belongs to gene/protein node. PI4KA is p...</td>\n",
       "      <td>MAAAPARGGGGGGGGGGGCSGSGSSASRGFYFNTVLSLARSLAVQR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29562</th>\n",
       "      <td>52445</td>\n",
       "      <td>83734</td>\n",
       "      <td>MAB21L4_(83734)</td>\n",
       "      <td>MAB21L4</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MAB21L4 belongs to gene/protein node. MAB21L4 ...</td>\n",
       "      <td>MPAPALPTSAMAVQVPLWHHYLQAIRSREAPRAQDFQRAENVLLTV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29563</th>\n",
       "      <td>52446</td>\n",
       "      <td>83735</td>\n",
       "      <td>PRR23D2_(83735)</td>\n",
       "      <td>PRR23D2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PRR23D2 belongs to gene/protein node. PRR23D2 ...</td>\n",
       "      <td>MYGYRRLRSPRDSQTEPQNDNEGETSLATTQMNPPKRRQVEQGPST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29564</th>\n",
       "      <td>52447</td>\n",
       "      <td>83740</td>\n",
       "      <td>C8orf86_(83740)</td>\n",
       "      <td>C8orf86</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>C8orf86 belongs to gene/protein node.</td>\n",
       "      <td>MRPLGKGLLPAEELIRSNLGVGRSLRDCLSQSGKLAEELGSKRLKP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29565</th>\n",
       "      <td>52448</td>\n",
       "      <td>83746</td>\n",
       "      <td>CRACDL_(83746)</td>\n",
       "      <td>CRACDL</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>CRACDL belongs to gene/protein node. CRACDL is...</td>\n",
       "      <td>MISTRVMDIKLREAAEGLGEDSTGKKKSKFKTFKKFFGKKKRKESP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29566</th>\n",
       "      <td>52449</td>\n",
       "      <td>83747</td>\n",
       "      <td>POTEA_(83747)</td>\n",
       "      <td>POTEA</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>POTEA belongs to gene/protein node. POTEA is P...</td>\n",
       "      <td>MVAEVSPKLAASPMKKPFGFRGKMGKWCCCCFPCCRGSGKNNMGAW...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18797 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_index  primekg_node_index          node_id node_name  \\\n",
       "10770           0                   0       PHYHIP_(0)    PHYHIP   \n",
       "10771           1                   1       GPANK1_(1)    GPANK1   \n",
       "10772           2                   2        ZRSR2_(2)     ZRSR2   \n",
       "10773           3                   3         NRF1_(3)      NRF1   \n",
       "10774           4                   4        PI4KA_(4)     PI4KA   \n",
       "...           ...                 ...              ...       ...   \n",
       "29562       52445               83734  MAB21L4_(83734)   MAB21L4   \n",
       "29563       52446               83735  PRR23D2_(83735)   PRR23D2   \n",
       "29564       52447               83740  C8orf86_(83740)   C8orf86   \n",
       "29565       52448               83746   CRACDL_(83746)    CRACDL   \n",
       "29566       52449               83747    POTEA_(83747)     POTEA   \n",
       "\n",
       "          node_type                                               desc  \\\n",
       "10770  gene/protein  PHYHIP belongs to gene/protein node. PHYHIP is...   \n",
       "10771  gene/protein  GPANK1 belongs to gene/protein node. GPANK1 is...   \n",
       "10772  gene/protein  ZRSR2 belongs to gene/protein node. ZRSR2 is z...   \n",
       "10773  gene/protein  NRF1 belongs to gene/protein node. NRF1 is nuc...   \n",
       "10774  gene/protein  PI4KA belongs to gene/protein node. PI4KA is p...   \n",
       "...             ...                                                ...   \n",
       "29562  gene/protein  MAB21L4 belongs to gene/protein node. MAB21L4 ...   \n",
       "29563  gene/protein  PRR23D2 belongs to gene/protein node. PRR23D2 ...   \n",
       "29564  gene/protein             C8orf86 belongs to gene/protein node.    \n",
       "29565  gene/protein  CRACDL belongs to gene/protein node. CRACDL is...   \n",
       "29566  gene/protein  POTEA belongs to gene/protein node. POTEA is P...   \n",
       "\n",
       "                                                    feat  \n",
       "10770  MELLSTPHSIEINNITCDSFRISWAMEDSDLERVTHYFIDLNKKEN...  \n",
       "10771  MSRPLLITFTPATDPSDLWKDGQQQPQPEKPESTLDGAAARAFYEA...  \n",
       "10772  MAAPEKMTFPEKPSHKKYRAALKKEKRKKRRQELARLRDSGLSQKE...  \n",
       "10773  MEEHGVTQTEHMATIEAHAVAQQVQQVHVATYTEHSMLSADEDSPS...  \n",
       "10774  MAAAPARGGGGGGGGGGGCSGSGSSASRGFYFNTVLSLARSLAVQR...  \n",
       "...                                                  ...  \n",
       "29562  MPAPALPTSAMAVQVPLWHHYLQAIRSREAPRAQDFQRAENVLLTV...  \n",
       "29563  MYGYRRLRSPRDSQTEPQNDNEGETSLATTQMNPPKRRQVEQGPST...  \n",
       "29564  MRPLGKGLLPAEELIRSNLGVGRSLRDCLSQSGKLAEELGSKRLKP...  \n",
       "29565  MISTRVMDIKLREAAEGLGEDSTGKKKSKFKTFKKFFGKKKRKESP...  \n",
       "29566  MVAEVSPKLAASPMKKPFGFRGKMGKWCCCCFPCCRGSGKNNMGAW...  \n",
       "\n",
       "[18797 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask = graph['nodes']['enrichment'].node_type == 'gene/protein'\n",
    "graph['nodes']['enrichment'][mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c35a753e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        False\n",
       "1        False\n",
       "2        False\n",
       "3        False\n",
       "4        False\n",
       "         ...  \n",
       "84976    False\n",
       "84977    False\n",
       "84978    False\n",
       "84979    False\n",
       "84980    False\n",
       "Name: node_type, Length: 84981, dtype: bool"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "eac0ab67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18797"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph['nodes']['embedding'][mask].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1cfabf99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((84981, 7), (84981, 3))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_enrichment_df.shape, nodes_embedding_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d7463852",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((3904610, 11), 16)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_enrichment_df.shape, len(edges_embedding_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "34192052",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-0.02087522, -0.00298006, -0.00025966, ..., -0.00838831,\n",
       "       -0.0213233 , -0.02149462])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e_prompt = cp.array(edges_embedding_df[0].loc[0, \"edge_emb\"])\n",
    "e_prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cf3ff1e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1536,)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e_prompt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "937a060f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cuvs.distance import pairwise_distance\n",
    "def _compute_sim_scores(features_a: cp.ndarray,\n",
    "                        features_b:  cp.ndarray,\n",
    "                        metric: str=\"cosine\"):\n",
    "    \"\"\"\n",
    "    Compute the similarity scores between two sets of features using the specified metric.\n",
    "\n",
    "    Args:\n",
    "        features_a: The first set of features.\n",
    "        features_b: The second set of features.\n",
    "        metric: The metric to use for computing the similarity scores.\n",
    "\n",
    "    Returns:\n",
    "        The similarity scores between the two sets of features.\n",
    "    \"\"\"\n",
    "    scores = pairwise_distance(features_a, features_b, metric=metric)\n",
    "    scores = 1 - cp.asarray(scores).ravel()\n",
    "    return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ace8d7a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(Index(['node_index', 'primekg_node_index', 'node_id', 'node_name', 'node_type',\n",
       "        'desc', 'feat'],\n",
       "       dtype='object'),\n",
       " Index(['node_id', 'desc_emb', 'feat_emb'], dtype='object'),\n",
       " Index(['triplet_index', 'primekg_head_index', 'primekg_tail_index', 'head_id',\n",
       "        'tail_id', 'display_relation', 'edge_type', 'edge_type_str',\n",
       "        'head_index', 'tail_index', 'feat'],\n",
       "       dtype='object'))"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_enrichment_df.columns, nodes_embedding_df.columns, edges_enrichment_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d90ee35d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['triplet_index', 'primekg_head_index', 'primekg_tail_index', 'head_id',\n",
       "       'tail_id', 'display_relation', 'edge_type', 'edge_type_str',\n",
       "       'head_index', 'tail_index', 'feat'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_enrichment_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0a049944",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>triplet_index</th>\n",
       "      <th>edge_emb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3050000</td>\n",
       "      <td>[-0.02087522, -0.002980056, -0.00025966353, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3050001</td>\n",
       "      <td>[-0.01760313, 0.014069217, 0.019861642, -0.011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3050002</td>\n",
       "      <td>[-0.019058425, 0.013303685, 0.019948881, -0.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3050003</td>\n",
       "      <td>[-0.019800434, 0.010459852, 0.014790365, -0.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3050004</td>\n",
       "      <td>[-0.022970848, 0.012780121, 0.0066903764, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249995</th>\n",
       "      <td>1599995</td>\n",
       "      <td>[-0.0012519549, -0.0014191568, 0.022983257, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249996</th>\n",
       "      <td>1599996</td>\n",
       "      <td>[-0.011657391, -0.0136824455, 0.017119711, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249997</th>\n",
       "      <td>1599997</td>\n",
       "      <td>[-0.010922454, -0.0057538515, 0.010422692, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249998</th>\n",
       "      <td>1599998</td>\n",
       "      <td>[-0.0056640725, -0.00014858306, 0.016633065, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249999</th>\n",
       "      <td>1599999</td>\n",
       "      <td>[-0.0070809745, -0.012916284, 0.005389001, -0....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250000 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        triplet_index                                           edge_emb\n",
       "0             3050000  [-0.02087522, -0.002980056, -0.00025966353, -0...\n",
       "1             3050001  [-0.01760313, 0.014069217, 0.019861642, -0.011...\n",
       "2             3050002  [-0.019058425, 0.013303685, 0.019948881, -0.00...\n",
       "3             3050003  [-0.019800434, 0.010459852, 0.014790365, -0.00...\n",
       "4             3050004  [-0.022970848, 0.012780121, 0.0066903764, -0.0...\n",
       "...               ...                                                ...\n",
       "249995        1599995  [-0.0012519549, -0.0014191568, 0.022983257, -0...\n",
       "249996        1599996  [-0.011657391, -0.0136824455, 0.017119711, -0....\n",
       "249997        1599997  [-0.010922454, -0.0057538515, 0.010422692, -0....\n",
       "249998        1599998  [-0.0056640725, -0.00014858306, 0.016633065, -...\n",
       "249999        1599999  [-0.0070809745, -0.012916284, 0.005389001, -0....\n",
       "\n",
       "[250000 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_embedding_df[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2d2527b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the data path\n",
    "DATA_PATH = \"../../../aiagents4pharma/talk2knowledgegraphs/tests/files\"\n",
    "\n",
    "# Define the agent state\n",
    "state = {\n",
    "    \"selections\": {\n",
    "        \"gene/protein\": [],\n",
    "        \"molecular_function\": [],\n",
    "        \"cellular_component\": [],\n",
    "        \"biological_process\": [],\n",
    "        \"drug\": [],\n",
    "        \"disease\": []\n",
    "    },\n",
    "    \"uploaded_files\": [],\n",
    "    \"topk_nodes\": 3,\n",
    "    \"topk_edges\": 3,\n",
    "    \"dic_source_graph\": [\n",
    "        {\n",
    "            \"name\": \"BioBridge\",\n",
    "            \"kg_pyg_path\": f\"{DATA_PATH}/biobridge_multimodal_pyg_graph.pkl\",\n",
    "            \"kg_text_path\": f\"{DATA_PATH}/biobridge_multimodal_text_graph.pkl\",\n",
    "        }\n",
    "    ],\n",
    "}\n",
    "\n",
    "# Update state\n",
    "state[\"uploaded_files\"] = [\n",
    "    {\n",
    "        \"file_name\": \"multimodal-analysis_single_gene.xlsx\",\n",
    "        \"file_path\": f\"{DATA_PATH}/multimodal-analysis_single_gene.xlsx\",\n",
    "        \"file_type\": \"multimodal\",\n",
    "        \"uploaded_by\": \"VPEUser\",\n",
    "        \"uploaded_timestamp\": \"2025-05-12 00:00:00\",\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b3cfeb84",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "multimodal_df = cudf.DataFrame({\"name\": [], \"node_type\": []})\n",
    "query_df = cudf.DataFrame({\"node_id\": [],\n",
    "                            \"node_type\": [],\n",
    "                            \"feat_emb\": [],\n",
    "                            \"desc_emb\": [],\n",
    "                            \"use_description\": []})\n",
    "\n",
    "# Loop over the uploaded files and find multimodal files\n",
    "# logger.log(logging.INFO, \"Looping over uploaded files\")\n",
    "for i in range(len(state[\"uploaded_files\"])):\n",
    "    # Check if multimodal file is uploaded\n",
    "    if state[\"uploaded_files\"][i][\"file_type\"] == \"multimodal\":\n",
    "        # Read the Excel file\n",
    "        multimodal_df = pd.read_excel(state[\"uploaded_files\"][i][\"file_path\"],\n",
    "                                        sheet_name=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f3570082",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if the multimodal_df is empty\n",
    "# logger.log(logging.INFO, \"Checking if multimodal_df is empty\")\n",
    "if len(multimodal_df) > 0:\n",
    "    # Prepare multimodal_df\n",
    "    # logger.log(logging.INFO, \"Preparing multimodal_df\")\n",
    "    # Merge all obtained dataframes into a single dataframe\n",
    "    multimodal_df = cudf.from_pandas(pd.concat(multimodal_df).reset_index())\n",
    "    multimodal_df.drop(columns=[\"level_1\"], inplace=True)\n",
    "    multimodal_df.rename(columns={\"level_0\": \"q_node_type\",\n",
    "                                    \"name\": \"q_node_name\"}, inplace=True)\n",
    "    # Since an excel sheet name could not contain a `/`,\n",
    "    # but the node type can be 'gene/protein' as exists in the PrimeKG\n",
    "    multimodal_df[\"q_node_type\"] = multimodal_df[\"q_node_type\"].str.replace('-', '/')\n",
    "\n",
    "    # Make and process a query dataframe by merging the graph_df and multimodal_df\n",
    "    # Merging with enrichment dataframe\n",
    "    # logger.log(logging.INFO, \"Merging with enrichment dataframe\")\n",
    "    query_df = graph['nodes']['enrichment'][\n",
    "        ['node_id', 'node_name', 'node_type', 'feat', 'desc']\n",
    "    ].merge(multimodal_df, how='cross')\n",
    "    # Post-process the query dataframe\n",
    "    # logger.log(logging.INFO, \"Lowering case for node names (q_node_name)\")\n",
    "    query_df['q_node_name'] = query_df['q_node_name'].str.lower()\n",
    "    # logger.log(logging.INFO, \"Lowering case for node names (node_name)\")\n",
    "    query_df['node_name'] = query_df['node_name'].str.lower()\n",
    "    # Get the mask for filtering based on the query\n",
    "    # logger.log(logging.INFO, \"Filtering based on the query\")\n",
    "    mask = (\n",
    "        query_df['node_name'].str.contains(query_df['q_node_name']) &\n",
    "        (query_df['node_type'] == query_df['q_node_type'])\n",
    "    )\n",
    "    query_df = query_df[mask]\n",
    "    # Merging the final query dataframe with embedding dataframe\n",
    "    # logger.log(logging.INFO, \"Merging with embedding dataframe\")\n",
    "    query_df = query_df.merge(\n",
    "        graph['nodes']['embedding'][['node_id', 'feat_emb', 'desc_emb']],\n",
    "        how='left',\n",
    "        on='node_id')\n",
    "    # Post-process the query dataframe\n",
    "    query_df = query_df[['node_id',\n",
    "                            'node_type',\n",
    "                            'feat',\n",
    "                            'feat_emb',\n",
    "                            'desc',\n",
    "                            'desc_emb']].reset_index(drop=True)\n",
    "    query_df['use_description'] = False # set to False for modal-specific embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "71cd4281",
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt = {}\n",
    "prompt[\"text\"] = \"What is the function of the gene/protein BRCA1?\"\n",
    "prompt[\"emb\"] = [emb_model.embed_query(prompt[\"text\"])]\n",
    "# emb_ = emb_model.embed_query(\"Hello world\")\n",
    "# cp.asarray(emb_).reshape(1, -1).astype('float32')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9e9f1314",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['node_id', 'node_type', 'feat', 'feat_emb', 'desc', 'desc_emb',\n",
       "       'use_description'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "55d52266",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>feat</th>\n",
       "      <th>feat_emb</th>\n",
       "      <th>desc</th>\n",
       "      <th>desc_emb</th>\n",
       "      <th>use_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RPS8_(11)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MGISRDNWHKRRKTGGKRKPYHKKRKYELGRPAANTKIGPRRIHTV...</td>\n",
       "      <td>[0.05257261171936989, -0.06346512585878372, -0...</td>\n",
       "      <td>RPS8 belongs to gene/protein node. RPS8 is rib...</td>\n",
       "      <td>[-0.017070618, -0.0182947, -0.017336722, -0.01...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FOS_(22)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MMFSGFNADYEASSSRCSSASPAGDSLSYYHSPADSFSSMGSPVNA...</td>\n",
       "      <td>[-0.02010018192231655, -0.0288854893296957, 0....</td>\n",
       "      <td>FOS belongs to gene/protein node. FOS is Fos p...</td>\n",
       "      <td>[-0.01809412, -0.011164032, -0.021542521, -0.0...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CALM1_(61)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTE...</td>\n",
       "      <td>[0.015500295907258987, 0.12202692776918411, -0...</td>\n",
       "      <td>CALM1 belongs to gene/protein node. CALM1 is c...</td>\n",
       "      <td>[-0.023683596, -0.008636789, -0.020697404, -0....</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SETD7_(103)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MDSDDEMVEEAVEGHLDDDGLPHGFCTVTYSSTDRFEGNFVHGEKN...</td>\n",
       "      <td>[-0.09826023876667023, -0.06921421736478806, 0...</td>\n",
       "      <td>SETD7 belongs to gene/protein node. SETD7 is S...</td>\n",
       "      <td>[-0.028181268, -0.0034074313, -0.021452002, -0...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HNRNPUL1_(159)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MDVRRLKVNELREELQRRGLDTRGLKAELAERLQAALEAEEPDDER...</td>\n",
       "      <td>[-0.07706909626722336, 0.004895209334790707, -...</td>\n",
       "      <td>HNRNPUL1 belongs to gene/protein node. HNRNPUL...</td>\n",
       "      <td>[-0.027910667, -0.0030972639, -0.020933, -0.01...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>PRPSAP2_(5003)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MFCVTPPELETKMNITKGGLVLFSANSNSSCMELSKKIAERLGVEM...</td>\n",
       "      <td>[-0.07528620213270187, 0.011673472821712494, 0...</td>\n",
       "      <td>PRPSAP2 belongs to gene/protein node. PRPSAP2 ...</td>\n",
       "      <td>[-0.007135198, -0.0068572033, 0.0063343085, -0...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ANXA5_(5004)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MAQVLRGTVTDFPGFDERADAETLRKAMKGLGTDEESILTLLTSRS...</td>\n",
       "      <td>[0.0765886902809143, 0.10536942631006241, 0.02...</td>\n",
       "      <td>ANXA5 belongs to gene/protein node. ANXA5 is a...</td>\n",
       "      <td>[-0.037901636, 0.0032147528, -0.007965701, -0....</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>LTBP4_(5024)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MPRPGTSGRRPLLLVLLLPLFAAATSAASPSPSPSQVVEVPGVPSR...</td>\n",
       "      <td>[-0.018929054960608482, -0.058839354664087296,...</td>\n",
       "      <td>LTBP4 belongs to gene/protein node. LTBP4 is l...</td>\n",
       "      <td>[-0.03300405, -0.008081024, -0.00728992, -0.03...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>GYPC_(5074)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MWSTRSPNSTAWPLSLEPDPGMASASTTMHTTTIAEPDPGMSGWPD...</td>\n",
       "      <td>[0.0609404481947422, -0.03754159063100815, -0....</td>\n",
       "      <td>GYPC belongs to gene/protein node. GYPC is gly...</td>\n",
       "      <td>[-0.002514031, 0.008406238, 0.017207902, -0.03...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>CD74_(4694)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MHRRRSRSCREDQKPVMDDQRDLISNNEQLPMLGRRPGAPESKCSR...</td>\n",
       "      <td>[0.04463711008429527, -0.0819774940609932, 0.0...</td>\n",
       "      <td>CD74 belongs to gene/protein node. CD74 is CD7...</td>\n",
       "      <td>[-0.023425357, -0.007765312, -0.015478855, -0....</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>411 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            node_id     node_type  \\\n",
       "0         RPS8_(11)  gene/protein   \n",
       "1          FOS_(22)  gene/protein   \n",
       "2        CALM1_(61)  gene/protein   \n",
       "3       SETD7_(103)  gene/protein   \n",
       "4    HNRNPUL1_(159)  gene/protein   \n",
       "..              ...           ...   \n",
       "406  PRPSAP2_(5003)  gene/protein   \n",
       "407    ANXA5_(5004)  gene/protein   \n",
       "408    LTBP4_(5024)  gene/protein   \n",
       "409     GYPC_(5074)  gene/protein   \n",
       "410     CD74_(4694)  gene/protein   \n",
       "\n",
       "                                                  feat  \\\n",
       "0    MGISRDNWHKRRKTGGKRKPYHKKRKYELGRPAANTKIGPRRIHTV...   \n",
       "1    MMFSGFNADYEASSSRCSSASPAGDSLSYYHSPADSFSSMGSPVNA...   \n",
       "2    MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTE...   \n",
       "3    MDSDDEMVEEAVEGHLDDDGLPHGFCTVTYSSTDRFEGNFVHGEKN...   \n",
       "4    MDVRRLKVNELREELQRRGLDTRGLKAELAERLQAALEAEEPDDER...   \n",
       "..                                                 ...   \n",
       "406  MFCVTPPELETKMNITKGGLVLFSANSNSSCMELSKKIAERLGVEM...   \n",
       "407  MAQVLRGTVTDFPGFDERADAETLRKAMKGLGTDEESILTLLTSRS...   \n",
       "408  MPRPGTSGRRPLLLVLLLPLFAAATSAASPSPSPSQVVEVPGVPSR...   \n",
       "409  MWSTRSPNSTAWPLSLEPDPGMASASTTMHTTTIAEPDPGMSGWPD...   \n",
       "410  MHRRRSRSCREDQKPVMDDQRDLISNNEQLPMLGRRPGAPESKCSR...   \n",
       "\n",
       "                                              feat_emb  \\\n",
       "0    [0.05257261171936989, -0.06346512585878372, -0...   \n",
       "1    [-0.02010018192231655, -0.0288854893296957, 0....   \n",
       "2    [0.015500295907258987, 0.12202692776918411, -0...   \n",
       "3    [-0.09826023876667023, -0.06921421736478806, 0...   \n",
       "4    [-0.07706909626722336, 0.004895209334790707, -...   \n",
       "..                                                 ...   \n",
       "406  [-0.07528620213270187, 0.011673472821712494, 0...   \n",
       "407  [0.0765886902809143, 0.10536942631006241, 0.02...   \n",
       "408  [-0.018929054960608482, -0.058839354664087296,...   \n",
       "409  [0.0609404481947422, -0.03754159063100815, -0....   \n",
       "410  [0.04463711008429527, -0.0819774940609932, 0.0...   \n",
       "\n",
       "                                                  desc  \\\n",
       "0    RPS8 belongs to gene/protein node. RPS8 is rib...   \n",
       "1    FOS belongs to gene/protein node. FOS is Fos p...   \n",
       "2    CALM1 belongs to gene/protein node. CALM1 is c...   \n",
       "3    SETD7 belongs to gene/protein node. SETD7 is S...   \n",
       "4    HNRNPUL1 belongs to gene/protein node. HNRNPUL...   \n",
       "..                                                 ...   \n",
       "406  PRPSAP2 belongs to gene/protein node. PRPSAP2 ...   \n",
       "407  ANXA5 belongs to gene/protein node. ANXA5 is a...   \n",
       "408  LTBP4 belongs to gene/protein node. LTBP4 is l...   \n",
       "409  GYPC belongs to gene/protein node. GYPC is gly...   \n",
       "410  CD74 belongs to gene/protein node. CD74 is CD7...   \n",
       "\n",
       "                                              desc_emb  use_description  \n",
       "0    [-0.017070618, -0.0182947, -0.017336722, -0.01...            False  \n",
       "1    [-0.01809412, -0.011164032, -0.021542521, -0.0...            False  \n",
       "2    [-0.023683596, -0.008636789, -0.020697404, -0....            False  \n",
       "3    [-0.028181268, -0.0034074313, -0.021452002, -0...            False  \n",
       "4    [-0.027910667, -0.0030972639, -0.020933, -0.01...            False  \n",
       "..                                                 ...              ...  \n",
       "406  [-0.007135198, -0.0068572033, 0.0063343085, -0...            False  \n",
       "407  [-0.037901636, 0.0032147528, -0.007965701, -0....            False  \n",
       "408  [-0.03300405, -0.008081024, -0.00728992, -0.03...            False  \n",
       "409  [-0.002514031, 0.008406238, 0.017207902, -0.03...            False  \n",
       "410  [-0.023425357, -0.007765312, -0.015478855, -0....            False  \n",
       "\n",
       "[411 rows x 7 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "79ca55c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>feat</th>\n",
       "      <th>feat_emb</th>\n",
       "      <th>desc</th>\n",
       "      <th>desc_emb</th>\n",
       "      <th>use_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>user_prompt</td>\n",
       "      <td>prompt</td>\n",
       "      <td>What is the function of the gene/protein BRCA1?</td>\n",
       "      <td>[-0.028648853302001953, -0.0061271535232663155...</td>\n",
       "      <td>What is the function of the gene/protein BRCA1?</td>\n",
       "      <td>[-0.028648853302001953, -0.0061271535232663155...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_id node_type                                             feat  \\\n",
       "0  user_prompt    prompt  What is the function of the gene/protein BRCA1?   \n",
       "\n",
       "                                            feat_emb  \\\n",
       "0  [-0.028648853302001953, -0.0061271535232663155...   \n",
       "\n",
       "                                              desc  \\\n",
       "0  What is the function of the gene/protein BRCA1?   \n",
       "\n",
       "                                            desc_emb  use_description  \n",
       "0  [-0.028648853302001953, -0.0061271535232663155...             True  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cudf.DataFrame({\n",
    "        'node_id': 'user_prompt',\n",
    "        'node_type': 'prompt',\n",
    "        'feat': prompt[\"text\"],\n",
    "        'feat_emb': prompt[\"emb\"],\n",
    "        'desc': prompt[\"text\"],\n",
    "        'desc_emb': prompt[\"emb\"],\n",
    "        'use_description': True # set to True for user prompt embedding\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8e068c23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>feat</th>\n",
       "      <th>feat_emb</th>\n",
       "      <th>desc</th>\n",
       "      <th>desc_emb</th>\n",
       "      <th>use_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RPS8_(11)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MGISRDNWHKRRKTGGKRKPYHKKRKYELGRPAANTKIGPRRIHTV...</td>\n",
       "      <td>[0.05257261171936989, -0.06346512585878372, -0...</td>\n",
       "      <td>RPS8 belongs to gene/protein node. RPS8 is rib...</td>\n",
       "      <td>[-0.017070618, -0.0182947, -0.017336722, -0.01...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FOS_(22)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MMFSGFNADYEASSSRCSSASPAGDSLSYYHSPADSFSSMGSPVNA...</td>\n",
       "      <td>[-0.02010018192231655, -0.0288854893296957, 0....</td>\n",
       "      <td>FOS belongs to gene/protein node. FOS is Fos p...</td>\n",
       "      <td>[-0.01809412, -0.011164032, -0.021542521, -0.0...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CALM1_(61)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTE...</td>\n",
       "      <td>[0.015500295907258987, 0.12202692776918411, -0...</td>\n",
       "      <td>CALM1 belongs to gene/protein node. CALM1 is c...</td>\n",
       "      <td>[-0.023683596, -0.008636789, -0.020697404, -0....</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SETD7_(103)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MDSDDEMVEEAVEGHLDDDGLPHGFCTVTYSSTDRFEGNFVHGEKN...</td>\n",
       "      <td>[-0.09826023876667023, -0.06921421736478806, 0...</td>\n",
       "      <td>SETD7 belongs to gene/protein node. SETD7 is S...</td>\n",
       "      <td>[-0.028181268, -0.0034074313, -0.021452002, -0...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HNRNPUL1_(159)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MDVRRLKVNELREELQRRGLDTRGLKAELAERLQAALEAEEPDDER...</td>\n",
       "      <td>[-0.07706909626722336, 0.004895209334790707, -...</td>\n",
       "      <td>HNRNPUL1 belongs to gene/protein node. HNRNPUL...</td>\n",
       "      <td>[-0.027910667, -0.0030972639, -0.020933, -0.01...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ANXA5_(5004)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MAQVLRGTVTDFPGFDERADAETLRKAMKGLGTDEESILTLLTSRS...</td>\n",
       "      <td>[0.0765886902809143, 0.10536942631006241, 0.02...</td>\n",
       "      <td>ANXA5 belongs to gene/protein node. ANXA5 is a...</td>\n",
       "      <td>[-0.037901636, 0.0032147528, -0.007965701, -0....</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>LTBP4_(5024)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MPRPGTSGRRPLLLVLLLPLFAAATSAASPSPSPSQVVEVPGVPSR...</td>\n",
       "      <td>[-0.018929054960608482, -0.058839354664087296,...</td>\n",
       "      <td>LTBP4 belongs to gene/protein node. LTBP4 is l...</td>\n",
       "      <td>[-0.03300405, -0.008081024, -0.00728992, -0.03...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>GYPC_(5074)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MWSTRSPNSTAWPLSLEPDPGMASASTTMHTTTIAEPDPGMSGWPD...</td>\n",
       "      <td>[0.0609404481947422, -0.03754159063100815, -0....</td>\n",
       "      <td>GYPC belongs to gene/protein node. GYPC is gly...</td>\n",
       "      <td>[-0.002514031, 0.008406238, 0.017207902, -0.03...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>CD74_(4694)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MHRRRSRSCREDQKPVMDDQRDLISNNEQLPMLGRRPGAPESKCSR...</td>\n",
       "      <td>[0.04463711008429527, -0.0819774940609932, 0.0...</td>\n",
       "      <td>CD74 belongs to gene/protein node. CD74 is CD7...</td>\n",
       "      <td>[-0.023425357, -0.007765312, -0.015478855, -0....</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>user_prompt</td>\n",
       "      <td>prompt</td>\n",
       "      <td>What is the function of the gene/protein BRCA1?</td>\n",
       "      <td>[-0.028648853302001953, -0.0061271535232663155...</td>\n",
       "      <td>What is the function of the gene/protein BRCA1?</td>\n",
       "      <td>[-0.028648853302001953, -0.0061271535232663155...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>412 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            node_id     node_type  \\\n",
       "0         RPS8_(11)  gene/protein   \n",
       "1          FOS_(22)  gene/protein   \n",
       "2        CALM1_(61)  gene/protein   \n",
       "3       SETD7_(103)  gene/protein   \n",
       "4    HNRNPUL1_(159)  gene/protein   \n",
       "..              ...           ...   \n",
       "407    ANXA5_(5004)  gene/protein   \n",
       "408    LTBP4_(5024)  gene/protein   \n",
       "409     GYPC_(5074)  gene/protein   \n",
       "410     CD74_(4694)  gene/protein   \n",
       "411     user_prompt        prompt   \n",
       "\n",
       "                                                  feat  \\\n",
       "0    MGISRDNWHKRRKTGGKRKPYHKKRKYELGRPAANTKIGPRRIHTV...   \n",
       "1    MMFSGFNADYEASSSRCSSASPAGDSLSYYHSPADSFSSMGSPVNA...   \n",
       "2    MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTE...   \n",
       "3    MDSDDEMVEEAVEGHLDDDGLPHGFCTVTYSSTDRFEGNFVHGEKN...   \n",
       "4    MDVRRLKVNELREELQRRGLDTRGLKAELAERLQAALEAEEPDDER...   \n",
       "..                                                 ...   \n",
       "407  MAQVLRGTVTDFPGFDERADAETLRKAMKGLGTDEESILTLLTSRS...   \n",
       "408  MPRPGTSGRRPLLLVLLLPLFAAATSAASPSPSPSQVVEVPGVPSR...   \n",
       "409  MWSTRSPNSTAWPLSLEPDPGMASASTTMHTTTIAEPDPGMSGWPD...   \n",
       "410  MHRRRSRSCREDQKPVMDDQRDLISNNEQLPMLGRRPGAPESKCSR...   \n",
       "411    What is the function of the gene/protein BRCA1?   \n",
       "\n",
       "                                              feat_emb  \\\n",
       "0    [0.05257261171936989, -0.06346512585878372, -0...   \n",
       "1    [-0.02010018192231655, -0.0288854893296957, 0....   \n",
       "2    [0.015500295907258987, 0.12202692776918411, -0...   \n",
       "3    [-0.09826023876667023, -0.06921421736478806, 0...   \n",
       "4    [-0.07706909626722336, 0.004895209334790707, -...   \n",
       "..                                                 ...   \n",
       "407  [0.0765886902809143, 0.10536942631006241, 0.02...   \n",
       "408  [-0.018929054960608482, -0.058839354664087296,...   \n",
       "409  [0.0609404481947422, -0.03754159063100815, -0....   \n",
       "410  [0.04463711008429527, -0.0819774940609932, 0.0...   \n",
       "411  [-0.028648853302001953, -0.0061271535232663155...   \n",
       "\n",
       "                                                  desc  \\\n",
       "0    RPS8 belongs to gene/protein node. RPS8 is rib...   \n",
       "1    FOS belongs to gene/protein node. FOS is Fos p...   \n",
       "2    CALM1 belongs to gene/protein node. CALM1 is c...   \n",
       "3    SETD7 belongs to gene/protein node. SETD7 is S...   \n",
       "4    HNRNPUL1 belongs to gene/protein node. HNRNPUL...   \n",
       "..                                                 ...   \n",
       "407  ANXA5 belongs to gene/protein node. ANXA5 is a...   \n",
       "408  LTBP4 belongs to gene/protein node. LTBP4 is l...   \n",
       "409  GYPC belongs to gene/protein node. GYPC is gly...   \n",
       "410  CD74 belongs to gene/protein node. CD74 is CD7...   \n",
       "411    What is the function of the gene/protein BRCA1?   \n",
       "\n",
       "                                              desc_emb  use_description  \n",
       "0    [-0.017070618, -0.0182947, -0.017336722, -0.01...            False  \n",
       "1    [-0.01809412, -0.011164032, -0.021542521, -0.0...            False  \n",
       "2    [-0.023683596, -0.008636789, -0.020697404, -0....            False  \n",
       "3    [-0.028181268, -0.0034074313, -0.021452002, -0...            False  \n",
       "4    [-0.027910667, -0.0030972639, -0.020933, -0.01...            False  \n",
       "..                                                 ...              ...  \n",
       "407  [-0.037901636, 0.0032147528, -0.007965701, -0....            False  \n",
       "408  [-0.03300405, -0.008081024, -0.00728992, -0.03...            False  \n",
       "409  [-0.002514031, 0.008406238, 0.017207902, -0.03...            False  \n",
       "410  [-0.023425357, -0.007765312, -0.015478855, -0....            False  \n",
       "411  [-0.028648853302001953, -0.0061271535232663155...             True  \n",
       "\n",
       "[412 rows x 7 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_df = cudf.concat([\n",
    "    query_df,\n",
    "    cudf.DataFrame({\n",
    "        'node_id': 'user_prompt',\n",
    "        'node_type': 'prompt',\n",
    "        'feat': prompt[\"text\"],\n",
    "        'feat_emb': prompt[\"emb\"],\n",
    "        'desc': prompt[\"text\"],\n",
    "        'desc_emb': prompt[\"emb\"],\n",
    "        'use_description': True # set to True for user prompt embedding\n",
    "    })\n",
    "]).reset_index(drop=True)\n",
    "query_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e9606a15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "2560\n",
      "1536\n"
     ]
    }
   ],
   "source": [
    "for i in query_df['feat_emb'].to_pandas().tolist():\n",
    "    print(len(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "cf63c222",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_emb = cp.asarray(graph['nodes']['embedding']['desc_emb'][0]).reshape(1, -1).astype(cp.float32)\n",
    "# ===\n",
    "sim_scores = cudf.Series(cp.zeros(len(graph['nodes']['embedding']), dtype=cp.float32))\n",
    "sim_scores[:] = _compute_sim_scores(\n",
    "                graph['nodes']['embedding']['desc_emb'].list.leaves.to_cupy().reshape(\n",
    "                    -1, len(graph['nodes']['embedding'][\"desc_emb\"][0])\n",
    "                ).astype(cp.float32),\n",
    "                query_emb\n",
    "            )  # shape [N, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "003d7892",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>desc_emb</th>\n",
       "      <th>feat_emb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Copper_(14012)</td>\n",
       "      <td>[-0.008437113, -0.0066760327, 0.017289385, -0....</td>\n",
       "      <td>[1.9533191919326782, 0.6867725253105164, 0.456...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Oxygen_(14013)</td>\n",
       "      <td>[0.0011131193, 0.00035166115, -0.011181605, -0...</td>\n",
       "      <td>[0.21793906390666962, 0.13471993803977966, 0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Flunisolide_(14014)</td>\n",
       "      <td>[-0.034678344, 0.0014620015, -0.00069094024, -...</td>\n",
       "      <td>[-0.12004103511571884, -0.36696380376815796, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Alclometasone_(14015)</td>\n",
       "      <td>[-0.00836069, 0.0018004619, 0.0005102673, -0.0...</td>\n",
       "      <td>[-0.24334891140460968, -0.6172693967819214, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Medrysone_(14016)</td>\n",
       "      <td>[-0.022494577, 0.00894156, -0.004813894, -0.02...</td>\n",
       "      <td>[-0.1530350148677826, -0.33404994010925293, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84976</th>\n",
       "      <td>structural constituent of tooth enamel_(124215)</td>\n",
       "      <td>[0.0027853793, 0.010976661, -0.0056875315, -0....</td>\n",
       "      <td>[-0.011537055, 0.008254736, 0.01594828, -0.017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84977</th>\n",
       "      <td>D-glucuronate transmembrane transporter activi...</td>\n",
       "      <td>[-0.0019518828, -0.024248326, 0.006703199, -0....</td>\n",
       "      <td>[-0.011491481, -0.013787123, 0.016679898, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84978</th>\n",
       "      <td>histone propionyltransferase activity_(124217)</td>\n",
       "      <td>[-0.010533179, -0.032219134, -0.0032646793, -0...</td>\n",
       "      <td>[-0.0020738854, -0.010127607, 0.021126406, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84979</th>\n",
       "      <td>sirohydrochlorin ferrochelatase activity_(124219)</td>\n",
       "      <td>[-0.0029098308, -0.0076754037, -0.0111139845, ...</td>\n",
       "      <td>[-0.005530768, 0.011203765, -0.0023338485, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84980</th>\n",
       "      <td>ethylene binding_(124220)</td>\n",
       "      <td>[-0.019622136, -0.0142971175, -0.02942627, -0....</td>\n",
       "      <td>[-0.029408008, -0.02011434, -0.017020838, -0.0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84981 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 node_id  \\\n",
       "0                                         Copper_(14012)   \n",
       "1                                         Oxygen_(14013)   \n",
       "2                                    Flunisolide_(14014)   \n",
       "3                                  Alclometasone_(14015)   \n",
       "4                                      Medrysone_(14016)   \n",
       "...                                                  ...   \n",
       "84976    structural constituent of tooth enamel_(124215)   \n",
       "84977  D-glucuronate transmembrane transporter activi...   \n",
       "84978     histone propionyltransferase activity_(124217)   \n",
       "84979  sirohydrochlorin ferrochelatase activity_(124219)   \n",
       "84980                          ethylene binding_(124220)   \n",
       "\n",
       "                                                desc_emb  \\\n",
       "0      [-0.008437113, -0.0066760327, 0.017289385, -0....   \n",
       "1      [0.0011131193, 0.00035166115, -0.011181605, -0...   \n",
       "2      [-0.034678344, 0.0014620015, -0.00069094024, -...   \n",
       "3      [-0.00836069, 0.0018004619, 0.0005102673, -0.0...   \n",
       "4      [-0.022494577, 0.00894156, -0.004813894, -0.02...   \n",
       "...                                                  ...   \n",
       "84976  [0.0027853793, 0.010976661, -0.0056875315, -0....   \n",
       "84977  [-0.0019518828, -0.024248326, 0.006703199, -0....   \n",
       "84978  [-0.010533179, -0.032219134, -0.0032646793, -0...   \n",
       "84979  [-0.0029098308, -0.0076754037, -0.0111139845, ...   \n",
       "84980  [-0.019622136, -0.0142971175, -0.02942627, -0....   \n",
       "\n",
       "                                                feat_emb  \n",
       "0      [1.9533191919326782, 0.6867725253105164, 0.456...  \n",
       "1      [0.21793906390666962, 0.13471993803977966, 0.0...  \n",
       "2      [-0.12004103511571884, -0.36696380376815796, -...  \n",
       "3      [-0.24334891140460968, -0.6172693967819214, 0....  \n",
       "4      [-0.1530350148677826, -0.33404994010925293, -0...  \n",
       "...                                                  ...  \n",
       "84976  [-0.011537055, 0.008254736, 0.01594828, -0.017...  \n",
       "84977  [-0.011491481, -0.013787123, 0.016679898, -0.0...  \n",
       "84978  [-0.0020738854, -0.010127607, 0.021126406, -0....  \n",
       "84979  [-0.005530768, 0.011203765, -0.0023338485, -0....  \n",
       "84980  [-0.029408008, -0.02011434, -0.017020838, -0.0...  \n",
       "\n",
       "[84981 rows x 3 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph['nodes']['embedding']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "08502169",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_index</th>\n",
       "      <th>primekg_node_index</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_type</th>\n",
       "      <th>desc</th>\n",
       "      <th>feat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10770</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PHYHIP_(0)</td>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PHYHIP belongs to gene/protein node. PHYHIP is...</td>\n",
       "      <td>MELLSTPHSIEINNITCDSFRISWAMEDSDLERVTHYFIDLNKKEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10771</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GPANK1_(1)</td>\n",
       "      <td>GPANK1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GPANK1 belongs to gene/protein node. GPANK1 is...</td>\n",
       "      <td>MSRPLLITFTPATDPSDLWKDGQQQPQPEKPESTLDGAAARAFYEA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10772</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>ZRSR2_(2)</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ZRSR2 belongs to gene/protein node. ZRSR2 is z...</td>\n",
       "      <td>MAAPEKMTFPEKPSHKKYRAALKKEKRKKRRQELARLRDSGLSQKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10773</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>NRF1_(3)</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NRF1 belongs to gene/protein node. NRF1 is nuc...</td>\n",
       "      <td>MEEHGVTQTEHMATIEAHAVAQQVQQVHVATYTEHSMLSADEDSPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10774</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>PI4KA_(4)</td>\n",
       "      <td>PI4KA</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PI4KA belongs to gene/protein node. PI4KA is p...</td>\n",
       "      <td>MAAAPARGGGGGGGGGGGCSGSGSSASRGFYFNTVLSLARSLAVQR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_index  primekg_node_index     node_id node_name     node_type  \\\n",
       "10770           0                   0  PHYHIP_(0)    PHYHIP  gene/protein   \n",
       "10771           1                   1  GPANK1_(1)    GPANK1  gene/protein   \n",
       "10772           2                   2   ZRSR2_(2)     ZRSR2  gene/protein   \n",
       "10773           3                   3    NRF1_(3)      NRF1  gene/protein   \n",
       "10774           4                   4   PI4KA_(4)     PI4KA  gene/protein   \n",
       "\n",
       "                                                    desc  \\\n",
       "10770  PHYHIP belongs to gene/protein node. PHYHIP is...   \n",
       "10771  GPANK1 belongs to gene/protein node. GPANK1 is...   \n",
       "10772  ZRSR2 belongs to gene/protein node. ZRSR2 is z...   \n",
       "10773  NRF1 belongs to gene/protein node. NRF1 is nuc...   \n",
       "10774  PI4KA belongs to gene/protein node. PI4KA is p...   \n",
       "\n",
       "                                                    feat  \n",
       "10770  MELLSTPHSIEINNITCDSFRISWAMEDSDLERVTHYFIDLNKKEN...  \n",
       "10771  MSRPLLITFTPATDPSDLWKDGQQQPQPEKPESTLDGAAARAFYEA...  \n",
       "10772  MAAPEKMTFPEKPSHKKYRAALKKEKRKKRRQELARLRDSGLSQKE...  \n",
       "10773  MEEHGVTQTEHMATIEAHAVAQQVQQVHVATYTEHSMLSADEDSPS...  \n",
       "10774  MAAAPARGGGGGGGGGGGCSGSGSSASRGFYFNTVLSLARSLAVQR...  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask = graph['nodes']['enrichment'].node_type == 'gene/protein'\n",
    "graph['nodes']['enrichment'][mask].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "daaa35f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>desc_emb</th>\n",
       "      <th>feat_emb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10770</th>\n",
       "      <td>PHYHIP_(0)</td>\n",
       "      <td>[-0.038923346, -0.022871112, -0.012125405, -0....</td>\n",
       "      <td>[0.04029838368296623, -0.018344514071941376, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10771</th>\n",
       "      <td>GPANK1_(1)</td>\n",
       "      <td>[-0.025375651, 0.012858219, 0.008264126, -0.00...</td>\n",
       "      <td>[-0.049913737922906876, -0.04380067065358162, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10772</th>\n",
       "      <td>ZRSR2_(2)</td>\n",
       "      <td>[-0.032085866, 0.0071205534, -0.017097335, -0....</td>\n",
       "      <td>[0.035360466688871384, -0.09613325446844101, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10773</th>\n",
       "      <td>NRF1_(3)</td>\n",
       "      <td>[-0.030888347, -0.024794728, -0.020263912, -0....</td>\n",
       "      <td>[-0.052261918783187866, -0.022747397422790527,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10774</th>\n",
       "      <td>PI4KA_(4)</td>\n",
       "      <td>[-0.029845022, -0.023542346, -0.01622012, -0.0...</td>\n",
       "      <td>[0.005174526944756508, -0.049968406558036804, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          node_id                                           desc_emb  \\\n",
       "10770  PHYHIP_(0)  [-0.038923346, -0.022871112, -0.012125405, -0....   \n",
       "10771  GPANK1_(1)  [-0.025375651, 0.012858219, 0.008264126, -0.00...   \n",
       "10772   ZRSR2_(2)  [-0.032085866, 0.0071205534, -0.017097335, -0....   \n",
       "10773    NRF1_(3)  [-0.030888347, -0.024794728, -0.020263912, -0....   \n",
       "10774   PI4KA_(4)  [-0.029845022, -0.023542346, -0.01622012, -0.0...   \n",
       "\n",
       "                                                feat_emb  \n",
       "10770  [0.04029838368296623, -0.018344514071941376, 0...  \n",
       "10771  [-0.049913737922906876, -0.04380067065358162, ...  \n",
       "10772  [0.035360466688871384, -0.09613325446844101, 0...  \n",
       "10773  [-0.052261918783187866, -0.022747397422790527,...  \n",
       "10774  [0.005174526944756508, -0.049968406558036804, ...  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph['nodes']['embedding'][mask].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6d8ac14c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        0.999999\n",
       "1        0.804206\n",
       "2        0.762189\n",
       "3        0.776730\n",
       "4        0.762661\n",
       "           ...   \n",
       "84976    0.777796\n",
       "84977    0.809125\n",
       "84978    0.777640\n",
       "84979    0.797852\n",
       "84980    0.779340\n",
       "Length: 84981, dtype: float32"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "93418eff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-0.00843711, -0.00667603,  0.01728939, ..., -0.02651415,\n",
       "       -0.02254812, -0.04615015])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph['nodes']['embedding']['desc_emb'].list.leaves.to_cupy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e3f4a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(384000000,)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph['edges']['embedding'][0]['edge_emb'].list.leaves.to_cupy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b689a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_nodes['embedding']['desc_emb'].list.leaves.to_cupy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c4d00d26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing edges chunk 1/16\n",
      "Processing edges chunk 2/16\n",
      "Processing edges chunk 3/16\n",
      "Processing edges chunk 4/16\n",
      "Processing edges chunk 5/16\n",
      "Processing edges chunk 6/16\n",
      "Processing edges chunk 7/16\n",
      "Processing edges chunk 8/16\n",
      "Processing edges chunk 9/16\n",
      "Processing edges chunk 10/16\n",
      "Processing edges chunk 11/16\n",
      "Processing edges chunk 12/16\n",
      "Processing edges chunk 13/16\n",
      "Processing edges chunk 14/16\n",
      "Processing edges chunk 15/16\n",
      "Processing edges chunk 16/16\n"
     ]
    }
   ],
   "source": [
    "sim_dict = {}\n",
    "sim_dict['index'] = []\n",
    "sim_dict['score'] = []\n",
    "for i, cur_edges_embedding_df in enumerate(edges_embedding_df):\n",
    "    print(f\"Processing edges chunk {i+1}/{len(edges_embedding_df)}\")\n",
    "    # Convert the edge embedding to a cupy array\n",
    "    cur_triplet_index = cur_edges_embedding_df['triplet_index'].to_cupy()\n",
    "    # cur_arr = cp.asarray(cur_edges_embedding_df[\"edge_emb\"].explode()).reshape(cur_edges_embedding_df.shape[0], -1)\n",
    "    cur_arr = cur_edges_embedding_df['edge_emb'].list.leaves.to_cupy().reshape(cur_edges_embedding_df.shape[0], -1).astype(cp.float32)\n",
    "\n",
    "    # Calculate the similarity score\n",
    "    # sim_score.append(cp.dot(cur_arr, e_prompt) / (cp.linalg.norm(cur_arr, axis=1) * cp.linalg.norm(e_prompt)))\n",
    "    sim_dict['score'].append(_compute_sim_scores(cur_arr, e_prompt.reshape(1, -1).astype(cp.float32)))\n",
    "\n",
    "    # Append the triplet indices\n",
    "    sim_dict['index'].append(cur_triplet_index)\n",
    "\n",
    "# Create a DataFrame with the results\n",
    "sim_scores_df = cudf.DataFrame({\n",
    "    'triplet_index': cp.concatenate(sim_dict['index']),\n",
    "    'sim_score': cp.concatenate(sim_dict['score'])\n",
    "})\n",
    "sim_scores_df = sim_scores_df.sort_values(\"triplet_index\").reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9a4896a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>desc_emb</th>\n",
       "      <th>feat_emb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Copper_(14012)</td>\n",
       "      <td>[-0.008437113, -0.0066760327, 0.017289385, -0....</td>\n",
       "      <td>[1.9533191919326782, 0.6867725253105164, 0.456...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Oxygen_(14013)</td>\n",
       "      <td>[0.0011131193, 0.00035166115, -0.011181605, -0...</td>\n",
       "      <td>[0.21793906390666962, 0.13471993803977966, 0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Flunisolide_(14014)</td>\n",
       "      <td>[-0.034678344, 0.0014620015, -0.00069094024, -...</td>\n",
       "      <td>[-0.12004103511571884, -0.36696380376815796, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Alclometasone_(14015)</td>\n",
       "      <td>[-0.00836069, 0.0018004619, 0.0005102673, -0.0...</td>\n",
       "      <td>[-0.24334891140460968, -0.6172693967819214, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Medrysone_(14016)</td>\n",
       "      <td>[-0.022494577, 0.00894156, -0.004813894, -0.02...</td>\n",
       "      <td>[-0.1530350148677826, -0.33404994010925293, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84976</th>\n",
       "      <td>structural constituent of tooth enamel_(124215)</td>\n",
       "      <td>[0.0027853793, 0.010976661, -0.0056875315, -0....</td>\n",
       "      <td>[-0.011537055, 0.008254736, 0.01594828, -0.017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84977</th>\n",
       "      <td>D-glucuronate transmembrane transporter activi...</td>\n",
       "      <td>[-0.0019518828, -0.024248326, 0.006703199, -0....</td>\n",
       "      <td>[-0.011491481, -0.013787123, 0.016679898, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84978</th>\n",
       "      <td>histone propionyltransferase activity_(124217)</td>\n",
       "      <td>[-0.010533179, -0.032219134, -0.0032646793, -0...</td>\n",
       "      <td>[-0.0020738854, -0.010127607, 0.021126406, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84979</th>\n",
       "      <td>sirohydrochlorin ferrochelatase activity_(124219)</td>\n",
       "      <td>[-0.0029098308, -0.0076754037, -0.0111139845, ...</td>\n",
       "      <td>[-0.005530768, 0.011203765, -0.0023338485, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84980</th>\n",
       "      <td>ethylene binding_(124220)</td>\n",
       "      <td>[-0.019622136, -0.0142971175, -0.02942627, -0....</td>\n",
       "      <td>[-0.029408008, -0.02011434, -0.017020838, -0.0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84981 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 node_id  \\\n",
       "0                                         Copper_(14012)   \n",
       "1                                         Oxygen_(14013)   \n",
       "2                                    Flunisolide_(14014)   \n",
       "3                                  Alclometasone_(14015)   \n",
       "4                                      Medrysone_(14016)   \n",
       "...                                                  ...   \n",
       "84976    structural constituent of tooth enamel_(124215)   \n",
       "84977  D-glucuronate transmembrane transporter activi...   \n",
       "84978     histone propionyltransferase activity_(124217)   \n",
       "84979  sirohydrochlorin ferrochelatase activity_(124219)   \n",
       "84980                          ethylene binding_(124220)   \n",
       "\n",
       "                                                desc_emb  \\\n",
       "0      [-0.008437113, -0.0066760327, 0.017289385, -0....   \n",
       "1      [0.0011131193, 0.00035166115, -0.011181605, -0...   \n",
       "2      [-0.034678344, 0.0014620015, -0.00069094024, -...   \n",
       "3      [-0.00836069, 0.0018004619, 0.0005102673, -0.0...   \n",
       "4      [-0.022494577, 0.00894156, -0.004813894, -0.02...   \n",
       "...                                                  ...   \n",
       "84976  [0.0027853793, 0.010976661, -0.0056875315, -0....   \n",
       "84977  [-0.0019518828, -0.024248326, 0.006703199, -0....   \n",
       "84978  [-0.010533179, -0.032219134, -0.0032646793, -0...   \n",
       "84979  [-0.0029098308, -0.0076754037, -0.0111139845, ...   \n",
       "84980  [-0.019622136, -0.0142971175, -0.02942627, -0....   \n",
       "\n",
       "                                                feat_emb  \n",
       "0      [1.9533191919326782, 0.6867725253105164, 0.456...  \n",
       "1      [0.21793906390666962, 0.13471993803977966, 0.0...  \n",
       "2      [-0.12004103511571884, -0.36696380376815796, -...  \n",
       "3      [-0.24334891140460968, -0.6172693967819214, 0....  \n",
       "4      [-0.1530350148677826, -0.33404994010925293, -0...  \n",
       "...                                                  ...  \n",
       "84976  [-0.011537055, 0.008254736, 0.01594828, -0.017...  \n",
       "84977  [-0.011491481, -0.013787123, 0.016679898, -0.0...  \n",
       "84978  [-0.0020738854, -0.010127607, 0.021126406, -0....  \n",
       "84979  [-0.005530768, 0.011203765, -0.0023338485, -0....  \n",
       "84980  [-0.029408008, -0.02011434, -0.017020838, -0.0...  \n",
       "\n",
       "[84981 rows x 3 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph[\"nodes\"][\"embedding\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "fa7f9dad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_index</th>\n",
       "      <th>primekg_node_index</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_type</th>\n",
       "      <th>desc</th>\n",
       "      <th>feat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PHYHIP_(0)</td>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PHYHIP belongs to gene/protein node. PHYHIP is...</td>\n",
       "      <td>MELLSTPHSIEINNITCDSFRISWAMEDSDLERVTHYFIDLNKKEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GPANK1_(1)</td>\n",
       "      <td>GPANK1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GPANK1 belongs to gene/protein node. GPANK1 is...</td>\n",
       "      <td>MSRPLLITFTPATDPSDLWKDGQQQPQPEKPESTLDGAAARAFYEA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>ZRSR2_(2)</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ZRSR2 belongs to gene/protein node. ZRSR2 is z...</td>\n",
       "      <td>MAAPEKMTFPEKPSHKKYRAALKKEKRKKRRQELARLRDSGLSQKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>NRF1_(3)</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NRF1 belongs to gene/protein node. NRF1 is nuc...</td>\n",
       "      <td>MEEHGVTQTEHMATIEAHAVAQQVQQVHVATYTEHSMLSADEDSPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>PI4KA_(4)</td>\n",
       "      <td>PI4KA</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PI4KA belongs to gene/protein node. PI4KA is p...</td>\n",
       "      <td>MAAAPARGGGGGGGGGGGCSGSGSSASRGFYFNTVLSLARSLAVQR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84976</th>\n",
       "      <td>84976</td>\n",
       "      <td>127430</td>\n",
       "      <td>host cell rough endoplasmic reticulum membrane...</td>\n",
       "      <td>host cell rough endoplasmic reticulum membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>host cell rough endoplasmic reticulum membrane...</td>\n",
       "      <td>The lipid bilayer surrounding the host cell ro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84977</th>\n",
       "      <td>84977</td>\n",
       "      <td>127431</td>\n",
       "      <td>collagen type VII anchoring fibril_(127431)</td>\n",
       "      <td>collagen type VII anchoring fibril</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>collagen type VII anchoring fibril belongs to ...</td>\n",
       "      <td>An antiparallel dimer of two collagen VII trim...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84978</th>\n",
       "      <td>84978</td>\n",
       "      <td>127432</td>\n",
       "      <td>cofilin-actin rod_(127432)</td>\n",
       "      <td>cofilin-actin rod</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>cofilin-actin rod belongs to cellular_componen...</td>\n",
       "      <td>A cellular structure consisting of parallel, h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84979</th>\n",
       "      <td>84979</td>\n",
       "      <td>127433</td>\n",
       "      <td>condensed chromosome, centromeric region_(127433)</td>\n",
       "      <td>condensed chromosome, centromeric region</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>condensed chromosome, centromeric region belon...</td>\n",
       "      <td>The region of a condensed chromosome that incl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84980</th>\n",
       "      <td>84980</td>\n",
       "      <td>127434</td>\n",
       "      <td>plastid acetyl-CoA carboxylase complex_(127434)</td>\n",
       "      <td>plastid acetyl-CoA carboxylase complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>plastid acetyl-CoA carboxylase complex belongs...</td>\n",
       "      <td>An acetyl-CoA carboxylase complex located in t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84981 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_index  primekg_node_index  \\\n",
       "0               0                   0   \n",
       "1               1                   1   \n",
       "2               2                   2   \n",
       "3               3                   3   \n",
       "4               4                   4   \n",
       "...           ...                 ...   \n",
       "84976       84976              127430   \n",
       "84977       84977              127431   \n",
       "84978       84978              127432   \n",
       "84979       84979              127433   \n",
       "84980       84980              127434   \n",
       "\n",
       "                                                 node_id  \\\n",
       "0                                             PHYHIP_(0)   \n",
       "1                                             GPANK1_(1)   \n",
       "2                                              ZRSR2_(2)   \n",
       "3                                               NRF1_(3)   \n",
       "4                                              PI4KA_(4)   \n",
       "...                                                  ...   \n",
       "84976  host cell rough endoplasmic reticulum membrane...   \n",
       "84977        collagen type VII anchoring fibril_(127431)   \n",
       "84978                         cofilin-actin rod_(127432)   \n",
       "84979  condensed chromosome, centromeric region_(127433)   \n",
       "84980    plastid acetyl-CoA carboxylase complex_(127434)   \n",
       "\n",
       "                                            node_name           node_type  \\\n",
       "0                                              PHYHIP        gene/protein   \n",
       "1                                              GPANK1        gene/protein   \n",
       "2                                               ZRSR2        gene/protein   \n",
       "3                                                NRF1        gene/protein   \n",
       "4                                               PI4KA        gene/protein   \n",
       "...                                               ...                 ...   \n",
       "84976  host cell rough endoplasmic reticulum membrane  cellular_component   \n",
       "84977              collagen type VII anchoring fibril  cellular_component   \n",
       "84978                               cofilin-actin rod  cellular_component   \n",
       "84979        condensed chromosome, centromeric region  cellular_component   \n",
       "84980          plastid acetyl-CoA carboxylase complex  cellular_component   \n",
       "\n",
       "                                                    desc  \\\n",
       "0      PHYHIP belongs to gene/protein node. PHYHIP is...   \n",
       "1      GPANK1 belongs to gene/protein node. GPANK1 is...   \n",
       "2      ZRSR2 belongs to gene/protein node. ZRSR2 is z...   \n",
       "3      NRF1 belongs to gene/protein node. NRF1 is nuc...   \n",
       "4      PI4KA belongs to gene/protein node. PI4KA is p...   \n",
       "...                                                  ...   \n",
       "84976  host cell rough endoplasmic reticulum membrane...   \n",
       "84977  collagen type VII anchoring fibril belongs to ...   \n",
       "84978  cofilin-actin rod belongs to cellular_componen...   \n",
       "84979  condensed chromosome, centromeric region belon...   \n",
       "84980  plastid acetyl-CoA carboxylase complex belongs...   \n",
       "\n",
       "                                                    feat  \n",
       "0      MELLSTPHSIEINNITCDSFRISWAMEDSDLERVTHYFIDLNKKEN...  \n",
       "1      MSRPLLITFTPATDPSDLWKDGQQQPQPEKPESTLDGAAARAFYEA...  \n",
       "2      MAAPEKMTFPEKPSHKKYRAALKKEKRKKRRQELARLRDSGLSQKE...  \n",
       "3      MEEHGVTQTEHMATIEAHAVAQQVQQVHVATYTEHSMLSADEDSPS...  \n",
       "4      MAAAPARGGGGGGGGGGGCSGSGSSASRGFYFNTVLSLARSLAVQR...  \n",
       "...                                                  ...  \n",
       "84976  The lipid bilayer surrounding the host cell ro...  \n",
       "84977  An antiparallel dimer of two collagen VII trim...  \n",
       "84978  A cellular structure consisting of parallel, h...  \n",
       "84979  The region of a condensed chromosome that incl...  \n",
       "84980  An acetyl-CoA carboxylase complex located in t...  \n",
       "\n",
       "[84981 rows x 7 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph[\"nodes\"][\"enrichment\"].sort_values(\"node_index\", ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5e12b89e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>triplet_index</th>\n",
       "      <th>primekg_head_index</th>\n",
       "      <th>primekg_tail_index</th>\n",
       "      <th>head_id</th>\n",
       "      <th>tail_id</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>edge_type_str</th>\n",
       "      <th>head_index</th>\n",
       "      <th>tail_index</th>\n",
       "      <th>feat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8889</td>\n",
       "      <td>PHYHIP_(0)</td>\n",
       "      <td>KIF15_(8889)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>0</td>\n",
       "      <td>8816</td>\n",
       "      <td>PHYHIP (gene/protein) has a direct relationshi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2798</td>\n",
       "      <td>GPANK1_(1)</td>\n",
       "      <td>PNMA1_(2798)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>1</td>\n",
       "      <td>2787</td>\n",
       "      <td>GPANK1 (gene/protein) has a direct relationshi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>5646</td>\n",
       "      <td>ZRSR2_(2)</td>\n",
       "      <td>TTC33_(5646)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>2</td>\n",
       "      <td>5610</td>\n",
       "      <td>ZRSR2 (gene/protein) has a direct relationship...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>11592</td>\n",
       "      <td>NRF1_(3)</td>\n",
       "      <td>MAN1B1_(11592)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>3</td>\n",
       "      <td>11467</td>\n",
       "      <td>NRF1 (gene/protein) has a direct relationship ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2122</td>\n",
       "      <td>PI4KA_(4)</td>\n",
       "      <td>RGS20_(2122)</td>\n",
       "      <td>ppi</td>\n",
       "      <td>[gene/protein, ppi, gene/protein]</td>\n",
       "      <td>gene/protein|ppi|gene/protein</td>\n",
       "      <td>4</td>\n",
       "      <td>2117</td>\n",
       "      <td>PI4KA (gene/protein) has a direct relationship...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904605</th>\n",
       "      <td>3904605</td>\n",
       "      <td>52855</td>\n",
       "      <td>34572</td>\n",
       "      <td>B cell receptor transport into membrane raft_(...</td>\n",
       "      <td>CD24_(34572)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>45323</td>\n",
       "      <td>27800</td>\n",
       "      <td>B cell receptor transport into membrane raft (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904606</th>\n",
       "      <td>3904606</td>\n",
       "      <td>113352</td>\n",
       "      <td>34572</td>\n",
       "      <td>chemokine receptor transport out of membrane r...</td>\n",
       "      <td>CD24_(34572)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>71241</td>\n",
       "      <td>27800</td>\n",
       "      <td>chemokine receptor transport out of membrane r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904607</th>\n",
       "      <td>3904607</td>\n",
       "      <td>42264</td>\n",
       "      <td>57675</td>\n",
       "      <td>negative regulation of cytoskeleton organizati...</td>\n",
       "      <td>IQCJ-SCHIP1_(57675)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>35166</td>\n",
       "      <td>49927</td>\n",
       "      <td>negative regulation of cytoskeleton organizati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904608</th>\n",
       "      <td>3904608</td>\n",
       "      <td>109904</td>\n",
       "      <td>58770</td>\n",
       "      <td>mesendoderm migration_(109904)</td>\n",
       "      <td>APELA_(58770)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>67928</td>\n",
       "      <td>50777</td>\n",
       "      <td>mesendoderm migration (biological_process) has...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904609</th>\n",
       "      <td>3904609</td>\n",
       "      <td>44810</td>\n",
       "      <td>57692</td>\n",
       "      <td>regulation of endoplasmic reticulum unfolded p...</td>\n",
       "      <td>PIGBOS1_(57692)</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "      <td>biological_process|interacts with|gene/protein</td>\n",
       "      <td>37608</td>\n",
       "      <td>49939</td>\n",
       "      <td>regulation of endoplasmic reticulum unfolded p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3904610 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         triplet_index  primekg_head_index  primekg_tail_index  \\\n",
       "0                    0                   0                8889   \n",
       "1                    1                   1                2798   \n",
       "2                    2                   2                5646   \n",
       "3                    3                   3               11592   \n",
       "4                    4                   4                2122   \n",
       "...                ...                 ...                 ...   \n",
       "3904605        3904605               52855               34572   \n",
       "3904606        3904606              113352               34572   \n",
       "3904607        3904607               42264               57675   \n",
       "3904608        3904608              109904               58770   \n",
       "3904609        3904609               44810               57692   \n",
       "\n",
       "                                                   head_id  \\\n",
       "0                                               PHYHIP_(0)   \n",
       "1                                               GPANK1_(1)   \n",
       "2                                                ZRSR2_(2)   \n",
       "3                                                 NRF1_(3)   \n",
       "4                                                PI4KA_(4)   \n",
       "...                                                    ...   \n",
       "3904605  B cell receptor transport into membrane raft_(...   \n",
       "3904606  chemokine receptor transport out of membrane r...   \n",
       "3904607  negative regulation of cytoskeleton organizati...   \n",
       "3904608                     mesendoderm migration_(109904)   \n",
       "3904609  regulation of endoplasmic reticulum unfolded p...   \n",
       "\n",
       "                     tail_id display_relation  \\\n",
       "0               KIF15_(8889)              ppi   \n",
       "1               PNMA1_(2798)              ppi   \n",
       "2               TTC33_(5646)              ppi   \n",
       "3             MAN1B1_(11592)              ppi   \n",
       "4               RGS20_(2122)              ppi   \n",
       "...                      ...              ...   \n",
       "3904605         CD24_(34572)   interacts with   \n",
       "3904606         CD24_(34572)   interacts with   \n",
       "3904607  IQCJ-SCHIP1_(57675)   interacts with   \n",
       "3904608        APELA_(58770)   interacts with   \n",
       "3904609      PIGBOS1_(57692)   interacts with   \n",
       "\n",
       "                                                 edge_type  \\\n",
       "0                        [gene/protein, ppi, gene/protein]   \n",
       "1                        [gene/protein, ppi, gene/protein]   \n",
       "2                        [gene/protein, ppi, gene/protein]   \n",
       "3                        [gene/protein, ppi, gene/protein]   \n",
       "4                        [gene/protein, ppi, gene/protein]   \n",
       "...                                                    ...   \n",
       "3904605  [biological_process, interacts with, gene/prot...   \n",
       "3904606  [biological_process, interacts with, gene/prot...   \n",
       "3904607  [biological_process, interacts with, gene/prot...   \n",
       "3904608  [biological_process, interacts with, gene/prot...   \n",
       "3904609  [biological_process, interacts with, gene/prot...   \n",
       "\n",
       "                                          edge_type_str  head_index  \\\n",
       "0                         gene/protein|ppi|gene/protein           0   \n",
       "1                         gene/protein|ppi|gene/protein           1   \n",
       "2                         gene/protein|ppi|gene/protein           2   \n",
       "3                         gene/protein|ppi|gene/protein           3   \n",
       "4                         gene/protein|ppi|gene/protein           4   \n",
       "...                                                 ...         ...   \n",
       "3904605  biological_process|interacts with|gene/protein       45323   \n",
       "3904606  biological_process|interacts with|gene/protein       71241   \n",
       "3904607  biological_process|interacts with|gene/protein       35166   \n",
       "3904608  biological_process|interacts with|gene/protein       67928   \n",
       "3904609  biological_process|interacts with|gene/protein       37608   \n",
       "\n",
       "         tail_index                                               feat  \n",
       "0              8816  PHYHIP (gene/protein) has a direct relationshi...  \n",
       "1              2787  GPANK1 (gene/protein) has a direct relationshi...  \n",
       "2              5610  ZRSR2 (gene/protein) has a direct relationship...  \n",
       "3             11467  NRF1 (gene/protein) has a direct relationship ...  \n",
       "4              2117  PI4KA (gene/protein) has a direct relationship...  \n",
       "...             ...                                                ...  \n",
       "3904605       27800  B cell receptor transport into membrane raft (...  \n",
       "3904606       27800  chemokine receptor transport out of membrane r...  \n",
       "3904607       49927  negative regulation of cytoskeleton organizati...  \n",
       "3904608       50777  mesendoderm migration (biological_process) has...  \n",
       "3904609       49939  regulation of endoplasmic reticulum unfolded p...  \n",
       "\n",
       "[3904610 rows x 11 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph[\"edges\"][\"enrichment\"].sort_values(\"triplet_index\", ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "e43b9c92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.9999999, 0.9467669, 0.9474174, ..., 0.7786807, 0.7878412,\n",
       "       0.7995058], dtype=float32)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_dict['score'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8d325379",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          0.842766\n",
       "1          0.845283\n",
       "2          0.853746\n",
       "3          0.852537\n",
       "4          0.841947\n",
       "             ...   \n",
       "3904605    0.866746\n",
       "3904606    0.856712\n",
       "3904607    0.854869\n",
       "3904608    0.838710\n",
       "3904609    0.869283\n",
       "Name: sim_score, Length: 3904610, dtype: float32"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_scores_df['sim_score']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "293c1377",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          0.842767\n",
       "1          0.845284\n",
       "2          0.853746\n",
       "3          0.852537\n",
       "4          0.841947\n",
       "             ...   \n",
       "3904605    0.866746\n",
       "3904606    0.856713\n",
       "3904607    0.854869\n",
       "3904608    0.838710\n",
       "3904609    0.869283\n",
       "Name: sim_score, Length: 3904610, dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_scores_df['sim_score']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "f66bb7c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          0.0\n",
       "1          0.0\n",
       "2          0.0\n",
       "3          0.0\n",
       "4          0.0\n",
       "          ... \n",
       "3904605    0.0\n",
       "3904606    0.0\n",
       "3904607    0.0\n",
       "3904608    0.0\n",
       "3904609    0.0\n",
       "Length: 3904610, dtype: float32"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cudf.Series(cp.zeros(3904610, dtype=cp.float32))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "84b63bcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>triplet_index</th>\n",
       "      <th>sim_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.842767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.845284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.853746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.852537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.841947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904605</th>\n",
       "      <td>3904605</td>\n",
       "      <td>0.866746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904606</th>\n",
       "      <td>3904606</td>\n",
       "      <td>0.856713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904607</th>\n",
       "      <td>3904607</td>\n",
       "      <td>0.854869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904608</th>\n",
       "      <td>3904608</td>\n",
       "      <td>0.838710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3904609</th>\n",
       "      <td>3904609</td>\n",
       "      <td>0.869283</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3904610 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         triplet_index  sim_score\n",
       "0                    0   0.842767\n",
       "1                    1   0.845284\n",
       "2                    2   0.853746\n",
       "3                    3   0.852537\n",
       "4                    4   0.841947\n",
       "...                ...        ...\n",
       "3904605        3904605   0.866746\n",
       "3904606        3904606   0.856713\n",
       "3904607        3904607   0.854869\n",
       "3904608        3904608   0.838710\n",
       "3904609        3904609   0.869283\n",
       "\n",
       "[3904610 rows x 2 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_scores_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "be989855",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([3050000, 3050001, 3050002, ..., 3099997, 3099998, 3099999]),\n",
       " array([2250000, 2250001, 2250002, ..., 2299997, 2299998, 2299999]),\n",
       " array([1150000, 1150001, 1150002, ..., 1199997, 1199998, 1199999]),\n",
       " array([600000, 600001, 600002, ..., 649997, 649998, 649999]),\n",
       " array([1550000, 1550001, 1550002, ..., 1599997, 1599998, 1599999]),\n",
       " array([1750000, 1750001, 1750002, ..., 1799997, 1799998, 1799999]),\n",
       " array([3100000, 3100001, 3100002, ..., 3149997, 3149998, 3149999]),\n",
       " array([1600000, 1600001, 1600002, ..., 1649997, 1649998, 1649999]),\n",
       " array([2750000, 2750001, 2750002, ..., 2799997, 2799998, 2799999]),\n",
       " array([2950000, 2950001, 2950002, ..., 2999997, 2999998, 2999999]),\n",
       " array([3700000, 3700001, 3700002, ..., 3749997, 3749998, 3749999]),\n",
       " array([500000, 500001, 500002, ..., 549997, 549998, 549999]),\n",
       " array([800000, 800001, 800002, ..., 849997, 849998, 849999]),\n",
       " array([1050000, 1050001, 1050002, ..., 1099997, 1099998, 1099999]),\n",
       " array([300000, 300001, 300002, ..., 349997, 349998, 349999]),\n",
       " array([250000, 250001, 250002, ..., 299997, 299998, 299999]),\n",
       " array([3600000, 3600001, 3600002, ..., 3649997, 3649998, 3649999]),\n",
       " array([900000, 900001, 900002, ..., 949997, 949998, 949999]),\n",
       " array([2400000, 2400001, 2400002, ..., 2449997, 2449998, 2449999]),\n",
       " array([1100000, 1100001, 1100002, ..., 1149997, 1149998, 1149999]),\n",
       " array([3450000, 3450001, 3450002, ..., 3499997, 3499998, 3499999]),\n",
       " array([400000, 400001, 400002, ..., 449997, 449998, 449999]),\n",
       " array([1500000, 1500001, 1500002, ..., 1549997, 1549998, 1549999]),\n",
       " array([    0,     1,     2, ..., 49997, 49998, 49999]),\n",
       " array([1250000, 1250001, 1250002, ..., 1299997, 1299998, 1299999]),\n",
       " array([3900000, 3900001, 3900002, ..., 3904607, 3904608, 3904609]),\n",
       " array([1400000, 1400001, 1400002, ..., 1449997, 1449998, 1449999]),\n",
       " array([50000, 50001, 50002, ..., 99997, 99998, 99999]),\n",
       " array([1450000, 1450001, 1450002, ..., 1499997, 1499998, 1499999]),\n",
       " array([1800000, 1800001, 1800002, ..., 1849997, 1849998, 1849999]),\n",
       " array([1000000, 1000001, 1000002, ..., 1049997, 1049998, 1049999]),\n",
       " array([1950000, 1950001, 1950002, ..., 1999997, 1999998, 1999999]),\n",
       " array([2100000, 2100001, 2100002, ..., 2149997, 2149998, 2149999]),\n",
       " array([2350000, 2350001, 2350002, ..., 2399997, 2399998, 2399999]),\n",
       " array([3500000, 3500001, 3500002, ..., 3549997, 3549998, 3549999]),\n",
       " array([3750000, 3750001, 3750002, ..., 3799997, 3799998, 3799999]),\n",
       " array([2800000, 2800001, 2800002, ..., 2849997, 2849998, 2849999]),\n",
       " array([1350000, 1350001, 1350002, ..., 1399997, 1399998, 1399999]),\n",
       " array([150000, 150001, 150002, ..., 199997, 199998, 199999]),\n",
       " array([2500000, 2500001, 2500002, ..., 2549997, 2549998, 2549999]),\n",
       " array([450000, 450001, 450002, ..., 499997, 499998, 499999]),\n",
       " array([850000, 850001, 850002, ..., 899997, 899998, 899999]),\n",
       " array([1850000, 1850001, 1850002, ..., 1899997, 1899998, 1899999]),\n",
       " array([3000000, 3000001, 3000002, ..., 3049997, 3049998, 3049999]),\n",
       " array([550000, 550001, 550002, ..., 599997, 599998, 599999]),\n",
       " array([200000, 200001, 200002, ..., 249997, 249998, 249999]),\n",
       " array([3300000, 3300001, 3300002, ..., 3349997, 3349998, 3349999]),\n",
       " array([2650000, 2650001, 2650002, ..., 2699997, 2699998, 2699999]),\n",
       " array([3350000, 3350001, 3350002, ..., 3399997, 3399998, 3399999]),\n",
       " array([100000, 100001, 100002, ..., 149997, 149998, 149999]),\n",
       " array([3400000, 3400001, 3400002, ..., 3449997, 3449998, 3449999]),\n",
       " array([3550000, 3550001, 3550002, ..., 3599997, 3599998, 3599999]),\n",
       " array([350000, 350001, 350002, ..., 399997, 399998, 399999]),\n",
       " array([3800000, 3800001, 3800002, ..., 3849997, 3849998, 3849999]),\n",
       " array([3200000, 3200001, 3200002, ..., 3249997, 3249998, 3249999]),\n",
       " array([2850000, 2850001, 2850002, ..., 2899997, 2899998, 2899999]),\n",
       " array([2200000, 2200001, 2200002, ..., 2249997, 2249998, 2249999]),\n",
       " array([3850000, 3850001, 3850002, ..., 3899997, 3899998, 3899999]),\n",
       " array([3250000, 3250001, 3250002, ..., 3299997, 3299998, 3299999]),\n",
       " array([2900000, 2900001, 2900002, ..., 2949997, 2949998, 2949999]),\n",
       " array([750000, 750001, 750002, ..., 799997, 799998, 799999]),\n",
       " array([1300000, 1300001, 1300002, ..., 1349997, 1349998, 1349999]),\n",
       " array([2300000, 2300001, 2300002, ..., 2349997, 2349998, 2349999]),\n",
       " array([2700000, 2700001, 2700002, ..., 2749997, 2749998, 2749999]),\n",
       " array([3150000, 3150001, 3150002, ..., 3199997, 3199998, 3199999]),\n",
       " array([1900000, 1900001, 1900002, ..., 1949997, 1949998, 1949999]),\n",
       " array([2150000, 2150001, 2150002, ..., 2199997, 2199998, 2199999]),\n",
       " array([950000, 950001, 950002, ..., 999997, 999998, 999999]),\n",
       " array([2450000, 2450001, 2450002, ..., 2499997, 2499998, 2499999]),\n",
       " array([1700000, 1700001, 1700002, ..., 1749997, 1749998, 1749999]),\n",
       " array([650000, 650001, 650002, ..., 699997, 699998, 699999]),\n",
       " array([3650000, 3650001, 3650002, ..., 3699997, 3699998, 3699999]),\n",
       " array([2550000, 2550001, 2550002, ..., 2599997, 2599998, 2599999]),\n",
       " array([700000, 700001, 700002, ..., 749997, 749998, 749999]),\n",
       " array([2000000, 2000001, 2000002, ..., 2049997, 2049998, 2049999]),\n",
       " array([2600000, 2600001, 2600002, ..., 2649997, 2649998, 2649999]),\n",
       " array([2050000, 2050001, 2050002, ..., 2099997, 2099998, 2099999]),\n",
       " array([1200000, 1200001, 1200002, ..., 1249997, 1249998, 1249999]),\n",
       " array([1650000, 1650001, 1650002, ..., 1699997, 1699998, 1699999])]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "triplet_indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "06f01452",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.        , 0.9927882 , 0.99180061])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = cp.concatenate(sim_score)\n",
    "a[cp.argsort(-a)[:3]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "5602c025",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.02087522, -0.00298006, -0.00025966, ..., -0.00838831,\n",
       "        -0.0213233 , -0.02149462],\n",
       "       [-0.01760313,  0.01406922,  0.01986164, ..., -0.01814783,\n",
       "        -0.02109718, -0.02145589],\n",
       "       [-0.01905843,  0.01330369,  0.01994888, ..., -0.01749016,\n",
       "        -0.02174308, -0.0217165 ],\n",
       "       ...,\n",
       "       [-0.03222188, -0.00610927,  0.00940235, ...,  0.00802325,\n",
       "        -0.010137  , -0.03085568],\n",
       "       [-0.02284207, -0.00256165,  0.01529403, ...,  0.00218062,\n",
       "        -0.00395876, -0.01533361],\n",
       "       [-0.01814055, -0.01461722,  0.010439  , ..., -0.00747888,\n",
       "        -0.00141784, -0.02034099]])"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur_arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6945593a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.02087522, -0.00298006, -0.00025966, ..., -0.00838831,\n",
       "        -0.0213233 , -0.02149462],\n",
       "       [-0.01760313,  0.01406922,  0.01986164, ..., -0.01814783,\n",
       "        -0.02109718, -0.02145589],\n",
       "       [-0.01905843,  0.01330369,  0.01994888, ..., -0.01749016,\n",
       "        -0.02174308, -0.0217165 ],\n",
       "       ...,\n",
       "       [-0.03222188, -0.00610927,  0.00940235, ...,  0.00802325,\n",
       "        -0.010137  , -0.03085568],\n",
       "       [-0.02284207, -0.00256165,  0.01529403, ...,  0.00218062,\n",
       "        -0.00395876, -0.01533361],\n",
       "       [-0.01814055, -0.01461722,  0.010439  , ..., -0.00747888,\n",
       "        -0.00141784, -0.02034099]])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cp.array(edges_embedding_df[0][\"edge_emb\"].to_arrow().to_pylist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed21b4d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import TensorAttr\n",
    "import torch\n",
    "from cugraph_pyg.data import GraphStore, TensorDictFeatureStore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0e128e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize FeatureStore and mapper\n",
    "graph_store = GraphStore()\n",
    "feature_store = TensorDictFeatureStore()\n",
    "mapper = {}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6746286",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop over group enrichment nodes by type\n",
    "for nt, nodes_df in nodes_enrichment_df.groupby('node_type'):\n",
    "    # print(f\"Node type: {nt}\")\n",
    "    # node_count = len(nodes_df)\n",
    "    # print(f\"Number of nodes: {node_count}\")\n",
    "\n",
    "    # Get node_ids\n",
    "    emb_df = nodes_embedding_df[nodes_embedding_df['node_id'].isin(nodes_df['node_id'])]\n",
    "\n",
    "    # Temporarily fix mismatched embedding\n",
    "    # Use desc_emb instead\n",
    "    emb_df['feat_emb'] = emb_df['desc_emb']\n",
    "\n",
    "    # Sort both by node_id for alignment\n",
    "    nodes_df_sorted = nodes_df.sort_values('node_id')\n",
    "    emb_df_sorted = emb_df.sort_values('node_id')\n",
    "\n",
    "    # Ensure sorted node_ids match\n",
    "    assert cudf.Series.equals(nodes_df_sorted['node_id'].reset_index(drop=True),\n",
    "                              emb_df_sorted['node_id'].reset_index(drop=True)), \\\n",
    "                                f\"Node ID mismatch in {nt} after sorting\"\n",
    "\n",
    "    # Get node_index as torch tensor directly\n",
    "    node_index_tensor = torch.tensor(nodes_df_sorted[\"node_index\"].to_numpy(),\n",
    "                                     dtype=torch.int64)\n",
    "    feature_store[TensorAttr(group_name=nt, attr_name=\"node_index\")] = node_index_tensor\n",
    "\n",
    "    # # Construct mapper for node_index\n",
    "    node_index_list = nodes_df_sorted[\"node_index\"].to_numpy().tolist()\n",
    "    mapper[nt] = {\n",
    "        'to_node_index': dict(enumerate(node_index_list)),\n",
    "        'from_node_index': {v: i for i, v in enumerate(node_index_list)}\n",
    "    }\n",
    "\n",
    "    # Convert embeddings as tensors and add to FeatureStore\n",
    "    for attr_name in [\"desc_emb\", \"feat_emb\"]:\n",
    "        emb_tensor = torch.tensor(emb_df_sorted[attr_name].to_arrow().to_pylist(),\n",
    "                                  dtype=torch.float32)\n",
    "        feature_store[TensorAttr(group_name=nt, attr_name=attr_name)] = emb_tensor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop over edge types\n",
    "for edge_type_str in edges_enrichment_df['edge_type_str'].unique().to_arrow().to_pylist():\n",
    "    if edge_type_str == 'gene/protein|ppi|gene/protein':\n",
    "        # print(f\"Processing edge type: {edge_type_str}\")\n",
    "        src_type, rel_type, tgt_type = edge_type_str.split('|')\n",
    "\n",
    "        # Filter edges for this edge_type_str once\n",
    "        filtered_df = edges_enrichment_df[\n",
    "            edges_enrichment_df['edge_type_str'] == edge_type_str\n",
    "        ][['triplet_index', 'head_index', 'tail_index']]\n",
    "\n",
    "        # Convert mapper dicts to cudf Series for vectorized mapping\n",
    "        src_map = cudf.Series(mapper[src_type]['from_node_index'])\n",
    "        tgt_map = cudf.Series(mapper[tgt_type]['from_node_index'])\n",
    "\n",
    "        # Vectorized mapping of head_index and tail_index using replace (works like dict lookup)\n",
    "        mapped_head = filtered_df['head_index'].replace(src_map).astype('int64')\n",
    "        mapped_tail = filtered_df['tail_index'].replace(tgt_map).astype('int64')\n",
    "\n",
    "        # Check if mapping was successful\n",
    "        if mapped_head.isnull().any() or mapped_tail.isnull().any():\n",
    "            raise ValueError(f\"Mapping failure for edge type {edge_type_str}\")\n",
    "\n",
    "        # Edge index\n",
    "        edge_index = torch.tensor(\n",
    "            cudf.concat([mapped_head, mapped_tail], axis=1).to_pandas().values.T,\n",
    "            dtype=torch.long\n",
    "        ).contiguous()\n",
    "\n",
    "        # Store edge index in the GraphStore\n",
    "        graph_store[(src_type, rel_type, tgt_type), \"coo\"] = edge_index\n",
    "\n",
    "        # Add triplet index to the FeatureStore\n",
    "        triplet_index_tensor = torch.tensor(filtered_df['triplet_index'].to_numpy(),\n",
    "                                            dtype=torch.long).unsqueeze(0)\n",
    "        feature_store[TensorAttr(group_name=(src_type, rel_type, tgt_type),\n",
    "                                 attr_name='triplet_index')] = triplet_index_tensor\n",
    "\n",
    "        # Store edge embeddings in the FeatureStore\n",
    "        # edge_emb_tensor = []\n",
    "        # for i, edge_df in edges_embedding_df:\n",
    "        #     edge_emb_df = edges_embedding_df[edges_embedding_df['edge_type_str'] == edge_type_str]\n",
    "\n",
    "        #     # Convert edge embeddings to torch tensor\n",
    "        #     edge_emb_tensor.append(torch.tensor(edge_emb_df['edge_emb'].to_arrow().to_pylist(),\n",
    "        #                                         dtype=torch.float32).unsqueeze(0))\n",
    "        # feature_store[TensorAttr(group_name=(src_type, rel_type, tgt_type),\n",
    "        #                         attr_name='edge_emb')] = edge_emb_tensor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e0a1845",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa4e64f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f11017a0-3a90-45d0-9d3f-f96069cbd31f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge nodes embedding into nodes enrichment\n",
    "graph_dict[\"nodes\"] = graph_dict[\"nodes\"][\"enrichment\"].merge(\n",
    "    graph_dict[\"nodes\"][\"embedding\"],\n",
    "    how=\"left\",\n",
    "    on=\"node_id\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5503babf-7cf2-422d-8fbc-1faceb23f702",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check head\n",
    "graph_dict[\"nodes\"].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3134c261",
   "metadata": {},
   "outputs": [],
   "source": [
    "# graph_dict[\"edges\"][\"enrichment\"] = graph_dict[\"edges\"][\"enrichment\"].merge(\n",
    "#     cudf.DataFrame({\n",
    "#         \"triplet_index\": graph_dict[\"edges\"][\"enrichment\"].triplet_index,\n",
    "#         \"feat_emb\": None\n",
    "#     }),\n",
    "#     how=\"left\",\n",
    "#     on=\"triplet_index\"\n",
    "# )\n",
    "graph_dict[\"edges\"][\"enrichment\"].drop(columns=[\"feat_emb_x\", \"feat_emb_y\", \"feat_emb\"], inplace=True)\n",
    "graph_dict[\"edges\"][\"enrichment\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e6d89e",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_dict[\"edges\"][\"embedding\"][0].loc[0, \"edge_emb\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "672b6fb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "[[]] * graph_dict[\"edges\"][\"enrichment\"].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9a6058c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a merged dataframe with additional column\n",
    "graph_dict[\"edges\"][\"enrichment\"] = graph_dict[\"edges\"][\"enrichment\"].merge(\n",
    "    cudf.DataFrame({\n",
    "        \"triplet_index\": graph_dict[\"edges\"][\"enrichment\"].triplet_index,\n",
    "        \"feat_emb\": [[]] * graph_dict[\"edges\"][\"enrichment\"].shape[0]\n",
    "    }),\n",
    "    how=\"left\",\n",
    "    on=\"triplet_index\"\n",
    ")\n",
    "graph_dict[\"edges\"][\"enrichment\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbee8170",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_dict[\"edges\"][\"enrichment\"]['feat_emb'].list.len() == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9466d4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_dict[\"edges\"][\"enrichment\"]['feat_emb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c5025ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_dict[\"edges\"][\"enrichment\"].drop(columns=[\"feat_emb_x\", \"feat_emb_y\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba4805a6-b755-4b03-b00a-ba6a66ff2e8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a merged dataframe with additional column\n",
    "graph_dict[\"edges\"][\"enrichment\"] = graph_dict[\"edges\"][\"enrichment\"].merge(\n",
    "    cudf.DataFrame({\n",
    "        \"triplet_index\": graph_dict[\"edges\"][\"enrichment\"].triplet_index,\n",
    "        \"feat_emb\": [[]] * graph_dict[\"edges\"][\"enrichment\"].shape[0]\n",
    "    }),\n",
    "    how=\"left\",\n",
    "    on=\"triplet_index\"\n",
    ")\n",
    "# graph_dict[\"edges\"][\"enrichment\"]\n",
    "\n",
    "for i, emb_df in enumerate(graph_dict[\"edges\"][\"embedding\"]):\n",
    "    # Merge the embeddings into the feature column dataframe\n",
    "    enrichment_df = graph_dict[\"edges\"][\"enrichment\"].merge(\n",
    "        emb_df,\n",
    "        on=\"triplet_index\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "\n",
    "    # Create mask for rows where feat_emb is empty list and triplet_index exists in current emb_df\n",
    "    mask = (\n",
    "        enrichment_df['feat_emb'].list.len() == 0\n",
    "    ) & (\n",
    "        enrichment_df['triplet_index'].isin(emb_df['triplet_index'])\n",
    "    )\n",
    "\n",
    "    # Assign edge_emb to feat_emb only where mask is True\n",
    "    enrichment_df.loc[mask, 'feat_emb'] = enrichment_df.loc[mask, 'edge_emb']\n",
    "\n",
    "    # Drop the edge_emb column\n",
    "    enrichment_df.drop(columns=['edge_emb'], inplace=True)\n",
    "\n",
    "    # Update enrichment in graph_dict\n",
    "    graph_dict[\"edges\"][\"enrichment\"] = enrichment_df\n",
    "\n",
    "    # Clean up\n",
    "    del emb_df\n",
    "    graph_dict[\"edges\"][\"embedding\"][i] = None\n",
    "\n",
    "\n",
    "# Merge the feature column with the edges enrichment dataframe\n",
    "# graph_dict[\"edges\"][\"enrichment\"] = graph_dict[\"edges\"][\"enrichment\"].merge(\n",
    "#     feat_col,\n",
    "#     on=\"triplet_index\",\n",
    "#     how=\"left\",\n",
    "# )\n",
    "\n",
    "# Store the edges enrichment\n",
    "# graph_dict[\"edges\"] = graph_dict[\"edges\"][\"enrichment\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c475d4a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "np.arange(10) + 1250000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c7c8670",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_dict[\"edges\"][\"enrichment\"][graph_dict[\"edges\"][\"enrichment\"].triplet_index.isin(np.arange(10) + 1350000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fb0fb69",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_dict[\"edges\"][\"enrichment\"][graph_dict[\"edges\"][\"enrichment\"].triplet_index.isin(np.arange(10) + 1350000)].iloc[0].feat_emb.to_arrow().to_pylist()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b15c0e9b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67ec9ee6",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "990a71ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Make a simple dataframe\n",
    "# feat_col = cudf.DataFrame(\n",
    "#     {\n",
    "#         \"triplet_index\": graph_dict[\"edges\"][\"enrichment\"].triplet_index,\n",
    "#         \"feat_emb\": None\n",
    "#     }\n",
    "# )\n",
    "\n",
    "# # Loop over a set of embeddings chunks\n",
    "# for i, emb_df in enumerate(graph_dict[\"edges\"][\"embedding\"]):\n",
    "#     # Merge the embeddings into the feature column dataframe\n",
    "#     feat_col = feat_col.merge(\n",
    "#         emb_df,\n",
    "#         on=\"triplet_index\",\n",
    "#         how=\"left\",\n",
    "#     )\n",
    "\n",
    "#     # Fill missing embeddings with edge embeddings\n",
    "#     mask = (feat_col['feat_emb'].isna()) & (feat_col['triplet_index'].isin(emb_df.triplet_index))\n",
    "#     feat_col.loc[mask, 'feat_emb'] = feat_col.loc[mask, 'edge_emb']\n",
    "\n",
    "#     # Drop the edge_emb column\n",
    "#     feat_col = feat_col.drop(columns=['edge_emb'])\n",
    "\n",
    "# # Merge the feature column with the edges enrichment dataframe\n",
    "# graph_dict[\"edges\"][\"enrichment\"] = graph_dict[\"edges\"][\"enrichment\"].merge(\n",
    "#     feat_col,\n",
    "#     on=\"triplet_index\",\n",
    "#     how=\"left\",\n",
    "# )\n",
    "\n",
    "# # Store the edges enrichment\n",
    "# graph_dict[\"edges\"] = graph_dict[\"edges\"][\"enrichment\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
