{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "81d74f39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/awmulyadi/Repositories/temp/office2/AIAgents4Pharma/venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/home/awmulyadi/Repositories/temp/office2/AIAgents4Pharma/venv/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/awmulyadi/Repositories/temp/office2/AIAgents4Pharma/venv/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "  warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\n",
      "/home/awmulyadi/Repositories/temp/office2/AIAgents4Pharma/venv/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/awmulyadi/Repositories/temp/office2/AIAgents4Pharma/venv/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "  warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\n"
     ]
    }
   ],
   "source": [
    "# Import necessary libraries\n",
    "import os\n",
    "import networkx as nx\n",
    "import openai \n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "from langchain_openai import ChatOpenAI, OpenAIEmbeddings\n",
    "import hydra\n",
    "import numpy as np\n",
    "# import pandas as pd\n",
    "import cudf\n",
    "import cupy as cp\n",
    "import cuvs\n",
    "import pickle\n",
    "import torch\n",
    "\n",
    "sys.path.append('../../..')\n",
    "from aiagents4pharma.talk2knowledgegraphs.utils.extractions.cu_multimodal_pcst import MultimodalPCSTPruning\n",
    "from aiagents4pharma.talk2knowledgegraphs.utils.embeddings.ollama import EmbeddingWithOllama\n",
    "from aiagents4pharma.talk2knowledgegraphs.utils.embeddings.sentence_transformer import EmbeddingWithSentenceTransformer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d3981d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"OPENAI_API_KEY\"] = \"XXX\"\n",
    "# Make sure to replace \"your_api_key\" with your actual API key."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ce163cd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# os.environ[\"CUDA_LAUNCH_BLOCKING\"] = \"1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "35dcb609",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'_target_': 'talk2knowledgegraphs.tools.multimodal_subgraph_extraction', 'ollama_embeddings': ['nomic-embed-text'], 'temperature': 0.1, 'streaming': False, 'topk': 5, 'topk_e': 5, 'cost_e': 0.5, 'c_const': 0.01, 'root': -1, 'num_clusters': 1, 'pruning': 'gw', 'verbosity_level': 0, 'node_id_column': 'node_id', 'node_attr_column': 'node_attr', 'edge_src_column': 'edge_src', 'edge_attr_column': 'edge_attr', 'edge_dst_column': 'edge_dst', 'node_colors_dict': {'gene/protein': '#6a79f7', 'molecular_function': '#82cafc', 'cellular_component': '#3f9b0b', 'biological_process': '#c5c9c7', 'drug': '#c4a661', 'disease': '#80013f'}}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load hydra configuration\n",
    "with hydra.initialize(version_base=None, config_path=\"../../../aiagents4pharma/talk2knowledgegraphs/configs\"):\n",
    "    cfg = hydra.compose(\n",
    "        config_name=\"config\", overrides=[\"tools/multimodal_subgraph_extraction=default\"]\n",
    "    )\n",
    "    cfg = cfg.tools.multimodal_subgraph_extraction\n",
    "cfg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "59ac1a64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define state\n",
    "state = {\n",
    "    \"llm_model\": ChatOpenAI(model=\"gpt-4o-mini\", temperature=0.0),\n",
    "    \"embedding_model\": OpenAIEmbeddings(model=\"text-embedding-3-small\"),\n",
    "    \"selected_genes\": [], #[\"IL6_(1567)\", \"IL21_(34967)\", \"TNF_(2329)\"],\n",
    "    \"selected_drugs\": [], #[\"Remdesivir_(15267)\", \"Mesalazine_(15876)\"],\n",
    "    \"uploaded_files\": [\n",
    "        {\n",
    "            \"file_name\": \"multimodal-analysis.csv\",\n",
    "            \"file_path\": '../../../aiagents4pharma/talk2knowledgegraphs/tests/files/multimodal-analysis.csv',\n",
    "            \"file_type\": \"multimodal\",\n",
    "            \"uploaded_by\": \"VPEUser\",\n",
    "            \"uploaded_timestamp\": \"2024-11-05 00:00:00\",\n",
    "        },\n",
    "    ],\n",
    "    \"topk_nodes\": 5,\n",
    "    \"topk_edges\": 5,\n",
    "    \"dic_source_graph\": [\n",
    "        {\n",
    "            \"name\": \"PrimeKG\",\n",
    "            \"kg_pyg_path\": \"../../../aiagents4pharma/talk2knowledgegraphs/tests/files/biobridge_multimodal_pyg_graph.pkl\",\n",
    "            \"kg_text_path\": \"../../../aiagents4pharma/talk2knowledgegraphs/tests/files/biobridge_multimodal_text_graph.pkl\",\n",
    "        }\n",
    "    ],\n",
    "    \"dic_extracted_graph\": []\n",
    "}\n",
    "\n",
    "# Define prompt\n",
    "prompt = \"\"\"\n",
    "DrugA is a human monoclonal antibody that binds to both the soluble and transmembrane bioactive forms of human TNFa (UniProt Acc: P01375). \n",
    "This interaction prevents the binding of TNFa to its receptors, thereby inhibiting the biological activity of TNFa (a cytokine protein).\n",
    "\n",
    "\n",
    "Please extract a subgraph and perform reasoning over it as evidence to explain these mechanisms of action of the given drug. Please set the extraction name for this process as `subkg_druga_tnfa`.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "90602837",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve source graph from the state\n",
    "initial_graph = {}\n",
    "initial_graph[\"source\"] = state[\"dic_source_graph\"][-1]  # The last source graph as of now\n",
    "# logger.log(logging.INFO, \"Source graph: %s\", source_graph)\n",
    "\n",
    "# Load the knowledge graph\n",
    "with open(initial_graph[\"source\"][\"kg_pyg_path\"], \"rb\") as f:\n",
    "    initial_graph[\"pyg\"] = pickle.load(f)\n",
    "# with open(initial_graph[\"source\"][\"kg_text_path\"], \"rb\") as f:\n",
    "#     initial_graph[\"text\"] = pickle.load(f)\n",
    "\n",
    "pyg_graph = initial_graph[\"pyg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "79fbc5f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: GET http://127.0.0.1:11434/api/tags \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST http://127.0.0.1:11434/api/embed \"HTTP/1.1 200 OK\"\n"
     ]
    }
   ],
   "source": [
    "prompt_emb = [EmbeddingWithOllama(model_name=cfg.ollama_embeddings[0]).embed_query(prompt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "61343c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data from the parquet files\n",
    "local_dir = '../../../aiagents4pharma/talk2knowledgegraphs/tests/files'\n",
    "nodes_df = cudf.read_parquet(os.path.join(local_dir, 'biobridge_nodes.parquet.gzip'))\n",
    "edges_df = cudf.read_parquet(os.path.join(local_dir, 'biobridge_edges.parquet.gzip'))\n",
    "\n",
    "graph_nodes = nodes_df.copy()\n",
    "graph_edges = edges_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "43109e17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize logger\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "logger = logging.getLogger(__name__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "255cc656",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:__main__:Initializing dataframes\n",
      "INFO:__main__:Looping over uploaded files\n",
      "INFO:__main__:Checking if multimodal_df is empty\n",
      "INFO:__main__:Preparing multimodal_df\n",
      "INFO:__main__:Processing query dataframe\n",
      "INFO:__main__:Lowering case for node names (q_node_name)\n",
      "INFO:__main__:Lowering case for node names (node_name)\n",
      "INFO:__main__:Filtering based on the query\n",
      "INFO:__main__:Updating state with selected node IDs\n",
      "INFO:__main__:Adding user prompt to query dataframe\n"
     ]
    }
   ],
   "source": [
    "# Initialize dataframes\n",
    "logger.log(logging.INFO, \"Initializing dataframes\")\n",
    "multimodal_df = cudf.DataFrame({\"name\": [], \"node_type\": []})\n",
    "query_df = cudf.DataFrame({\"node_id\": [],\n",
    "                            \"node_type\": [],\n",
    "                            \"x\": [],\n",
    "                            \"desc_x\": [],\n",
    "                            \"use_description\": []})\n",
    "\n",
    "# Loop over the uploaded files and find multimodal files\n",
    "logger.log(logging.INFO, \"Looping over uploaded files\")\n",
    "for i in range(len(state[\"uploaded_files\"])):\n",
    "    # Check if multimodal file is uploaded\n",
    "    if state[\"uploaded_files\"][i][\"file_type\"] == \"multimodal\":\n",
    "        # Read the csv file\n",
    "        multimodal_df = cudf.read_csv(state[\"uploaded_files\"][i][\"file_path\"])\n",
    "\n",
    "# Check if the multimodal_df is empty\n",
    "logger.log(logging.INFO, \"Checking if multimodal_df is empty\")\n",
    "if len(multimodal_df) > 0:\n",
    "    # Prepare multimodal_df\n",
    "    logger.log(logging.INFO, \"Preparing multimodal_df\")\n",
    "    multimodal_df.rename(columns={\"name\": \"q_node_name\",\n",
    "                                    \"node_type\": \"q_node_type\"}, inplace=True)\n",
    "\n",
    "    # Make and process a query dataframe by merging the graph_df and multimodal_df\n",
    "    logger.log(logging.INFO, \"Processing query dataframe\")\n",
    "    query_df = graph_nodes[\n",
    "        ['node_id', 'node_name', 'node_type', 'enriched_node', 'x', 'desc', 'desc_x']\n",
    "    ].merge(multimodal_df, how='cross')\n",
    "    logger.log(logging.INFO, \"Lowering case for node names (q_node_name)\")\n",
    "    query_df['q_node_name'] = query_df['q_node_name'].str.lower()\n",
    "    logger.log(logging.INFO, \"Lowering case for node names (node_name)\")\n",
    "    query_df['node_name'] = query_df['node_name'].str.lower()\n",
    "    # Get the mask for filtering based on the query\n",
    "    logger.log(logging.INFO, \"Filtering based on the query\")\n",
    "    mask = (\n",
    "        query_df['node_name'].str.contains(query_df['q_node_name']) &\n",
    "        (query_df['node_type'] == query_df['q_node_type'])\n",
    "    )\n",
    "    query_df = query_df[mask]\n",
    "    query_df = query_df[['node_id',\n",
    "                            'node_type', \n",
    "                            'enriched_node', \n",
    "                            'x', \n",
    "                            'desc', \n",
    "                            'desc_x']].reset_index(drop=True)\n",
    "    query_df['use_description'] = False # set to False for modal-specific embeddings\n",
    "\n",
    "    # Update the state by adding the the selected node IDs\n",
    "    logger.log(logging.INFO, \"Updating state with selected node IDs\")\n",
    "    state[\"selections\"] = query_df.to_pandas().groupby(\n",
    "        \"node_type\"\n",
    "    )[\"node_id\"].apply(list).to_dict()\n",
    "\n",
    "# Append a user prompt to the query dataframe\n",
    "logger.log(logging.INFO, \"Adding user prompt to query dataframe\")\n",
    "query_df = cudf.concat([\n",
    "    query_df,\n",
    "    cudf.DataFrame({\n",
    "        'node_id': 'user_prompt',\n",
    "        'node_type': 'prompt',\n",
    "        # 'enriched_node': prompt,\n",
    "        'x': prompt_emb,\n",
    "        # 'desc': prompt,\n",
    "        'desc_x': prompt_emb,\n",
    "        'use_description': True # set to True for user prompt embedding\n",
    "    })\n",
    "]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "8074bfa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from aiagents4pharma.talk2knowledgegraphs.utils.extractions.cu_multimodal_pcst import MultimodalPCSTPruning\n",
    "graph = {}\n",
    "graph[\"nodes\"] = nodes_df\n",
    "graph[\"edges\"] = edges_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "76e33cdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>enriched_node</th>\n",
       "      <th>x</th>\n",
       "      <th>desc</th>\n",
       "      <th>desc_x</th>\n",
       "      <th>use_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL7R_(625)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MTILGTTFGMVFSLLQVVSGESGYAQNGDLEDAELDDYSFSCYSQL...</td>\n",
       "      <td>[0.06364653259515762, 0.06951971352100372, 0.0...</td>\n",
       "      <td>IL7R belongs to gene/protein node. IL7R is int...</td>\n",
       "      <td>[0.04506902, 0.008911126, -0.17318207, -0.0157...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCF7_(5195)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MPQLDSGGGGAGGGDDLGAPDELLAFQDEGEEQDDKSRDSAAGPER...</td>\n",
       "      <td>[0.028321774676442146, 0.003539376426488161, 0...</td>\n",
       "      <td>TCF7 belongs to gene/protein node. TCF7 is tra...</td>\n",
       "      <td>[0.036997586, 0.038098544, -0.19027671, -0.006...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>user_prompt</td>\n",
       "      <td>prompt</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>[0.045646045, 0.016633496, -0.14382866, -0.020...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>[0.045646045, 0.016633496, -0.14382866, -0.020...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_id     node_type  \\\n",
       "0   IL7R_(625)  gene/protein   \n",
       "1  TCF7_(5195)  gene/protein   \n",
       "2  user_prompt        prompt   \n",
       "\n",
       "                                       enriched_node  \\\n",
       "0  MTILGTTFGMVFSLLQVVSGESGYAQNGDLEDAELDDYSFSCYSQL...   \n",
       "1  MPQLDSGGGGAGGGDDLGAPDELLAFQDEGEEQDDKSRDSAAGPER...   \n",
       "2                                               <NA>   \n",
       "\n",
       "                                                   x  \\\n",
       "0  [0.06364653259515762, 0.06951971352100372, 0.0...   \n",
       "1  [0.028321774676442146, 0.003539376426488161, 0...   \n",
       "2  [0.045646045, 0.016633496, -0.14382866, -0.020...   \n",
       "\n",
       "                                                desc  \\\n",
       "0  IL7R belongs to gene/protein node. IL7R is int...   \n",
       "1  TCF7 belongs to gene/protein node. TCF7 is tra...   \n",
       "2                                               <NA>   \n",
       "\n",
       "                                              desc_x  use_description  \n",
       "0  [0.04506902, 0.008911126, -0.17318207, -0.0157...            False  \n",
       "1  [0.036997586, 0.038098544, -0.19027671, -0.006...            False  \n",
       "2  [0.045646045, 0.016633496, -0.14382866, -0.020...             True  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "c14b51e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize the subgraph dictionary\n",
    "subgraphs = {}\n",
    "subgraphs[\"nodes\"] = []\n",
    "subgraphs[\"edges\"] = []\n",
    "\n",
    "# Loop over query embeddings and modalities\n",
    "for q in query_df.to_pandas().iterrows():\n",
    "    # Prepare the PCSTPruning object and extract the subgraph\n",
    "    # Parameters were set in the configuration file obtained from Hydra\n",
    "    subgraph = MultimodalPCSTPruning(\n",
    "        topk=state[\"topk_nodes\"],\n",
    "        topk_e=state[\"topk_edges\"],\n",
    "        cost_e=cfg.cost_e,\n",
    "        c_const=cfg.c_const,\n",
    "        root=cfg.root,\n",
    "        num_clusters=cfg.num_clusters,\n",
    "        pruning=cfg.pruning,\n",
    "        verbosity_level=cfg.verbosity_level,\n",
    "        use_description=q[1]['use_description'],\n",
    "    ).extract_subgraph(graph,\n",
    "                       cp.array(q[1]['desc_x']).reshape(1, -1).astype(cp.float32),\n",
    "                       cp.array(q[1]['x']).reshape(1, -1).astype(cp.float32),\n",
    "                       q[1]['node_type'])\n",
    "\n",
    "    # Append the extracted subgraph to the dictionary\n",
    "    subgraphs[\"nodes\"].append(subgraph[\"nodes\"].tolist())\n",
    "    subgraphs[\"edges\"].append(subgraph[\"edges\"].tolist())\n",
    "\n",
    "# Concatenate and get unique node and edge indices\n",
    "subgraphs[\"nodes\"] = np.unique(\n",
    "    np.concatenate([np.array(list_) for list_ in subgraphs[\"nodes\"]])\n",
    ")\n",
    "subgraphs[\"edges\"] = np.unique(\n",
    "    np.concatenate([np.array(list_) for list_ in subgraphs[\"edges\"]])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "cedd9094",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nodes': array([   5,    9,   12,   15,   22,   24,   27,   30,   49,   54,   72,\n",
       "          82,  264,  547,  610,  706,  708,  741,  742,  743,  840,  845,\n",
       "         846,  854,  855, 1320, 1415, 1804, 1863, 1952, 2102, 2263, 2479,\n",
       "        2779, 2827]),\n",
       " 'edges': array([ 123,  125,  167,  176,  288,  291,  888,  920,  921,  942, 1004,\n",
       "        1064, 1083, 1088, 1191, 1216, 1268, 1650, 1760, 1846, 2486, 4005,\n",
       "        4557, 4558, 4958, 5110, 5801, 5802, 5803, 5958, 6613, 6762, 7397,\n",
       "        7400, 9642])}"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subgraphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "bb806d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "subgraph = subgraphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "bc6f8328",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35,)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subgraph[\"nodes\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "350af011",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the dict to a cudf DataFrame\n",
    "node_colors = {n: cfg.node_colors_dict[k]\n",
    "                for k, v in state[\"selections\"].items() for n in v}\n",
    "color_df = cudf.DataFrame(list(node_colors.items()), columns=[\"node_id\", \"color\"])\n",
    "\n",
    "# Prepare graph dataframes\n",
    "# Nodes\n",
    "graph_nodes = graph[\"nodes\"].copy()\n",
    "graph_nodes = graph_nodes.iloc[subgraph['nodes']][\n",
    "    ['node_id', 'node_name', 'node_type', 'desc', 'enriched_node']\n",
    "]\n",
    "graph_nodes = graph_nodes.merge(color_df, on=\"node_id\", how=\"left\")\n",
    "# Edges\n",
    "graph_edges = graph[\"edges\"].copy()\n",
    "graph_edges = graph_edges.iloc[subgraph['edges']][\n",
    "    ['head_id', 'tail_id', 'edge_type']\n",
    "]\n",
    "\n",
    "# Prepare lists for visualization\n",
    "graph_dict = {}\n",
    "graph_dict[\"nodes\"] = [(\n",
    "    row.node_id,\n",
    "    {'desc': row.desc,\n",
    "        'node_type': row.node_type,\n",
    "        'node_name': row.node_name,\n",
    "        'enriched_node': row.enriched_node,\n",
    "        'color': row.color})\n",
    "        for row in graph_nodes.to_arrow().to_pandas().itertuples(index=False)]\n",
    "graph_dict[\"edges\"] = [(\n",
    "    row.head_id, \n",
    "    row.tail_id,\n",
    "    {'label': tuple(row.edge_type)})\n",
    "    for row in graph_edges.to_arrow().to_pandas().itertuples(index=False)]\n",
    "\n",
    "# Prepare the textualized subgraph\n",
    "graph_dict[\"text\"] = (\n",
    "    graph_nodes[\n",
    "        ['node_id', 'desc']\n",
    "    ].rename(columns={'desc': 'node_attr'}).to_arrow().to_pandas().to_csv(index=False)\n",
    "    + \"\\n\"\n",
    "    + graph_edges[\n",
    "        ['head_id', 'edge_type', 'tail_id']\n",
    "    ].to_arrow().to_pandas().to_csv(index=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "6683edfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_type</th>\n",
       "      <th>desc</th>\n",
       "      <th>enriched_node</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7...</td>\n",
       "      <td>(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7...</td>\n",
       "      <td>drug</td>\n",
       "      <td>(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7...</td>\n",
       "      <td>[H][C@](O)(CCC(O)=O)\\C=C\\C=C/C\\C=C/C\\C=C/C\\C=C...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Talmapimod_(17589)</td>\n",
       "      <td>Talmapimod</td>\n",
       "      <td>drug</td>\n",
       "      <td>Talmapimod belongs to drug node. Talmapimod is...</td>\n",
       "      <td>[H][C@]1(C)CN(C(=O)C2=C(Cl)C=C3N(C)C=C(C(=O)C(...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VX-702_(17590)</td>\n",
       "      <td>VX-702</td>\n",
       "      <td>drug</td>\n",
       "      <td>VX-702 belongs to drug node. VX-702 is a small...</td>\n",
       "      <td>NC(=O)N(C1=CC=C(C(N)=O)C(=N1)C1=CC=C(F)C=C1F)C...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Atiprimod_(17591)</td>\n",
       "      <td>Atiprimod</td>\n",
       "      <td>drug</td>\n",
       "      <td>Atiprimod belongs to drug node.     Investigat...</td>\n",
       "      <td>CCCC1(CCC)CCC2(CCN(CCCN(CC)CC)C2)CC1</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>inflammatory bowel disease</td>\n",
       "      <td>disease</td>\n",
       "      <td>inflammatory bowel disease belongs to disease ...</td>\n",
       "      <td>Any inflammatory bowel disease in which the ca...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IL23R_(34778)</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>IL23R belongs to gene/protein node. IL23R is i...</td>\n",
       "      <td>MNQVTIQWDAVIALYILFSWCHGGITNINCSGHIWVEPATIFKMGM...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NKX2-3_(34779)</td>\n",
       "      <td>NKX2-3</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NKX2-3 belongs to gene/protein node. NKX2-3 is...</td>\n",
       "      <td>MMLPSPVTSTPFSVKDILNLEQQHQHFHGAHLQADLEHHFHSAPCM...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Crohn disease_(37784)</td>\n",
       "      <td>Crohn disease</td>\n",
       "      <td>disease</td>\n",
       "      <td>Crohn disease belongs to disease node. A gastr...</td>\n",
       "      <td>A gastrointestinal disorder characterized by c...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ulcerative colitis (disease)_(37785)</td>\n",
       "      <td>ulcerative colitis (disease)</td>\n",
       "      <td>disease</td>\n",
       "      <td>ulcerative colitis (disease) belongs to diseas...</td>\n",
       "      <td>An inflammatory bowel disease involving the mu...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>regulation of gamma-delta T cell differentiati...</td>\n",
       "      <td>regulation of gamma-delta T cell differentiation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>regulation of gamma-delta T cell differentiati...</td>\n",
       "      <td>Any process that modulates the frequency, rate...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>canonical Wnt signaling pathway_(46706)</td>\n",
       "      <td>canonical Wnt signaling pathway</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>canonical Wnt signaling pathway belongs to bio...</td>\n",
       "      <td>The series of molecular signals initiated by b...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>positive regulation of peptidyl-serine phospho...</td>\n",
       "      <td>positive regulation of peptidyl-serine phospho...</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>positive regulation of peptidyl-serine phospho...</td>\n",
       "      <td>Any process that activates or increases the fr...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>protein binding_(53699)</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>protein binding belongs to molecular_function ...</td>\n",
       "      <td>Binding to a protein.</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cytokine receptor activity_(54621)</td>\n",
       "      <td>cytokine receptor activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>cytokine receptor activity belongs to molecula...</td>\n",
       "      <td>Combining with a cytokine and transmitting the...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>beta-catenin-TCF complex_(56282)</td>\n",
       "      <td>beta-catenin-TCF complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>beta-catenin-TCF complex belongs to cellular_c...</td>\n",
       "      <td>A protein complex that contains beta-catenin a...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>beta-catenin-TCF complex assembly_(104150)</td>\n",
       "      <td>beta-catenin-TCF complex assembly</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>beta-catenin-TCF complex assembly belongs to b...</td>\n",
       "      <td>The aggregation, arrangement and bonding toget...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>interleukin-7-mediated signaling pathway_(108951)</td>\n",
       "      <td>interleukin-7-mediated signaling pathway</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>interleukin-7-mediated signaling pathway belon...</td>\n",
       "      <td>The series of molecular signals initiated by i...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>beta-catenin binding_(119055)</td>\n",
       "      <td>beta-catenin binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>beta-catenin binding belongs to molecular_func...</td>\n",
       "      <td>Binding to a catenin beta subunit.</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>interleukin-7 receptor activity_(122046)</td>\n",
       "      <td>interleukin-7 receptor activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>interleukin-7 receptor activity belongs to mol...</td>\n",
       "      <td>Combining with interleukin-7 and transmitting ...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>IL7R_(625)</td>\n",
       "      <td>IL7R</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>IL7R belongs to gene/protein node. IL7R is int...</td>\n",
       "      <td>MTILGTTFGMVFSLLQVVSGESGYAQNGDLEDAELDDYSFSCYSQL...</td>\n",
       "      <td>#6a79f7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TCF7_(5195)</td>\n",
       "      <td>TCF7</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TCF7 belongs to gene/protein node. TCF7 is tra...</td>\n",
       "      <td>MPQLDSGGGGAGGGDDLGAPDELLAFQDEGEEQDDKSRDSAAGPER...</td>\n",
       "      <td>#6a79f7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PPARG_(989)</td>\n",
       "      <td>PPARG</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PPARG belongs to gene/protein node. PPARG is p...</td>\n",
       "      <td>MGETLGDSPIDPESDSFTDTLSANISQEMTMVDTEMPFWPTNFGIS...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>IL10RA_(1299)</td>\n",
       "      <td>IL10RA</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>IL10RA belongs to gene/protein node. IL10RA is...</td>\n",
       "      <td>MLPCLVVLLAALLSLRLGSDAHGTELPSPPSVWFEAEFFHHILHWT...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>JAK2_(1618)</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>JAK2 belongs to gene/protein node. JAK2 is Jan...</td>\n",
       "      <td>MGMACLTMTEMEGTSTSSIYQNGDISGNANSMKQIDPVLQVYLYHS...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>LRRK2_(2111)</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>LRRK2 belongs to gene/protein node. LRRK2 is l...</td>\n",
       "      <td>MASGSCQGCEEDEETLKKLIVRLNNVQEGKQIETLVQILEDLLVFT...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TNF_(2329)</td>\n",
       "      <td>TNF</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TNF belongs to gene/protein node. TNF is tumor...</td>\n",
       "      <td>MSTESMIRDVELAEEALPKKTGGPQGSRRCLFLSLFSFLIVAGATT...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>LSP1_(2643)</td>\n",
       "      <td>LSP1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>LSP1 belongs to gene/protein node. LSP1 is lym...</td>\n",
       "      <td>MAEASSDPGAEEREELLGPTAQWSVEDEEEAVHEQCQHERDRQLQA...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PRDM1_(2874)</td>\n",
       "      <td>PRDM1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PRDM1 belongs to gene/protein node. PRDM1 is P...</td>\n",
       "      <td>MLDICLEKRVGTTLAAPKCNSSTVRFQGLAEGTKGTMKMDMEDADM...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>RIPK2_(4731)</td>\n",
       "      <td>RIPK2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>RIPK2 belongs to gene/protein node. RIPK2 is r...</td>\n",
       "      <td>MNGEAICSALPTIPYHKLADLRYLSRGASGTVSSARHADWRVQVAV...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>INAVA_(9104)</td>\n",
       "      <td>INAVA</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INAVA belongs to gene/protein node. INAVA is i...</td>\n",
       "      <td>MLQMPKLNEIPPGRAGRREARGEGRWPGQTGPEAARLEWRAQGQAG...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>IL18RAP_(11588)</td>\n",
       "      <td>IL18RAP</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>IL18RAP belongs to gene/protein node. IL18RAP ...</td>\n",
       "      <td>MLCLGWIFLWLVAGERIKGFNISGCSTKKLLWTYSTRSEEEFVLFC...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Tofacitinib_(14407)</td>\n",
       "      <td>Tofacitinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Tofacitinib belongs to drug node. Tofacitinib ...</td>\n",
       "      <td>[H][C@@]1(C)CCN(C[C@]1([H])N(C)C1=NC=NC2=C1C=C...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Binimetinib_(15391)</td>\n",
       "      <td>Binimetinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Binimetinib belongs to drug node. Binimetinib,...</td>\n",
       "      <td>CN1C=NC2=C(F)C(NC3=CC=C(Br)C=C3F)=C(C=C12)C(=O...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Fostamatinib_(15674)</td>\n",
       "      <td>Fostamatinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Fostamatinib belongs to drug node. Fostamatini...</td>\n",
       "      <td>COC1=CC(NC2=NC=C(F)C(NC3=NC4=C(OC(C)(C)C(=O)N4...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic a...</td>\n",
       "      <td>(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic acid</td>\n",
       "      <td>drug</td>\n",
       "      <td>(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic a...</td>\n",
       "      <td>[H][C@](O)(CCCCC)\\C=C\\C=C/CCCCCCCC(O)=O</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              node_id  \\\n",
       "0   (4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7...   \n",
       "1                                  Talmapimod_(17589)   \n",
       "2                                      VX-702_(17590)   \n",
       "3                                   Atiprimod_(17591)   \n",
       "4                  inflammatory bowel disease_(28158)   \n",
       "5                                       IL23R_(34778)   \n",
       "6                                      NKX2-3_(34779)   \n",
       "7                               Crohn disease_(37784)   \n",
       "8                ulcerative colitis (disease)_(37785)   \n",
       "9   regulation of gamma-delta T cell differentiati...   \n",
       "10            canonical Wnt signaling pathway_(46706)   \n",
       "11  positive regulation of peptidyl-serine phospho...   \n",
       "12                            protein binding_(53699)   \n",
       "13                 cytokine receptor activity_(54621)   \n",
       "14                   beta-catenin-TCF complex_(56282)   \n",
       "15         beta-catenin-TCF complex assembly_(104150)   \n",
       "16  interleukin-7-mediated signaling pathway_(108951)   \n",
       "17                      beta-catenin binding_(119055)   \n",
       "18           interleukin-7 receptor activity_(122046)   \n",
       "19                                         IL7R_(625)   \n",
       "20                                        TCF7_(5195)   \n",
       "21                                        PPARG_(989)   \n",
       "22                                      IL10RA_(1299)   \n",
       "23                                        JAK2_(1618)   \n",
       "24                                       LRRK2_(2111)   \n",
       "25                                         TNF_(2329)   \n",
       "26                                        LSP1_(2643)   \n",
       "27                                       PRDM1_(2874)   \n",
       "28                                       RIPK2_(4731)   \n",
       "29                                       INAVA_(9104)   \n",
       "30                                    IL18RAP_(11588)   \n",
       "31                                Tofacitinib_(14407)   \n",
       "32                                Binimetinib_(15391)   \n",
       "33                               Fostamatinib_(15674)   \n",
       "34  (9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic a...   \n",
       "\n",
       "                                            node_name           node_type  \\\n",
       "0   (4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7...                drug   \n",
       "1                                          Talmapimod                drug   \n",
       "2                                              VX-702                drug   \n",
       "3                                           Atiprimod                drug   \n",
       "4                          inflammatory bowel disease             disease   \n",
       "5                                               IL23R        gene/protein   \n",
       "6                                              NKX2-3        gene/protein   \n",
       "7                                       Crohn disease             disease   \n",
       "8                        ulcerative colitis (disease)             disease   \n",
       "9    regulation of gamma-delta T cell differentiation  biological_process   \n",
       "10                    canonical Wnt signaling pathway  biological_process   \n",
       "11  positive regulation of peptidyl-serine phospho...  biological_process   \n",
       "12                                    protein binding  molecular_function   \n",
       "13                         cytokine receptor activity  molecular_function   \n",
       "14                           beta-catenin-TCF complex  cellular_component   \n",
       "15                  beta-catenin-TCF complex assembly  biological_process   \n",
       "16           interleukin-7-mediated signaling pathway  biological_process   \n",
       "17                               beta-catenin binding  molecular_function   \n",
       "18                    interleukin-7 receptor activity  molecular_function   \n",
       "19                                               IL7R        gene/protein   \n",
       "20                                               TCF7        gene/protein   \n",
       "21                                              PPARG        gene/protein   \n",
       "22                                             IL10RA        gene/protein   \n",
       "23                                               JAK2        gene/protein   \n",
       "24                                              LRRK2        gene/protein   \n",
       "25                                                TNF        gene/protein   \n",
       "26                                               LSP1        gene/protein   \n",
       "27                                              PRDM1        gene/protein   \n",
       "28                                              RIPK2        gene/protein   \n",
       "29                                              INAVA        gene/protein   \n",
       "30                                            IL18RAP        gene/protein   \n",
       "31                                        Tofacitinib                drug   \n",
       "32                                        Binimetinib                drug   \n",
       "33                                       Fostamatinib                drug   \n",
       "34  (9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic acid                drug   \n",
       "\n",
       "                                                 desc  \\\n",
       "0   (4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7...   \n",
       "1   Talmapimod belongs to drug node. Talmapimod is...   \n",
       "2   VX-702 belongs to drug node. VX-702 is a small...   \n",
       "3   Atiprimod belongs to drug node.     Investigat...   \n",
       "4   inflammatory bowel disease belongs to disease ...   \n",
       "5   IL23R belongs to gene/protein node. IL23R is i...   \n",
       "6   NKX2-3 belongs to gene/protein node. NKX2-3 is...   \n",
       "7   Crohn disease belongs to disease node. A gastr...   \n",
       "8   ulcerative colitis (disease) belongs to diseas...   \n",
       "9   regulation of gamma-delta T cell differentiati...   \n",
       "10  canonical Wnt signaling pathway belongs to bio...   \n",
       "11  positive regulation of peptidyl-serine phospho...   \n",
       "12  protein binding belongs to molecular_function ...   \n",
       "13  cytokine receptor activity belongs to molecula...   \n",
       "14  beta-catenin-TCF complex belongs to cellular_c...   \n",
       "15  beta-catenin-TCF complex assembly belongs to b...   \n",
       "16  interleukin-7-mediated signaling pathway belon...   \n",
       "17  beta-catenin binding belongs to molecular_func...   \n",
       "18  interleukin-7 receptor activity belongs to mol...   \n",
       "19  IL7R belongs to gene/protein node. IL7R is int...   \n",
       "20  TCF7 belongs to gene/protein node. TCF7 is tra...   \n",
       "21  PPARG belongs to gene/protein node. PPARG is p...   \n",
       "22  IL10RA belongs to gene/protein node. IL10RA is...   \n",
       "23  JAK2 belongs to gene/protein node. JAK2 is Jan...   \n",
       "24  LRRK2 belongs to gene/protein node. LRRK2 is l...   \n",
       "25  TNF belongs to gene/protein node. TNF is tumor...   \n",
       "26  LSP1 belongs to gene/protein node. LSP1 is lym...   \n",
       "27  PRDM1 belongs to gene/protein node. PRDM1 is P...   \n",
       "28  RIPK2 belongs to gene/protein node. RIPK2 is r...   \n",
       "29  INAVA belongs to gene/protein node. INAVA is i...   \n",
       "30  IL18RAP belongs to gene/protein node. IL18RAP ...   \n",
       "31  Tofacitinib belongs to drug node. Tofacitinib ...   \n",
       "32  Binimetinib belongs to drug node. Binimetinib,...   \n",
       "33  Fostamatinib belongs to drug node. Fostamatini...   \n",
       "34  (9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic a...   \n",
       "\n",
       "                                        enriched_node    color  \n",
       "0   [H][C@](O)(CCC(O)=O)\\C=C\\C=C/C\\C=C/C\\C=C/C\\C=C...    black  \n",
       "1   [H][C@]1(C)CN(C(=O)C2=C(Cl)C=C3N(C)C=C(C(=O)C(...    black  \n",
       "2   NC(=O)N(C1=CC=C(C(N)=O)C(=N1)C1=CC=C(F)C=C1F)C...    black  \n",
       "3                CCCC1(CCC)CCC2(CCN(CCCN(CC)CC)C2)CC1    black  \n",
       "4   Any inflammatory bowel disease in which the ca...    black  \n",
       "5   MNQVTIQWDAVIALYILFSWCHGGITNINCSGHIWVEPATIFKMGM...    black  \n",
       "6   MMLPSPVTSTPFSVKDILNLEQQHQHFHGAHLQADLEHHFHSAPCM...    black  \n",
       "7   A gastrointestinal disorder characterized by c...    black  \n",
       "8   An inflammatory bowel disease involving the mu...    black  \n",
       "9   Any process that modulates the frequency, rate...    black  \n",
       "10  The series of molecular signals initiated by b...    black  \n",
       "11  Any process that activates or increases the fr...    black  \n",
       "12                              Binding to a protein.    black  \n",
       "13  Combining with a cytokine and transmitting the...    black  \n",
       "14  A protein complex that contains beta-catenin a...    black  \n",
       "15  The aggregation, arrangement and bonding toget...    black  \n",
       "16  The series of molecular signals initiated by i...    black  \n",
       "17                 Binding to a catenin beta subunit.    black  \n",
       "18  Combining with interleukin-7 and transmitting ...    black  \n",
       "19  MTILGTTFGMVFSLLQVVSGESGYAQNGDLEDAELDDYSFSCYSQL...  #6a79f7  \n",
       "20  MPQLDSGGGGAGGGDDLGAPDELLAFQDEGEEQDDKSRDSAAGPER...  #6a79f7  \n",
       "21  MGETLGDSPIDPESDSFTDTLSANISQEMTMVDTEMPFWPTNFGIS...    black  \n",
       "22  MLPCLVVLLAALLSLRLGSDAHGTELPSPPSVWFEAEFFHHILHWT...    black  \n",
       "23  MGMACLTMTEMEGTSTSSIYQNGDISGNANSMKQIDPVLQVYLYHS...    black  \n",
       "24  MASGSCQGCEEDEETLKKLIVRLNNVQEGKQIETLVQILEDLLVFT...    black  \n",
       "25  MSTESMIRDVELAEEALPKKTGGPQGSRRCLFLSLFSFLIVAGATT...    black  \n",
       "26  MAEASSDPGAEEREELLGPTAQWSVEDEEEAVHEQCQHERDRQLQA...    black  \n",
       "27  MLDICLEKRVGTTLAAPKCNSSTVRFQGLAEGTKGTMKMDMEDADM...    black  \n",
       "28  MNGEAICSALPTIPYHKLADLRYLSRGASGTVSSARHADWRVQVAV...    black  \n",
       "29  MLQMPKLNEIPPGRAGRREARGEGRWPGQTGPEAARLEWRAQGQAG...    black  \n",
       "30  MLCLGWIFLWLVAGERIKGFNISGCSTKKLLWTYSTRSEEEFVLFC...    black  \n",
       "31  [H][C@@]1(C)CCN(C[C@]1([H])N(C)C1=NC=NC2=C1C=C...    black  \n",
       "32  CN1C=NC2=C(F)C(NC3=CC=C(Br)C=C3F)=C(C=C12)C(=O...    black  \n",
       "33  COC1=CC(NC2=NC=C(F)C(NC3=NC4=C(OC(C)(C)C(=O)N4...    black  \n",
       "34            [H][C@](O)(CCCCC)\\C=C\\C=C/CCCCCCCC(O)=O    black  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_nodes['color'].fillna('black', inplace=True)\n",
    "graph_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "840a5cdc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_type</th>\n",
       "      <th>desc</th>\n",
       "      <th>enriched_node</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>TNF_(2329)</td>\n",
       "      <td>TNF</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TNF belongs to gene/protein node. TNF is tumor...</td>\n",
       "      <td>MSTESMIRDVELAEEALPKKTGGPQGSRRCLFLSLFSFLIVAGATT...</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_id node_name     node_type  \\\n",
       "53  TNF_(2329)       TNF  gene/protein   \n",
       "\n",
       "                                                 desc  \\\n",
       "53  TNF belongs to gene/protein node. TNF is tumor...   \n",
       "\n",
       "                                        enriched_node  color  \n",
       "53  MSTESMIRDVELAEEALPKKTGGPQGSRRCLFLSLFSFLIVAGATT...  black  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_nodes[graph_nodes.node_name=='TNF']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "bc6aaa55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('KIF21B_(8564)',\n",
       "  {'desc': 'KIF21B belongs to gene/protein node. KIF21B is kinesin family member 21B. This gene encodes a member of the kinesin superfamily. Kinesins are ATP-dependent microtubule-based motor proteins that are involved in the intracellular transport of membranous organelles. Single nucleotide polymorphisms in this gene are associated with inflammatory bowel disease and multiple sclerosis. Alternatively spliced transcript variants encoding multiple isoforms have been observed for this gene. [provided by RefSeq, Nov 2011].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'KIF21B',\n",
       "   'enriched_node': 'MAGQGDCCVKVAVRIRPQLSKEKIEGCHICTSVTPGEPQVLLGKDKAFTYDFVFDLDTWQEQIYSTCVSKLIEGCFEGYNATVLAYGQTGAGKTYTMGTGFDMATSEEEQGIIPRAIAHLFGGIAERKRRAQEQGVAGPEFKVSAQFLELYNEEILDLFDSTRDPDTRHRRSNIKIHEDANGGIYTTGVTSRLIHSQEELIQCLKQGALSRTTASTQMNVQSSRSHAIFTIHLCQMRMCTQPDLVNEAVTGLPDGTPPSSEYETLTAKFHFVDLAGSERLKRTGATGERAKEGISINCGLLALGNVISALGDQSKKVVHVPYRDSKLTRLLQDSLGGNSQTIMIACVSPSDRDFMETLNTLKYANRARNIKNKVVVNQDKTSQQISALRAEIARLQMELMEYKAGKRVIGEDGAEGYSDLFRENAMLQKENGALRLRVKAMQEAIDAINNRVTQLMSQEANLLLAKAGDGNEAIGALIQNYIREIEELRTKLLESEAMNESLRRSLSRASARSPYSLGASPAAPAFGGSPASSMEDASEVIRRAKQDLERLKKKEVRQRRKSPEKEAFKKRAKLQQENSEETDENEAEEEEEERDESGCEEEEGREDEDEDSGSEESLVDSDSDPEEKEVNFQADLADLTCEIEIKQKLIDELENSQRRLQTLKHQYEEKLILLQNKIRDTQLERDRVLQNLSTMECYTEEKANKIKADYEKRLREMNRDLQKLQAAQKEHARLLKNQSRYERELKKLQAEVAEMKKAKVALMKQMREEQQRRRLVETKRNREIAQLKKEQRRQEFQIRALESQKRQQEMVLRRKTQEVSALRRLAKPMSERVAGRAGLKPPMLDSGAEVSASTTSSEAESGARSVSSIVRQWNRKINHFLGDHPAPTVNGTRPARKKFQKKGASQSFSKAARLKWQSLERRIIDIVMQRMTIVNLEADMERLIKKREELFLLQEALRRKRERLQAESPEEEKGLQELAEEIEVLAANIDYINDGITDCQATIVQLEETKEELDSTDTSVVISSCSLAEARLLLDNFLKASIDKGLQVAQKEAQIRLLEGRLRQTDMAGSSQNHLLLDALREKAEAHPELQALIYNVQQENGYASTDEEISEFSEGSFSQSFTMKGSTSHDDFKFKSEPKLSAQMKAVSAECLGPPLDISTKNITKSLASLVEIKEDGVGFSVRDPYYRDRVSRTVSLPTRGSTFPRQSRATETSPLTRRKSYDRGQPIRSTDVGFTPPSSPPTRPRNDRNVFSRLTSNQSQGSALDKSDDSDSSLSEVLRGIISPVGGAKGARTAPLQCVSMAEGHTKPILCLDATDELLFTGSKDRSCKMWNLVTGQEIAALKGHPNNVVSIKYCSHSGLVFSVSTSYIKVWDIRDSAKCIRTLTSSGQVISGDACAATSTRAITSAQGEHQINQIALSPSGTMLYAASGNAVRIWELSRFQPVGKLTGHIGPVMCLTVTQTASQHDLVVTGSKDHYVKMFELGECVTGTIGPTHNFEPPHYDGIECLAIQGDILFSGSRDNGIKKWDLDQQELIQQIPNAHKDWVCALAFIPGRPMLLSACRAGVIKVWNVDNFTPIGEIKGHDSPINAICTNAKHIFTASSDCRVKLWNYVPGLTPCLPRRVLAIKGRATTLP',\n",
       "   'color': <NA>}),\n",
       " ('INAVA_(9104)',\n",
       "  {'desc': 'INAVA belongs to gene/protein node. INAVA is innate immunity activator. Involved in several processes, including nucleotide-binding activity oligomerization domain containing 2 signaling pathway; positive regulation of cytokine production; and positive regulation of intracellular signal transduction. Located in cytoplasm and nucleus. Implicated in inflammatory bowel disease 29. [provided by Alliance of Genome Resources, Apr 2022]',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'INAVA',\n",
       "   'enriched_node': 'MLQMPKLNEIPPGRAGRREARGEGRWPGQTGPEAARLEWRAQGQAGGARAPWDSWGSSRLPTQPGPGWSRCPPSLLCALSFQKSTMESKDEVSDTDSGIILQSGPDSPVSPMKELTHAVHKQQRALEARLEACLEELRRLCLREAELTGTLPAEYPLKPGEKAPKVRRRIGAAYKLDDWALHREDPLSSLERQLALQLQITEAARRLCLEENLSRQARRQRKHSMLQEEKKLQELQRCLVERRRNSEPPPAAALPLGRELSASDDSSLSDGLLLEEEESQVPKPPPESPAPPSRPLPPQTLEGLQPTGPEAGSPERAPVQNSPWKETSLDHPYEKPRKSSEPWSESSSPATTPQDGPSASSLWLLEPASYHVVPIRGVPGQWQGRTSAPATPEIQGRRGQSQSLRVDSFRAGPEGRGRSAFPRRRPTHYTVTVPDSCFPATKPPLPHAACHSCSEDSGSDVSSISHPTSPGSSSPDISFLQPLSPPKTHRHRGAWVPAGSRELVAHHPKLLLPPGYFPAGRYVVVAESPLPPGEWELRRAAPGPAYEEEGTPLRYQRLVPSRSRIVRTPSLKDSPAGRGLSKAAVSEELKWWHERARLRSTRPHSLDRQGAFRVRSLPLGREGFGRALGPRAQVPTVCVLRRSPDGAPVQVFVPEKGEIISQV',\n",
       "   'color': <NA>}),\n",
       " ('IL18RAP_(11588)',\n",
       "  {'desc': \"IL18RAP belongs to gene/protein node. IL18RAP is interleukin 18 receptor accessory protein. The protein encoded by this gene is an accessory subunit of the heterodimeric receptor for interleukin 18 (IL18), a proinflammatory cytokine involved in inducing cell-mediated immunity. This protein enhances the IL18-binding activity of the IL18 receptor and plays a role in signaling by IL18. Mutations in this gene are associated with Crohn's disease and inflammatory bowel disease, and susceptibility to celiac disease and leprosy. Alternatively spliced transcript variants of this gene have been described, but their full-length nature is not known. [provided by RefSeq, Feb 2014].\",\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'IL18RAP',\n",
       "   'enriched_node': 'MLCLGWIFLWLVAGERIKGFNISGCSTKKLLWTYSTRSEEEFVLFCDLPEPQKSHFCHRNRLSPKQVPEHLPFMGSNDLSDVQWYQQPSNGDPLEDIRKSYPHIIQDKCTLHFLTPGVNNSGSYICRPKMIKSPYDVACCVKMILEVKPQTNASCEYSASHKQDLLLGSTGSISCPSLSCQSDAQSPAVTWYKNGKLLSVERSNRIVVDEVYDYHQGTYVCDYTQSDTVSSWTVRAVVQVRTIVGDTKLKPDILDPVEDTLEVELGKPLTISCKARFGFERVFNPVIKWYIKDSDLEWEVSVPEAKSIKSTLKDEIIERNIILEKVTQRDLRRKFVCFVQNSIGNTTQSVQLKEKRGVVLLYILLGTIGTLVAVLAASALLYRHWIEIVLLYRTYQSKDQTLGDKKDFDAFVSYAKWSSFPSEATSSLSEEHLALSLFPDVLENKYGYSLCLLERDVAPGGVYAEDIVSIIKRSRRGIFILSPNYVNGPSIFELQAAVNLALDDQTLKLILIKFCYFQEPESLPHLVKKALRVLPTVTWRGLKSVPPNSRFWAKMRYHMPVKNSQGFTWNQLRITSRIFQWKGLSRTETTGRSSQPKEW',\n",
       "   'color': <NA>}),\n",
       " ('PHOX2B_(13014)',\n",
       "  {'desc': 'PHOX2B belongs to gene/protein node. PHOX2B is paired like homeobox 2B. The DNA-associated protein encoded by this gene is a member of the paired family of homeobox proteins localized to the nucleus. The protein functions as a transcription factor involved in the development of several major noradrenergic neuron populations and the determination of neurotransmitter phenotype. The gene product is linked to enhancement of second messenger-mediated activation of the dopamine beta-hydroylase, c-fos promoters and several enhancers, including cyclic amp-response element and serum-response element. Expansion of a 20 amino acid polyalanine tract in this protein by 5-13 aa has been associated with congenital central hypoventilation syndrome. [provided by RefSeq, Jul 2016].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'PHOX2B',\n",
       "   'enriched_node': 'MYKMEYSYLNSSAYESCMAGMDTSSLASAYADFSSCSQASGFQYNPIRTTFGATSGCPSLTPGSCSLGTLRDHQSSPYAAVPYKLFTDHGGLNEKRKQRRIRTTFTSAQLKELERVFAETHYPDIYTREELALKIDLTEARVQVWFQNRRAKFRKQERAAAAAAAAAKNGSSGKKSDSSRDDESKEAKSTDPDSTGGPGPNPNPTPSCGANGGGGGGPSPAGAPGAAGPGGPGGEPGKGGAAAAAAAAAAAAAAAAAAAAGGLAAAGGPGQGWAPGPGPITSIPDSLGGPFASVLSSLQRPNGAKAALVKSSMF',\n",
       "   'color': <NA>}),\n",
       " ('Diclofenac_(14061)',\n",
       "  {'desc': 'Diclofenac belongs to drug node. Diclofenac is a phenylacetic acid derivative and non-steroidal anti-inflammatory drug (NSAID). NSAIDs inhibit cyclooxygenase (COX)-1 and-2 which are the enzyme responsible for producing prostaglandins (PGs). PGs contribute to inflammation and pain signalling. Diclofenac, like other NSAIDs, is often used as first line therapy for acute and chronic pain and inflammation from a variety of causes. Diclofenac was the product of rational drug design based on the structures of , , and. The addition of two chlorine groups in the ortho position of the phenyl ring locks the ring in maximal torsion which appears to be related to increased potency. It is often used in combination with to prevent NSAID-induced gastric ulcers. Diclofenac was first approved by the FDA in July 1988 under the trade name Voltaren, marketed by Novartis (previously Ciba-Geigy). Diclofenac inhibits cyclooxygenase-1 and -2, the enzymes responsible for production of prostaglandin (PG) G<sub>2</sub> which is the precursor to other PGs. These molecules have broad activity in pain and inflammation and the inhibition of their production is the common mechanism linking each effect of diclofenac. Diclofenac is over 99.7% bound to serum proteins, primarily albumin. It is undergoes limited binding to lipoproteins as well with 1.1% bound to HDL, 0.3% to LDL, and 0.15% to VLDL. Diclofenac reduces inflammation and by extension reduces nociceptive pain and combats fever. It also increases the risk of developing a gastrointestinal ulcer by inhibiting the production of protective mucus in the stomach. Diclofenac is indicated for use in the treatment of pain and inflammation from varying sources including inflammatory conditions such as osteoarthritis, rheumatoid arthritis, and akylosing spondylitis, as well as injury-related inflammation due to surgery and physical trauma. It is often used in combination with as a gastro-protective agent in patients with high risk of developing NSAID-induced ulcers.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Diclofenac',\n",
       "   'enriched_node': 'OC(=O)CC1=CC=CC=C1NC1=C(Cl)C=CC=C1Cl',\n",
       "   'color': <NA>}),\n",
       " ('Flufenamic acid_(14069)',\n",
       "  {'desc': 'Flufenamic acid belongs to drug node. An anthranilic acid derivative with analgesic, anti-inflammatory, and antipyretic properties. It is used in musculoskeletal and joint disorders and administered by mouth and topically. (From Martindale, The Extra Pharmacopoeia, 30th ed, p16)    ',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Flufenamic acid',\n",
       "   'enriched_node': 'OC(=O)C1=CC=CC=C1NC1=CC(=CC=C1)C(F)(F)F',\n",
       "   'color': <NA>}),\n",
       " ('Glycyrrhizic acid_(14107)',\n",
       "  {'desc': 'Glycyrrhizic acid belongs to drug node. Glycyrrhizic acid is extracted from the root of the licorice plant; _Glycyrrhiza glabra_. It is a triterpene glycoside with glycyrrhetinic acid that possesses a wide range of pharmacological and biological activities. When extracted from the plant, it can be obtained in the form of ammonium glycyrrhizin and mono-ammonium glycyrrhizin. Glycyrrhizic acid has been developed in Japan and China as a hepatoprotective drug in cases of chronic hepatitis. From January 2014, glycyrrhizic acid as part of the licorice extract was approved by the FDA as an existing food sweetener. It was approved by Health Canada to be used in over-the-counter products but all the products are currently on the status canceled post marketed. Glycyrrhizic acid can be found in the alpha and beta forms. The alpha form is predomit in the liver and duodenum and thus, it is thought that the anti-inflammatory liver effect of this drug are mainly due to the action of this isomer. Glycyrrhizic acid anti-inflammatory effect is generated via suppression of TNF alpha and caspase 3. It also inhibits the translocation of NFkB into the nuclei and conjugates free radicals. Some studies have shown a glycyrrhizic-driven inhibition of CD4+ T cell proliferation via JNK, ERK and PI3K/AKT. Glycyrrhizic acid does not bind to any plasma proteins as it is not absorbed systemically. On the other hand, its main active metabolite, glycyrrhetinic acid presents a very large binding to serum proteins such as albumin. Glycyrrhizic acid was reported to present antiallergic, antiviral and anti-inflammatory activities as well as improvements in glucose tolerance.  Glycyrrhizic acid is widely applied in foods as a natural sweetener. As a therapeutic agent, is has been used in a vast variety of formulations as it is reported to be anti-inflammatory, anti-ulcer, anti-allergic, antioxidant, anti-tumor, anti-diabetic and hepatoprotective. Due to this properties, its indications have been: treatment of premenstrual syndrome, treatment of viral infections, anti-lipidemic and antihyperglycemic. It is also known to be used as a remedy for peptic ulcer and other stomach diseases.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Glycyrrhizic acid',\n",
       "   'enriched_node': '[H][C@@]12C[C@](C)(CC[C@]1(C)CC[C@]1(C)C2=CC(=O)[C@]2([H])[C@@]3(C)CC[C@H](O[C@H]4O[C@@H]([C@@H](O)[C@H](O)[C@H]4O[C@@H]4O[C@@H]([C@@H](O)[C@H](O)[C@H]4O)C(O)=O)C(O)=O)C(C)(C)[C@]3([H])CC[C@@]12C)C(O)=O',\n",
       "   'color': <NA>}),\n",
       " ('Tofacitinib_(14407)',\n",
       "  {'desc': 'Tofacitinib belongs to drug node. Tofacitinib is an inhibitor of Janus kinases, a group of intracellular enzymes involved in signalling pathways that affect hematopoiesis and immune cell function. It is approved by the FDA for treatment of moderate to severe rheumatoid arthritis that responds inadequately to methotrexate or in those who are intolerant to methotrexate.  Rheumatoid arthritis is an autoimmune disease characterized by a dysregulation of pro-inflammatory cytokines including IL7, IL15, IL21, IL6, IFN-alpha, and IFN-beta. (3) Cytokines signalling results in tissue inflammation and joint damage by stimulating the recruitment and activation of immune cells via the janus kinase signalling pathway. 40%, mostly bound to albumin. Tofacitinib targets inflammation present in rheumatoid arthritis by inhibiting the janus kinases involved in the inflammatory response pathway.  For the treatment of moderate to severe rheumatoid arthritis which is resistant or intolerant to methotrexate therapy. It may also be used as an adjunct to methotrexate therapy, or other non-biologic disease-modifying antirheumatic drugs (DMARDS), when methotrexate alone is not sufficient. ',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Tofacitinib',\n",
       "   'enriched_node': '[H][C@@]1(C)CCN(C[C@]1([H])N(C)C1=NC=NC2=C1C=CN2)C(=O)CC#N',\n",
       "   'color': <NA>}),\n",
       " ('Binimetinib_(15391)',\n",
       "  {'desc': 'Binimetinib belongs to drug node. Binimetinib, also known as _Mektovi_, is a potent is a potent and selective oral mitogen-activated protein kinase 1/2  (MEK 1/2) inhibitor which is combined with ,.  Binimetinib is 97% bound to human plasma proteins and the blood-to-plasma ratio is 0.72  Binimetinib is a MEK inhibitor. MEK is an enzyme that regulates the biosynthesis of the inflammatory cytokines TNF, IL-6 and IL-1. MEK inhibitors interfere with these biosynthetic processes.  It is a chemotherapeutic agent that has anti-tumor activity ,. ',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Binimetinib',\n",
       "   'enriched_node': 'CN1C=NC2=C(F)C(NC3=CC=C(Br)C=C3F)=C(C=C12)C(=O)NOCCO',\n",
       "   'color': <NA>}),\n",
       " ('Fostamatinib_(15674)',\n",
       "  {'desc': 'Fostamatinib belongs to drug node. Fostamatinib has been investigated for the treatment and basic science of Rheumatoid Arthritis and Immune Thrombocytopenic Purpura (ITP). It was approved on April 17, 2018 under the trade name Tavalisse for use in ITP. Fostamatinib has also been granted orphan drug status by the FDA. The active metabolite of fostamatinib, R406, is an inhibitor of spleen tyrosine kinase (Syk). It binds reversibly to the ATP binding pocket with high affinity (Ki = 30nM), inhibiting the kinase activity with an IC50 of 41nM. R406 is 98.3% bound to plasma proteins. The active metabolite of fostamatinib, R406, inhibits signal transduction by Fc receptors involved in the antibody-mediated destruction of platelets by immune cells in chronic ITP. This results in increased platelet counts in this population. Fostamatinib is indicated for use in the treatment of chronic immune thrombocytopenia (ITP) in patients who have had insufficient response to previous therapy.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Fostamatinib',\n",
       "   'enriched_node': 'COC1=CC(NC2=NC=C(F)C(NC3=NC4=C(OC(C)(C)C(=O)N4COP(O)(O)=O)C=C3)=N2)=CC(OC)=C1OC',\n",
       "   'color': <NA>}),\n",
       " ('Sulfasalazine_(15982)',\n",
       "  {'desc': \"Sulfasalazine belongs to drug node. A drug that is used in the management of inflammatory bowel diseases. Its activity is generally considered to lie in its metabolic breakdown product, 5-aminosalicylic acid (see mesalamine) released in the colon. (From Martindale, The Extra Pharmacopoeia, 30th ed, p907) The mode of action of Sulfasalazine or its metabolites, 5-aminosalicylic acid (5-ASA) and sulfapyridine (SP), is still under investigation, but may be related to the anti-inflammatory and/or immunomodulatory properties that have been observed in animal and <i>in vitro</i> models, to its affinity for connective tissue, and/or to the relatively high concentration it reaches in serous fluids, the liver and intestinal walls, as demonstrated in autoradiographic studies in animals. In ulcerative colitis, clinical studies utilizing rectal administration of Sulfasalazine, SP and 5-ASA have indicated that the major therapeutic action may reside in the 5-ASA moiety. The relative contribution of the parent drug and the major metabolites in rheumatoid arthritis is unknown.  Sulfasalazine is an anti-inflammatory indicated for the treatment of ulcerative colitis and rheumatoid arthritis. For the treatment of Crohn's disease and rheumatoid arthritis as a second-line agent.\",\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Sulfasalazine',\n",
       "   'enriched_node': 'OC(=O)C1=CC(=CC=C1O)\\\\N=N\\\\C1=CC=C(C=C1)S(=O)(=O)NC1=NC=CC=C1',\n",
       "   'color': <NA>}),\n",
       " ('Glucosamine_(16195)',\n",
       "  {'desc': 'Glucosamine belongs to drug node. Glucosamine is commonly used as a treatment for osteoarthritis, although its acceptance as a medical therapy varies. It is an amino sugar and precursor of glycosylated proteins and lipids. Since glucosamine is a precursor for glycosaminoglycans, and glycosaminoglycans are a major component of joint cartilage, supplemental glucosamine may help to rebuild cartilage and treat arthritis. Glucosamine is a precursor of glycosylated proteins and lipids. Oral glucosamine is commonly used for the treatment of osteoarthritis. Since glucosamine is a precursor for glycosaminoglycans, and glycosaminoglycans are a major component of joint cartilage, supplemental glucosamine may help to rebuild cartilage and treat arthritis. Its use as a therapy for osteoarthritis appears safe, but there is conflicting evidence as to its effectiveness with more recent studies showing limited to no clinical benefit of use. Glucosamine is not FDA approved for use in humans. Since glucosamine is classified as a dietary supplement, safety and formulation are solely the responsibility of the manufacturer; evidence of safety and efficacy is not required as long as it is not advertised as a treatment for a medical condition.  Osteoarthritis is characterized by the progressive degeneration of cartilage glycosaminoglycans. The formation of glucosamine is the rate limiting step in glycosaminoglycans synthesis thus the addition is glucosamine, would in theory provide a building block towards the synthesis of glycosaminoglycans and thus slow down the progression of osteoarthritis. Thus far however, the results have not been conclusive. Glucosamine is usually used in the treatment of osteoarthritis, although its efficacy is still in question.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Glucosamine',\n",
       "   'enriched_node': 'N[C@H]1C(O)O[C@H](CO)[C@@H](O)[C@@H]1O',\n",
       "   'color': <NA>}),\n",
       " ('2-Amino-N,3,3-Trimethylbutanamide_(16196)',\n",
       "  {'desc': '2-Amino-N,3,3-Trimethylbutanamide belongs to drug node.     ',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': '2-Amino-N,3,3-Trimethylbutanamide',\n",
       "   'enriched_node': 'CNC(=O)[C@@H](N)C(C)(C)C',\n",
       "   'color': <NA>}),\n",
       " ('(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic acid_(16613)',\n",
       "  {'desc': '(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic acid belongs to drug node.     ',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': '(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic acid',\n",
       "   'enriched_node': '[H][C@](O)(CCCCC)\\\\C=C\\\\C=C/CCCCCCCC(O)=O',\n",
       "   'color': <NA>}),\n",
       " ('(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7,10,13,16,19-hexaenoic acid_(16615)',\n",
       "  {'desc': '(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7,10,13,16,19-hexaenoic acid belongs to drug node.     ',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': '(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7,10,13,16,19-hexaenoic acid',\n",
       "   'enriched_node': '[H][C@](O)(CCC(O)=O)\\\\C=C\\\\C=C/C\\\\C=C/C\\\\C=C/C\\\\C=C/C\\\\C=C/CC',\n",
       "   'color': <NA>}),\n",
       " ('Talmapimod_(17589)',\n",
       "  {'desc': 'Talmapimod belongs to drug node. Talmapimod is the first-generation oral p38 MAP kinase inhibitor developed by Scios. It has shown to be effective to cure inflammatory diseases such as Rheumatoid Arthritis. SCIO-469 inhibits p38 kinase, a stimulatory modulator of pro-inflammatory factors including tumor necrosis factor-alpha (TNFa), interleukin-1 (IL-1), and cyclooxygenase-2 (COX-2), all of which are known to contribute to both symptoms and disease progression in patients with Rheumatoid Arthritis (RA). Existing protein-based products that antagonize TNFa have been shown to markedly relieve the symptoms and retard the progression of RA. It also has the potential for additional benefits associated with its inhibition of IL-1 and COX-2.   Investigated for use/treatment in pain (acute or chronic) and rheumatoid arthritis.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Talmapimod',\n",
       "   'enriched_node': '[H][C@]1(C)CN(C(=O)C2=C(Cl)C=C3N(C)C=C(C(=O)C(=O)N(C)C)C3=C2)[C@]([H])(C)CN1CC1=CC=C(F)C=C1',\n",
       "   'color': <NA>}),\n",
       " ('VX-702_(17590)',\n",
       "  {'desc': 'VX-702 belongs to drug node. VX-702 is a small molecule investigational oral anti-cytokine therapy for treatment of inflammatory diseases, specifically rheumatoid arthritis (RA). It acts as a p38 MAP kinase inhibitor. In the future, VX-702 may be investigated for combination with methotrexate, a commonly used therapy for RA. This p38 MAP kinase inhibitor effectively inhibits LPS-stimulated TNF||, IL-6 and IL-1|| production.  VX-703 is an anti-cytokine therapy in which p38 MAP kinase inhibitor effectively inhibits LPS-stimulated TNF||, IL-6 and IL-1|| production. Investigated for use/treatment in coronary artery disease, inflammatory disorders (unspecified), and rheumatoid arthritis.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'VX-702',\n",
       "   'enriched_node': 'NC(=O)N(C1=CC=C(C(N)=O)C(=N1)C1=CC=C(F)C=C1F)C1=C(F)C=CC=C1F',\n",
       "   'color': <NA>}),\n",
       " ('Atiprimod_(17591)',\n",
       "  {'desc': 'Atiprimod belongs to drug node.     Investigated for use/treatment in cancer/tumors (unspecified), multiple myeloma, and rheumatoid arthritis.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Atiprimod',\n",
       "   'enriched_node': 'CCCC1(CCC)CCC2(CCN(CCCN(CC)CC)C2)CC1',\n",
       "   'color': <NA>}),\n",
       " ('Andrographolide_(17594)',\n",
       "  {'desc': 'Andrographolide belongs to drug node. Andrographolide (HMPL-004) is a botanical product extracted from a herb that occurs naturally in China. The herb has an extensive history of use in TCM for the treatment of upper respiratory tract infections and other inflammatory and infectious diseases. HMPL-004 acts on multiple cellular targets in the inflammatory signal transduction pathways resulting in suppressed inflammation cytokine expression including TNF-, IL-1 and IL-6. HMPL-004 was demonstrated to inhibit TNF- and IL-1 production in cell-based assays. HMPL-004 is also able to inhibit NF-kB activation. NF-kB is a family of transcriptional factors that regulate a wide spectrum of genes critically involved in host defence and inflammation. The mechanism of action of HMPL-004 was further supported in laboratory IBD animal models. Treatment of IBD rats with HMPL-004 caused a significant drop in plasma cytokine concentrations, including TNF- and IL-1.   Investigated for use/treatment in ulcerative colitis.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Andrographolide',\n",
       "   'enriched_node': '[H][C@]12CCC(=C)[C@@H](C\\\\C=C3/[C@H](O)COC3=O)[C@]1(C)CC[C@@H](O)[C@@]2(C)CO',\n",
       "   'color': <NA>}),\n",
       " ('Dilmapimod_(17602)',\n",
       "  {'desc': 'Dilmapimod belongs to drug node. Dilmapimod has been used in trials studying the treatment and diagnostic of Nerve Trauma, Inflammation, Pain, Neuropathic, Arthritis, Rheumatoid, and Coronary Heart Disease, among others. Dilmapimod (SB-681323) is a p38 MAP-kinase inhibitor that has potential uses in inflammatory conditions such as RA (Rheumatoid Arthritis). Previous p38 MAP-kinase inhibitors have been hindered in development by liver toxicity. Methotrexate (common treatment for RA patients) also has potential liver toxicity.} Dilmapimod reduces the levels of proinflammatory cytokines and chemokines and reduce cellular infiltration to sites of inflammation, thereby reducing local damage.    Dilmapimod has been used in trials studying the treatment and diagnostic of Nerve Trauma, Inflammation, Pain, Neuropathic, Arthritis, Rheumatoid, and Coronary Heart Disease, among others.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Dilmapimod',\n",
       "   'enriched_node': 'CC1=CC(F)=CC=C1C1=C2C=CC(=O)N(C2=NC(NC(CO)CO)=N1)C1=C(F)C=CC=C1F',\n",
       "   'color': <NA>}),\n",
       " ('inflammatory bowel disease_(28158)',\n",
       "  {'desc': 'inflammatory bowel disease belongs to disease node. Any inflammatory bowel disease in which the cause of the disease is a mutation in the NOD2 gene.  ',\n",
       "   'node_type': 'disease',\n",
       "   'node_name': 'inflammatory bowel disease',\n",
       "   'enriched_node': 'Any inflammatory bowel disease in which the cause of the disease is a mutation in the NOD2 gene.',\n",
       "   'color': <NA>}),\n",
       " ('IL23R_(34778)',\n",
       "  {'desc': 'IL23R belongs to gene/protein node. IL23R is interleukin 23 receptor. The protein encoded by this gene is a subunit of the receptor for IL23A/IL23. This protein pairs with the receptor molecule IL12RB1/IL12Rbeta1, and both are required for IL23A signaling. This protein associates constitutively with Janus kinase 2 (JAK2), and also binds to transcription activator STAT3 in a ligand-dependent manner. [provided by RefSeq, Jul 2008].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'IL23R',\n",
       "   'enriched_node': 'MNQVTIQWDAVIALYILFSWCHGGITNINCSGHIWVEPATIFKMGMNISIYCQAAIKNCQPRKLHFYKNGIKERFQITRINKTTARLWYKNFLEPHASMYCTAECPKHFQETLICGKDISSGYPPDIPDEVTCVIYEYSGNMTCTWNAGKLTYIDTKYVVHVKSLETEEEQQYLTSSYINISTDSLQGGKKYLVWVQAANALGMEESKQLQIHLDDIVIPSAAVISRAETINATVPKTIIYWDSQTTIEKVSCEMRYKATTNQTWNVKEFDTNFTYVQQSEFYLEPNIKYVFQVRCQETGKRYWQPWSSLFFHKTPETVPQVTSKAFQHDTWNSGLTVASISTGHLTSDNRGDIGLLLGMIVFAVMLSILSLIGIFNRSFRTGIKRRILLLIPKWLYEDIPNMKNSNVVKMLQENSELMNNNSSEQVLYVDPMITEIKEIFIPEHKPTDYKKENTGPLETRDYPQNSLFDNTTVVYIPDLNTGYKPQISNFLPEGSHLSNNNEITSLTLKPPVDSLDSGNNPRLQKHPNFAFSVSSVNSLSNTIFLGELSLILNQGECSSPDIQNSVEEETTMLLENDSPSETIPEQTLLPDEFVSCLGIVNEELPSINTYFPQNILESHFNRISLLEK',\n",
       "   'color': <NA>}),\n",
       " ('NKX2-3_(34779)',\n",
       "  {'desc': 'NKX2-3 belongs to gene/protein node. NKX2-3 is NK2 homeobox 3. This gene encodes a homeodomain-containing transcription factor. The encoded protein is a member of the NKX family of homeodomain transcription factors. Studies of similar proteins in mouse and rat have indicated a potential role in cellular differentiation.[provided by RefSeq, Mar 2010].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'NKX2-3',\n",
       "   'enriched_node': 'MMLPSPVTSTPFSVKDILNLEQQHQHFHGAHLQADLEHHFHSAPCMLAAAEGTQFSDGGEEDEEDEGEKLSYLNSLAAADGHGDSGLCPQGYVHTVLRDSCSEPKEHEEEPEVVRDRSQKSCQLKKSLETAGDCKAAEESERPKPRSRRKPRVLFSQAQVFELERRFKQQRYLSAPEREHLASSLKLTSTQVKIWFQNRRYKCKRQRQDKSLELGAHAPPPPPRRVAVPVLVRDGKPCVTPSAQAYGAPYSVGASAYSYNSFPAYGYGNSAAAAAAAAAAAAAAAAYSSSYGCAYPAGGGGGGGGTSAATTAMQPACSAAGGGPFVNVSNLGGFGSGGSAQPLHQGTAAGAACAQGTLQGIRAW',\n",
       "   'color': <NA>}),\n",
       " ('TAGAP_(34814)',\n",
       "  {'desc': 'TAGAP belongs to gene/protein node. TAGAP is T cell activation RhoGTPase activating protein. This gene encodes a member of the Rho GTPase-activator protein superfamily. The encoded protein may function as a Rho GTPase-activating protein. Alterations in this gene may be associated with several diseases, including rheumatoid arthritis, celiac disease, and multiple sclerosis. Alternate splicing results in multiple transcript variants encoding distinct isoforms. [provided by RefSeq, Jul 2013].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'TAGAP',\n",
       "   'enriched_node': 'MKLRSSHNASKTLNANNMETLIECQSEGDIKEHPLLASCESEDSICQLIEVKKRKKVLSWPFLMRRLSPASDFSGALETDLKASLFDQPLSIICGDSDTLPRPIQDILTILCLKGPSTEGIFRRAANEKARKELKEELNSGDAVDLERLPVHLLAVVFKDFLRSIPRKLLSSDLFEEWMGALEMQDEEDRIEALKQVADKLPRPNLLLLKHLVYVLHLISKNSEVNRMDSSNLAICIGPNMLTLENDQSLSFEAQKDLNNKVKTLVEFLIDNCFEIFGENIPVHSSITSDDSLEHTDSSDVSTLQNDSAYDSNDPDVESNSSSGISSPSRQPQVPMATAAGLDSAGPQDAREVSPEPIVSTVARLKSSLAQPDRRYSEPSMPSSQECLESRVTNQTLTKSEGDFPVPRVGSRLESEEAEDPFPEEVFPAVQGKTKRPVDLKIKNLAPGSVLPRALVLKAFSSSSLDASSDSSPVASPSSPKRNFFSRHQSFTTKTEKGKPSREIKKHSMSFTFAPHKKVLTKNLSAGSGKSQDFTRDHVPRGVRKESQLAGRIVQENGCETHNQTARGFCLRPHALSVDDVFQGADWERPGSPPSYEEAMQGPAARLVASESQTVGSMTVGSMRARMLEAHCLLPPLPPAHHVEDSRHRGSKEPLPGHGLSPLPERWKQSRTVHASGDSLGHVSGPGRPELLPLRTVSESVQRNKRDCLVRRCSQPVFEADQFQYAKESYI',\n",
       "   'color': <NA>}),\n",
       " ('DENND1B_(34887)',\n",
       "  {'desc': 'DENND1B belongs to gene/protein node. DENND1B is DENN domain containing 1B. Clathrin (see MIM 118955)-mediated endocytosis is a major mechanism for internalization of proteins and lipids. Members of the connecdenn family, such as DENND1B, function as guanine nucleotide exchange factors (GEFs) for the early endosomal small GTPase RAB35 (MIM 604199) and bind to clathrin and clathrin adaptor protein-2 (AP2; see MIM 601024). Thus, connecdenns link RAB35 activation with the clathrin machinery (Marat and McPherson, 2010 [PubMed 20154091]).[supplied by OMIM, Nov 2010].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'DENND1B',\n",
       "   'enriched_node': 'MDCRTKANPDRTFDLVLKVKCHASENEDPVVLWKFPEDFGDQEILQSVPKFCFPFDVERVSQNQVGQHFTFVLTDIESKQRFGFCRLTSGGTICLCILSYLPWFEVYYKLLNTLADYLAKELENDLNETLRSLYNHPVPKANTPVNLSVNQEIFIACEQVLKDQPALVPHSYFIAPDVTGLPTIPESRNLTEYFVAVDVNNMLQLYASMLHERRIVIISSKLSTLTACIHGSAALLYPMYWQHIYIPVLPPHLLDYCCAPMPYLIGIHSSLIERVKNKSLEDVVMLNVDTNTLESPFSDLNNLPSDVVSALKNKLKKQSTATGDGVARAFLRAQAALFGSYRDALRYKPGEPITFCEESFVKHRSSVMKQFLETAINLQLFKQFIDGRLAKLNAGRGFSDVFEEEITSGGFCGGNPRSYQQWVHTVKKGGALFNTAMTKATPAVRTAYKFAKNHAKLGLKEVKSKLKHKENEEDYGTCSSSVQYTPVYKLHNEKGGNSEKRKLAQARLKRPLKSLDGALYDDEDDDDIERASKLSSEDGEEASAYLYESDDSVETRVKTPYSGEMDLLGEILDTLSTHSSDQGKLAAAKSLDFFRSMDDIDYKPTNKSNAPSENNLAFLCGGSGDQAEWNLGQDDSALHGKHLPPSPRKRVSSSGLTDSLFILKEENSNKHLGADNVSDPTSGLDFQLTSPEVSQTDKGKTEKRETLSQISDDLLIPGLGRHSSTFVPWEKEGKEAKETSEDIGLLHEVVSLCHMTSDFQQSLNISDKNTNGNQT',\n",
       "   'color': <NA>}),\n",
       " ('Crohn disease_(37784)',\n",
       "  {'desc': \"Crohn disease belongs to disease node. A gastrointestinal disorder characterized by chronic inflammation involving all layers of the intestinal wall, noncaseating granulomas affecting the intestinal wall and regional lymph nodes, and transmural fibrosis. Crohn disease most commonly involves the terminal ileum; the colon is the second most common site of involvement. In Crohn's disease, any part of your small or large intestine can be involved, and it may be continuous or may involve multiple segments. In some people, the disease is confined to the colon, which is part of the large intestine. Signs and symptoms of Crohn's disease can range from mild to severe. They usually develop gradually, but sometimes will come on suddenly, without warning. You may also have periods of time when you have no signs or symptoms . When the disease is active, signs and symptoms may include: Diarrhea, Fever, Fatigue, Abdominal pain and cramping, Blood in your stool, Mouth sores, Reduced appetite and weight loss, Pain or drainage near or around the anus due to inflammation from a tunnel into the skin, Other signs and symptoms, People with severe Crohn's disease may also experience: Inflammation of skin, eyes and joints, Inflammation of the liver or bile ducts, Kidney stones, Iron deficiency, Delayed growth or sexual development, in children The exact cause of Crohn's disease remains unknown. Previously, diet and stress were suspected, but now doctors know that these factors may aggravate, but don't cause, Crohn's disease. Several factors, such as heredity and a malfunctioning immune system, likely play a role in its development. Immune system. It's possible that a virus or bacterium may trigger Crohn's disease; however, scientists have yet to identify such a trigger. When your immune system tries to fight off the invading microorganism, an abnormal immune response causes the immune system to attack the cells in the digestive tract, too. Heredity. Crohn's is more common in people who have family members with the disease, so genes may play a role in making people more susceptible. However, most people with Crohn's disease don't have a family history of the disease.\",\n",
       "   'node_type': 'disease',\n",
       "   'node_name': 'Crohn disease',\n",
       "   'enriched_node': 'A gastrointestinal disorder characterized by chronic inflammation involving all layers of the intestinal wall, noncaseating granulomas affecting the intestinal wall and regional lymph nodes, and transmural fibrosis. Crohn disease most commonly involves the terminal ileum; the colon is the second most common site of involvement.',\n",
       "   'color': <NA>}),\n",
       " ('ulcerative colitis (disease)_(37785)',\n",
       "  {'desc': \"ulcerative colitis (disease) belongs to disease node. An inflammatory bowel disease involving the mucosal surface of the large intestine and rectum. It may present with an acute or slow onset and follows an intermittent or continuous course. Signs and symptoms include abdominal pain, diarrhea, fever, weight loss, and intestinal hemorrhage. Ulcerative colitis symptoms can vary, depending on the severity of inflammation and where it occurs. Signs and symptoms may include: Diarrhea, often with blood or pus, Abdominal pain and cramping, Rectal pain, Rectal bleeding  passing small amount of blood with stool, Urgency to defecate, Inability to defecate despite urgency, Weight loss, Fatigue, Fever, In children, failure to grow, Most people with ulcerative colitis have mild to moderate symptoms. The course of ulcerative colitis may vary, with some people having long periods of remission. Types, Doctors often classify ulcerative colitis according to its location. Types of ulcerative colitis include: Ulcerative proctitis. Inflammation is confined to the area closest to the anus, and rectal bleeding may be the only sign of the disease. Proctosigmoiditis. Inflammation involves the rectum and sigmoid colon  the lower end of the colon. Signs and symptoms include bloody diarrhea, abdominal cramps and pain, and an inability to move the bowels in spite of the urge to do so . Left-sided colitis. Inflammation extends from the rectum up through the sigmoid and descending colon. Signs and symptoms include bloody diarrhea, abdominal cramping and pain on the left side, and urgency to defecate. Pancolitis. This type often affects the entire colon and causes bouts of bloody diarrhea that may be severe, abdominal cramps and pain, fatigue, and significant weight loss. The exact cause of ulcerative colitis remains unknown. Previously, diet and stress were suspected, but now doctors know that these factors may aggravate but don't cause ulcerative colitis. One possible cause is an immune system malfunction. When your immune system tries to fight off an invading virus or bacterium, an abnormal immune response causes the immune system to attack the cells in the digestive tract, too. Heredity also seems to play a role in that ulcerative colitis is more common in people who have family members with the disease. However, most people with ulcerative colitis don't have this family history.\",\n",
       "   'node_type': 'disease',\n",
       "   'node_name': 'ulcerative colitis (disease)',\n",
       "   'enriched_node': 'An inflammatory bowel disease involving the mucosal surface of the large intestine and rectum. It may present with an acute or slow onset and follows an intermittent or continuous course. Signs and symptoms include abdominal pain, diarrhea, fever, weight loss, and intestinal hemorrhage.',\n",
       "   'color': <NA>}),\n",
       " ('immune response_(39995)',\n",
       "  {'desc': 'immune response belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'immune response',\n",
       "   'enriched_node': 'Any immune system process that functions in the calibrated response of an organism to a potential internal or invasive threat.',\n",
       "   'color': <NA>}),\n",
       " ('regulation of DNA recombination_(43811)',\n",
       "  {'desc': 'regulation of DNA recombination belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'regulation of DNA recombination',\n",
       "   'enriched_node': 'Any process that modulates the frequency, rate or extent of DNA recombination, a DNA metabolic process in which a new genotype is formed by reassortment of genes resulting in gene combinations different from those that were present in the parents.',\n",
       "   'color': <NA>}),\n",
       " ('regulation of gamma-delta T cell differentiation_(45624)',\n",
       "  {'desc': 'regulation of gamma-delta T cell differentiation belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'regulation of gamma-delta T cell differentiation',\n",
       "   'enriched_node': 'Any process that modulates the frequency, rate or extent of gamma-delta T cell differentiation.',\n",
       "   'color': <NA>}),\n",
       " ('cell surface receptor signaling pathway_(46168)',\n",
       "  {'desc': 'cell surface receptor signaling pathway belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'cell surface receptor signaling pathway',\n",
       "   'enriched_node': 'The series of molecular signals initiated by activation of a receptor on the surface of a cell. The pathway begins with binding of an extracellular ligand to a cell surface receptor, or for receptors that signal in the absence of a ligand, by ligand-withdrawal or the activity of a constitutively active receptor. The pathway ends with regulation of a downstream cellular process, e.g. transcription.',\n",
       "   'color': <NA>}),\n",
       " ('canonical Wnt signaling pathway_(46706)',\n",
       "  {'desc': 'canonical Wnt signaling pathway belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'canonical Wnt signaling pathway',\n",
       "   'enriched_node': 'The series of molecular signals initiated by binding of a Wnt protein to a frizzled family receptor on the surface of the target cell, followed by propagation of the signal via beta-catenin, and ending with a change in transcription of target genes. In this pathway, the activated receptor signals via downstream effectors that result in the inhibition of beta-catenin phosphorylation, thereby preventing degradation of beta-catenin. Stabilized beta-catenin can then accumulate and travel to the nucleus to trigger changes in transcription of target genes.',\n",
       "   'color': <NA>}),\n",
       " ('positive regulation of receptor signaling pathway via STAT_(51389)',\n",
       "  {'desc': 'positive regulation of receptor signaling pathway via STAT belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'positive regulation of receptor signaling pathway via STAT',\n",
       "   'enriched_node': 'Any process that activates or increases the frequency, rate or extent of receptor signaling pathway via STAT.',\n",
       "   'color': <NA>}),\n",
       " ('gamma-delta T cell differentiation_(51532)',\n",
       "  {'desc': 'gamma-delta T cell differentiation belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'gamma-delta T cell differentiation',\n",
       "   'enriched_node': 'The process in which a relatively unspecialized hemopoietic cell acquires specialized features of a gamma-delta T cell. A gamma-delta T cell is a T cell that expresses a gamma-delta T cell receptor complex.',\n",
       "   'color': <NA>}),\n",
       " ('positive regulation of peptidyl-serine phosphorylation_(52659)',\n",
       "  {'desc': 'positive regulation of peptidyl-serine phosphorylation belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'positive regulation of peptidyl-serine phosphorylation',\n",
       "   'enriched_node': 'Any process that activates or increases the frequency, rate or extent of the phosphorylation of peptidyl-serine.',\n",
       "   'color': <NA>}),\n",
       " ('transcription cis-regulatory region binding_(54288)',\n",
       "  {'desc': 'transcription cis-regulatory region binding belongs to molecular_function node. ',\n",
       "   'node_type': 'molecular_function',\n",
       "   'node_name': 'transcription cis-regulatory region binding',\n",
       "   'enriched_node': 'Binding to a specific sequence of DNA that is part of a regulatory region that controls transcription of that section of the DNA. The transcribed region might be described as a gene, cistron, or operon.',\n",
       "   'color': <NA>}),\n",
       " ('DNA-binding transcription repressor activity_(54389)',\n",
       "  {'desc': 'DNA-binding transcription repressor activity belongs to molecular_function node. ',\n",
       "   'node_type': 'molecular_function',\n",
       "   'node_name': 'DNA-binding transcription repressor activity',\n",
       "   'enriched_node': 'A DNA-binding transcription factor activity that represses or decreases the transcription of specific gene sets.',\n",
       "   'color': <NA>}),\n",
       " ('cytokine receptor activity_(54621)',\n",
       "  {'desc': 'cytokine receptor activity belongs to molecular_function node. ',\n",
       "   'node_type': 'molecular_function',\n",
       "   'node_name': 'cytokine receptor activity',\n",
       "   'enriched_node': 'Combining with a cytokine and transmitting the signal from one side of the membrane to the other to initiate a change in cell activity.',\n",
       "   'color': <NA>}),\n",
       " ('beta-catenin-TCF complex_(56282)',\n",
       "  {'desc': 'beta-catenin-TCF complex belongs to cellular_component node. ',\n",
       "   'node_type': 'cellular_component',\n",
       "   'node_name': 'beta-catenin-TCF complex',\n",
       "   'enriched_node': 'A protein complex that contains beta-catenin and a member of the T-cell factor (TCF)/lymphoid enhancer binding factor (LEF) family of transcription factors.',\n",
       "   'color': <NA>}),\n",
       " ('beta-catenin-TCF complex assembly_(104150)',\n",
       "  {'desc': 'beta-catenin-TCF complex assembly belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'beta-catenin-TCF complex assembly',\n",
       "   'enriched_node': 'The aggregation, arrangement and bonding together of a set of components to form a beta-catenin-TCF complex.',\n",
       "   'color': <NA>}),\n",
       " ('lymph node development_(107899)',\n",
       "  {'desc': 'lymph node development belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'lymph node development',\n",
       "   'enriched_node': 'The process whose specific outcome is the progression of lymph nodes over time, from their formation to the mature structure. A lymph node is a round, oval, or bean shaped structure localized in clusters along the lymphatic vessels, with a distinct internal structure including specialized vasculature and B- and T-zones for the activation of lymphocytes.',\n",
       "   'color': <NA>}),\n",
       " ('T cell receptor V(D)J recombination_(108222)',\n",
       "  {'desc': 'T cell receptor V(D)J recombination belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'T cell receptor V(D)J recombination',\n",
       "   'enriched_node': 'The process in which T cell receptor V, D, and J, or V and J gene segments, depending on the specific locus, are recombined within a single locus utilizing the conserved heptamer and nonomer recombination signal sequences (RSS).',\n",
       "   'color': <NA>}),\n",
       " ('interleukin-7-mediated signaling pathway_(108951)',\n",
       "  {'desc': 'interleukin-7-mediated signaling pathway belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'interleukin-7-mediated signaling pathway',\n",
       "   'enriched_node': 'The series of molecular signals initiated by interleukin-7 binding to its receptor on the surface of a cell, and ending with the regulation of a downstream cellular process, e.g. transcription.',\n",
       "   'color': <NA>}),\n",
       " ('beta-catenin binding_(119055)',\n",
       "  {'desc': 'beta-catenin binding belongs to molecular_function node. ',\n",
       "   'node_type': 'molecular_function',\n",
       "   'node_name': 'beta-catenin binding',\n",
       "   'enriched_node': 'Binding to a catenin beta subunit.',\n",
       "   'color': <NA>}),\n",
       " ('interleukin-7 receptor activity_(122046)',\n",
       "  {'desc': 'interleukin-7 receptor activity belongs to molecular_function node. ',\n",
       "   'node_type': 'molecular_function',\n",
       "   'node_name': 'interleukin-7 receptor activity',\n",
       "   'enriched_node': 'Combining with interleukin-7 and transmitting the signal from one side of the membrane to the other to initiate a change in cell activity.',\n",
       "   'color': <NA>}),\n",
       " ('nucleoplasm_(124249)',\n",
       "  {'desc': 'nucleoplasm belongs to cellular_component node. ',\n",
       "   'node_type': 'cellular_component',\n",
       "   'node_name': 'nucleoplasm',\n",
       "   'enriched_node': 'That part of the nuclear content other than the chromosomes or the nucleolus.',\n",
       "   'color': <NA>}),\n",
       " ('IL7R_(625)',\n",
       "  {'desc': 'IL7R belongs to gene/protein node. IL7R is interleukin 7 receptor. The protein encoded by this gene is a receptor for interleukin 7 (IL7). The function of this receptor requires the interleukin 2 receptor, gamma chain (IL2RG), which is a common gamma chain shared by the receptors of various cytokines, including interleukins 2, 4, 7, 9, and 15. This protein has been shown to play a critical role in V(D)J recombination during lymphocyte development. Defects in this gene may be associated with severe combined immunodeficiency (SCID). Alternatively spliced transcript variants have been found. [provided by RefSeq, Dec 2015].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'IL7R',\n",
       "   'enriched_node': 'MTILGTTFGMVFSLLQVVSGESGYAQNGDLEDAELDDYSFSCYSQLEVNGSQHSLTCAFEDPDVNITNLEFEICGALVEVKCLNFRKLQEIYFIETKKFLLIGKSNICVKVGEKSLTCKKIDLTTIVKPEAPFDLSVVYREGANDFVVTFNTSHLQKKYVKVLMHDVAYRQEKDENKWTHVNLSSTKLTLLQRKLQPAAMYEIKVRSIPDHYFKGFWSEWSPSYYFRTPEINNSSGEMDPILLTISILSFFSVALLVILACVLWKKRIKPIVWPSLPDHKKTLEHLCKKPRKNLNVSFNPESFLDCQIHRVDDIQARDEVEGFLQDTFPQQLEESEKQRLGGDVQSPNCPSEDVVITPESFGRDSSLTCLAGNVSACDAPILSSSRSLDCRESGKNGPHVYQDLLLSLGTTNSTLPPPFSLQSGILTLNPVAQGQPILTSLGSNQEEAYVTMSSFYQNQ',\n",
       "   'color': '#6a79f7'}),\n",
       " ('TCF7_(5195)',\n",
       "  {'desc': 'TCF7 belongs to gene/protein node. TCF7 is transcription factor 7. This gene encodes a member of the T-cell factor/lymphoid enhancer-binding factor family of high mobility group (HMG) box transcriptional activators. This gene is expressed predominantly in T-cells and plays a critical role in natural killer cell and innate lymphoid cell development. The encoded protein forms a complex with beta-catenin and activates transcription through a Wnt/beta-catenin signaling pathway. Mice with a knockout of this gene are viable and fertile, but display a block in T-lymphocyte differentiation. Alternative splicing results in multiple transcript variants. Naturally-occurring isoforms lacking the N-terminal beta-catenin interaction domain may act as dominant negative regulators of Wnt signaling. [provided by RefSeq, Oct 2016].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'TCF7',\n",
       "   'enriched_node': 'MPQLDSGGGGAGGGDDLGAPDELLAFQDEGEEQDDKSRDSAAGPERDLAELKSSLVNESEGAAGGAGIPGVPGAGAGARGEAEALGREHAAQRLFPDKLPEPLEDGLKAPECTSGMYKETVYSAFNLLMHYPPPSGAGQHPQPQPPLHKANQPPHGVPQLSLYEHFNSPHPTPAPADISQKQVHRPLQTPDLSGFYSLTSGSMGQLPHTVSWFTHPSLMLGSGVPGHPAAIPHPAIVPPSGKQELQPFDRNLKTQAESKAEKEAKKPTIKKPLNAFMLYMKEMRAKVIAECTLKESAAINQILGRRWHALSREEQAKYYELARKERQLHMQLYPGWSARDNYGKKKRRSREKHQESTTETNWPRELKDGNGQESLSMSSSSSPA',\n",
       "   'color': '#6a79f7'}),\n",
       " ('HNF4A_(279)',\n",
       "  {'desc': 'HNF4A belongs to gene/protein node. HNF4A is hepatocyte nuclear factor 4 alpha. The protein encoded by this gene is a nuclear transcription factor which binds DNA as a homodimer. The encoded protein controls the expression of several genes, including hepatocyte nuclear factor 1 alpha, a transcription factor which regulates the expression of several hepatic genes. This gene may play a role in development of the liver, kidney, and intestines. Mutations in this gene have been associated with monogenic autosomal dominant non-insulin-dependent diabetes mellitus type I. Alternative splicing of this gene results in multiple transcript variants encoding several different isoforms. [provided by RefSeq, Apr 2012].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'HNF4A',\n",
       "   'enriched_node': 'MRLSKTLVDMDMADYSAALDPAYTTLEFENVQVLTMGNDTSPSEGTNLNAPNSLGVSALCAICGDRATGKHYGASSCDGCKGFFRRSVRKNHMYSCRFSRQCVVDKDKRNQCRYCRLKKCFRAGMKKEAVQNERDRISTRRSSYEDSSLPSINALLQAEVLSRQITSPVSGINGDIRAKKIASIADVCESMKEQLLVLVEWAKYIPAFCELPLDDQVALLRAHAGEHLLLGATKRSMVFKDVLLLGNDYIVPRHCPELAEMSRVSIRILDELVLPFQELQIDDNEYAYLKAIIFFDPDAKGLSDPGKIKRLRSQVQVSLEDYINDRQYDSRGRFGELLLLLPTLQSITWQMIEQIQFIKLFGMAKIDNLLQEMLLGGSPSDAPHAHHPLHPHLMQEHMGTNVIVANTMPTHLSNGQMCEWPRPRGQAATPETPQPSPPGGSGSEPYKLLPGAVATIVKPLSAIPQPTITKQEVI',\n",
       "   'color': <NA>}),\n",
       " ('PPARG_(989)',\n",
       "  {'desc': 'PPARG belongs to gene/protein node. PPARG is peroxisome proliferator activated receptor gamma. This gene encodes a member of the peroxisome proliferator-activated receptor (PPAR) subfamily of nuclear receptors. PPARs form heterodimers with retinoid X receptors (RXRs) and these heterodimers regulate transcription of various genes. Three subtypes of PPARs are known: PPAR-alpha, PPAR-delta, and PPAR-gamma. The protein encoded by this gene is PPAR-gamma and is a regulator of adipocyte differentiation. Additionally, PPAR-gamma has been implicated in the pathology of numerous diseases including obesity, diabetes, atherosclerosis and cancer. Alternatively spliced transcript variants that encode different isoforms have been described. [provided by RefSeq, Jul 2008].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'PPARG',\n",
       "   'enriched_node': 'MGETLGDSPIDPESDSFTDTLSANISQEMTMVDTEMPFWPTNFGISSVDLSVMEDHSHSFDIKPFTTVDFSSISTPHYEDIPFTRTDPVVADYKYDLKLQEYQSAIKVEPASPPYYSEKTQLYNKPHEEPSNSLMAIECRVCGDKASGFHYGVHACEGCKGFFRRTIRLKLIYDRCDLNCRIHKKSRNKCQYCRFQKCLAVGMSHNAIRFGRMPQAEKEKLLAEISSDIDQLNPESADLRALAKHLYDSYIKSFPLTKAKARAILTGKTTDKSPFVIYDMNSLMMGEDKIKFKHITPLQEQSKEVAIRIFQGCQFRSVEAVQEITEYAKSIPGFVNLDLNDQVTLLKYGVHEIIYTMLASLMNKDGVLISEGQGFMTREFLKSLRKPFGDFMEPKFEFAVKFNALELDDSDLAIFIAVIILSGDRPGLLNVKPIEDIQDNLLQALELQLKLNHPESSQLFAKLLQKMTDLRQIVTEHVQLLQVIKKTETDMSLHPLLQEIYKDLY',\n",
       "   'color': <NA>}),\n",
       " ('IL10RA_(1299)',\n",
       "  {'desc': 'IL10RA belongs to gene/protein node. IL10RA is interleukin 10 receptor subunit alpha. The protein encoded by this gene is a receptor for interleukin 10. This protein is structurally related to interferon receptors. It has been shown to mediate the immunosuppressive signal of interleukin 10, and thus inhibits the synthesis of proinflammatory cytokines. This receptor is reported to promote survival of progenitor myeloid cells through the insulin receptor substrate-2/PI 3-kinase/AKT pathway. Activation of this receptor leads to tyrosine phosphorylation of JAK1 and TYK2 kinases. Two transcript variants, one protein-coding and the other not protein-coding, have been found for this gene. [provided by RefSeq, Jan 2009].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'IL10RA',\n",
       "   'enriched_node': 'MLPCLVVLLAALLSLRLGSDAHGTELPSPPSVWFEAEFFHHILHWTPIPNQSESTCYEVALLRYGIESWNSISNCSQTLSYDLTAVTLDLYHSNGYRARVRAVDGSRHSNWTVTNTRFSVDEVTLTVGSVNLEIHNGFILGKIQLPRPKMAPANDTYESIFSHFREYEIAIRKVPGNFTFTHKKVKHENFSLLTSGEVGEFCVQVKPSVASRSNKGMWSKEECISLTRQYFTVTNVIIFFAFVLLLSGALAYCLALQLYVRRRKKLPSVLLFKKPSPFIFISQRPSPETQDTIHPLDEEAFLKVSPELKNLDLHGSTDSGFGSTKPSLQTEEPQFLLPDPHPQADRTLGNREPPVLGDSCSSGSSNSTDSGICLQEPSLSPSTGPTWEQQVGSNSRGQDDSGIDLVQNSEGRAGDTQGGSALGHHSPPEPEVPGEEDPAAVAFQGYLRQTRCAEEKATKTGCLEEESPLTDGLGPKFGRCLVDEAGLHPPALAKGYLKQDPLEMTLASSGAPTGQWNQPTEEWSLLALSSCSDLGISDWSFAHDLAPLGCVAAPGGLLGSFNSDLVTLPLISSLQSSE',\n",
       "   'color': <NA>}),\n",
       " ('JAK2_(1618)',\n",
       "  {'desc': 'JAK2 belongs to gene/protein node. JAK2 is Janus kinase 2. This gene encodes a non-receptor tyrosine kinase that plays a central role in cytokine and growth factor signalling. The primary isoform of this protein has an N-terminal FERM domain that is required for erythropoietin receptor association, an SH2 domain that binds STAT transcription factors, a pseudokinase domain and a C-terminal tyrosine kinase domain. Cytokine binding induces autophosphorylation and activation of this kinase. This kinase then recruits and phosphorylates signal transducer and activator of transcription (STAT) proteins. Growth factors like TGF-beta 1 also induce phosphorylation and activation of this kinase and translocation of downstream STAT proteins to the nucleus where they influence gene transcription. Mutations in this gene are associated with numerous inflammatory diseases and malignancies. This gene is a downstream target of the pleiotropic cytokine IL6 that is produced by B cells, T cells, dendritic cells and macrophages to produce an immune response or inflammation. Disregulation of the IL6/JAK2/STAT3 signalling pathways produces increased cellular proliferation and myeloproliferative neoplasms of hematopoietic stem cells. A nonsynonymous mutation in the pseudokinase domain of this gene disrupts the domains inhibitory effect and results in constitutive tyrosine phosphorylation activity and hypersensitivity to cytokine signalling. This gene and the IL6/JAK2/STAT3 signalling pathway is a therapeutic target for the treatment of excessive inflammatory responses to viral infections. Alternative splicing results in multiple transcript variants encoding distinct isoforms. [provided by RefSeq, Jul 2020].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'JAK2',\n",
       "   'enriched_node': 'MGMACLTMTEMEGTSTSSIYQNGDISGNANSMKQIDPVLQVYLYHSLGKSEADYLTFPSGEYVAEEICIAASKACGITPVYHNMFALMSETERIWYPPNHVFHIDESTRHNVLYRIRFYFPRWYCSGSNRAYRHGISRGAEAPLLDDFVMSYLFAQWRHDFVHGWIKVPVTHETQEECLGMAVLDMMRIAKENDQTPLAIYNSISYKTFLPKCIRAKIQDYHILTRKRIRYRFRRFIQQFSQCKATARNLKLKYLINLETLQSAFYTEKFEVKEPGSGPSGEEIFATIIITGNGGIQWSRGKHKESETLTEQDLQLYCDFPNIIDVSIKQANQEGSNESRVVTIHKQDGKNLEIELSSLREALSFVSLIDGYYRLTADAHHYLCKEVAPPAVLENIQSNCHGPISMDFAISKLKKAGNQTGLYVLRCSPKDFNKYFLTFAVERENVIEYKHCLITKNENEEYNLSGTKKNFSSLKDLLNCYQMETVRSDNIIFQFTKCCPPKPKDKSNLLVFRTNGVSDVPTSPTLQRPTHMNQMVFHKIRNEDLIFNESLGQGTFTKIFKGVRREVGDYGQLHETEVLLKVLDKAHRNYSESFFEAASMMSKLSHKHLVLNYGVCVCGDENILVQEFVKFGSLDTYLKKNKNCINILWKLEVAKQLAWAMHFLEENTLIHGNVCAKNILLIREEDRKTGNPPFIKLSDPGISITVLPKDILQERIPWVPPECIENPKNLNLATDKWSFGTTLWEICSGGDKPLSALDSQRKLQFYEDRHQLPAPKWAELANLINNCMDYEPDFRPSFRAIIRDLNSLFTPDYELLTENDMLPNMRIGALGFSGAFEDRDPTQFEERHLKFLQQLGKGNFGSVEMCRYDPLQDNTGEVVAVKKLQHSTEEHLRDFEREIEILKSLQHDNIVKYKGVCYSAGRRNLKLIMEYLPYGSLRDYLQKHKERIDHIKLLQYTSQICKGMEYLGTKRYIHRDLATRNILVENENRVKIGDFGLTKVLPQDKEYYKVKEPGESPIFWYAPESLTESKFSVASDVWSFGVVLYELFTYIEKSKSPPAEFMRMIGNDKQGQMIVFHLIELLKNNGRLPRPDGCPDEIYMIMTECWNNNVNQRPSFRDLALRVDQIRDNMAG',\n",
       "   'color': <NA>}),\n",
       " ('LRRK2_(2111)',\n",
       "  {'desc': 'LRRK2 belongs to gene/protein node. LRRK2 is leucine rich repeat kinase 2. This gene is a member of the leucine-rich repeat kinase family and encodes a protein with an ankryin repeat region, a leucine-rich repeat (LRR) domain, a kinase domain, a DFG-like motif, a RAS domain, a GTPase domain, a MLK-like domain, and a WD40 domain. The protein is present largely in the cytoplasm but also associates with the mitochondrial outer membrane. Mutations in this gene have been associated with Parkinson disease-8. [provided by RefSeq, Jul 2008].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'LRRK2',\n",
       "   'enriched_node': 'MASGSCQGCEEDEETLKKLIVRLNNVQEGKQIETLVQILEDLLVFTYSERASKLFQGKNIHVPLLIVLDSYMRVASVQQVGWSLLCKLIEVCPGTMQSLMGPQDVGNDWEVLGVHQLILKMLTVHNASVNLSVIGLKTLDLLLTSGKITLLILDEESDIFMLIFDAMHSFPANDEVQKLGCKALHVLFERVSEEQLTEFVENKDYMILLSALTNFKDEEEIVLHVLHCLHSLAIPCNNVEVLMSGNVRCYNIVVEAMKAFPMSERIQEVSCCLLHRLTLGNFFNILVLNEVHEFVVKAVQQYPENAALQISALSCLALLTETIFLNQDLEEKNENQENDDEGEEDKLFWLEACYKALTWHRKNKHVQEAACWALNNLLMYQNSLHEKIGDEDGHFPAHREVMLSMLMHSSSKEVFQASANALSTLLEQNVNFRKILLSKGIHLNVLELMQKHIHSPEVAESGCKMLNHLFEGSNTSLDIMAAVVPKILTVMKRHETSLPVQLEALRAILHFIVPGMPEESREDTEFHHKLNMVKKQCFKNDIHKLVLAALNRFIGNPGIQKCGLKVISSIVHFPDALEMLSLEGAMDSVLHTLQMYPDDQEIQCLGLSLIGYLITKKNVFIGTGHLLAKILVSSLYRFKDVAEIQTKGFQTILAILKLSASFSKLLVHHSFDLVIFHQMSSNIMEQKDQQFLNLCCKCFAKVAMDDYLKNVMLERACDQNNSIMVECLLLLGADANQAKEGSSLICQVCEKESSPKLVELLLNSGSREQDVRKALTISIGKGDSQIISLLLRRLALDVANNSICLGGFCIGKVEPSWLGPLFPDKTSNLRKQTNIASTLARMVIRYQMKSAVEEGTASGSDGNFSEDVLSKFDEWTFIPDSSMDSVFAQSDDLDSEGSEGSFLVKKKSNSISVGEFYRDAVLQRCSPNLQRHSNSLGPIFDHEDLLKRKRKILSSDDSLRSSKLQSHMRHSDSISSLASEREYITSLDLSANELRDIDALSQKCCISVHLEHLEKLELHQNALTSFPQQLCETLKSLTHLDLHSNKFTSFPSYLLKMSCIANLDVSRNDIGPSVVLDPTVKCPTLKQFNLSYNQLSFVPENLTDVVEKLEQLILEGNKISGICSPLRLKELKILNLSKNHISSLSENFLEACPKVESFSARMNFLAAMPFLPPSMTILKLSQNKFSCIPEAILNLPHLRSLDMSSNDIQYLPGPAHWKSLNLRELLFSHNQISILDLSEKAYLWSRVEKLHLSHNKLKEIPPEIGCLENLTSLDVSYNLELRSFPNEMGKLSKIWDLPLDELHLNFDFKHIGCKAKDIIRFLQQRLKKAVPYNRMKLMIVGNTGSGKTTLLQQLMKTKKSDLGMQSATVGIDVKDWPIQIRDKRKRDLVLNVWDFAGREEFYSTHPHFMTQRALYLAVYDLSKGQAEVDAMKPWLFNIKARASSSPVILVGTHLDVSDEKQRKACMSKITKELLNKRGFPAIRDYHFVNATEESDALAKLRKTIINESLNFKIRDQLVVGQLIPDCYVELEKIILSERKNVPIEFPVIDRKRLLQLVRENQLQLDENELPHAVHFLNESGVLLHFQDPALQLSDLYFVEPKWLCKIMAQILTVKVEGCPKHPKGIISRRDVEKFLSKKRKFPKNYMSQYFKLLEKFQIALPIGEEYLLVPSSLSDHRPVIELPHCENSEIIIRLYEMPYFPMGFWSRLINRLLEISPYMLSGRERALRPNRMYWRQGIYLNWSPEAYCLVGSEVLDNHPESFLKITVPSCRKGCILLGQVVDHIDSLMEEWFPGLLEIDICGEGETLLKKWALYSFNDGEEHQKILLDDLMKKAEEGDLLVNPDQPRLTIPISQIAPDLILADLPRNIMLNNDELEFEQAPEFLLGDGSFGSVYRAAYEGEEVAVKIFNKHTSLRLLRQELVVLCHLHHPSLISLLAAGIRPRMLVMELASKGSLDRLLQQDKASLTRTLQHRIALHVADGLRYLHSAMIIYRDLKPHNVLLFTLYPNAAIIAKIADYGIAQYCCRMGIKTSEGTPGFRAPEVARGNVIYNQQADVYSFGLLLYDILTTGGRIVEGLKFPNEFDELEIQGKLPDPVKEYGCAPWPMVEKLIKQCLKENPQERPTSAQVFDILNSAELVCLTRRILLPKNVIVECMVATHHNSRNASIWLGCGHTDRGQLSFLDLNTEGYTSEEVADSRILCLALVHLPVEKESWIVSGTQSGTLLVINTEDGKKRHTLEKMTDSVTCLYCNSFSKQSKQKNFLLVGTADGKLAIFEDKTVKLKGAAPLKILNIGNVSTPLMCLSESTNSTERNVMWGGCGTKIFSFSNDFTIQKLIETRTSQLFSYAAFSDSNIITVVVDTALYIAKQNSPVVEVWDKKTEKLCGLIDCVHFLREVMVKENKESKHKMSYSGRVKTLCLQKNTALWIGTGGGHILLLDLSTRRLIRVIYNFCNSVRVMMTAQLGSLKNVMLVLGYNRKNTEGTQKQKEIQSCLTVWDINLPHEVQNLEKHIEVRKELAEKMRRTSVE',\n",
       "   'color': <NA>}),\n",
       " ('TNF_(2329)',\n",
       "  {'desc': 'TNF belongs to gene/protein node. TNF is tumor necrosis factor. This gene encodes a multifunctional proinflammatory cytokine that belongs to the tumor necrosis factor (TNF) superfamily. This cytokine is mainly secreted by macrophages. It can bind to, and thus functions through its receptors TNFRSF1A/TNFR1 and TNFRSF1B/TNFBR. This cytokine is involved in the regulation of a wide spectrum of biological processes including cell proliferation, differentiation, apoptosis, lipid metabolism, and coagulation. This cytokine has been implicated in a variety of diseases, including autoimmune diseases, insulin resistance, psoriasis, rheumatoid arthritis ankylosing spondylitis, tuberculosis, autosomal dominant polycystic kidney disease, and cancer. Mutations in this gene affect susceptibility to cerebral malaria, septic shock, and Alzheimer disease. Knockout studies in mice also suggested the neuroprotective function of this cytokine. [provided by RefSeq, Aug 2020].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'TNF',\n",
       "   'enriched_node': 'MSTESMIRDVELAEEALPKKTGGPQGSRRCLFLSLFSFLIVAGATTLFCLLHFGVIGPQREEFPRDLSLISPLAQAVRSSSRTPSDKPVAHVVANPQAEGQLQWLNRRANALLANGVELRDNQLVVPSEGLYLIYSQVLFKGQGCPSTHVLLTHTISRIAVSYQTKVNLLSAIKSPCQRETPEGAEAKPWYEPIYLGGVFQLEKGDRLSAEINRPDYLDFAESGQVYFGIIAL',\n",
       "   'color': <NA>}),\n",
       " ('LSP1_(2643)',\n",
       "  {'desc': 'LSP1 belongs to gene/protein node. LSP1 is lymphocyte specific protein 1. This gene encodes an intracellular F-actin binding protein. The protein is expressed in lymphocytes, neutrophils, macrophages, and endothelium and may regulate neutrophil motility, adhesion to fibrinogen matrix proteins, and transendothelial migration. Alternative splicing results in multiple transcript variants encoding different isoforms. [provided by RefSeq, Jul 2008].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'LSP1',\n",
       "   'enriched_node': 'MAEASSDPGAEEREELLGPTAQWSVEDEEEAVHEQCQHERDRQLQAQDEEGGGHVPERPKQEMLLSLKPSEAPELDEDEGFGDWSQRPEQRQQHEGAQGALDSGEPPQCRSPEGEQEDRPGLHAYEKEDSDEVHLEELSLSKEGPGPEDTVQDNLGAAGAEEEQEEHQKCQQPRTPSPLVLEGTIEQSSPPLSPTTKLIDRTESLNRSIEKSNSVKKSQPDLPISKIDQWLEQYTQAIETAGRTPKLARQASIELPSMAVASTKSRWETGEVQAQSAAKTPSCKDIVAGDMSKKSLWEQKGGSKTSSTIKSTPSGKRYKFVATGHGKYEKVLVEGGPAP',\n",
       "   'color': <NA>}),\n",
       " ('PRDM1_(2874)',\n",
       "  {'desc': 'PRDM1 belongs to gene/protein node. PRDM1 is PR/SET domain 1. This gene encodes a protein that acts as a repressor of beta-interferon gene expression. The protein binds specifically to the PRDI (positive regulatory domain I element) of the beta-IFN gene promoter. Transcription of this gene increases upon virus induction. Two alternatively spliced transcript variants that encode different isoforms have been reported. [provided by RefSeq, Jul 2008].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'PRDM1',\n",
       "   'enriched_node': 'MLDICLEKRVGTTLAAPKCNSSTVRFQGLAEGTKGTMKMDMEDADMTLWTEAEFEEKCTYIVNDHPWDSGADGGTSVQAEASLPRNLLFKYATNSEEVIGVMSKEYIPKGTRFGPLIGEIYTNDTVPKNANRKYFWRIYSRGELHHFIDGFNEEKSNWMRYVNPAHSPREQNLAACQNGMNIYFYTIKPIPANQELLVWYCRDFAERLHYPYPGELTMMNLTQTQSSLKQPSTEKNELCPKNVPKREYSVKEILKLDSNPSKGKDLYRSNISPLTSEKDLDDFRRRGSPEMPFYPRVVYPIRAPLPEDFLKASLAYGIERPTYITRSPIPSSTTPSPSARSSPDQSLKSSSPHSSPGNTVSPVGPGSQEHRDSYAYLNASYGTEGLGSYPGYAPLPHLPPAFIPSYNAHYPKFLLPPYGMNCNGLSAVSSMNGINNFGLFPRLCPVYSNLLGGGSLPHPMLNPTSLPSSLPSDGARRLLQPEHPREVLVPAPHSAFSFTGAAASMKDKACSPTSGSPTAGTAATAEHVVQPKATSAAMAAPSSDEAMNLIKNKRNMTGYKTLPYPLKKQNGKIKYECNVCAKTFGQLSNLKVHLRVHSGERPFKCQTCNKGFTQLAHLQKHYLVHTGEKPHECQVCHKRFSSTSNLKTHLRLHSGEKPYQCKVCPAKFTQFVHLKLHKRLHTRERPHKCSQCHKNYIHLCSLKVHLKGNCAAAPAPGLPLEDLTRINEEIEKFDISDNADRLEDVEDDISVISVVEKEILAVVRKEKEETGLKVSLQRNMGNGLLSSGCSLYESSDLPLMKLPPSNPLPLVPVKVKQETVEPMDP',\n",
       "   'color': <NA>}),\n",
       " ('CARD9_(3460)',\n",
       "  {'desc': 'CARD9 belongs to gene/protein node. CARD9 is caspase recruitment domain family member 9. The protein encoded by this gene is a member of the CARD protein family, which is defined by the presence of a characteristic caspase-associated recruitment domain (CARD). CARD is a protein interaction domain known to participate in activation or suppression of CARD containing members of the caspase family, and thus plays an important regulatory role in cell apoptosis. This protein was identified by its selective association with the CARD domain of BCL10, a postive regulator of apoptosis and NF-kappaB activation, and is thought to function as a molecular scaffold for the assembly of a BCL10 signaling complex that activates NF-kappaB. Several alternatively spliced transcript variants have been observed, but their full-length nature is not clearly defined. [provided by RefSeq, Jul 2008].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'CARD9',\n",
       "   'enriched_node': 'MSDYENDDECWSVLEGFRVTLTSVIDPSRITPYLRQCKVLNPDDEEQVLSDPNLVIRKRKVGVLLDILQRTGHKGYVAFLESLELYYPQLYKKVTGKEPARVFSMIIDASGESGLTQLLMTEVMKLQKKVQDLTALLSSKDDFIKELRVKDSLLRKHQERVQRLKEECEAGSRELKRCKEENYDLAMRLAHQSEEKGAALMRNRDLQLEIDQLKHSLMKAEDDCKVERKHTLKLRHAMEQRPSQELLWELQQEKALLQARVQELEASVQEGKLDRSSPYIQVLEEDWRQALRDHQEQANTIFSLRKDLRQGEARRLRCMEEKEMFELQCLALRKDSKMYKDRIEAILLQMEEVAIERDQAIATREELHAQHARGLQEKDALRKQVRELGEKADELQLQVFQCEAQLLAVEGRLRRQQLETLVLSSDLEDGSPRRSQELSLPQDLEDTQLSDKGCLAGGGSPKQPFAALHQEQVLRNPHDAGLSSGEPPEKERRRLKESFENYRRKRALRKMQKGWRQGEEDRENTTGSDNTDTEGS',\n",
       "   'color': <NA>}),\n",
       " ('MMP9_(3474)',\n",
       "  {'desc': \"MMP9 belongs to gene/protein node. MMP9 is matrix metallopeptidase 9. Proteins of the matrix metalloproteinase (MMP) family are involved in the breakdown of extracellular matrix in normal physiological processes, such as embryonic development, reproduction, and tissue remodeling, as well as in disease processes, such as arthritis and metastasis. Most MMP's are secreted as inactive proproteins which are activated when cleaved by extracellular proteinases. The enzyme encoded by this gene degrades type IV and V collagens. Studies in rhesus monkeys suggest that the enzyme is involved in IL-8-induced mobilization of hematopoietic progenitor cells from bone marrow, and murine studies suggest a role in tumor-associated tissue remodeling. [provided by RefSeq, Jul 2008].\",\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'MMP9',\n",
       "   'enriched_node': 'MSLWQPLVLVLLVLGCCFAAPRQRQSTLVLFPGDLRTNLTDRQLAEEYLYRYGYTRVAEMRGESKSLGPALLLLQKQLSLPETGELDSATLKAMRTPRCGVPDLGRFQTFEGDLKWHHHNITYWIQNYSEDLPRAVIDDAFARAFALWSAVTPLTFTRVYSRDADIVIQFGVAEHGDGYPFDGKDGLLAHAFPPGPGIQGDAHFDDDELWSLGKGVVVPTRFGNADGAACHFPFIFEGRSYSACTTDGRSDGLPWCSTTANYDTDDRFGFCPSERLYTQDGNADGKPCQFPFIFQGQSYSACTTDGRSDGYRWCATTANYDRDKLFGFCPTRADSTVMGGNSAGELCVFPFTFLGKEYSTCTSEGRGDGRLWCATTSNFDSDKKWGFCPDQGYSLFLVAAHEFGHALGLDHSSVPEALMYPMYRFTEGPPLHKDDVNGIRHLYGPRPEPEPRPPTTTTPQPTAPPTVCPTGPPTVHPSERPTAGPTGPPSAGPTGPPTAGPSTATTVPLSPVDDACNVNIFDAIAEIGNQLYLFKDGKYWRFSEGRGSRPQGPFLIADKWPALPRKLDSVFEERLSKKLFFFSGRQVWVYTGASVLGPRRLDKLGLGADVAQVTGALRSGRGKMLLFSGRRLWRFDVKAQMVDPRSASEVDRMFPGVPLDTHDVFQYREKAYFCQDRFYWRVSSRSELNQVDQVGYVTYDILQCPED',\n",
       "   'color': <NA>}),\n",
       " ('ABCB1_(4152)',\n",
       "  {'desc': 'ABCB1 belongs to gene/protein node. ABCB1 is ATP binding cassette subfamily B member 1. The membrane-associated protein encoded by this gene is a member of the superfamily of ATP-binding cassette (ABC) transporters. ABC proteins transport various molecules across extra- and intra-cellular membranes. ABC genes are divided into seven distinct subfamilies (ABC1, MDR/TAP, MRP, ALD, OABP, GCN20, White). This protein is a member of the MDR/TAP subfamily. Members of the MDR/TAP subfamily are involved in multidrug resistance. The protein encoded by this gene is an ATP-dependent drug efflux pump for xenobiotic compounds with broad substrate specificity. It is responsible for decreased drug accumulation in multidrug-resistant cells and often mediates the development of resistance to anticancer drugs. This protein also functions as a transporter in the blood-brain barrier. Mutations in this gene are associated with colchicine resistance and Inflammatory bowel disease 13. Alternative splicing and the use of alternative promoters results in multiple transcript variants. [provided by RefSeq, Feb 2017].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'ABCB1',\n",
       "   'enriched_node': 'MDLEGDRNGGAKKKNFFKLNNKSEKDKKEKKPTVSVFSMFRYSNWLDKLYMVVGTLAAIIHGAGLPLMMLVFGEMTDIFANAGNLEDLMSNITNRSDINDTGFFMNLEEDMTRYAYYYSGIGAGVLVAAYIQVSFWCLAAGRQIHKIRKQFFHAIMRQEIGWFDVHDVGELNTRLTDDVSKINEGIGDKIGMFFQSMATFFTGFIVGFTRGWKLTLVILAISPVLGLSAAVWAKILSSFTDKELLAYAKAGAVAEEVLAAIRTVIAFGGQKKELERYNKNLEEAKRIGIKKAITANISIGAAFLLIYASYALAFWYGTTLVLSGEYSIGQVLTVFFSVLIGAFSVGQASPSIEAFANARGAAYEIFKIIDNKPSIDSYSKSGHKPDNIKGNLEFRNVHFSYPSRKEVKILKGLNLKVQSGQTVALVGNSGCGKSTTVQLMQRLYDPTEGMVSVDGQDIRTINVRFLREIIGVVSQEPVLFATTIAENIRYGRENVTMDEIEKAVKEANAYDFIMKLPHKFDTLVGERGAQLSGGQKQRIAIARALVRNPKILLLDEATSALDTESEAVVQVALDKARKGRTTIVIAHRLSTVRNADVIAGFDDGVIVEKGNHDELMKEKGIYFKLVTMQTAGNEVELENAADESKSEIDALEMSSNDSRSSLIRKRSTRRSVRGSQAQDRKLSTKEALDESIPPVSFWRIMKLNLTEWPYFVVGVFCAIINGGLQPAFAIIFSKIIGVFTRIDDPETKRQNSNLFSLLFLALGIISFITFFLQGFTFGKAGEILTKRLRYMVFRSMLRQDVSWFDDPKNTTGALTTRLANDAAQVKGAIGSRLAVITQNIANLGTGIIISFIYGWQLTLLLLAIVPIIAIAGVVEMKMLSGQALKDKKELEGSGKIATEAIENFRTVVSLTQEQKFEHMYAQSLQVPYRNSLRKAHIFGITFSFTQAMMYFSYAGCFRFGAYLVAHKLMSFEDVLLVFSAVVFGAMAVGQVSSFAPDYAKAKISAAHIIMIIEKTPLIDSYSTEGLMPNTLEGNVTFGEVVFNYPTRPDIPVLQGLSLEVKKGQTLALVGSSGCGKSTVVQLLERFYDPLAGKVLLDGKEIKRLNVQWLRAHLGIVSQEPILFDCSIAENIAYGDNSRVVSQEEIVRAAKEANIHAFIESLPNKYSTKVGDKGTQLSGGQKQRIAIARALVRQPHILLLDEATSALDTESEKVVQEALDKAREGRTCIVIAHRLSTIQNADLIVVFQNGRVKEHGTHQQLLAQKGIYFSMVSVQAGTKRQ',\n",
       "   'color': <NA>}),\n",
       " ('RIPK2_(4731)',\n",
       "  {'desc': 'RIPK2 belongs to gene/protein node. RIPK2 is receptor interacting serine/threonine kinase 2. This gene encodes a member of the receptor-interacting protein (RIP) family of serine/threonine protein kinases. The encoded protein contains a C-terminal caspase activation and recruitment domain (CARD), and is a component of signaling complexes in both the innate and adaptive immune pathways. It is a potent activator of NF-kappaB and inducer of apoptosis in response to various stimuli. [provided by RefSeq, Jul 2008].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'RIPK2',\n",
       "   'enriched_node': 'MNGEAICSALPTIPYHKLADLRYLSRGASGTVSSARHADWRVQVAVKHLHIHTPLLDSERKDVLREAEILHKARFSYILPILGICNEPEFLGIVTEYMPNGSLNELLHRKTEYPDVAWPLRFRILHEIALGVNYLHNMTPPLLHHDLKTQNILLDNEFHVKIADFGLSKWRMMSLSQSRSSKSAPEGGTIIYMPPENYEPGQKSRASIKHDIYSYAVITWEVLSRKQPFEDVTNPLQIMYSVSQGHRPVINEESLPYDIPHRARMISLIESGWAQNPDERPSFLKCLIELEPVLRTFEEITFLEAVIQLKKTKLQSVSSAIHLCDKKKMELSLNIPVNHGPQEESCGSSQLHENSGSPETSRSLPAPQDNDFLSRKAQDCYFMKLHHCPGNHSWDSTISGSQRAAFCDHKTTPCSSAIINPLSTAGNSERLQPGIAQQWIQSKREDIVNQMTEACLNQSLDALLSRDLIMKEDYELVSTKPTRTSKVRQLLDTTDIQGEEFAKVIVQKLKDNKQMGLQPYPEILVVSRSPSLNLLQNKSM',\n",
       "   'color': <NA>}),\n",
       " ('IL12B_(6168)',\n",
       "  {'desc': 'IL12B belongs to gene/protein node. IL12B is interleukin 12B. This gene encodes a subunit of interleukin 12, a cytokine that acts on T and natural killer cells, and has a broad array of biological activities. Interleukin 12 is a disulfide-linked heterodimer composed of the 40 kD cytokine receptor like subunit encoded by this gene, and a 35 kD subunit encoded by IL12A. This cytokine is expressed by activated macrophages that serve as an essential inducer of Th1 cells development. This cytokine has been found to be important for sustaining a sufficient number of memory/effector Th1 cells to mediate long-term protection to an intracellular pathogen. Overexpression of this gene was observed in the central nervous system of patients with multiple sclerosis (MS), suggesting a role of this cytokine in the pathogenesis of the disease. The promoter polymorphism of this gene has been reported to be associated with the severity of atopic and non-atopic asthma in children. [provided by RefSeq, Jul 2008].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'IL12B',\n",
       "   'enriched_node': 'MCHQQLVISWFSLVFLASPLVAIWELKKDVYVVELDWYPDAPGEMVVLTCDTPEEDGITWTLDQSSEVLGSGKTLTIQVKEFGDAGQYTCHKGGEVLSHSLLLLHKKEDGIWSTDILKDQKEPKNKTFLRCEAKNYSGRFTCWWLTTISTDLTFSVKSSRGSSDPQGVTCGAATLSAERVRGDNKEYEYSVECQEDSACPAAEESLPIEVMVDAVHKLKYENYTSSFFIRDIIKPDPPKNLQLKPLKNSRQVEVSWEYPDTWSTPHSYFSLTFCVQVQGKSKREKKDRVFTDKTSATVICRKNASISVRAQDRYYSSSWSEWASVPCS',\n",
       "   'color': <NA>}),\n",
       " ('TYK2_(6428)',\n",
       "  {'desc': 'TYK2 belongs to gene/protein node. TYK2 is tyrosine kinase 2. This gene encodes a member of the tyrosine kinase and, more specifically, the Janus kinases (JAKs) protein families. This protein associates with the cytoplasmic domain of type I and type II cytokine receptors and promulgate cytokine signals by phosphorylating receptor subunits. It is also a component of both the type I and type III interferon signaling pathways. As such, it may play a role in anti-viral immunity. A mutation in this gene has been associated with Immunodeficiency 35. [provided by RefSeq, Sep 2020].',\n",
       "   'node_type': 'gene/protein',\n",
       "   'node_name': 'TYK2',\n",
       "   'enriched_node': 'MPLRHWGMARGSKPVGDGAQPMAAMGGLKVLLHWAGPGGGEPWVTFSESSLTAEEVCIHIAHKVGITPPCFNLFALFDAQAQVWLPPNHILEIPRDASLMLYFRIRFYFRNWHGMNPREPAVYRCGPPGTEASSDQTAQGMQLLDPASFEYLFEQGKHEFVNDVASLWELSTEEEIHHFKNESLGMAFLHLCHLALRHGIPLEEVAKKTSFKDCIPRSFRRHIRQHSALTRLRLRNVFRRFLRDFQPGRLSQQMVMVKYLATLERLAPRFGTERVPVCHLRLLAQAEGEPCYIRDSGVAPTDPGPESAAGPPTHEVLVTGTGGIQWWPVEEEVNKEEGSSGSSGRNPQASLFGKKAKAHKAVGQPADRPREPLWAYFCDFRDITHVVLKEHCVSIHRQDNKCLELSLPSRAAALSFVSLVDGYFRLTADSSHYLCHEVAPPRLVMSIRDGIHGPLLEPFVQAKLRPEDGLYLIHWSTSHPYRLILTVAQRSQAPDGMQSLRLRKFPIEQQDGAFVLEGWGRSFPSVRELGAALQGCLLRAGDDCFSLRRCCLPQPGETSNLIIMRGARASPRTLNLSQLSFHRVDQKEITQLSHLGQGTRTNVYEGRLRVEGSGDPEEGKMDDEDPLVPGRDRGQELRVVLKVLDPSHHDIALAFYETASLMSQVSHTHLAFVHGVCVRGPENIMVTEYVEHGPLDVWLRRERGHVPMAWKMVVAQQLASALSYLENKNLVHGNVCGRNILLARLGLAEGTSPFIKLSDPGVGLGALSREERVERIPWLAPECLPGGANSLSTAMDKWGFGATLLEICFDGEAPLQSRSPSEKEHFYQRQHRLPEPSCPQLATLTSQCLTYEPTQRPSFRTILRDLTRLQPHNLADVLTVNPDSPASDPTVFHKRYLKKIRDLGEGHFGKVSLYCYDPTNDGTGEMVAVKALKADCGPQHRSGWKQEIDILRTLYHEHIIKYKGCCEDQGEKSLQLVMEYVPLGSLRDYLPRHSIGLAQLLLFAQQICEGMAYLHAQHYIHRDLAARNVLLDNDRLVKIGDFGLAKAVPEGHEYYRVREDGDSPVFWYAPECLKEYKFYYASDVWSFGVTLYELLTHCDSSQSPPTKFLELIGIAQGQMTVLRLTELLERGERLPRPDKCPCEVYHLMKNCWETEASFRPTFENLIPILKTVHEKYQGQAPSVFSVC',\n",
       "   'color': <NA>})]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_dict[\"nodes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8941b320",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_index</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>desc</th>\n",
       "      <th>desc_x</th>\n",
       "      <th>enriched_node</th>\n",
       "      <th>x</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1296</td>\n",
       "      <td>negative regulation of calcium ion transport</td>\n",
       "      <td>negative regulation of calcium ion transport_(...</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>negative regulation of calcium ion transport b...</td>\n",
       "      <td>[0.06554254, 0.081316106, -0.1285265, 0.049076...</td>\n",
       "      <td>Any process that stops, prevents, or reduces t...</td>\n",
       "      <td>[0.087609835, 0.070211984, -0.14004038, 0.0102...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1297</td>\n",
       "      <td>vasoconstriction</td>\n",
       "      <td>vasoconstriction_(45312)</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>vasoconstriction belongs to biological_process...</td>\n",
       "      <td>[0.09431399, 0.04267256, -0.14150876, -0.01277...</td>\n",
       "      <td>A decrease in the diameter of blood vessels, e...</td>\n",
       "      <td>[0.124152906, 0.08743669, -0.14511469, 0.00278...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1298</td>\n",
       "      <td>positive regulation of vascular permeability</td>\n",
       "      <td>positive regulation of vascular permeability_(...</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>positive regulation of vascular permeability b...</td>\n",
       "      <td>[0.07449961, 0.05626029, -0.166832, 0.00161353...</td>\n",
       "      <td>Any process that increases the extent to which...</td>\n",
       "      <td>[0.08488041, 0.08616824, -0.16508923, -0.05333...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1299</td>\n",
       "      <td>negative regulation of cation channel activity</td>\n",
       "      <td>negative regulation of cation channel activity...</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>negative regulation of cation channel activity...</td>\n",
       "      <td>[0.052357886, 0.057002943, -0.13946539, 0.0354...</td>\n",
       "      <td>Any process that stops, prevents or reduces th...</td>\n",
       "      <td>[0.07234384, 0.062300272, -0.14855722, -0.0117...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1300</td>\n",
       "      <td>regulation of cell cycle</td>\n",
       "      <td>regulation of cell cycle_(45381)</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>regulation of cell cycle belongs to biological...</td>\n",
       "      <td>[0.05315926, 0.031067748, -0.14542376, 0.02276...</td>\n",
       "      <td>Any process that modulates the rate or extent ...</td>\n",
       "      <td>[0.049430866, 0.04014543, -0.15489803, 0.00624...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2986</th>\n",
       "      <td>11</td>\n",
       "      <td>PPARA</td>\n",
       "      <td>PPARA_(1122)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PPARA belongs to gene/protein node. PPARA is p...</td>\n",
       "      <td>[0.036692616, 0.03641927, -0.15898165, 0.00553...</td>\n",
       "      <td>MVDTESPLCPLSPLEAGDLESPLSEEFLQEMGNIQEISQSIGEDSS...</td>\n",
       "      <td>[-0.00225809495896101, 0.03649000823497772, 0....</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2987</th>\n",
       "      <td>12</td>\n",
       "      <td>IL10RA</td>\n",
       "      <td>IL10RA_(1299)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>IL10RA belongs to gene/protein node. IL10RA is...</td>\n",
       "      <td>[0.01976429, 0.020241434, -0.16177836, -0.0315...</td>\n",
       "      <td>MLPCLVVLLAALLSLRLGSDAHGTELPSPPSVWFEAEFFHHILHWT...</td>\n",
       "      <td>[0.01952783204615116, 0.059050511568784714, 0....</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2988</th>\n",
       "      <td>13</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>ADIPOQ_(1480)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ADIPOQ belongs to gene/protein node. ADIPOQ is...</td>\n",
       "      <td>[0.06728461, 0.036462087, -0.14274466, -0.0324...</td>\n",
       "      <td>MLLLGAVLLLLALPGHDQETTTQGPGVLLPLPKGACTGWMAGIPGH...</td>\n",
       "      <td>[-0.02511977031826973, -0.12880179286003113, -...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2989</th>\n",
       "      <td>14</td>\n",
       "      <td>IL6</td>\n",
       "      <td>IL6_(1567)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>IL6 belongs to gene/protein node. IL6 is inter...</td>\n",
       "      <td>[0.0419631, 0.02979343, -0.17528865, -0.003821...</td>\n",
       "      <td>MNSFSTSAFGPVAFSLGLLLVLPAAFPAPVPPGEDSKDVAAPHRQP...</td>\n",
       "      <td>[0.05400313809514046, 0.00949622318148613, 0.0...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2990</th>\n",
       "      <td>15</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>JAK2_(1618)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>JAK2 belongs to gene/protein node. JAK2 is Jan...</td>\n",
       "      <td>[0.028720772, 0.026752697, -0.16897796, 0.0070...</td>\n",
       "      <td>MGMACLTMTEMEGTSTSSIYQNGDISGNANSMKQIDPVLQVYLYHS...</td>\n",
       "      <td>[0.01645437814295292, -0.05675109475851059, -0...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2991 rows  9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      node_index                                       node_name  \\\n",
       "0           1296    negative regulation of calcium ion transport   \n",
       "1           1297                                vasoconstriction   \n",
       "2           1298    positive regulation of vascular permeability   \n",
       "3           1299  negative regulation of cation channel activity   \n",
       "4           1300                        regulation of cell cycle   \n",
       "...          ...                                             ...   \n",
       "2986          11                                           PPARA   \n",
       "2987          12                                          IL10RA   \n",
       "2988          13                                          ADIPOQ   \n",
       "2989          14                                             IL6   \n",
       "2990          15                                            JAK2   \n",
       "\n",
       "                                                node_id           node_type  \\\n",
       "0     negative regulation of calcium ion transport_(...  biological_process   \n",
       "1                              vasoconstriction_(45312)  biological_process   \n",
       "2     positive regulation of vascular permeability_(...  biological_process   \n",
       "3     negative regulation of cation channel activity...  biological_process   \n",
       "4                      regulation of cell cycle_(45381)  biological_process   \n",
       "...                                                 ...                 ...   \n",
       "2986                                       PPARA_(1122)        gene/protein   \n",
       "2987                                      IL10RA_(1299)        gene/protein   \n",
       "2988                                      ADIPOQ_(1480)        gene/protein   \n",
       "2989                                         IL6_(1567)        gene/protein   \n",
       "2990                                        JAK2_(1618)        gene/protein   \n",
       "\n",
       "                                                   desc  \\\n",
       "0     negative regulation of calcium ion transport b...   \n",
       "1     vasoconstriction belongs to biological_process...   \n",
       "2     positive regulation of vascular permeability b...   \n",
       "3     negative regulation of cation channel activity...   \n",
       "4     regulation of cell cycle belongs to biological...   \n",
       "...                                                 ...   \n",
       "2986  PPARA belongs to gene/protein node. PPARA is p...   \n",
       "2987  IL10RA belongs to gene/protein node. IL10RA is...   \n",
       "2988  ADIPOQ belongs to gene/protein node. ADIPOQ is...   \n",
       "2989  IL6 belongs to gene/protein node. IL6 is inter...   \n",
       "2990  JAK2 belongs to gene/protein node. JAK2 is Jan...   \n",
       "\n",
       "                                                 desc_x  \\\n",
       "0     [0.06554254, 0.081316106, -0.1285265, 0.049076...   \n",
       "1     [0.09431399, 0.04267256, -0.14150876, -0.01277...   \n",
       "2     [0.07449961, 0.05626029, -0.166832, 0.00161353...   \n",
       "3     [0.052357886, 0.057002943, -0.13946539, 0.0354...   \n",
       "4     [0.05315926, 0.031067748, -0.14542376, 0.02276...   \n",
       "...                                                 ...   \n",
       "2986  [0.036692616, 0.03641927, -0.15898165, 0.00553...   \n",
       "2987  [0.01976429, 0.020241434, -0.16177836, -0.0315...   \n",
       "2988  [0.06728461, 0.036462087, -0.14274466, -0.0324...   \n",
       "2989  [0.0419631, 0.02979343, -0.17528865, -0.003821...   \n",
       "2990  [0.028720772, 0.026752697, -0.16897796, 0.0070...   \n",
       "\n",
       "                                          enriched_node  \\\n",
       "0     Any process that stops, prevents, or reduces t...   \n",
       "1     A decrease in the diameter of blood vessels, e...   \n",
       "2     Any process that increases the extent to which...   \n",
       "3     Any process that stops, prevents or reduces th...   \n",
       "4     Any process that modulates the rate or extent ...   \n",
       "...                                                 ...   \n",
       "2986  MVDTESPLCPLSPLEAGDLESPLSEEFLQEMGNIQEISQSIGEDSS...   \n",
       "2987  MLPCLVVLLAALLSLRLGSDAHGTELPSPPSVWFEAEFFHHILHWT...   \n",
       "2988  MLLLGAVLLLLALPGHDQETTTQGPGVLLPLPKGACTGWMAGIPGH...   \n",
       "2989  MNSFSTSAFGPVAFSLGLLLVLPAAFPAPVPPGEDSKDVAAPHRQP...   \n",
       "2990  MGMACLTMTEMEGTSTSSIYQNGDISGNANSMKQIDPVLQVYLYHS...   \n",
       "\n",
       "                                                      x color  \n",
       "0     [0.087609835, 0.070211984, -0.14004038, 0.0102...  <NA>  \n",
       "1     [0.124152906, 0.08743669, -0.14511469, 0.00278...  <NA>  \n",
       "2     [0.08488041, 0.08616824, -0.16508923, -0.05333...  <NA>  \n",
       "3     [0.07234384, 0.062300272, -0.14855722, -0.0117...  <NA>  \n",
       "4     [0.049430866, 0.04014543, -0.15489803, 0.00624...  <NA>  \n",
       "...                                                 ...   ...  \n",
       "2986  [-0.00225809495896101, 0.03649000823497772, 0....  <NA>  \n",
       "2987  [0.01952783204615116, 0.059050511568784714, 0....  <NA>  \n",
       "2988  [-0.02511977031826973, -0.12880179286003113, -...  <NA>  \n",
       "2989  [0.05400313809514046, 0.00949622318148613, 0.0...  <NA>  \n",
       "2990  [0.01645437814295292, -0.05675109475851059, -0...  <NA>  \n",
       "\n",
       "[2991 rows x 9 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a69c5653",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "node_id,node_attr\n",
      "KIF21B_(8564),\"KIF21B belongs to gene/protein node. KIF21B is kinesin family member 21B. This gene encodes a member of the kinesin superfamily. Kinesins are ATP-dependent microtubule-based motor proteins that are involved in the intracellular transport of membranous organelles. Single nucleotide polymorphisms in this gene are associated with inflammatory bowel disease and multiple sclerosis. Alternatively spliced transcript variants encoding multiple isoforms have been observed for this gene. [provided by RefSeq, Nov 2011].\"\n",
      "INAVA_(9104),\"INAVA belongs to gene/protein node. INAVA is innate immunity activator. Involved in several processes, including nucleotide-binding activity oligomerization domain containing 2 signaling pathway; positive regulation of cytokine production; and positive regulation of intracellular signal transduction. Located in cytoplasm and nucleus. Implicated in inflammatory bowel disease 29. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "IL18RAP_(11588),\"IL18RAP belongs to gene/protein node. IL18RAP is interleukin 18 receptor accessory protein. The protein encoded by this gene is an accessory subunit of the heterodimeric receptor for interleukin 18 (IL18), a proinflammatory cytokine involved in inducing cell-mediated immunity. This protein enhances the IL18-binding activity of the IL18 receptor and plays a role in signaling by IL18. Mutations in this gene are associated with Crohn's disease and inflammatory bowel disease, and susceptibility to celiac disease and leprosy. Alternatively spliced transcript variants of this gene have been described, but their full-length nature is not known. [provided by RefSeq, Feb 2014].\"\n",
      "PHOX2B_(13014),\"PHOX2B belongs to gene/protein node. PHOX2B is paired like homeobox 2B. The DNA-associated protein encoded by this gene is a member of the paired family of homeobox proteins localized to the nucleus. The protein functions as a transcription factor involved in the development of several major noradrenergic neuron populations and the determination of neurotransmitter phenotype. The gene product is linked to enhancement of second messenger-mediated activation of the dopamine beta-hydroylase, c-fos promoters and several enhancers, including cyclic amp-response element and serum-response element. Expansion of a 20 amino acid polyalanine tract in this protein by 5-13 aa has been associated with congenital central hypoventilation syndrome. [provided by RefSeq, Jul 2016].\"\n",
      "Diclofenac_(14061),\"Diclofenac belongs to drug node. Diclofenac is a phenylacetic acid derivative and non-steroidal anti-inflammatory drug (NSAID). NSAIDs inhibit cyclooxygenase (COX)-1 and-2 which are the enzyme responsible for producing prostaglandins (PGs). PGs contribute to inflammation and pain signalling. Diclofenac, like other NSAIDs, is often used as first line therapy for acute and chronic pain and inflammation from a variety of causes. Diclofenac was the product of rational drug design based on the structures of , , and. The addition of two chlorine groups in the ortho position of the phenyl ring locks the ring in maximal torsion which appears to be related to increased potency. It is often used in combination with to prevent NSAID-induced gastric ulcers. Diclofenac was first approved by the FDA in July 1988 under the trade name Voltaren, marketed by Novartis (previously Ciba-Geigy). Diclofenac inhibits cyclooxygenase-1 and -2, the enzymes responsible for production of prostaglandin (PG) G<sub>2</sub> which is the precursor to other PGs. These molecules have broad activity in pain and inflammation and the inhibition of their production is the common mechanism linking each effect of diclofenac. Diclofenac is over 99.7% bound to serum proteins, primarily albumin. It is undergoes limited binding to lipoproteins as well with 1.1% bound to HDL, 0.3% to LDL, and 0.15% to VLDL. Diclofenac reduces inflammation and by extension reduces nociceptive pain and combats fever. It also increases the risk of developing a gastrointestinal ulcer by inhibiting the production of protective mucus in the stomach. Diclofenac is indicated for use in the treatment of pain and inflammation from varying sources including inflammatory conditions such as osteoarthritis, rheumatoid arthritis, and akylosing spondylitis, as well as injury-related inflammation due to surgery and physical trauma. It is often used in combination with as a gastro-protective agent in patients with high risk of developing NSAID-induced ulcers.\"\n",
      "Flufenamic acid_(14069),\"Flufenamic acid belongs to drug node. An anthranilic acid derivative with analgesic, anti-inflammatory, and antipyretic properties. It is used in musculoskeletal and joint disorders and administered by mouth and topically. (From Martindale, The Extra Pharmacopoeia, 30th ed, p16)    \"\n",
      "Glycyrrhizic acid_(14107),\"Glycyrrhizic acid belongs to drug node. Glycyrrhizic acid is extracted from the root of the licorice plant; _Glycyrrhiza glabra_. It is a triterpene glycoside with glycyrrhetinic acid that possesses a wide range of pharmacological and biological activities. When extracted from the plant, it can be obtained in the form of ammonium glycyrrhizin and mono-ammonium glycyrrhizin. Glycyrrhizic acid has been developed in Japan and China as a hepatoprotective drug in cases of chronic hepatitis. From January 2014, glycyrrhizic acid as part of the licorice extract was approved by the FDA as an existing food sweetener. It was approved by Health Canada to be used in over-the-counter products but all the products are currently on the status canceled post marketed. Glycyrrhizic acid can be found in the alpha and beta forms. The alpha form is predomit in the liver and duodenum and thus, it is thought that the anti-inflammatory liver effect of this drug are mainly due to the action of this isomer. Glycyrrhizic acid anti-inflammatory effect is generated via suppression of TNF alpha and caspase 3. It also inhibits the translocation of NFkB into the nuclei and conjugates free radicals. Some studies have shown a glycyrrhizic-driven inhibition of CD4+ T cell proliferation via JNK, ERK and PI3K/AKT. Glycyrrhizic acid does not bind to any plasma proteins as it is not absorbed systemically. On the other hand, its main active metabolite, glycyrrhetinic acid presents a very large binding to serum proteins such as albumin. Glycyrrhizic acid was reported to present antiallergic, antiviral and anti-inflammatory activities as well as improvements in glucose tolerance.  Glycyrrhizic acid is widely applied in foods as a natural sweetener. As a therapeutic agent, is has been used in a vast variety of formulations as it is reported to be anti-inflammatory, anti-ulcer, anti-allergic, antioxidant, anti-tumor, anti-diabetic and hepatoprotective. Due to this properties, its indications have been: treatment of premenstrual syndrome, treatment of viral infections, anti-lipidemic and antihyperglycemic. It is also known to be used as a remedy for peptic ulcer and other stomach diseases.\"\n",
      "Tofacitinib_(14407),\"Tofacitinib belongs to drug node. Tofacitinib is an inhibitor of Janus kinases, a group of intracellular enzymes involved in signalling pathways that affect hematopoiesis and immune cell function. It is approved by the FDA for treatment of moderate to severe rheumatoid arthritis that responds inadequately to methotrexate or in those who are intolerant to methotrexate.  Rheumatoid arthritis is an autoimmune disease characterized by a dysregulation of pro-inflammatory cytokines including IL7, IL15, IL21, IL6, IFN-alpha, and IFN-beta. (3) Cytokines signalling results in tissue inflammation and joint damage by stimulating the recruitment and activation of immune cells via the janus kinase signalling pathway. 40%, mostly bound to albumin. Tofacitinib targets inflammation present in rheumatoid arthritis by inhibiting the janus kinases involved in the inflammatory response pathway.  For the treatment of moderate to severe rheumatoid arthritis which is resistant or intolerant to methotrexate therapy. It may also be used as an adjunct to methotrexate therapy, or other non-biologic disease-modifying antirheumatic drugs (DMARDS), when methotrexate alone is not sufficient. \"\n",
      "Binimetinib_(15391),\"Binimetinib belongs to drug node. Binimetinib, also known as _Mektovi_, is a potent is a potent and selective oral mitogen-activated protein kinase 1/2  (MEK 1/2) inhibitor which is combined with ,.  Binimetinib is 97% bound to human plasma proteins and the blood-to-plasma ratio is 0.72  Binimetinib is a MEK inhibitor. MEK is an enzyme that regulates the biosynthesis of the inflammatory cytokines TNF, IL-6 and IL-1. MEK inhibitors interfere with these biosynthetic processes.  It is a chemotherapeutic agent that has anti-tumor activity ,. \"\n",
      "Fostamatinib_(15674),\"Fostamatinib belongs to drug node. Fostamatinib has been investigated for the treatment and basic science of Rheumatoid Arthritis and Immune Thrombocytopenic Purpura (ITP). It was approved on April 17, 2018 under the trade name Tavalisse for use in ITP. Fostamatinib has also been granted orphan drug status by the FDA. The active metabolite of fostamatinib, R406, is an inhibitor of spleen tyrosine kinase (Syk). It binds reversibly to the ATP binding pocket with high affinity (Ki = 30nM), inhibiting the kinase activity with an IC50 of 41nM. R406 is 98.3% bound to plasma proteins. The active metabolite of fostamatinib, R406, inhibits signal transduction by Fc receptors involved in the antibody-mediated destruction of platelets by immune cells in chronic ITP. This results in increased platelet counts in this population. Fostamatinib is indicated for use in the treatment of chronic immune thrombocytopenia (ITP) in patients who have had insufficient response to previous therapy.\"\n",
      "Sulfasalazine_(15982),\"Sulfasalazine belongs to drug node. A drug that is used in the management of inflammatory bowel diseases. Its activity is generally considered to lie in its metabolic breakdown product, 5-aminosalicylic acid (see mesalamine) released in the colon. (From Martindale, The Extra Pharmacopoeia, 30th ed, p907) The mode of action of Sulfasalazine or its metabolites, 5-aminosalicylic acid (5-ASA) and sulfapyridine (SP), is still under investigation, but may be related to the anti-inflammatory and/or immunomodulatory properties that have been observed in animal and <i>in vitro</i> models, to its affinity for connective tissue, and/or to the relatively high concentration it reaches in serous fluids, the liver and intestinal walls, as demonstrated in autoradiographic studies in animals. In ulcerative colitis, clinical studies utilizing rectal administration of Sulfasalazine, SP and 5-ASA have indicated that the major therapeutic action may reside in the 5-ASA moiety. The relative contribution of the parent drug and the major metabolites in rheumatoid arthritis is unknown.  Sulfasalazine is an anti-inflammatory indicated for the treatment of ulcerative colitis and rheumatoid arthritis. For the treatment of Crohn's disease and rheumatoid arthritis as a second-line agent.\"\n",
      "Glucosamine_(16195),\"Glucosamine belongs to drug node. Glucosamine is commonly used as a treatment for osteoarthritis, although its acceptance as a medical therapy varies. It is an amino sugar and precursor of glycosylated proteins and lipids. Since glucosamine is a precursor for glycosaminoglycans, and glycosaminoglycans are a major component of joint cartilage, supplemental glucosamine may help to rebuild cartilage and treat arthritis. Glucosamine is a precursor of glycosylated proteins and lipids. Oral glucosamine is commonly used for the treatment of osteoarthritis. Since glucosamine is a precursor for glycosaminoglycans, and glycosaminoglycans are a major component of joint cartilage, supplemental glucosamine may help to rebuild cartilage and treat arthritis. Its use as a therapy for osteoarthritis appears safe, but there is conflicting evidence as to its effectiveness with more recent studies showing limited to no clinical benefit of use. Glucosamine is not FDA approved for use in humans. Since glucosamine is classified as a dietary supplement, safety and formulation are solely the responsibility of the manufacturer; evidence of safety and efficacy is not required as long as it is not advertised as a treatment for a medical condition.  Osteoarthritis is characterized by the progressive degeneration of cartilage glycosaminoglycans. The formation of glucosamine is the rate limiting step in glycosaminoglycans synthesis thus the addition is glucosamine, would in theory provide a building block towards the synthesis of glycosaminoglycans and thus slow down the progression of osteoarthritis. Thus far however, the results have not been conclusive. Glucosamine is usually used in the treatment of osteoarthritis, although its efficacy is still in question.\"\n",
      "\"2-Amino-N,3,3-Trimethylbutanamide_(16196)\",\"2-Amino-N,3,3-Trimethylbutanamide belongs to drug node.     \"\n",
      "\"(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic acid_(16613)\",\"(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic acid belongs to drug node.     \"\n",
      "\"(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7,10,13,16,19-hexaenoic acid_(16615)\",\"(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7,10,13,16,19-hexaenoic acid belongs to drug node.     \"\n",
      "Talmapimod_(17589),\"Talmapimod belongs to drug node. Talmapimod is the first-generation oral p38 MAP kinase inhibitor developed by Scios. It has shown to be effective to cure inflammatory diseases such as Rheumatoid Arthritis. SCIO-469 inhibits p38 kinase, a stimulatory modulator of pro-inflammatory factors including tumor necrosis factor-alpha (TNFa), interleukin-1 (IL-1), and cyclooxygenase-2 (COX-2), all of which are known to contribute to both symptoms and disease progression in patients with Rheumatoid Arthritis (RA). Existing protein-based products that antagonize TNFa have been shown to markedly relieve the symptoms and retard the progression of RA. It also has the potential for additional benefits associated with its inhibition of IL-1 and COX-2.   Investigated for use/treatment in pain (acute or chronic) and rheumatoid arthritis.\"\n",
      "VX-702_(17590),\"VX-702 belongs to drug node. VX-702 is a small molecule investigational oral anti-cytokine therapy for treatment of inflammatory diseases, specifically rheumatoid arthritis (RA). It acts as a p38 MAP kinase inhibitor. In the future, VX-702 may be investigated for combination with methotrexate, a commonly used therapy for RA. This p38 MAP kinase inhibitor effectively inhibits LPS-stimulated TNF||, IL-6 and IL-1|| production.  VX-703 is an anti-cytokine therapy in which p38 MAP kinase inhibitor effectively inhibits LPS-stimulated TNF||, IL-6 and IL-1|| production. Investigated for use/treatment in coronary artery disease, inflammatory disorders (unspecified), and rheumatoid arthritis.\"\n",
      "Atiprimod_(17591),\"Atiprimod belongs to drug node.     Investigated for use/treatment in cancer/tumors (unspecified), multiple myeloma, and rheumatoid arthritis.\"\n",
      "Andrographolide_(17594),\"Andrographolide belongs to drug node. Andrographolide (HMPL-004) is a botanical product extracted from a herb that occurs naturally in China. The herb has an extensive history of use in TCM for the treatment of upper respiratory tract infections and other inflammatory and infectious diseases. HMPL-004 acts on multiple cellular targets in the inflammatory signal transduction pathways resulting in suppressed inflammation cytokine expression including TNF-, IL-1 and IL-6. HMPL-004 was demonstrated to inhibit TNF- and IL-1 production in cell-based assays. HMPL-004 is also able to inhibit NF-kB activation. NF-kB is a family of transcriptional factors that regulate a wide spectrum of genes critically involved in host defence and inflammation. The mechanism of action of HMPL-004 was further supported in laboratory IBD animal models. Treatment of IBD rats with HMPL-004 caused a significant drop in plasma cytokine concentrations, including TNF- and IL-1.   Investigated for use/treatment in ulcerative colitis.\"\n",
      "Dilmapimod_(17602),\"Dilmapimod belongs to drug node. Dilmapimod has been used in trials studying the treatment and diagnostic of Nerve Trauma, Inflammation, Pain, Neuropathic, Arthritis, Rheumatoid, and Coronary Heart Disease, among others. Dilmapimod (SB-681323) is a p38 MAP-kinase inhibitor that has potential uses in inflammatory conditions such as RA (Rheumatoid Arthritis). Previous p38 MAP-kinase inhibitors have been hindered in development by liver toxicity. Methotrexate (common treatment for RA patients) also has potential liver toxicity.} Dilmapimod reduces the levels of proinflammatory cytokines and chemokines and reduce cellular infiltration to sites of inflammation, thereby reducing local damage.    Dilmapimod has been used in trials studying the treatment and diagnostic of Nerve Trauma, Inflammation, Pain, Neuropathic, Arthritis, Rheumatoid, and Coronary Heart Disease, among others.\"\n",
      "inflammatory bowel disease_(28158),inflammatory bowel disease belongs to disease node. Any inflammatory bowel disease in which the cause of the disease is a mutation in the NOD2 gene.  \n",
      "IL23R_(34778),\"IL23R belongs to gene/protein node. IL23R is interleukin 23 receptor. The protein encoded by this gene is a subunit of the receptor for IL23A/IL23. This protein pairs with the receptor molecule IL12RB1/IL12Rbeta1, and both are required for IL23A signaling. This protein associates constitutively with Janus kinase 2 (JAK2), and also binds to transcription activator STAT3 in a ligand-dependent manner. [provided by RefSeq, Jul 2008].\"\n",
      "NKX2-3_(34779),\"NKX2-3 belongs to gene/protein node. NKX2-3 is NK2 homeobox 3. This gene encodes a homeodomain-containing transcription factor. The encoded protein is a member of the NKX family of homeodomain transcription factors. Studies of similar proteins in mouse and rat have indicated a potential role in cellular differentiation.[provided by RefSeq, Mar 2010].\"\n",
      "TAGAP_(34814),\"TAGAP belongs to gene/protein node. TAGAP is T cell activation RhoGTPase activating protein. This gene encodes a member of the Rho GTPase-activator protein superfamily. The encoded protein may function as a Rho GTPase-activating protein. Alterations in this gene may be associated with several diseases, including rheumatoid arthritis, celiac disease, and multiple sclerosis. Alternate splicing results in multiple transcript variants encoding distinct isoforms. [provided by RefSeq, Jul 2013].\"\n",
      "DENND1B_(34887),\"DENND1B belongs to gene/protein node. DENND1B is DENN domain containing 1B. Clathrin (see MIM 118955)-mediated endocytosis is a major mechanism for internalization of proteins and lipids. Members of the connecdenn family, such as DENND1B, function as guanine nucleotide exchange factors (GEFs) for the early endosomal small GTPase RAB35 (MIM 604199) and bind to clathrin and clathrin adaptor protein-2 (AP2; see MIM 601024). Thus, connecdenns link RAB35 activation with the clathrin machinery (Marat and McPherson, 2010 [PubMed 20154091]).[supplied by OMIM, Nov 2010].\"\n",
      "Crohn disease_(37784),\"Crohn disease belongs to disease node. A gastrointestinal disorder characterized by chronic inflammation involving all layers of the intestinal wall, noncaseating granulomas affecting the intestinal wall and regional lymph nodes, and transmural fibrosis. Crohn disease most commonly involves the terminal ileum; the colon is the second most common site of involvement. In Crohn's disease, any part of your small or large intestine can be involved, and it may be continuous or may involve multiple segments. In some people, the disease is confined to the colon, which is part of the large intestine. Signs and symptoms of Crohn's disease can range from mild to severe. They usually develop gradually, but sometimes will come on suddenly, without warning. You may also have periods of time when you have no signs or symptoms . When the disease is active, signs and symptoms may include: Diarrhea, Fever, Fatigue, Abdominal pain and cramping, Blood in your stool, Mouth sores, Reduced appetite and weight loss, Pain or drainage near or around the anus due to inflammation from a tunnel into the skin, Other signs and symptoms, People with severe Crohn's disease may also experience: Inflammation of skin, eyes and joints, Inflammation of the liver or bile ducts, Kidney stones, Iron deficiency, Delayed growth or sexual development, in children The exact cause of Crohn's disease remains unknown. Previously, diet and stress were suspected, but now doctors know that these factors may aggravate, but don't cause, Crohn's disease. Several factors, such as heredity and a malfunctioning immune system, likely play a role in its development. Immune system. It's possible that a virus or bacterium may trigger Crohn's disease; however, scientists have yet to identify such a trigger. When your immune system tries to fight off the invading microorganism, an abnormal immune response causes the immune system to attack the cells in the digestive tract, too. Heredity. Crohn's is more common in people who have family members with the disease, so genes may play a role in making people more susceptible. However, most people with Crohn's disease don't have a family history of the disease.\"\n",
      "ulcerative colitis (disease)_(37785),\"ulcerative colitis (disease) belongs to disease node. An inflammatory bowel disease involving the mucosal surface of the large intestine and rectum. It may present with an acute or slow onset and follows an intermittent or continuous course. Signs and symptoms include abdominal pain, diarrhea, fever, weight loss, and intestinal hemorrhage. Ulcerative colitis symptoms can vary, depending on the severity of inflammation and where it occurs. Signs and symptoms may include: Diarrhea, often with blood or pus, Abdominal pain and cramping, Rectal pain, Rectal bleeding  passing small amount of blood with stool, Urgency to defecate, Inability to defecate despite urgency, Weight loss, Fatigue, Fever, In children, failure to grow, Most people with ulcerative colitis have mild to moderate symptoms. The course of ulcerative colitis may vary, with some people having long periods of remission. Types, Doctors often classify ulcerative colitis according to its location. Types of ulcerative colitis include: Ulcerative proctitis. Inflammation is confined to the area closest to the anus, and rectal bleeding may be the only sign of the disease. Proctosigmoiditis. Inflammation involves the rectum and sigmoid colon  the lower end of the colon. Signs and symptoms include bloody diarrhea, abdominal cramps and pain, and an inability to move the bowels in spite of the urge to do so . Left-sided colitis. Inflammation extends from the rectum up through the sigmoid and descending colon. Signs and symptoms include bloody diarrhea, abdominal cramping and pain on the left side, and urgency to defecate. Pancolitis. This type often affects the entire colon and causes bouts of bloody diarrhea that may be severe, abdominal cramps and pain, fatigue, and significant weight loss. The exact cause of ulcerative colitis remains unknown. Previously, diet and stress were suspected, but now doctors know that these factors may aggravate but don't cause ulcerative colitis. One possible cause is an immune system malfunction. When your immune system tries to fight off an invading virus or bacterium, an abnormal immune response causes the immune system to attack the cells in the digestive tract, too. Heredity also seems to play a role in that ulcerative colitis is more common in people who have family members with the disease. However, most people with ulcerative colitis don't have this family history.\"\n",
      "immune response_(39995),immune response belongs to biological_process node. \n",
      "regulation of DNA recombination_(43811),regulation of DNA recombination belongs to biological_process node. \n",
      "regulation of gamma-delta T cell differentiation_(45624),regulation of gamma-delta T cell differentiation belongs to biological_process node. \n",
      "cell surface receptor signaling pathway_(46168),cell surface receptor signaling pathway belongs to biological_process node. \n",
      "canonical Wnt signaling pathway_(46706),canonical Wnt signaling pathway belongs to biological_process node. \n",
      "positive regulation of receptor signaling pathway via STAT_(51389),positive regulation of receptor signaling pathway via STAT belongs to biological_process node. \n",
      "gamma-delta T cell differentiation_(51532),gamma-delta T cell differentiation belongs to biological_process node. \n",
      "positive regulation of peptidyl-serine phosphorylation_(52659),positive regulation of peptidyl-serine phosphorylation belongs to biological_process node. \n",
      "transcription cis-regulatory region binding_(54288),transcription cis-regulatory region binding belongs to molecular_function node. \n",
      "DNA-binding transcription repressor activity_(54389),DNA-binding transcription repressor activity belongs to molecular_function node. \n",
      "cytokine receptor activity_(54621),cytokine receptor activity belongs to molecular_function node. \n",
      "beta-catenin-TCF complex_(56282),beta-catenin-TCF complex belongs to cellular_component node. \n",
      "beta-catenin-TCF complex assembly_(104150),beta-catenin-TCF complex assembly belongs to biological_process node. \n",
      "lymph node development_(107899),lymph node development belongs to biological_process node. \n",
      "T cell receptor V(D)J recombination_(108222),T cell receptor V(D)J recombination belongs to biological_process node. \n",
      "interleukin-7-mediated signaling pathway_(108951),interleukin-7-mediated signaling pathway belongs to biological_process node. \n",
      "beta-catenin binding_(119055),beta-catenin binding belongs to molecular_function node. \n",
      "interleukin-7 receptor activity_(122046),interleukin-7 receptor activity belongs to molecular_function node. \n",
      "nucleoplasm_(124249),nucleoplasm belongs to cellular_component node. \n",
      "IL7R_(625),\"IL7R belongs to gene/protein node. IL7R is interleukin 7 receptor. The protein encoded by this gene is a receptor for interleukin 7 (IL7). The function of this receptor requires the interleukin 2 receptor, gamma chain (IL2RG), which is a common gamma chain shared by the receptors of various cytokines, including interleukins 2, 4, 7, 9, and 15. This protein has been shown to play a critical role in V(D)J recombination during lymphocyte development. Defects in this gene may be associated with severe combined immunodeficiency (SCID). Alternatively spliced transcript variants have been found. [provided by RefSeq, Dec 2015].\"\n",
      "TCF7_(5195),\"TCF7 belongs to gene/protein node. TCF7 is transcription factor 7. This gene encodes a member of the T-cell factor/lymphoid enhancer-binding factor family of high mobility group (HMG) box transcriptional activators. This gene is expressed predominantly in T-cells and plays a critical role in natural killer cell and innate lymphoid cell development. The encoded protein forms a complex with beta-catenin and activates transcription through a Wnt/beta-catenin signaling pathway. Mice with a knockout of this gene are viable and fertile, but display a block in T-lymphocyte differentiation. Alternative splicing results in multiple transcript variants. Naturally-occurring isoforms lacking the N-terminal beta-catenin interaction domain may act as dominant negative regulators of Wnt signaling. [provided by RefSeq, Oct 2016].\"\n",
      "HNF4A_(279),\"HNF4A belongs to gene/protein node. HNF4A is hepatocyte nuclear factor 4 alpha. The protein encoded by this gene is a nuclear transcription factor which binds DNA as a homodimer. The encoded protein controls the expression of several genes, including hepatocyte nuclear factor 1 alpha, a transcription factor which regulates the expression of several hepatic genes. This gene may play a role in development of the liver, kidney, and intestines. Mutations in this gene have been associated with monogenic autosomal dominant non-insulin-dependent diabetes mellitus type I. Alternative splicing of this gene results in multiple transcript variants encoding several different isoforms. [provided by RefSeq, Apr 2012].\"\n",
      "PPARG_(989),\"PPARG belongs to gene/protein node. PPARG is peroxisome proliferator activated receptor gamma. This gene encodes a member of the peroxisome proliferator-activated receptor (PPAR) subfamily of nuclear receptors. PPARs form heterodimers with retinoid X receptors (RXRs) and these heterodimers regulate transcription of various genes. Three subtypes of PPARs are known: PPAR-alpha, PPAR-delta, and PPAR-gamma. The protein encoded by this gene is PPAR-gamma and is a regulator of adipocyte differentiation. Additionally, PPAR-gamma has been implicated in the pathology of numerous diseases including obesity, diabetes, atherosclerosis and cancer. Alternatively spliced transcript variants that encode different isoforms have been described. [provided by RefSeq, Jul 2008].\"\n",
      "IL10RA_(1299),\"IL10RA belongs to gene/protein node. IL10RA is interleukin 10 receptor subunit alpha. The protein encoded by this gene is a receptor for interleukin 10. This protein is structurally related to interferon receptors. It has been shown to mediate the immunosuppressive signal of interleukin 10, and thus inhibits the synthesis of proinflammatory cytokines. This receptor is reported to promote survival of progenitor myeloid cells through the insulin receptor substrate-2/PI 3-kinase/AKT pathway. Activation of this receptor leads to tyrosine phosphorylation of JAK1 and TYK2 kinases. Two transcript variants, one protein-coding and the other not protein-coding, have been found for this gene. [provided by RefSeq, Jan 2009].\"\n",
      "JAK2_(1618),\"JAK2 belongs to gene/protein node. JAK2 is Janus kinase 2. This gene encodes a non-receptor tyrosine kinase that plays a central role in cytokine and growth factor signalling. The primary isoform of this protein has an N-terminal FERM domain that is required for erythropoietin receptor association, an SH2 domain that binds STAT transcription factors, a pseudokinase domain and a C-terminal tyrosine kinase domain. Cytokine binding induces autophosphorylation and activation of this kinase. This kinase then recruits and phosphorylates signal transducer and activator of transcription (STAT) proteins. Growth factors like TGF-beta 1 also induce phosphorylation and activation of this kinase and translocation of downstream STAT proteins to the nucleus where they influence gene transcription. Mutations in this gene are associated with numerous inflammatory diseases and malignancies. This gene is a downstream target of the pleiotropic cytokine IL6 that is produced by B cells, T cells, dendritic cells and macrophages to produce an immune response or inflammation. Disregulation of the IL6/JAK2/STAT3 signalling pathways produces increased cellular proliferation and myeloproliferative neoplasms of hematopoietic stem cells. A nonsynonymous mutation in the pseudokinase domain of this gene disrupts the domains inhibitory effect and results in constitutive tyrosine phosphorylation activity and hypersensitivity to cytokine signalling. This gene and the IL6/JAK2/STAT3 signalling pathway is a therapeutic target for the treatment of excessive inflammatory responses to viral infections. Alternative splicing results in multiple transcript variants encoding distinct isoforms. [provided by RefSeq, Jul 2020].\"\n",
      "LRRK2_(2111),\"LRRK2 belongs to gene/protein node. LRRK2 is leucine rich repeat kinase 2. This gene is a member of the leucine-rich repeat kinase family and encodes a protein with an ankryin repeat region, a leucine-rich repeat (LRR) domain, a kinase domain, a DFG-like motif, a RAS domain, a GTPase domain, a MLK-like domain, and a WD40 domain. The protein is present largely in the cytoplasm but also associates with the mitochondrial outer membrane. Mutations in this gene have been associated with Parkinson disease-8. [provided by RefSeq, Jul 2008].\"\n",
      "TNF_(2329),\"TNF belongs to gene/protein node. TNF is tumor necrosis factor. This gene encodes a multifunctional proinflammatory cytokine that belongs to the tumor necrosis factor (TNF) superfamily. This cytokine is mainly secreted by macrophages. It can bind to, and thus functions through its receptors TNFRSF1A/TNFR1 and TNFRSF1B/TNFBR. This cytokine is involved in the regulation of a wide spectrum of biological processes including cell proliferation, differentiation, apoptosis, lipid metabolism, and coagulation. This cytokine has been implicated in a variety of diseases, including autoimmune diseases, insulin resistance, psoriasis, rheumatoid arthritis ankylosing spondylitis, tuberculosis, autosomal dominant polycystic kidney disease, and cancer. Mutations in this gene affect susceptibility to cerebral malaria, septic shock, and Alzheimer disease. Knockout studies in mice also suggested the neuroprotective function of this cytokine. [provided by RefSeq, Aug 2020].\"\n",
      "LSP1_(2643),\"LSP1 belongs to gene/protein node. LSP1 is lymphocyte specific protein 1. This gene encodes an intracellular F-actin binding protein. The protein is expressed in lymphocytes, neutrophils, macrophages, and endothelium and may regulate neutrophil motility, adhesion to fibrinogen matrix proteins, and transendothelial migration. Alternative splicing results in multiple transcript variants encoding different isoforms. [provided by RefSeq, Jul 2008].\"\n",
      "PRDM1_(2874),\"PRDM1 belongs to gene/protein node. PRDM1 is PR/SET domain 1. This gene encodes a protein that acts as a repressor of beta-interferon gene expression. The protein binds specifically to the PRDI (positive regulatory domain I element) of the beta-IFN gene promoter. Transcription of this gene increases upon virus induction. Two alternatively spliced transcript variants that encode different isoforms have been reported. [provided by RefSeq, Jul 2008].\"\n",
      "CARD9_(3460),\"CARD9 belongs to gene/protein node. CARD9 is caspase recruitment domain family member 9. The protein encoded by this gene is a member of the CARD protein family, which is defined by the presence of a characteristic caspase-associated recruitment domain (CARD). CARD is a protein interaction domain known to participate in activation or suppression of CARD containing members of the caspase family, and thus plays an important regulatory role in cell apoptosis. This protein was identified by its selective association with the CARD domain of BCL10, a postive regulator of apoptosis and NF-kappaB activation, and is thought to function as a molecular scaffold for the assembly of a BCL10 signaling complex that activates NF-kappaB. Several alternatively spliced transcript variants have been observed, but their full-length nature is not clearly defined. [provided by RefSeq, Jul 2008].\"\n",
      "MMP9_(3474),\"MMP9 belongs to gene/protein node. MMP9 is matrix metallopeptidase 9. Proteins of the matrix metalloproteinase (MMP) family are involved in the breakdown of extracellular matrix in normal physiological processes, such as embryonic development, reproduction, and tissue remodeling, as well as in disease processes, such as arthritis and metastasis. Most MMP's are secreted as inactive proproteins which are activated when cleaved by extracellular proteinases. The enzyme encoded by this gene degrades type IV and V collagens. Studies in rhesus monkeys suggest that the enzyme is involved in IL-8-induced mobilization of hematopoietic progenitor cells from bone marrow, and murine studies suggest a role in tumor-associated tissue remodeling. [provided by RefSeq, Jul 2008].\"\n",
      "ABCB1_(4152),\"ABCB1 belongs to gene/protein node. ABCB1 is ATP binding cassette subfamily B member 1. The membrane-associated protein encoded by this gene is a member of the superfamily of ATP-binding cassette (ABC) transporters. ABC proteins transport various molecules across extra- and intra-cellular membranes. ABC genes are divided into seven distinct subfamilies (ABC1, MDR/TAP, MRP, ALD, OABP, GCN20, White). This protein is a member of the MDR/TAP subfamily. Members of the MDR/TAP subfamily are involved in multidrug resistance. The protein encoded by this gene is an ATP-dependent drug efflux pump for xenobiotic compounds with broad substrate specificity. It is responsible for decreased drug accumulation in multidrug-resistant cells and often mediates the development of resistance to anticancer drugs. This protein also functions as a transporter in the blood-brain barrier. Mutations in this gene are associated with colchicine resistance and Inflammatory bowel disease 13. Alternative splicing and the use of alternative promoters results in multiple transcript variants. [provided by RefSeq, Feb 2017].\"\n",
      "RIPK2_(4731),\"RIPK2 belongs to gene/protein node. RIPK2 is receptor interacting serine/threonine kinase 2. This gene encodes a member of the receptor-interacting protein (RIP) family of serine/threonine protein kinases. The encoded protein contains a C-terminal caspase activation and recruitment domain (CARD), and is a component of signaling complexes in both the innate and adaptive immune pathways. It is a potent activator of NF-kappaB and inducer of apoptosis in response to various stimuli. [provided by RefSeq, Jul 2008].\"\n",
      "IL12B_(6168),\"IL12B belongs to gene/protein node. IL12B is interleukin 12B. This gene encodes a subunit of interleukin 12, a cytokine that acts on T and natural killer cells, and has a broad array of biological activities. Interleukin 12 is a disulfide-linked heterodimer composed of the 40 kD cytokine receptor like subunit encoded by this gene, and a 35 kD subunit encoded by IL12A. This cytokine is expressed by activated macrophages that serve as an essential inducer of Th1 cells development. This cytokine has been found to be important for sustaining a sufficient number of memory/effector Th1 cells to mediate long-term protection to an intracellular pathogen. Overexpression of this gene was observed in the central nervous system of patients with multiple sclerosis (MS), suggesting a role of this cytokine in the pathogenesis of the disease. The promoter polymorphism of this gene has been reported to be associated with the severity of atopic and non-atopic asthma in children. [provided by RefSeq, Jul 2008].\"\n",
      "TYK2_(6428),\"TYK2 belongs to gene/protein node. TYK2 is tyrosine kinase 2. This gene encodes a member of the tyrosine kinase and, more specifically, the Janus kinases (JAKs) protein families. This protein associates with the cytoplasmic domain of type I and type II cytokine receptors and promulgate cytokine signals by phosphorylating receptor subunits. It is also a component of both the type I and type III interferon signaling pathways. As such, it may play a role in anti-viral immunity. A mutation in this gene has been associated with Immunodeficiency 35. [provided by RefSeq, Sep 2020].\"\n",
      "\n",
      "head_id,edge_type,tail_id\n",
      "Glucosamine_(16195),['drug' 'target' 'gene/protein'],MMP9_(3474)\n",
      "\"2-Amino-N,3,3-Trimethylbutanamide_(16196)\",['drug' 'target' 'gene/protein'],MMP9_(3474)\n",
      "\"(9Z,11E,13S)-13-hydroxyoctadeca-9,11-dienoic acid_(16613)\",['drug' 'target' 'gene/protein'],PPARG_(989)\n",
      "\"(4S,5E,7Z,10Z,13Z,16Z,19Z)-4-hydroxydocosa-5,7,10,13,16,19-hexaenoic acid_(16615)\",['drug' 'target' 'gene/protein'],PPARG_(989)\n",
      "Glucosamine_(16195),['drug' 'target' 'gene/protein'],TNF_(2329)\n",
      "Talmapimod_(17589),['drug' 'target' 'gene/protein'],TNF_(2329)\n",
      "Atiprimod_(17591),['drug' 'target' 'gene/protein'],TNF_(2329)\n",
      "Binimetinib_(15391),['drug' 'target' 'gene/protein'],TNF_(2329)\n",
      "Dilmapimod_(17602),['drug' 'target' 'gene/protein'],TNF_(2329)\n",
      "Tofacitinib_(14407),['drug' 'target' 'gene/protein'],JAK2_(1618)\n",
      "Fostamatinib_(15674),['drug' 'target' 'gene/protein'],JAK2_(1618)\n",
      "Diclofenac_(14061),['drug' 'transporter' 'gene/protein'],ABCB1_(4152)\n",
      "PRDM1_(2874),['gene/protein' 'associated with' 'disease'],ulcerative colitis (disease)_(37785)\n",
      "HNF4A_(279),['gene/protein' 'associated with' 'disease'],ulcerative colitis (disease)_(37785)\n",
      "HNF4A_(279),['gene/protein' 'associated with' 'disease'],inflammatory bowel disease_(28158)\n",
      "IL7R_(625),['gene/protein' 'associated with' 'disease'],ulcerative colitis (disease)_(37785)\n",
      "IL7R_(625),['gene/protein' 'associated with' 'disease'],inflammatory bowel disease_(28158)\n",
      "LSP1_(2643),['gene/protein' 'associated with' 'disease'],ulcerative colitis (disease)_(37785)\n",
      "MMP9_(3474),['gene/protein' 'associated with' 'disease'],inflammatory bowel disease_(28158)\n",
      "KIF21B_(8564),['gene/protein' 'associated with' 'disease'],ulcerative colitis (disease)_(37785)\n",
      "CARD9_(3460),['gene/protein' 'associated with' 'disease'],ulcerative colitis (disease)_(37785)\n",
      "NKX2-3_(34779),['gene/protein' 'associated with' 'disease'],ulcerative colitis (disease)_(37785)\n",
      "PPARG_(989),['gene/protein' 'associated with' 'disease'],inflammatory bowel disease_(28158)\n",
      "TCF7_(5195),['gene/protein' 'associated with' 'disease'],Crohn disease_(37784)\n",
      "TYK2_(6428),['gene/protein' 'associated with' 'disease'],inflammatory bowel disease_(28158)\n",
      "PHOX2B_(13014),['gene/protein' 'associated with' 'disease'],Crohn disease_(37784)\n",
      "INAVA_(9104),['gene/protein' 'associated with' 'disease'],inflammatory bowel disease_(28158)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'molecular_function'],transcription cis-regulatory region binding_(54288)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'molecular_function'],beta-catenin binding_(119055)\n",
      "IL7R_(625),['gene/protein' 'interacts with' 'molecular_function'],cytokine receptor activity_(54621)\n",
      "IL7R_(625),['gene/protein' 'interacts with' 'molecular_function'],interleukin-7 receptor activity_(122046)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'molecular_function'],DNA-binding transcription repressor activity_(54389)\n",
      "PPARG_(989),['gene/protein' 'interacts with' 'cellular_component'],nucleoplasm_(124249)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'cellular_component'],beta-catenin-TCF complex_(56282)\n",
      "IL7R_(625),['gene/protein' 'interacts with' 'biological_process'],cell surface receptor signaling pathway_(46168)\n",
      "TNF_(2329),['gene/protein' 'interacts with' 'biological_process'],positive regulation of peptidyl-serine phosphorylation_(52659)\n",
      "IL7R_(625),['gene/protein' 'interacts with' 'biological_process'],immune response_(39995)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'biological_process'],beta-catenin-TCF complex assembly_(104150)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'biological_process'],canonical Wnt signaling pathway_(46706)\n",
      "IL7R_(625),['gene/protein' 'interacts with' 'biological_process'],lymph node development_(107899)\n",
      "IL7R_(625),['gene/protein' 'interacts with' 'biological_process'],interleukin-7-mediated signaling pathway_(108951)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'biological_process'],regulation of gamma-delta T cell differentiation_(45624)\n",
      "IL7R_(625),['gene/protein' 'interacts with' 'biological_process'],positive regulation of receptor signaling pathway via STAT_(51389)\n",
      "IL7R_(625),['gene/protein' 'interacts with' 'biological_process'],regulation of DNA recombination_(43811)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'biological_process'],gamma-delta T cell differentiation_(51532)\n",
      "TCF7_(5195),['gene/protein' 'interacts with' 'biological_process'],T cell receptor V(D)J recombination_(108222)\n",
      "PPARG_(989),['gene/protein' 'target' 'drug'],Sulfasalazine_(15982)\n",
      "PPARG_(989),['gene/protein' 'target' 'drug'],Flufenamic acid_(14069)\n",
      "TNF_(2329),['gene/protein' 'target' 'drug'],Talmapimod_(17589)\n",
      "TNF_(2329),['gene/protein' 'target' 'drug'],VX-702_(17590)\n",
      "TNF_(2329),['gene/protein' 'target' 'drug'],Atiprimod_(17591)\n",
      "TNF_(2329),['gene/protein' 'target' 'drug'],Andrographolide_(17594)\n",
      "TNF_(2329),['gene/protein' 'target' 'drug'],Glycyrrhizic acid_(14107)\n",
      "RIPK2_(4731),['gene/protein' 'target' 'drug'],Fostamatinib_(15674)\n",
      "inflammatory bowel disease_(28158),['disease' 'associated with' 'gene/protein'],JAK2_(1618)\n",
      "inflammatory bowel disease_(28158),['disease' 'associated with' 'gene/protein'],PPARG_(989)\n",
      "inflammatory bowel disease_(28158),['disease' 'associated with' 'gene/protein'],TAGAP_(34814)\n",
      "inflammatory bowel disease_(28158),['disease' 'associated with' 'gene/protein'],LRRK2_(2111)\n",
      "inflammatory bowel disease_(28158),['disease' 'associated with' 'gene/protein'],DENND1B_(34887)\n",
      "transcription cis-regulatory region binding_(54288),['molecular_function' 'interacts with' 'gene/protein'],HNF4A_(279)\n",
      "cytokine receptor activity_(54621),['molecular_function' 'interacts with' 'gene/protein'],IL10RA_(1299)\n",
      "cytokine receptor activity_(54621),['molecular_function' 'interacts with' 'gene/protein'],IL12B_(6168)\n",
      "cytokine receptor activity_(54621),['molecular_function' 'interacts with' 'gene/protein'],IL23R_(34778)\n",
      "nucleoplasm_(124249),['cellular_component' 'interacts with' 'gene/protein'],ABCB1_(4152)\n",
      "beta-catenin-TCF complex_(56282),['cellular_component' 'interacts with' 'gene/protein'],TCF7_(5195)\n",
      "positive regulation of peptidyl-serine phosphorylation_(52659),['biological_process' 'interacts with' 'gene/protein'],RIPK2_(4731)\n",
      "immune response_(39995),['biological_process' 'interacts with' 'gene/protein'],IL18RAP_(11588)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(graph_dict['text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3a5f4dc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head_id</th>\n",
       "      <th>tail_id</th>\n",
       "      <th>edge_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10757</th>\n",
       "      <td>macrophage derived foam cell differentiation_(...</td>\n",
       "      <td>TGFB1_(2889)</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10508</th>\n",
       "      <td>positive regulation of tyrosine phosphorylatio...</td>\n",
       "      <td>IL21_(34967)</td>\n",
       "      <td>[biological_process, interacts with, gene/prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4864</th>\n",
       "      <td>IFNG_(3495)</td>\n",
       "      <td>positive regulation of tyrosine phosphorylatio...</td>\n",
       "      <td>[gene/protein, interacts with, biological_proc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5120</th>\n",
       "      <td>PPARG_(989)</td>\n",
       "      <td>macrophage derived foam cell differentiation_(...</td>\n",
       "      <td>[gene/protein, interacts with, biological_proc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1190</th>\n",
       "      <td>TGFB1_(2889)</td>\n",
       "      <td>inflammatory bowel disease, immunodeficiency, ...</td>\n",
       "      <td>[gene/protein, associated with, disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1192</th>\n",
       "      <td>IL21_(34967)</td>\n",
       "      <td>IL21-related infantile inflammatory bowel dise...</td>\n",
       "      <td>[gene/protein, associated with, disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6769</th>\n",
       "      <td>Crohn disease_(37784)</td>\n",
       "      <td>NKX2-3_(34779)</td>\n",
       "      <td>[disease, associated with, gene/protein]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6812</th>\n",
       "      <td>Crohn ileitis and jejunitis_(35814)</td>\n",
       "      <td>ATG16L1_(6661)</td>\n",
       "      <td>[disease, associated with, gene/protein]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6555</th>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>IL1B_(1004)</td>\n",
       "      <td>[disease, associated with, gene/protein]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>ICOSLG_(9454)</td>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>[gene/protein, associated with, disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1005</th>\n",
       "      <td>NKX2-3_(34779)</td>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>[gene/protein, associated with, disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1030</th>\n",
       "      <td>IFNG_(3495)</td>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>[gene/protein, associated with, disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1031</th>\n",
       "      <td>IFNG_(3495)</td>\n",
       "      <td>Crohn's colitis_(83770)</td>\n",
       "      <td>[gene/protein, associated with, disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1065</th>\n",
       "      <td>PPARG_(989)</td>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>[gene/protein, associated with, disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1151</th>\n",
       "      <td>IFNG_(3495)</td>\n",
       "      <td>Crohn ileitis and jejunitis_(35814)</td>\n",
       "      <td>[gene/protein, associated with, disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6629</th>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>ICOSLG_(9454)</td>\n",
       "      <td>[disease, associated with, gene/protein]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6641</th>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>NKX2-3_(34779)</td>\n",
       "      <td>[disease, associated with, gene/protein]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6666</th>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>IFNG_(3495)</td>\n",
       "      <td>[disease, associated with, gene/protein]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6701</th>\n",
       "      <td>inflammatory bowel disease_(28158)</td>\n",
       "      <td>PPARG_(989)</td>\n",
       "      <td>[disease, associated with, gene/protein]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 head_id  \\\n",
       "10757  macrophage derived foam cell differentiation_(...   \n",
       "10508  positive regulation of tyrosine phosphorylatio...   \n",
       "4864                                         IFNG_(3495)   \n",
       "5120                                         PPARG_(989)   \n",
       "1190                                        TGFB1_(2889)   \n",
       "1192                                        IL21_(34967)   \n",
       "6769                               Crohn disease_(37784)   \n",
       "6812                 Crohn ileitis and jejunitis_(35814)   \n",
       "6555                  inflammatory bowel disease_(28158)   \n",
       "993                                        ICOSLG_(9454)   \n",
       "1005                                      NKX2-3_(34779)   \n",
       "1030                                         IFNG_(3495)   \n",
       "1031                                         IFNG_(3495)   \n",
       "1065                                         PPARG_(989)   \n",
       "1151                                         IFNG_(3495)   \n",
       "6629                  inflammatory bowel disease_(28158)   \n",
       "6641                  inflammatory bowel disease_(28158)   \n",
       "6666                  inflammatory bowel disease_(28158)   \n",
       "6701                  inflammatory bowel disease_(28158)   \n",
       "\n",
       "                                                 tail_id  \\\n",
       "10757                                       TGFB1_(2889)   \n",
       "10508                                       IL21_(34967)   \n",
       "4864   positive regulation of tyrosine phosphorylatio...   \n",
       "5120   macrophage derived foam cell differentiation_(...   \n",
       "1190   inflammatory bowel disease, immunodeficiency, ...   \n",
       "1192   IL21-related infantile inflammatory bowel dise...   \n",
       "6769                                      NKX2-3_(34779)   \n",
       "6812                                      ATG16L1_(6661)   \n",
       "6555                                         IL1B_(1004)   \n",
       "993                   inflammatory bowel disease_(28158)   \n",
       "1005                  inflammatory bowel disease_(28158)   \n",
       "1030                  inflammatory bowel disease_(28158)   \n",
       "1031                             Crohn's colitis_(83770)   \n",
       "1065                  inflammatory bowel disease_(28158)   \n",
       "1151                 Crohn ileitis and jejunitis_(35814)   \n",
       "6629                                       ICOSLG_(9454)   \n",
       "6641                                      NKX2-3_(34779)   \n",
       "6666                                         IFNG_(3495)   \n",
       "6701                                         PPARG_(989)   \n",
       "\n",
       "                                               edge_type  \n",
       "10757  [biological_process, interacts with, gene/prot...  \n",
       "10508  [biological_process, interacts with, gene/prot...  \n",
       "4864   [gene/protein, interacts with, biological_proc...  \n",
       "5120   [gene/protein, interacts with, biological_proc...  \n",
       "1190            [gene/protein, associated with, disease]  \n",
       "1192            [gene/protein, associated with, disease]  \n",
       "6769            [disease, associated with, gene/protein]  \n",
       "6812            [disease, associated with, gene/protein]  \n",
       "6555            [disease, associated with, gene/protein]  \n",
       "993             [gene/protein, associated with, disease]  \n",
       "1005            [gene/protein, associated with, disease]  \n",
       "1030            [gene/protein, associated with, disease]  \n",
       "1031            [gene/protein, associated with, disease]  \n",
       "1065            [gene/protein, associated with, disease]  \n",
       "1151            [gene/protein, associated with, disease]  \n",
       "6629            [disease, associated with, gene/protein]  \n",
       "6641            [disease, associated with, gene/protein]  \n",
       "6666            [disease, associated with, gene/protein]  \n",
       "6701            [disease, associated with, gene/protein]  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "814cb74b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('protein kinase A binding_(54006)',\n",
       "  {'desc': 'protein kinase A binding belongs to molecular_function node. ',\n",
       "   'node_type': 'molecular_function',\n",
       "   'node_name': 'protein kinase A binding',\n",
       "   'enriched_node': 'Binding to a protein kinase A.'}),\n",
       " ('RNA polymerase II-specific DNA-binding transcription factor binding_(54015)',\n",
       "  {'desc': 'RNA polymerase II-specific DNA-binding transcription factor binding belongs to molecular_function node. ',\n",
       "   'node_type': 'molecular_function',\n",
       "   'node_name': 'RNA polymerase II-specific DNA-binding transcription factor binding',\n",
       "   'enriched_node': 'Binding to a sequence-specific DNA binding RNA polymerase II transcription factor, any of the factors that interact selectively and non-covalently with a specific DNA sequence in order to modulate transcription.'}),\n",
       " ('enzyme binding_(54290)',\n",
       "  {'desc': 'enzyme binding belongs to molecular_function node. ',\n",
       "   'node_type': 'molecular_function',\n",
       "   'node_name': 'enzyme binding',\n",
       "   'enriched_node': 'Binding to an enzyme, a protein with catalytic activity.'}),\n",
       " ('antibacterial humoral response_(45652)',\n",
       "  {'desc': 'antibacterial humoral response belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'antibacterial humoral response',\n",
       "   'enriched_node': 'An immune response against bacteria mediated through a body fluid. Examples of this process are the antibacterial humoral responses in Mus musculus and Drosophila melanogaster.'}),\n",
       " ('5-[4-(1-Carboxymethyl-2-Oxo-Propylcarbamoyl)-Benzylsulfamoyl]-2-Hydroxy-Benzoic Acid_(18703)',\n",
       "  {'desc': '5-[4-(1-Carboxymethyl-2-Oxo-Propylcarbamoyl)-Benzylsulfamoyl]-2-Hydroxy-Benzoic Acid belongs to drug node.     ',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': '5-[4-(1-Carboxymethyl-2-Oxo-Propylcarbamoyl)-Benzylsulfamoyl]-2-Hydroxy-Benzoic Acid',\n",
       "   'enriched_node': '[H][C@@](CC(O)=O)(NC(=O)C1=CC=C(CNS(=O)(=O)C2=CC(C(O)=O)=C(O)C=C2)C=C1)C(C)=O'}),\n",
       " ('microtubule-based movement_(45481)',\n",
       "  {'desc': 'microtubule-based movement belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'microtubule-based movement',\n",
       "   'enriched_node': 'A microtubule-based process that results in the movement of organelles, other microtubules, or other cellular components. Examples include motor-driven movement along microtubules and movement driven by polymerization or depolymerization of microtubules.'}),\n",
       " ('SPOTS complex_(126815)',\n",
       "  {'desc': 'SPOTS complex belongs to cellular_component node. ',\n",
       "   'node_type': 'cellular_component',\n",
       "   'node_name': 'SPOTS complex',\n",
       "   'enriched_node': 'A multiprotein complex at least composed of serine palmitoyltransferases and ORM proteins (known as ORMDL proteins in mammals and other higher vertebrates) that plays a key role in sphingolipid homeostasis.'}),\n",
       " ('microtubule cytoskeleton_(126859)',\n",
       "  {'desc': 'microtubule cytoskeleton belongs to cellular_component node. ',\n",
       "   'node_type': 'cellular_component',\n",
       "   'node_name': 'microtubule cytoskeleton',\n",
       "   'enriched_node': 'The part of the cytoskeleton (the internal framework of a cell) composed of microtubules and associated proteins.'}),\n",
       " ('focal adhesion_(126873)',\n",
       "  {'desc': 'focal adhesion belongs to cellular_component node. ',\n",
       "   'node_type': 'cellular_component',\n",
       "   'node_name': 'focal adhesion',\n",
       "   'enriched_node': 'A cell-substrate junction that anchors the cell to the extracellular matrix and that forms a point of termination of actin filaments. In insects focal adhesion has also been referred to as hemi-adherens junction (HAJ).'}),\n",
       " ('NF-kappaB p50/p65 complex_(127022)',\n",
       "  {'desc': 'NF-kappaB p50/p65 complex belongs to cellular_component node. ',\n",
       "   'node_type': 'cellular_component',\n",
       "   'node_name': 'NF-kappaB p50/p65 complex',\n",
       "   'enriched_node': 'A heterodimer of NF-kappa B p50 and p65 subunits.'}),\n",
       " ('negative regulation of late endosome to lysosome transport_(111364)',\n",
       "  {'desc': 'negative regulation of late endosome to lysosome transport belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'negative regulation of late endosome to lysosome transport',\n",
       "   'enriched_node': 'Any process that stops, prevents or reduces the frequency, rate or extent of late endosome to lysosome transport.'}),\n",
       " ('positive regulation of synaptic vesicle endocytosis_(111372)',\n",
       "  {'desc': 'positive regulation of synaptic vesicle endocytosis belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'positive regulation of synaptic vesicle endocytosis',\n",
       "   'enriched_node': 'Any process that activates or increases the frequency, rate or extent of synaptic vesicle endocytosis.'}),\n",
       " ('antigen processing and presentation of endogenous peptide antigen via MHC class II_(111399)',\n",
       "  {'desc': 'antigen processing and presentation of endogenous peptide antigen via MHC class II belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'antigen processing and presentation of endogenous peptide antigen via MHC class II',\n",
       "   'enriched_node': 'The process in which an antigen-presenting cell expresses a peptide antigen of endogenous origin on its cell surface in association with an MHC class II protein complex. The peptide antigen is typically, but not always, processed from a whole protein.'}),\n",
       " ('regulation of proteasomal ubiquitin-dependent protein catabolic process_(49789)',\n",
       "  {'desc': 'regulation of proteasomal ubiquitin-dependent protein catabolic process belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'regulation of proteasomal ubiquitin-dependent protein catabolic process',\n",
       "   'enriched_node': 'Any process that modulates the frequency, rate or extent of the breakdown of a protein or peptide by hydrolysis of its peptide bonds, initiated by the covalent attachment of ubiquitin, and mediated by the proteasome.'}),\n",
       " ('Ziprasidone_(14140)',\n",
       "  {'desc': 'Ziprasidone belongs to drug node. Disorders such as schizophrenia and bipolar disorder can significantly impair mood, cognition, and behavior. These mental illnesses can often be accompanied by comorbidities such as depression and substance abuse, and can significantly impact the quality of life of patients and caregivers. Luckily, several treatment options for psychotic disorders have been introduced to market since the realization of chlorpromazine\\'s antipsychotic properties in 1952.  The effects of ziprasidone are differentiated from other antispychotics based on its preference and affinity for certain receptors. Ziprasidone binds to serotonin-2A (5-HT2A) and dopamine D2 receptors in a similar fashion to other atypical antipsychotics; however, one key difference is that ziprasidone has a higher 5-HT2A/D2 receptor affinity ratio when compared to other antipsychotics such as olanzapine, quetiapine, risperidone, and aripiprazole. Ziprasidone offers enhanced modulation of mood, notable negative symptom relief, overall cognitive improvement and reduced motor dysfunction which is linked to it\\'s potent interaction with 5-HT2C, 5-HT1D, and 5-HT1A receptors in brain tissue. Ziprasidone can bind moderately to norepinephrine and serotonin reuptake sites which may contribute to its antidepressant and anxiolytic activity. Patient\\'s taking ziprasidone will likely experience a lower incidence of orthostatic hypotension, cognitive disturbance, sedation, weight gain, and disruption in prolactin levels since ziprasidone has a lower affinity for histamine H1, muscarinic M1, and alpha1-adrenoceptors.  Ziprasidone is extensively protein bound with over 99% of the drug bound to plasma proteins, primarily albumin and alpha1-acid glycoprotein. Ziprasidone is classified as a \\\\second generation\\\\\" or \\\\\"atypical\\\\\" antipsychotic and is a dopamine and 5HT2A receptor antagonist with a unique receptor binding profile. As previously mentioned, ziprasidone has a very high 5-HT2A/D2 affinity ratio, binds to multiple serotonin receptors in addition to 5-HT2A, and blocks monoamine transporters which prevents 5HT and NE reuptake. On the other hand, ziprasidone has a low affinity for muscarinic cholinergic M1, histamine H1, and alpha1-adrenergic receptors.\" In its oral form, ziprasidone is approved for the treatment of schizophrenia, as monotherapy for acute treatment of manic or mixed episodes related to bipolar I disorder, and as adjunctive therapy to lithium or valproate for maintece treatment of bipolar I disorder. The injectable formulation is approved only for treatment of acute agitation in schizophrenia.',\n",
       "   'node_type': 'drug',\n",
       "   'node_name': 'Ziprasidone',\n",
       "   'enriched_node': 'ClC1=C(CCN2CCN(CC2)C2=NSC3=CC=CC=C23)C=C2CC(=O)NC2=C1'}),\n",
       " ('behavioral response to nicotine_(110055)',\n",
       "  {'desc': 'behavioral response to nicotine belongs to biological_process node. ',\n",
       "   'node_type': 'biological_process',\n",
       "   'node_name': 'behavioral response to nicotine',\n",
       "   'enriched_node': 'Any process that results in a change in the behavior of an organism as a result of a nicotine stimulus.'})]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_dict[\"nodes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34bfa385",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f5443b44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>enriched_node</th>\n",
       "      <th>x</th>\n",
       "      <th>desc</th>\n",
       "      <th>desc_x</th>\n",
       "      <th>use_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL7R_(625)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MTILGTTFGMVFSLLQVVSGESGYAQNGDLEDAELDDYSFSCYSQL...</td>\n",
       "      <td>[0.06364653259515762, 0.06951971352100372, 0.0...</td>\n",
       "      <td>IL7R belongs to gene/protein node. IL7R is int...</td>\n",
       "      <td>[0.04506902, 0.008911126, -0.17318207, -0.0157...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCF7_(5195)</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MPQLDSGGGGAGGGDDLGAPDELLAFQDEGEEQDDKSRDSAAGPER...</td>\n",
       "      <td>[0.028321774676442146, 0.003539376426488161, 0...</td>\n",
       "      <td>TCF7 belongs to gene/protein node. TCF7 is tra...</td>\n",
       "      <td>[0.036997586, 0.038098544, -0.19027671, -0.006...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>user_prompt</td>\n",
       "      <td>prompt</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>[0.047626022, 0.05007147, -0.16455309, -0.0442...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>[0.047626022, 0.05007147, -0.16455309, -0.0442...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_id     node_type  \\\n",
       "0   IL7R_(625)  gene/protein   \n",
       "1  TCF7_(5195)  gene/protein   \n",
       "2  user_prompt        prompt   \n",
       "\n",
       "                                       enriched_node  \\\n",
       "0  MTILGTTFGMVFSLLQVVSGESGYAQNGDLEDAELDDYSFSCYSQL...   \n",
       "1  MPQLDSGGGGAGGGDDLGAPDELLAFQDEGEEQDDKSRDSAAGPER...   \n",
       "2                                               <NA>   \n",
       "\n",
       "                                                   x  \\\n",
       "0  [0.06364653259515762, 0.06951971352100372, 0.0...   \n",
       "1  [0.028321774676442146, 0.003539376426488161, 0...   \n",
       "2  [0.047626022, 0.05007147, -0.16455309, -0.0442...   \n",
       "\n",
       "                                                desc  \\\n",
       "0  IL7R belongs to gene/protein node. IL7R is int...   \n",
       "1  TCF7 belongs to gene/protein node. TCF7 is tra...   \n",
       "2                                               <NA>   \n",
       "\n",
       "                                              desc_x  use_description  \n",
       "0  [0.04506902, 0.008911126, -0.17318207, -0.0157...            False  \n",
       "1  [0.036997586, 0.038098544, -0.19027671, -0.006...            False  \n",
       "2  [0.047626022, 0.05007147, -0.16455309, -0.0442...             True  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4b34aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cuvs.distance import pairwise_distance\n",
    "def _compute_sim_scores(features_a: cp.ndarray,\n",
    "                            features_b:  cp.ndarray,\n",
    "                            metric: str=\"cosine\"):\n",
    "        \"\"\"\n",
    "        Compute the similarity scores between two sets of features using the specified metric.\n",
    "\n",
    "        Args:\n",
    "            features_a: The first set of features.\n",
    "            features_b: The second set of features.\n",
    "            metric: The metric to use for computing the similarity scores.\n",
    "        \n",
    "        Returns:\n",
    "            The similarity scores between the two sets of features.\n",
    "        \"\"\"\n",
    "        scores = pairwise_distance(features_a, features_b, metric=metric)\n",
    "        scores = 1 - cp.asarray(scores).ravel()\n",
    "        return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59e5f2b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_emb = cp.array(query_df.iloc[2]['x'][2]).reshape(1, -1).astype(cp.float32)\n",
    "query_emb.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d92ce08e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_scores = cudf.Series(cp.zeros(len(graph_nodes), dtype=cp.float32))\n",
    "sim_scores[:] = _compute_sim_scores(\n",
    "                graph_nodes[\"desc_x\"].list.leaves.to_cupy().reshape(\n",
    "                    -1, len(graph_nodes[\"desc_x\"][0])\n",
    "                ).astype(cp.float32),\n",
    "                query_emb\n",
    "            )  # shape [N, 1]\n",
    "sim_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f776c4e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "isinstance(sim_scores, cudf.Series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "323c234d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize dataframes\n",
    "multimodal_df = cudf.DataFrame({\"name\": [], \"node_type\": []})\n",
    "query_df = cudf.DataFrame({\"node_id\": [],\n",
    "                            \"node_type\": [],\n",
    "                            \"x\": [],\n",
    "                            \"desc_x\": [],\n",
    "                            \"use_description\": []})\n",
    "\n",
    "# Loop over the uploaded files and find multimodal files\n",
    "for i in range(len(state[\"uploaded_files\"])):\n",
    "    # Check if multimodal file is uploaded\n",
    "    if state[\"uploaded_files\"][i][\"file_type\"] == \"multimodal\":\n",
    "        # Read the Excel file\n",
    "        multimodal_df = cudf.read_csv(state[\"uploaded_files\"][i][\"file_path\"])\n",
    "\n",
    "# Check if the multimodal_df is empty\n",
    "if len(multimodal_df) > 0:\n",
    "    # Merge all obtained dataframes into a single dataframe\n",
    "    multimodal_df.rename(columns={\"name\": \"q_node_name\", \"node_type\": \"q_node_type\"}, inplace=True)\n",
    "\n",
    "    # Make and process a query dataframe by merging the graph_df and multimodal_df\n",
    "    query_df = nodes_df[['node_id', 'node_name', 'node_type', 'enriched_node', 'x', 'desc', 'desc_x']].merge(multimodal_df, how='cross')\n",
    "    query_df['q_node_name'] = query_df['q_node_name'].str.lower()\n",
    "    query_df['node_name'] = query_df['node_name'].str.lower()\n",
    "    # Get the mask for filtering based on the query\n",
    "    mask = (\n",
    "        query_df['node_name'].str.contains(query_df['q_node_name']) &\n",
    "        (query_df['node_type'] == query_df['q_node_type'])\n",
    "    )\n",
    "    query_df = query_df[mask]\n",
    "    query_df = query_df[['node_id', 'node_type', 'enriched_node', 'x', 'desc', 'desc_x']].reset_index(drop=True)\n",
    "    query_df['use_description'] = False # set to False for modal-specific embeddings\n",
    "\n",
    "    # Update the state by adding the the selected node IDs\n",
    "    state[\"selections\"] = query_df.to_pandas().groupby(\"node_type\")[\"node_id\"].apply(list).to_dict()\n",
    "\n",
    "# Append a user prompt to the query dataframe\n",
    "query_df = cudf.concat([\n",
    "    query_df,\n",
    "    cudf.DataFrame({\n",
    "        'node_id': 'user_prompt',\n",
    "        'node_type': 'prompt',\n",
    "        # 'enriched_node': prompt,\n",
    "        'x': prompt_emb,\n",
    "        # 'desc': prompt,\n",
    "        'desc_x': prompt_emb,\n",
    "        'use_description': True # set to True for user prompt embedding\n",
    "    })\n",
    "]).reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "180989a4",
   "metadata": {},
   "source": [
    "### Before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d95dbd01",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import Data\n",
    "\n",
    "topk = state[\"topk_nodes\"]  \n",
    "topk_e = state[\"topk_edges\"]\n",
    "c_const = 0.01\n",
    "\n",
    "def _compute_node_prizes(graph: Data,\n",
    "                         query_emb: torch.Tensor,\n",
    "                         modality: str,\n",
    "                         use_description: bool=False) :\n",
    "    \"\"\"\n",
    "    Compute the node prizes based on the cosine similarity between the query and nodes.\n",
    "\n",
    "    Args:\n",
    "        graph: The knowledge graph in PyTorch Geometric Data format.\n",
    "        query_emb: The query embedding in PyTorch Tensor format. This can be an embedding of\n",
    "            a prompt, sequence, or any other feature to be used for the subgraph extraction.\n",
    "        modality: The modality to use for the subgraph extraction based on the node type.\n",
    "\n",
    "    Returns:\n",
    "        The prizes of the nodes.\n",
    "    \"\"\"\n",
    "    # Convert PyG graph to a DataFrame\n",
    "    graph_df = cudf.DataFrame({\n",
    "        \"node_type\": graph.node_type,\n",
    "        \"desc_x\": [x.tolist() for x in graph.desc_x],\n",
    "        \"x\": [list(x) for x in graph.x],\n",
    "        \"score\": [0.0 for _ in range(len(graph.node_id))],\n",
    "    })\n",
    "\n",
    "    # Calculate cosine similarity for text features and update the score\n",
    "    if use_description:\n",
    "        graph_df.loc[:, \"score\"] = torch.nn.CosineSimilarity(dim=-1)(\n",
    "                query_emb,\n",
    "                torch.tensor(list(graph_df.desc_x.values)) # Using textual description features\n",
    "            ).tolist()\n",
    "    else:\n",
    "        graph_df.loc[graph_df[\"node_type\"] == modality,\n",
    "                        \"score\"] = torch.nn.CosineSimilarity(dim=-1)(\n",
    "                query_emb,\n",
    "                torch.tensor(list(graph_df[graph_df[\"node_type\"]== modality].x.values))\n",
    "            ).tolist()\n",
    "\n",
    "    # Set the prizes for nodes based on the similarity scores\n",
    "    n_prizes = torch.tensor(graph_df.score.values, dtype=torch.float32)\n",
    "    # n_prizes = torch.nn.CosineSimilarity(dim=-1)(query_emb, graph.x)\n",
    "    topk = min(topk, graph.num_nodes)\n",
    "    _, topk_n_indices = torch.topk(n_prizes, topk, largest=True)\n",
    "    n_prizes = torch.zeros_like(n_prizes)\n",
    "    n_prizes[topk_n_indices] = torch.arange(topk, 0, -1).float()\n",
    "\n",
    "    return n_prizes\n",
    "\n",
    "def _compute_edge_prizes(graph: Data,\n",
    "                         text_emb: torch.Tensor) :\n",
    "    \"\"\"\n",
    "    Compute the node prizes based on the cosine similarity between the query and nodes.\n",
    "\n",
    "    Args:\n",
    "        graph: The knowledge graph in PyTorch Geometric Data format.\n",
    "        text_emb: The textual description embedding in PyTorch Tensor format.\n",
    "\n",
    "    Returns:\n",
    "        The prizes of the nodes.\n",
    "    \"\"\"\n",
    "    # Note that as of now, the edge features are based on textual features\n",
    "    # Compute prizes for edges\n",
    "    e_prizes = torch.nn.CosineSimilarity(dim=-1)(text_emb, graph.edge_attr)\n",
    "    unique_prizes, inverse_indices = e_prizes.unique(return_inverse=True)\n",
    "    topk_e = min(topk_e, unique_prizes.size(0))\n",
    "    topk_e_values, _ = torch.topk(unique_prizes, topk_e, largest=True)\n",
    "    e_prizes[e_prizes < topk_e_values[-1]] = 0.0\n",
    "    last_topk_e_value = topk_e\n",
    "    for k in range(topk_e):\n",
    "        indices = inverse_indices == (\n",
    "            unique_prizes == topk_e_values[k]\n",
    "        ).nonzero(as_tuple=True)[0]\n",
    "        value = min((topk_e - k) / indices.sum().item(), last_topk_e_value)\n",
    "        e_prizes[indices] = value\n",
    "        last_topk_e_value = value * (1 - c_const)\n",
    "\n",
    "    return e_prizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af9a52c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "graph = initial_graph[\"pyg\"]\n",
    "text_emb = torch.tensor(query_df.iloc[0]['desc_x'][0])\n",
    "query_emb = torch.tensor(query_df.iloc[0]['x'][0])\n",
    "modality = query_df.iloc[0]['node_type'][0]\n",
    "\n",
    "# Convert PyG graph to a DataFrame\n",
    "graph_df = pd.DataFrame({\n",
    "    \"node_type\": graph.node_type,\n",
    "    \"desc_x\": [x.tolist() for x in graph.desc_x],\n",
    "    \"x\": [list(x) for x in graph.x],\n",
    "    \"score\": [0.0 for _ in range(len(graph.node_id))],\n",
    "})\n",
    "\n",
    "graph_df.loc[graph_df[\"node_type\"] == modality, \"score\"]  = torch.nn.CosineSimilarity(dim=-1)(\n",
    "        query_emb,\n",
    "        torch.tensor(list(graph_df[graph_df[\"node_type\"]== modality].x.values))\n",
    "    ).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04658e5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the prizes for nodes based on the similarity scores\n",
    "n_prizes = torch.tensor(graph_df.score.values, dtype=torch.float32)\n",
    "topk = min(topk, graph.num_nodes)\n",
    "_, topk_n_indices = torch.topk(n_prizes, topk, largest=True)\n",
    "n_prizes = torch.zeros_like(n_prizes)\n",
    "n_prizes[topk_n_indices] = torch.arange(topk, 0, -1).float()\n",
    "n_prizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "312b9608",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_edges = edges_df\n",
    "c_const = cfg.c_const\n",
    "\n",
    "e_prizes = torch.nn.CosineSimilarity(dim=-1)(torch.tensor(text_emb), torch.tensor([list(f) for f in graph_edges.edge_attr.to_arrow().to_pandas().values]))\n",
    "unique_prizes, inverse_indices = e_prizes.unique(return_inverse=True)\n",
    "topk_e = min(topk_e, unique_prizes.size(0))\n",
    "topk_e_values, _ = torch.topk(unique_prizes, topk_e, largest=True)\n",
    "e_prizes[e_prizes < topk_e_values[-1]] = 0.0\n",
    "last_topk_e_value = topk_e\n",
    "for k in range(topk_e):\n",
    "    print(k, (unique_prizes == topk_e_values[k]).nonzero())\n",
    "    indices = inverse_indices == (\n",
    "        unique_prizes == topk_e_values[k]\n",
    "    ).nonzero(as_tuple=True)[0]\n",
    "    value = min((topk_e - k) / indices.sum().item(), last_topk_e_value)\n",
    "    e_prizes[indices] = value\n",
    "    last_topk_e_value = value * (1 - c_const)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "197851ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "prizes = {\"nodes\": n_prizes, \"edges\": e_prizes}\n",
    "cost_e = cfg.cost_e\n",
    "c_const = cfg.c_const\n",
    "\n",
    "# Logic to reduce the cost of the edges such that at least one edge is selected\n",
    "updated_cost_e = min(\n",
    "    cost_e,\n",
    "    prizes[\"edges\"].max().item() * (1 - c_const / 2),\n",
    ")\n",
    "\n",
    "# Initialize variables\n",
    "edges = []\n",
    "costs = []\n",
    "virtual = {\n",
    "    \"n_prizes\": [],\n",
    "    \"edges\": [],\n",
    "    \"costs\": [],\n",
    "}\n",
    "mapping = {\"nodes\": {}, \"edges\": {}}\n",
    "\n",
    "# Compute the costs, edges, and virtual variables based on the prizes\n",
    "for i, (src, dst) in enumerate(graph.edge_index.T.numpy()):\n",
    "    prize_e = prizes[\"edges\"][i]\n",
    "    if prize_e <= updated_cost_e:\n",
    "        mapping[\"edges\"][len(edges)] = i\n",
    "        edges.append((src, dst))\n",
    "        costs.append(updated_cost_e - prize_e)\n",
    "    else:\n",
    "        virtual_node_id = graph.num_nodes + len(virtual[\"n_prizes\"])\n",
    "        mapping[\"nodes\"][virtual_node_id] = i\n",
    "        virtual[\"edges\"].append((src, virtual_node_id))\n",
    "        virtual[\"edges\"].append((virtual_node_id, dst))\n",
    "        virtual[\"costs\"].append(0)\n",
    "        virtual[\"costs\"].append(0)\n",
    "        virtual[\"n_prizes\"].append(prize_e - updated_cost_e)\n",
    "prizes = np.concatenate([prizes[\"nodes\"], np.array(virtual[\"n_prizes\"])])\n",
    "edges_dict = {}\n",
    "edges_dict[\"edges\"] = edges\n",
    "edges_dict[\"num_prior_edges\"] = len(edges)\n",
    "# Final computation of the costs and edges based on the virtual costs and virtual edges\n",
    "if len(virtual[\"costs\"]) > 0:\n",
    "    costs = np.array(costs + virtual[\"costs\"])\n",
    "    edges = np.array(edges + virtual[\"edges\"])\n",
    "    edges_dict[\"edges\"] = edges\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15b7afc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.unique(prizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f70ada5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.unique(prizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e88bb02e",
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_dict[\"edges\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fee4de95",
   "metadata": {},
   "outputs": [],
   "source": [
    "costs + virtual[\"costs\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95b5ab4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(virtual[\"costs\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcc3382e",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(virtual[\"costs\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10f44592",
   "metadata": {},
   "outputs": [],
   "source": [
    "costs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd21c253",
   "metadata": {},
   "outputs": [],
   "source": [
    "e_prizes.nonzero()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b2e359",
   "metadata": {},
   "outputs": [],
   "source": [
    "e_prizes[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adee5a0e",
   "metadata": {},
   "source": [
    "### After"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adbf20a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize dataframes\n",
    "multimodal_df = cudf.DataFrame({\"name\": [], \"node_type\": []})\n",
    "query_df = cudf.DataFrame({\"node_id\": [],\n",
    "                            \"node_type\": [],\n",
    "                            \"x\": [],\n",
    "                            \"desc_x\": [],\n",
    "                            \"use_description\": []})\n",
    "\n",
    "# Loop over the uploaded files and find multimodal files\n",
    "for i in range(len(state[\"uploaded_files\"])):\n",
    "    # Check if multimodal file is uploaded\n",
    "    if state[\"uploaded_files\"][i][\"file_type\"] == \"multimodal\":\n",
    "        # Read the csv file\n",
    "        multimodal_df = cudf.read_csv(state[\"uploaded_files\"][i][\"file_path\"])\n",
    "\n",
    "# Check if the multimodal_df is empty\n",
    "if len(multimodal_df) > 0:\n",
    "    # Prepare multimodal_df\n",
    "    multimodal_df.rename(columns={\"name\": \"q_node_name\",\n",
    "                                    \"node_type\": \"q_node_type\"}, inplace=True)\n",
    "\n",
    "    # Make and process a query dataframe by merging the graph_df and multimodal_df\n",
    "    query_df = graph_nodes[\n",
    "        ['node_id', 'node_name', 'node_type', 'enriched_node', 'x', 'desc', 'desc_x']\n",
    "    ].merge(multimodal_df, how='cross')\n",
    "#     query_df['q_node_name'] = query_df['q_node_name'].str.lower()\n",
    "#     query_df['node_name'] = query_df['node_name'].str.lower()\n",
    "#     # Get the mask for filtering based on the query\n",
    "#     mask = (\n",
    "#         query_df['node_name'].str.contains(query_df['q_node_name']) &\n",
    "#         (query_df['node_type'] == query_df['q_node_type'])\n",
    "#     )\n",
    "#     query_df = query_df[mask]\n",
    "#     query_df = query_df[['node_id',\n",
    "#                             'node_type', \n",
    "#                             'enriched_node', \n",
    "#                             'x', \n",
    "#                             'desc', \n",
    "#                             'desc_x']].reset_index(drop=True)\n",
    "#     query_df['use_description'] = False # set to False for modal-specific embeddings\n",
    "\n",
    "#     # Update the state by adding the the selected node IDs\n",
    "#     state[\"selections\"] = query_df.to_pandas().groupby(\n",
    "#         \"node_type\"\n",
    "#     )[\"node_id\"].apply(list).to_dict()\n",
    "\n",
    "# # Append a user prompt to the query dataframe\n",
    "# query_df = cudf.concat([\n",
    "#     query_df,\n",
    "#     cudf.DataFrame({\n",
    "#         'node_id': 'user_prompt',\n",
    "#         'node_type': 'prompt',\n",
    "#         # 'enriched_node': prompt,\n",
    "#         'x': prompt_emb,\n",
    "#         # 'desc': prompt,\n",
    "#         'desc_x': prompt_emb,\n",
    "#         'use_description': True # set to True for user prompt embedding\n",
    "#     })\n",
    "# ]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64a2c611",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ec29691",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _compute_sim_scores(features_a, features_b, metric=\"cosine\"):\n",
    "    scores = cuvs.distance.pairwise_distance(features_a, features_b, metric=metric)\n",
    "    scores = 1 - cp.asarray(scores).ravel()\n",
    "    return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27f72c0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_nodes = nodes_df\n",
    "graph_edges = edges_df\n",
    "topk = cfg.topk\n",
    "topk_e = cfg.topk_e\n",
    "use_description = False\n",
    "# query_emb = torch.tensor(query_df.iloc[0]['x'][0]) # torch.Size([2560])\n",
    "text_emb = cp.array(query_df.iloc[1]['desc_x'][1]).reshape(1, -1).astype(cp.float32)\n",
    "query_emb = cp.array(query_df.iloc[1]['x'][1]).reshape(1, -1).astype(cp.float32)\n",
    "\n",
    "# Initialize variables\n",
    "sim_scores = cudf.Series(cp.zeros(len(graph_nodes), dtype=cp.float32))\n",
    "mask = (graph_nodes.node_type == modality)\n",
    "\n",
    "# Calculate cosine similarity for text features and update the score\n",
    "if use_description:\n",
    "    sim_scores = _compute_sim_scores(\n",
    "        graph_nodes[\"desc_x\"].list.leaves.to_cupy().reshape(-1, len(graph_nodes[\"desc_x\"][0])).astype(cp.float32),\n",
    "        query_emb\n",
    "    )  # shape [N, 1]\n",
    "else:\n",
    "    sim_scores[mask] = _compute_sim_scores(\n",
    "        graph_nodes[mask][\"x\"].list.leaves.to_cupy().reshape(-1, len(graph_nodes[mask][\"x\"][0])).astype(cp.float32),\n",
    "        query_emb\n",
    "    )  # shape [N, 1]\n",
    "\n",
    "# Set the prizes for nodes based on the similarity scores\n",
    "# n_prizes = torch.tensor(graph_df.score.values, dtype=torch.float32)\n",
    "topk = min(topk, sim_scores.size)\n",
    "n_prizes = cudf.Series(0.0, index=cp.arange(sim_scores.size))\n",
    "n_prizes[(-sim_scores).sort_values()[:topk].index] = cp.arange(topk, 0, -1).astype(cp.float32)\n",
    "n_prizes = n_prizes.to_cupy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10bbd838",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sim_scores = cudf.Series(cp.zeros(len(graph_edges), dtype=cp.float32))\n",
    "\n",
    "e_prizes = _compute_sim_scores(\n",
    "    graph_edges[\"edge_attr\"].list.leaves.to_cupy().reshape(-1, len(graph_edges[\"edge_attr\"][0])).astype(cp.float32),\n",
    "    text_emb)\n",
    "\n",
    "unique_prizes, inverse_indices = cp.unique(e_prizes, return_inverse=True)\n",
    "topk_e = min(topk_e, sim_scores.size) \n",
    "topk_e_values = unique_prizes[cp.argsort(-unique_prizes)[:topk_e]]\n",
    "e_prizes[e_prizes < topk_e_values[-1]] = 0.0\n",
    "last_topk_e_value = topk_e\n",
    "for k in range(topk_e):\n",
    "    indices = inverse_indices == (unique_prizes == topk_e_values[k]).nonzero()[0]\n",
    "    value = min((topk_e - k) / indices.sum().item(), last_topk_e_value)\n",
    "    e_prizes[indices] = value\n",
    "    last_topk_e_value = value * (1 - c_const)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf72d5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "prizes = {\"nodes\": n_prizes, \"edges\": e_prizes}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a8a9407",
   "metadata": {},
   "outputs": [],
   "source": [
    "cost_e = cfg.cost_e\n",
    "\n",
    "# Logic to reduce the cost of the edges such that at least one edge is selected\n",
    "updated_cost_e = min(\n",
    "    cost_e,\n",
    "    prizes[\"edges\"].max().item() * (1 - c_const / 2),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b50f9175",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def _create_edge_index(graph_nodes, graph_edges):\n",
    "#     # Create and additional node_index column\n",
    "#     graph_nodes = graph_nodes.reset_index(drop=True)\n",
    "#     graph_nodes['node_index'] = graph_nodes.index\n",
    "\n",
    "#     # Get head_index and tail_index\n",
    "#     edges = graph_edges.merge(graph_nodes[['node_id', 'node_index']],\n",
    "#                             left_on='head_id', right_on='node_id',\n",
    "#                             how='left').rename(columns={'node_index': 'head_index'}).drop(columns=['node_id'])\n",
    "#     edges = edges.merge(graph_nodes[['node_id', 'node_index']],\n",
    "#                         left_on='tail_id', right_on='node_id',\n",
    "#                         how='left').rename(columns={'node_index': 'tail_index'}).drop(columns=['node_id'])\n",
    "\n",
    "#     # Stacking to get into edge_index\n",
    "#     edge_index = cp.stack([\n",
    "#         edges['head_index'].to_cupy(),\n",
    "#         edges['tail_index'].to_cupy()\n",
    "#     ])\n",
    "\n",
    "#     return edge_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b9d327f",
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_index = cp.stack([\n",
    "    graph_edges['head_index'].to_cupy(),\n",
    "    graph_edges['tail_index'].to_cupy()\n",
    "])\n",
    "edge_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502198f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "prizes = {\"nodes\": n_prizes, \"edges\": e_prizes}\n",
    "\n",
    "cost_e = cfg.cost_e\n",
    "\n",
    "# Logic to reduce the cost of the edges such that at least one edge is selected\n",
    "updated_cost_e = min(\n",
    "    cost_e,\n",
    "    prizes[\"edges\"].max().item() * (1 - c_const / 2),\n",
    ")\n",
    "\n",
    "# Initialize variables\n",
    "edges = []\n",
    "costs = []\n",
    "virtual = {\n",
    "    \"n_prizes\": [],\n",
    "    \"edges\": [],\n",
    "    \"costs\": [],\n",
    "}\n",
    "mapping = {\"nodes\": {}, \"edges\": {}}\n",
    "\n",
    "# Compute the costs, edges, and virtual variables based on the prizes\n",
    "for i, (src, dst) in enumerate(edge_index.T):\n",
    "    prize_e = prizes[\"edges\"][i].item()\n",
    "    if prize_e <= updated_cost_e:\n",
    "        mapping[\"edges\"][len(edges)] = i\n",
    "        edges.append((src.item(), dst.item()))\n",
    "        costs.append(updated_cost_e - prize_e)\n",
    "    else:\n",
    "        virtual_node_id = graph_nodes.shape[0] + len(virtual[\"n_prizes\"])\n",
    "        mapping[\"nodes\"][virtual_node_id] = i\n",
    "        virtual[\"edges\"].append((src.item(), virtual_node_id))\n",
    "        virtual[\"edges\"].append((virtual_node_id, dst.item()))\n",
    "        virtual[\"costs\"].append(0)\n",
    "        virtual[\"costs\"].append(0)\n",
    "        virtual[\"n_prizes\"].append(prize_e - updated_cost_e)\n",
    "prizes = cp.concatenate([prizes[\"nodes\"], cp.array(virtual[\"n_prizes\"])])\n",
    "edges_dict = {}\n",
    "edges_dict[\"edges\"] = edges\n",
    "edges_dict[\"num_prior_edges\"] = len(edges)\n",
    "# Final computation of the costs and edges based on the virtual costs and virtual edges\n",
    "if len(virtual[\"costs\"]) > 0:\n",
    "    costs = cp.array(costs + virtual[\"costs\"])\n",
    "    edges = cp.array(edges + virtual[\"edges\"])\n",
    "    edges_dict[\"edges\"] = edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67bbe3cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pcst_fast\n",
    "root = cfg.root\n",
    "num_clusters = cfg.num_clusters\n",
    "pruning = cfg.pruning\n",
    "verbosity_level = cfg.verbosity_level\n",
    "\n",
    "# Retrieve the subgraph using the PCST algorithm\n",
    "result_vertices, result_edges = pcst_fast.pcst_fast(\n",
    "    edges_dict[\"edges\"].get(),\n",
    "    prizes.get(),\n",
    "    costs.get(),\n",
    "    root,\n",
    "    num_clusters,\n",
    "    pruning,\n",
    "    verbosity_level,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c04c748",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edge_index = _create_edge_index(graph_nodes, graph_edges)\n",
    "vertices = cp.asarray(result_vertices)\n",
    "edges_dict = {\"edges\": cp.asarray(result_edges), \n",
    "              \"num_prior_edges\": edges_dict[\"num_prior_edges\"]}\n",
    "mapping = mapping\n",
    "\n",
    "# Get edges information\n",
    "edges = edges_dict[\"edges\"]\n",
    "num_prior_edges = edges_dict[\"num_prior_edges\"]\n",
    "# Retrieve the selected nodes and edges based on the given vertices and edges\n",
    "subgraph_nodes = vertices[vertices < len(graph_nodes)]\n",
    "subgraph_edges = [mapping[\"edges\"][e.item()] for e in edges if e < num_prior_edges]\n",
    "virtual_vertices = vertices[vertices >= graph.num_nodes]\n",
    "if len(virtual_vertices) > 0:\n",
    "    virtual_vertices = vertices[vertices >= len(graph_nodes)]\n",
    "    virtual_edges = [mapping[\"nodes\"][i.item()] for i in virtual_vertices]\n",
    "    subgraph_edges = cp.array(subgraph_edges + virtual_edges)\n",
    "edge_index = edge_index[:, subgraph_edges]\n",
    "subgraph_nodes = cp.unique(\n",
    "    cp.concatenate(\n",
    "        [subgraph_nodes, edge_index[0], edge_index[1]]\n",
    "    )\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f173a2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "subgraph = {\"nodes\": subgraph_nodes, \"edges\": subgraph_edges}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf8ad2b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "subgraphs = {}\n",
    "subgraphs[\"nodes\"] = []\n",
    "subgraphs[\"edges\"] = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38deaabd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Append the extracted subgraph to the dictionary\n",
    "subgraphs[\"nodes\"].append(subgraph[\"nodes\"].tolist())\n",
    "subgraphs[\"edges\"].append(subgraph[\"edges\"].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c2c5cd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate and get unique node and edge indices\n",
    "subgraphs[\"nodes\"] = cp.unique(\n",
    "    cp.concatenate([cp.array(list_) for list_ in subgraphs[\"nodes\"]])\n",
    ")\n",
    "subgraphs[\"edges\"] = np.unique(\n",
    "    cp.concatenate([cp.array(list_) for list_ in subgraphs[\"edges\"]])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c01422e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "state[\"selections\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a0ef705",
   "metadata": {},
   "outputs": [],
   "source": [
    "subgraphs['nodes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7d7240c",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.unique(np.array(graph_nodes.iloc[subgraphs['nodes']][['node_id', 'node_name', 'node_type', 'desc', 'enriched_node']].node_id.to_arrow().to_pylist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50d98944",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.unique(np.array([graph_edges.iloc[subgraphs['edges']][['head_id', 'tail_id']].head_id.to_arrow().to_pylist() +\\\n",
    "graph_edges.iloc[subgraphs['edges']][['head_id', 'tail_id']].tail_id.to_arrow().to_pylist()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83e849b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cugraph as cg\n",
    "import nx_cugraph as nxcg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4a3dbf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "nxcg.Graph(directed=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "026cdff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_nodes = nodes_df\n",
    "graph_edges = edges_df\n",
    "\n",
    "# Networkx DiGraph construction to be visualized in the frontend\n",
    "nx_graph = nx.DiGraph()\n",
    "# Add nodes with attributes\n",
    "node_colors = {n: cfg.node_colors_dict[k]\n",
    "                for k, v in state[\"selections\"].items() for n in v}\n",
    "for n in graph_nodes.iloc[subgraphs['nodes']][['node_id', 'node_name', 'node_type', 'desc', 'enriched_node']].to_arrow().to_pandas().node_name:\n",
    "    nx_graph.add_node(n, color=node_colors.get(n, None))\n",
    "\n",
    "\n",
    "# Add edges with attributes\n",
    "edges = zip(\n",
    "    pyg_graph.edge_index[0].tolist(),\n",
    "    pyg_graph.edge_index[1].tolist(),\n",
    "    pyg_graph.edge_type\n",
    ")\n",
    "for src, dst, edge_type in edges:\n",
    "    nx_graph.add_edge(\n",
    "        pyg_graph.node_name[src],\n",
    "        pyg_graph.node_name[dst],\n",
    "        relation=edge_type,\n",
    "        label=edge_type,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b35820",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_nodes = nodes_df\n",
    "graph_edges = edges_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77074953",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_edges[['head_id', 'tail_id', 'edge_type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "403cbc12",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = {n: i for i, n in enumerate(subgraph[\"nodes\"].tolist())}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d3503d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_nodes.iloc[subgraphs['nodes']][['node_id', 'node_name', 'node_type', 'desc', 'enriched_node']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "784767e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_edges.iloc[subgraphs[\"edges\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ed49da9",
   "metadata": {},
   "outputs": [],
   "source": [
    "subgr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9468a9e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cugraph.experimental import PropertyGraph\n",
    "\n",
    "# Create PropertyGraph\n",
    "pG = PropertyGraph()\n",
    "pG.add_vertex_data(graph_nodes.iloc[subgraphs['nodes']][['node_id', 'node_name', 'node_type', 'desc', 'enriched_node']], vertex_col_name=\"node_id\")\n",
    "pG.add_edge_data(graph_edges.iloc[subgraphs['edges']][['head_id', 'tail_id', 'edge_type']], vertex_col_names=(\"head_id\", \"tail_id\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4100cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pG.get_edge_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6ac23df",
   "metadata": {},
   "outputs": [],
   "source": [
    "(row._0, row._1, {'edge_type': row.edge_type})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24ab58b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "864786ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "node_colors = {n: cfg.node_colors_dict[k]\n",
    "                for k, v in state[\"selections\"].items() for n in v}\n",
    "node_colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa2bd44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the dict to a cudf DataFrame\n",
    "node_colors = {n: cfg.node_colors_dict[k]\n",
    "                for k, v in state[\"selections\"].items() for n in v}\n",
    "color_df = cudf.DataFrame(list(node_colors.items()), columns=[\"node_id\", \"color\"])\n",
    "\n",
    "# Merge the color_df with graph_nodes\n",
    "graph_nodes = graph_nodes.merge(color_df, on=\"node_id\", how=\"left\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d283a90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "[x for x in graph_nodes.iloc[subgraphs[\"nodes\"]].to_arrow().to_pandas().itertuples(index=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a93d6001",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to list of tuples with node ID and a dictionary of properties\n",
    "node_tuples = [(row.node_id, {'desc': row.desc, 'enriched_node': row.enriched_node}) \n",
    "               for row in graph_nodes.iloc[subgraphs[\"nodes\"]].to_arrow().to_pandas().itertuples(index=False)]\n",
    "edge_tuples = [(row.head_id, row.tail_id, {'label': tuple(row.edge_type)})\n",
    "                for row in graph_edges.iloc[subgraphs[\"edges\"]].to_arrow().to_pandas().itertuples(index=False)]\n",
    "node_tuples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09129425",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Textualized graph\n",
    "# Prepare the textualized subgraph\n",
    "graph_text = (\n",
    "    graph_nodes.iloc[subgraphs[\"nodes\"]][['node_id', 'desc']].rename(columns={'desc': 'node_attr'}).to_arrow().to_pandas().to_csv(index=False)\n",
    "    + \"\\n\"\n",
    "    + graph_edges.iloc[subgraph[\"edges\"]][['head_id', 'edge_type', 'tail_id']].to_arrow().to_pandas().to_csv(index=False)\n",
    ")\n",
    "print(graph_text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72f7886a",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_edges.iloc[subgraph[\"edges\"]][['head_id', 'edge_type', 'tail_id']].to_arrow().to_pandas().to_csv(index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a3113dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the vertex DataFrame from the PropertyGraph\n",
    "vertex_df = pG.get_vertex_data()\n",
    "\n",
    "# Convert to list of tuples with node ID and a dictionary of properties\n",
    "node_tuples = [(row._0, {'desc': row.desc, 'enriched_node': row.enriched_node}) \n",
    "               for row in pG.get_vertex_data().to_arrow().to_pandas().itertuples(index=False)]\n",
    "edge_tuples = [(row._2, row._3, {'label': tuple(row.edge_type)})\n",
    "                for row in pG.get_edge_data().to_arrow().to_pandas().itertuples(index=False)]\n",
    "edge_tuples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2026c830",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(nx_graph.edges(data=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc1e89aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(nx_graph.nodes(data=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4abc7d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Networkx DiGraph construction to be visualized in the frontend\n",
    "nxcg_graph = cg.Graph(directed=True)\n",
    "# Add nodes with attributes\n",
    "node_colors = {n: cfg.node_colors_dict[k]\n",
    "                for k, v in state[\"selections\"].items() for n in v}\n",
    "for n in graph_nodes.iloc[subgraphs['nodes']][['node_id', 'node_name', 'node_type', 'desc', 'enriched_node']].to_arrow().to_pandas().node_name:\n",
    "    nxcg_graph.add_node(n, color=node_colors.get(n, None))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61ca060c",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_nodes.iloc[subgraphs['nodes']][['node_id', 'node_name', 'node_type', 'desc', 'enriched_node']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb776f3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_graph = graph_nodes.iloc[subgraphs['nodes']][['node_id', 'node_name', 'node_type', 'desc', 'enriched_node']]\n",
    "cg_graph.add_nodes_from(my_graph.node_id.to_arrow().to_pylist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec9268e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cugraph.experimental import PropertyGraph\n",
    "import cudf\n",
    "\n",
    "# Example node DataFrame with properties\n",
    "graph_nodes = cudf.DataFrame({\n",
    "    \"node_id\": [\"A\", \"B\", \"C\"],\n",
    "    \"color\": [\"red\", \"blue\", \"green\"],\n",
    "    \"description\": [\"start\", \"middle\", \"end\"]\n",
    "})\n",
    "\n",
    "# Example edge DataFrame\n",
    "graph_edges = cudf.DataFrame({\n",
    "    \"src\": [\"A\", \"B\", \"C\"],\n",
    "    \"dst\": [\"B\", \"C\", \"A\"],\n",
    "    \"edge_type\": [\"link\", \"link\", \"loop\"],\n",
    "    \"weight\": [1.0, 2.0, 3.0]\n",
    "})\n",
    "\n",
    "# Create PropertyGraph\n",
    "pG = PropertyGraph()\n",
    "pG.add_vertex_data(graph_nodes, vertex_col_name=\"node_id\")\n",
    "pG.add_edge_data(graph_edges, vertex_col_names=(\"src\", \"dst\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a7f4ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pG.get_vertex_data().to_arrow().to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd2958ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
