{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b2178a53",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "### Generate ground truth results for T2B prompts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4aab4242",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\code\\benchmark\\AIAgents4Pharma\\.venv\\Lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import basico\n",
    "import pickle\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "331fb94e",
   "metadata": {},
   "source": [
    "#### Set 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb06713d",
   "metadata": {},
   "source": [
    "##### Precision Medicine Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "15f7e64d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dwivedi2014 - Crohns IL6 Disease model - Anti-IL6R Antibody\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2.2691308858431047"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate expected answers for the basic prompts\n",
    "# precision medicine model\n",
    "#mod_precision = basico.load_biomodel('BIOMD0000000537')\n",
    "mod_precision = basico.load_model(r'./models/BIOMD0000000537_url.xml')\n",
    "print(basico.get_model_name())\n",
    "# run simulation for set 1 and get the ground truth answer\n",
    "tc = basico.run_time_course(2016, 2016)\n",
    "tc.iloc[-1][\"CRP{serum}\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2522f690",
   "metadata": {},
   "source": [
    "##### Epidemiology Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d5b82489",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tang2020 - Estimation of transmission risk of COVID-19 and impact of public health interventions\n"
     ]
    }
   ],
   "source": [
    "# epidemiology mode\n",
    "#mod_epidemiology = basico.load_biomodel('BIOMD0000000971')\n",
    "mod_epidemiology = basico.load_model(r'./models/Tang2020.xml')\n",
    "print(basico.get_model_name())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d67e9ad2",
   "metadata": {},
   "source": [
    "##### Theoritical Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1fea2d4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Markevich2004 - MAPK double phosphorylation,  ordered Michaelis-Menton\n"
     ]
    }
   ],
   "source": [
    "# epidemiology mode\n",
    "#mod_theory = basico.load_biomodel('BIOMD0000000027')\n",
    "mod_theory = basico.load_model(r'./models/BIOMD0000000027_url.xml')\n",
    "print(basico.get_model_name())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6e1078a",
   "metadata": {},
   "source": [
    "### Set 2 vary simulation time, interval and species to be simulated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6472afe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_precision_model(interval, time, species, concentration):\n",
    "    basico.set_species(name = species, model = mod_precision, initial_concentration = concentration)\n",
    "    # check if the species was set correctly\n",
    "    #print(basico.get_species(name = species, model = mod_precision)['initial_concentration'])\n",
    "    # run simulations\n",
    "    tc = basico.run_time_course(duration = time, intervals = interval, model = mod_precision, update_model = False)\n",
    "    return tc.iloc[-1][[\"CRP{serum}\",  \"CRP{liver}\", \"IL6{serum}\", \"STAT3{gut}\"]]\n",
    "\n",
    "#run_precision_model(time = 100, interval = 2016)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b2664f9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_epidemiology_model(interval, time):\n",
    "    tc = basico.run_time_course(duration = time, intervals = interval, model = mod_epidemiology, update_model = False)\n",
    "    return tc.iloc[-1][[\"Infected\", \"Susceptible\", \"Recovered\", \"Hospitalised\"]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd168012",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_theory_model(interval, time):\n",
    "    tc = basico.run_time_course(duration = time, intervals = interval, model = mod_theory, update_model = False)\n",
    "    return tc.iloc[-1][[\"Mpp\", \"M\", \"Mp\"]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65788649",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_theory_model_set_parameters(interval, time, parameter, value):\n",
    "    basico.set_parameters(name = parameter, initial_value = value, model = mod_theory)\n",
    "    tc = basico.run_time_course(duration = time, intervals = interval, model = mod_theory, update_model = False)\n",
    "    return tc.iloc[-1][[\"Mpp\", \"M\", \"Mp\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe3cc9df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mpp    499.080250\n",
       "M        0.113662\n",
       "Mp       0.806087\n",
       "Name: 1000.0, dtype: float64"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_theory_model_set_parameters(1000, 1000, \"Values[k1cat]\", 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c648a36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.01\n",
      "1000\n",
      "2.6\n",
      "435628.8965511659\n",
      "1583.2584678161063\n",
      "6.106360135082123e-08\n",
      "6.106360135082122e-10\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'interval_2016_time_2016_species_CRP{serum}_concentration_0.01': CRP{serum}    10.301993\n",
       " CRP{liver}     6.707914\n",
       " IL6{serum}     0.000641\n",
       " STAT3{gut}     9.124980\n",
       " Name: 2016.0, dtype: float64,\n",
       " 'interval_100_time_20_species_CRP{serum}_concentration_1000': CRP{serum}    243.477247\n",
       " CRP{liver}    162.273954\n",
       " IL6{serum}      0.003962\n",
       " STAT3{gut}      0.291445\n",
       " Name: 20.0, dtype: float64,\n",
       " 'interval_4032_time_2016_species_CRP{serum}_concentration_2.6': CRP{serum}    10.302718\n",
       " CRP{liver}     6.708469\n",
       " IL6{serum}     0.000641\n",
       " STAT3{gut}     9.124957\n",
       " Name: 2016.0, dtype: float64,\n",
       " 'interval_1000_time_1000_species_IL6{serum}_concentration_435628.8965511659': CRP{serum}    91.299723\n",
       " CRP{liver}    60.894795\n",
       " IL6{serum}     0.000642\n",
       " STAT3{gut}     6.813039\n",
       " Name: 1000.0, dtype: float64,\n",
       " 'interval_1000_time_2016_species_CRP{liver}_concentration_1583.2584678161063': CRP{serum}    10.328675\n",
       " CRP{liver}     6.728273\n",
       " IL6{serum}     0.000641\n",
       " STAT3{gut}     9.124149\n",
       " Name: 2016.0, dtype: float64,\n",
       " 'interval_500_time_500_species_STAT3{gut}_concentration_6.106360135082123e-08': CRP{serum}    215.468283\n",
       " CRP{liver}    154.750841\n",
       " IL6{serum}      0.000441\n",
       " STAT3{gut}      0.884170\n",
       " Name: 500.0, dtype: float64,\n",
       " 'interval_2000_time_1500_species_STAT3{gut}_concentration_6.106360135082122e-10': CRP{serum}    159.484293\n",
       " CRP{liver}    110.490735\n",
       " IL6{serum}      0.000643\n",
       " STAT3{gut}      4.080634\n",
       " Name: 1500.0, dtype: float64}"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run with different random conditions and store in dictionary\n",
    "results = {}\n",
    "# adjust initial concentrations of species\n",
    "initial_concentr_CRP_serum = basico.get_species(name = \"CRP{serum}\", model = mod_precision)['initial_concentration'].iloc[0]\n",
    "initial_concentr_CRP_liver = basico.get_species(name = \"CRP{liver}\", model = mod_precision)['initial_concentration'].iloc[0]\n",
    "initial_concentr_IL6_serum = basico.get_species(name = \"IL6{serum}\", model = mod_precision)['initial_concentration'].iloc[0]\n",
    "initial_concentr_STAT3_gut = basico.get_species(name = \"STAT3{gut}\", model = mod_precision)['initial_concentration'].iloc[0]\n",
    "# time \n",
    "conditions = [\n",
    "    (2016, 2016, \"CRP{serum}\", 0.01),\n",
    "    (20, 100, \"CRP{serum}\", 1000),\n",
    "    (2016, 4032, \"CRP{serum}\", 2.6),\n",
    "    (1000, 1000, \"IL6{serum}\", initial_concentr_IL6_serum*10), \n",
    "    (2016, 1000, \"CRP{liver}\", initial_concentr_CRP_liver*10),\n",
    "    (500, 500, \"STAT3{gut}\", initial_concentr_STAT3_gut*10),\n",
    "    (1500, 2000, \"STAT3{gut}\", initial_concentr_STAT3_gut*0.10)\n",
    "]\n",
    "\n",
    "for time, interval, species, concentration in conditions:\n",
    "    key = f\"interval_{interval}_time_{time}_species_{species}_concentration_{concentration}\"\n",
    "    print (concentration)\n",
    "    results[key] = run_precision_model(interval, time, species, concentration)\n",
    "\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b9551d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'interval_100_time_50': Infected        1.043385e+05\n",
       " Susceptible     1.017891e+06\n",
       " Recovered       2.231583e+06\n",
       " Hospitalised    1.325140e+05\n",
       " Name: 50.0, dtype: float64,\n",
       " 'interval_200_time_100': Infected        7.143353e+04\n",
       " Susceptible     1.055609e+06\n",
       " Recovered       4.586298e+06\n",
       " Hospitalised    8.688262e+04\n",
       " Name: 100.0, dtype: float64,\n",
       " 'interval_400_time_180': Infected        4.031662e+04\n",
       " Susceptible     1.055585e+06\n",
       " Recovered       6.955857e+06\n",
       " Hospitalised    4.901975e+04\n",
       " Name: 180.0, dtype: float64,\n",
       " 'interval_1000_time_500': Infected        4.090069e+03\n",
       " Susceptible     1.055585e+06\n",
       " Recovered       9.714202e+06\n",
       " Hospitalised    4.972991e+03\n",
       " Name: 500.0, dtype: float64,\n",
       " 'interval_400_time_20': Infected        1.206681e+05\n",
       " Susceptible     3.167693e+06\n",
       " Recovered       1.634611e+05\n",
       " Hospitalised    4.076004e+04\n",
       " Name: 20.0, dtype: float64,\n",
       " 'interval_10_time_10': Infected        3.413654e+03\n",
       " Susceptible     1.079620e+07\n",
       " Recovered       2.980390e+03\n",
       " Hospitalised    8.065950e+02\n",
       " Name: 10.0, dtype: float64}"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run with different random conditions and store in dictionary\n",
    "results = {}\n",
    "# time \n",
    "conditions = [\n",
    "    (50, 100),\n",
    "    (100, 200),\n",
    "    (180, 400),\n",
    "    (500, 1000), \n",
    "    (20, 400),\n",
    "    (10, 10)\n",
    "]\n",
    "\n",
    "for time, interval in conditions:\n",
    "    key = f\"interval_{interval}_time_{time}\"\n",
    "    results[key] = run_epidemiology_model(interval, time)\n",
    "\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3b3881e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'interval_100_time_1000': Mpp    493.786732\n",
       " M        0.782237\n",
       " Mp       5.431030\n",
       " Name: 1000.0, dtype: float64,\n",
       " 'interval_1000_time_1000': Mpp    493.786732\n",
       " M        0.782237\n",
       " Mp       5.431030\n",
       " Name: 1000.0, dtype: float64,\n",
       " 'interval_400_time_180': Mpp    491.137882\n",
       " M        3.033640\n",
       " Mp       5.828478\n",
       " Name: 180.0, dtype: float64,\n",
       " 'interval_1000_time_500': Mpp    493.786732\n",
       " M        0.782237\n",
       " Mp       5.431030\n",
       " Name: 500.0, dtype: float64,\n",
       " 'interval_400_time_20': Mpp     45.234214\n",
       " M      424.991709\n",
       " Mp      29.774077\n",
       " Name: 20.0, dtype: float64,\n",
       " 'interval_10_time_10': Mpp     17.029310\n",
       " M      460.393300\n",
       " Mp      22.577389\n",
       " Name: 10.0, dtype: float64}"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run with different random conditions and store in dictionary\n",
    "results = {}\n",
    "# time \n",
    "conditions = [\n",
    "    (1000, 100),\n",
    "    (1000, 1000),\n",
    "    (180, 400),\n",
    "    (500, 1000), \n",
    "    (20, 400),\n",
    "    (10, 10)\n",
    "]\n",
    "\n",
    "for time, interval in conditions:\n",
    "    key = f\"interval_{interval}_time_{time}\"\n",
    "    results[key] = run_theory_model(interval, time)\n",
    "\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f18df2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'time_100_parameter_k1cat_value_0.1': Mpp    317.356267\n",
       " M      154.982325\n",
       " Mp      27.661408\n",
       " Name: 100.0, dtype: float64,\n",
       " 'time_1000_parameter_k1cat_value_1': Mpp    494.565011\n",
       " M        0.076041\n",
       " Mp       5.358948\n",
       " Name: 1000.0, dtype: float64,\n",
       " 'time_180_parameter_k1cat_value_100': Mpp    494.647985\n",
       " M        0.000758\n",
       " Mp       5.351257\n",
       " Name: 180.0, dtype: float64,\n",
       " 'time_500_parameter_k2cat_value_10': Mpp    491.982136\n",
       " M        0.001140\n",
       " Mp       8.016723\n",
       " Name: 500.0, dtype: float64,\n",
       " 'time_20_parameter_k2cat_value_0.1': Mpp     42.542954\n",
       " M        0.103107\n",
       " Mp     457.353939\n",
       " Name: 20.0, dtype: float64,\n",
       " 'time_10_parameter_k2cat_value_100': Mpp    499.195565\n",
       " M        0.000113\n",
       " Mp       0.804322\n",
       " Name: 10.0, dtype: float64}"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run theory model with different parameters\n",
    "results = {}\n",
    "# time \n",
    "interval = 1000\n",
    "conditions = [\n",
    "    (100, 'k1cat', 0.1),\n",
    "    (1000, 'k1cat', 1),\n",
    "    (180, 'k1cat', 100),\n",
    "    (500, 'k2cat', 10), \n",
    "    (20, 'k2cat', 0.1),\n",
    "    (10, 'k2cat', 100)\n",
    "]\n",
    "\n",
    "for time, parameter, value in conditions:\n",
    "    key = f\"time_{time}_parameter_{parameter}_value_{value}\"\n",
    "    results[key] = run_theory_model_set_parameters(interval, time, parameter, value)\n",
    "\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd87cdc7",
   "metadata": {},
   "source": [
    "### Steady state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5fdef477",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error while running the simulation: >WARNING 2025-09-19T17:39:31<\n",
      "  CSteadyState (2): No steady state was found.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>compartment</th>\n",
       "      <th>type</th>\n",
       "      <th>unit</th>\n",
       "      <th>initial_concentration</th>\n",
       "      <th>initial_particle_number</th>\n",
       "      <th>initial_expression</th>\n",
       "      <th>expression</th>\n",
       "      <th>concentration</th>\n",
       "      <th>particle_number</th>\n",
       "      <th>rate</th>\n",
       "      <th>particle_number_rate</th>\n",
       "      <th>key</th>\n",
       "      <th>sbml_id</th>\n",
       "      <th>transition_time</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IL6</th>\n",
       "      <td>serum</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>0.000436</td>\n",
       "      <td>2.623419e+11</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.000436</td>\n",
       "      <td>2.623419e+11</td>\n",
       "      <td>-1.621620e-17</td>\n",
       "      <td>-0.009766</td>\n",
       "      <td>Metabolite_114</td>\n",
       "      <td>mwf626e95e_543f_41e4_aad4_c6bf60ab345b</td>\n",
       "      <td>0.028719</td>\n",
       "      <td>IL6{serum}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgp130</th>\n",
       "      <td>serum</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>3.900000</td>\n",
       "      <td>2.348635e+15</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3.900000</td>\n",
       "      <td>2.348635e+15</td>\n",
       "      <td>4.167564e-16</td>\n",
       "      <td>0.250977</td>\n",
       "      <td>Metabolite_115</td>\n",
       "      <td>mwbbbce920_e8dd_4320_9386_fc94bfb2fc99</td>\n",
       "      <td>0.997567</td>\n",
       "      <td>sgp130{serum}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sR_IL6_sgp130</th>\n",
       "      <td>serum</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>0.087406</td>\n",
       "      <td>5.263716e+13</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.087406</td>\n",
       "      <td>5.263716e+13</td>\n",
       "      <td>-2.367565e-16</td>\n",
       "      <td>-0.142578</td>\n",
       "      <td>Metabolite_116</td>\n",
       "      <td>mw810ff751_fa4e_4143_bd50_169b3e325e1e</td>\n",
       "      <td>9.190533</td>\n",
       "      <td>sR_IL6_sgp130{serum}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRP</th>\n",
       "      <td>serum</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>221.063676</td>\n",
       "      <td>1.331277e+17</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>221.063676</td>\n",
       "      <td>1.331277e+17</td>\n",
       "      <td>6.243627e-13</td>\n",
       "      <td>376.000000</td>\n",
       "      <td>Metabolite_117</td>\n",
       "      <td>mw114aa90f_5f5b_4fe8_9406_361c8489b6a1</td>\n",
       "      <td>2.777778</td>\n",
       "      <td>CRP{serum}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sR</th>\n",
       "      <td>serum</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>4.253507</td>\n",
       "      <td>2.561522e+15</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.253507</td>\n",
       "      <td>2.561522e+15</td>\n",
       "      <td>1.328431e-14</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>Metabolite_118</td>\n",
       "      <td>mw30ae63db_6cd3_4b6f_93ad_3350cd360bcc</td>\n",
       "      <td>3.283772</td>\n",
       "      <td>sR{serum}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              compartment       type    unit  initial_concentration  \\\n",
       "name                                                                  \n",
       "IL6                 serum  reactions  nmol/l               0.000436   \n",
       "sgp130              serum  reactions  nmol/l               3.900000   \n",
       "sR_IL6_sgp130       serum  reactions  nmol/l               0.087406   \n",
       "CRP                 serum  reactions  nmol/l             221.063676   \n",
       "sR                  serum  reactions  nmol/l               4.253507   \n",
       "\n",
       "               initial_particle_number initial_expression expression  \\\n",
       "name                                                                   \n",
       "IL6                       2.623419e+11                                 \n",
       "sgp130                    2.348635e+15                                 \n",
       "sR_IL6_sgp130             5.263716e+13                                 \n",
       "CRP                       1.331277e+17                                 \n",
       "sR                        2.561522e+15                                 \n",
       "\n",
       "               concentration  particle_number          rate  \\\n",
       "name                                                          \n",
       "IL6                 0.000436     2.623419e+11 -1.621620e-17   \n",
       "sgp130              3.900000     2.348635e+15  4.167564e-16   \n",
       "sR_IL6_sgp130       0.087406     5.263716e+13 -2.367565e-16   \n",
       "CRP               221.063676     1.331277e+17  6.243627e-13   \n",
       "sR                  4.253507     2.561522e+15  1.328431e-14   \n",
       "\n",
       "               particle_number_rate             key  \\\n",
       "name                                                  \n",
       "IL6                       -0.009766  Metabolite_114   \n",
       "sgp130                     0.250977  Metabolite_115   \n",
       "sR_IL6_sgp130             -0.142578  Metabolite_116   \n",
       "CRP                      376.000000  Metabolite_117   \n",
       "sR                         8.000000  Metabolite_118   \n",
       "\n",
       "                                              sbml_id  transition_time  \\\n",
       "name                                                                     \n",
       "IL6            mwf626e95e_543f_41e4_aad4_c6bf60ab345b         0.028719   \n",
       "sgp130         mwbbbce920_e8dd_4320_9386_fc94bfb2fc99         0.997567   \n",
       "sR_IL6_sgp130  mw810ff751_fa4e_4143_bd50_169b3e325e1e         9.190533   \n",
       "CRP            mw114aa90f_5f5b_4fe8_9406_361c8489b6a1         2.777778   \n",
       "sR             mw30ae63db_6cd3_4b6f_93ad_3350cd360bcc         3.283772   \n",
       "\n",
       "                       display_name  \n",
       "name                                 \n",
       "IL6                      IL6{serum}  \n",
       "sgp130                sgp130{serum}  \n",
       "sR_IL6_sgp130  sR_IL6_sgp130{serum}  \n",
       "CRP                      CRP{serum}  \n",
       "sR                        sR{serum}  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_precision = basico.load_model(r'./models/BIOMD0000000537_url.xml')\n",
    "basico.task_steadystate.run_steadystate()\n",
    "basico.get_species().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "88a82c9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>compartment</th>\n",
       "      <th>type</th>\n",
       "      <th>unit</th>\n",
       "      <th>initial_concentration</th>\n",
       "      <th>initial_particle_number</th>\n",
       "      <th>initial_expression</th>\n",
       "      <th>expression</th>\n",
       "      <th>concentration</th>\n",
       "      <th>particle_number</th>\n",
       "      <th>rate</th>\n",
       "      <th>particle_number_rate</th>\n",
       "      <th>key</th>\n",
       "      <th>sbml_id</th>\n",
       "      <th>transition_time</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>uVol</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>500.0</td>\n",
       "      <td>3.011070e+17</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>437.729984</td>\n",
       "      <td>2.636072e+17</td>\n",
       "      <td>1.037837e-16</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>Metabolite_158</td>\n",
       "      <td>M</td>\n",
       "      <td>978.030464</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mp</th>\n",
       "      <td>uVol</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>12.852481</td>\n",
       "      <td>7.739945e+15</td>\n",
       "      <td>-4.151348e-16</td>\n",
       "      <td>-0.2500</td>\n",
       "      <td>Metabolite_159</td>\n",
       "      <td>Mp</td>\n",
       "      <td>5.313708</td>\n",
       "      <td>Mp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mpp</th>\n",
       "      <td>uVol</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>49.417535</td>\n",
       "      <td>2.975994e+16</td>\n",
       "      <td>4.151348e-16</td>\n",
       "      <td>0.2500</td>\n",
       "      <td>Metabolite_160</td>\n",
       "      <td>Mpp</td>\n",
       "      <td>25.070051</td>\n",
       "      <td>Mpp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAPKK</th>\n",
       "      <td>uVol</td>\n",
       "      <td>fixed</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>50.0</td>\n",
       "      <td>3.011070e+16</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>50.000000</td>\n",
       "      <td>3.011070e+16</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Metabolite_161</td>\n",
       "      <td>MAPKK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAPKK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MKP3</th>\n",
       "      <td>uVol</td>\n",
       "      <td>fixed</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>100.0</td>\n",
       "      <td>6.022141e+16</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>100.000000</td>\n",
       "      <td>6.022141e+16</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Metabolite_162</td>\n",
       "      <td>MKP3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MKP3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      compartment       type    unit  initial_concentration  \\\n",
       "name                                                          \n",
       "M            uVol  reactions  nmol/l                  500.0   \n",
       "Mp           uVol  reactions  nmol/l                    0.0   \n",
       "Mpp          uVol  reactions  nmol/l                    0.0   \n",
       "MAPKK        uVol      fixed  nmol/l                   50.0   \n",
       "MKP3         uVol      fixed  nmol/l                  100.0   \n",
       "\n",
       "       initial_particle_number initial_expression expression  concentration  \\\n",
       "name                                                                          \n",
       "M                 3.011070e+17                                   437.729984   \n",
       "Mp                0.000000e+00                                    12.852481   \n",
       "Mpp               0.000000e+00                                    49.417535   \n",
       "MAPKK             3.011070e+16                                    50.000000   \n",
       "MKP3              6.022141e+16                                   100.000000   \n",
       "\n",
       "       particle_number          rate  particle_number_rate             key  \\\n",
       "name                                                                         \n",
       "M         2.636072e+17  1.037837e-16                0.0625  Metabolite_158   \n",
       "Mp        7.739945e+15 -4.151348e-16               -0.2500  Metabolite_159   \n",
       "Mpp       2.975994e+16  4.151348e-16                0.2500  Metabolite_160   \n",
       "MAPKK     3.011070e+16  0.000000e+00                0.0000  Metabolite_161   \n",
       "MKP3      6.022141e+16  0.000000e+00                0.0000  Metabolite_162   \n",
       "\n",
       "      sbml_id  transition_time display_name  \n",
       "name                                         \n",
       "M           M       978.030464            M  \n",
       "Mp         Mp         5.313708           Mp  \n",
       "Mpp       Mpp        25.070051          Mpp  \n",
       "MAPKK   MAPKK              NaN        MAPKK  \n",
       "MKP3     MKP3              NaN         MKP3  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_theory = basico.load_model(r'./models/BIOMD0000000027_url.xml')\n",
    "basico.task_steadystate.run_steadystate()\n",
    "basico.get_species()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3b6a979f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>compartment</th>\n",
       "      <th>type</th>\n",
       "      <th>unit</th>\n",
       "      <th>initial_concentration</th>\n",
       "      <th>initial_particle_number</th>\n",
       "      <th>initial_expression</th>\n",
       "      <th>expression</th>\n",
       "      <th>concentration</th>\n",
       "      <th>particle_number</th>\n",
       "      <th>rate</th>\n",
       "      <th>particle_number_rate</th>\n",
       "      <th>key</th>\n",
       "      <th>sbml_id</th>\n",
       "      <th>transition_time</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Susceptible</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>1.108100e+07</td>\n",
       "      <td>1.108100e+07</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.107825e+07</td>\n",
       "      <td>1.107825e+07</td>\n",
       "      <td>4.237148e-24</td>\n",
       "      <td>4.237148e-24</td>\n",
       "      <td>Metabolite_163</td>\n",
       "      <td>Susceptible</td>\n",
       "      <td>9.765576e+29</td>\n",
       "      <td>Susceptible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exposed</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>1.051000e+02</td>\n",
       "      <td>1.051000e+02</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.997059e-24</td>\n",
       "      <td>5.997059e-24</td>\n",
       "      <td>-1.452434e-25</td>\n",
       "      <td>-1.452434e-25</td>\n",
       "      <td>Metabolite_164</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>7.000007e+00</td>\n",
       "      <td>Exposed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Infected</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>2.767900e+01</td>\n",
       "      <td>2.767900e+01</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.137373e-24</td>\n",
       "      <td>1.137373e-24</td>\n",
       "      <td>2.173586e-25</td>\n",
       "      <td>2.173586e-25</td>\n",
       "      <td>Metabolite_165</td>\n",
       "      <td>Infected</td>\n",
       "      <td>1.528880e+00</td>\n",
       "      <td>Infected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Asymptomatic</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>5.353900e+01</td>\n",
       "      <td>5.353900e+01</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-9.305782e-25</td>\n",
       "      <td>-9.305782e-25</td>\n",
       "      <td>2.428722e-25</td>\n",
       "      <td>2.428722e-25</td>\n",
       "      <td>Metabolite_166</td>\n",
       "      <td>Asymptomatic</td>\n",
       "      <td>3.831555e+00</td>\n",
       "      <td>Asymptomatic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Susceptible_quarantined</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>7.390000e+02</td>\n",
       "      <td>7.390000e+02</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.588187e-22</td>\n",
       "      <td>1.588187e-22</td>\n",
       "      <td>-4.948627e-24</td>\n",
       "      <td>-4.948627e-24</td>\n",
       "      <td>Metabolite_167</td>\n",
       "      <td>Susceptible_quarantined</td>\n",
       "      <td>1.400001e+01</td>\n",
       "      <td>Susceptible_quarantined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exposed_quarantined</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>1.164200e+00</td>\n",
       "      <td>1.164200e+00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-1.085675e-24</td>\n",
       "      <td>-1.085675e-24</td>\n",
       "      <td>1.366866e-25</td>\n",
       "      <td>1.366866e-25</td>\n",
       "      <td>Metabolite_168</td>\n",
       "      <td>Exposed_quarantined</td>\n",
       "      <td>7.942804e+00</td>\n",
       "      <td>Exposed_quarantined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hospitalised</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-1.240771e-24</td>\n",
       "      <td>-1.240771e-24</td>\n",
       "      <td>1.584469e-25</td>\n",
       "      <td>1.584469e-25</td>\n",
       "      <td>Metabolite_169</td>\n",
       "      <td>Hospitalised</td>\n",
       "      <td>4.204104e+00</td>\n",
       "      <td>Hospitalised</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Recovered</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2.030779e+01</td>\n",
       "      <td>2.030779e+01</td>\n",
       "      <td>1.013596e-25</td>\n",
       "      <td>1.013596e-25</td>\n",
       "      <td>Metabolite_170</td>\n",
       "      <td>Recovered</td>\n",
       "      <td>5.405853e+25</td>\n",
       "      <td>Recovered</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        compartment       type  unit  initial_concentration  \\\n",
       "name                                                                          \n",
       "Susceptible                   Wuhan  reactions  #/ml           1.108100e+07   \n",
       "Exposed                       Wuhan  reactions  #/ml           1.051000e+02   \n",
       "Infected                      Wuhan  reactions  #/ml           2.767900e+01   \n",
       "Asymptomatic                  Wuhan  reactions  #/ml           5.353900e+01   \n",
       "Susceptible_quarantined       Wuhan  reactions  #/ml           7.390000e+02   \n",
       "Exposed_quarantined           Wuhan  reactions  #/ml           1.164200e+00   \n",
       "Hospitalised                  Wuhan  reactions  #/ml           1.000000e+00   \n",
       "Recovered                     Wuhan  reactions  #/ml           2.000000e+00   \n",
       "\n",
       "                         initial_particle_number initial_expression  \\\n",
       "name                                                                  \n",
       "Susceptible                         1.108100e+07                      \n",
       "Exposed                             1.051000e+02                      \n",
       "Infected                            2.767900e+01                      \n",
       "Asymptomatic                        5.353900e+01                      \n",
       "Susceptible_quarantined             7.390000e+02                      \n",
       "Exposed_quarantined                 1.164200e+00                      \n",
       "Hospitalised                        1.000000e+00                      \n",
       "Recovered                           2.000000e+00                      \n",
       "\n",
       "                        expression  concentration  particle_number  \\\n",
       "name                                                                 \n",
       "Susceptible                          1.107825e+07     1.107825e+07   \n",
       "Exposed                              5.997059e-24     5.997059e-24   \n",
       "Infected                             1.137373e-24     1.137373e-24   \n",
       "Asymptomatic                        -9.305782e-25    -9.305782e-25   \n",
       "Susceptible_quarantined              1.588187e-22     1.588187e-22   \n",
       "Exposed_quarantined                 -1.085675e-24    -1.085675e-24   \n",
       "Hospitalised                        -1.240771e-24    -1.240771e-24   \n",
       "Recovered                            2.030779e+01     2.030779e+01   \n",
       "\n",
       "                                 rate  particle_number_rate             key  \\\n",
       "name                                                                          \n",
       "Susceptible              4.237148e-24          4.237148e-24  Metabolite_163   \n",
       "Exposed                 -1.452434e-25         -1.452434e-25  Metabolite_164   \n",
       "Infected                 2.173586e-25          2.173586e-25  Metabolite_165   \n",
       "Asymptomatic             2.428722e-25          2.428722e-25  Metabolite_166   \n",
       "Susceptible_quarantined -4.948627e-24         -4.948627e-24  Metabolite_167   \n",
       "Exposed_quarantined      1.366866e-25          1.366866e-25  Metabolite_168   \n",
       "Hospitalised             1.584469e-25          1.584469e-25  Metabolite_169   \n",
       "Recovered                1.013596e-25          1.013596e-25  Metabolite_170   \n",
       "\n",
       "                                         sbml_id  transition_time  \\\n",
       "name                                                                \n",
       "Susceptible                          Susceptible     9.765576e+29   \n",
       "Exposed                                  Exposed     7.000007e+00   \n",
       "Infected                                Infected     1.528880e+00   \n",
       "Asymptomatic                        Asymptomatic     3.831555e+00   \n",
       "Susceptible_quarantined  Susceptible_quarantined     1.400001e+01   \n",
       "Exposed_quarantined          Exposed_quarantined     7.942804e+00   \n",
       "Hospitalised                        Hospitalised     4.204104e+00   \n",
       "Recovered                              Recovered     5.405853e+25   \n",
       "\n",
       "                                    display_name  \n",
       "name                                              \n",
       "Susceptible                          Susceptible  \n",
       "Exposed                                  Exposed  \n",
       "Infected                                Infected  \n",
       "Asymptomatic                        Asymptomatic  \n",
       "Susceptible_quarantined  Susceptible_quarantined  \n",
       "Exposed_quarantined          Exposed_quarantined  \n",
       "Hospitalised                        Hospitalised  \n",
       "Recovered                              Recovered  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_epidemiology = basico.load_model(r'./models/Tang2020.xml')\n",
    "basico.task_steadystate.run_steadystate()\n",
    "basico.get_species()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d677fa89",
   "metadata": {},
   "source": [
    "#### Search models\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "10cf8401",
   "metadata": {},
   "outputs": [],
   "source": [
    "from basico import biomodels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "61cd8287",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'format': 'SBML',\n",
       "  'id': 'BIOMD0000000583',\n",
       "  'lastModified': None,\n",
       "  'name': 'Leber2015 - Mucosal immunity and gut microbiome interaction during C. difficile infection',\n",
       "  'submissionDate': '2015-07-19T23:00:00Z',\n",
       "  'submitter': 'Andrew Leber',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000583'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000962',\n",
       "  'lastModified': None,\n",
       "  'name': 'Zhao2020 - SUQC model of COVID-19 transmission dynamics in Wuhan, Hubei, and China',\n",
       "  'submissionDate': '2020-08-25T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000962'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2106070001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Montagud2022 - Prostate cancer Boolean model',\n",
       "  'submissionDate': '2021-06-06T23:00:00Z',\n",
       "  'submitter': 'Arnau Montagud',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2106070001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2203300001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Bannerman2022 Whole Genome Metabolism - Mycobacterium leprae (strain Br4923)_1',\n",
       "  'submissionDate': '2022-03-29T23:00:00Z',\n",
       "  'submitter': 'Bridget Bannerman',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2203300001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2203300002',\n",
       "  'lastModified': None,\n",
       "  'name': 'Bannerman2022 Whole Genome Metabolism - M. abscessus (strain ATCC 19977 / DSM 44196)',\n",
       "  'submissionDate': '2022-03-29T23:00:00Z',\n",
       "  'submitter': 'Bridget Bannerman',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2203300002'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2402290001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Gulhane2022 - Sphingolipid metabolism and PI3K-AKT axis in NSCLC',\n",
       "  'submissionDate': '2024-02-29T00:00:00Z',\n",
       "  'submitter': 'Shailza Singh',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2402290001'},\n",
       " {'format': 'Open Neural Network Exchange',\n",
       "  'id': 'MODEL2407230001',\n",
       "  'lastModified': None,\n",
       "  'name': 'YangziChen2024 - Metabolic Machine Learning Predictor Model for Diagnosis of Gastric Cancer',\n",
       "  'submissionDate': '2024-07-22T23:00:00Z',\n",
       "  'submitter': 'Akshat Pandey',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2407230001'}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Search model ussing basico\n",
    "results = biomodels.search_for_model(query = \"precision medicine\",\n",
    "                                      num_results = 100)\n",
    "print(len(results))\n",
    "\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "27603036",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'format': 'SBML',\n",
       "  'id': 'BIOMD0000000534',\n",
       "  'lastModified': None,\n",
       "  'name': 'Dwivedi2014 - Healthy Volunteer IL6 Model',\n",
       "  'submissionDate': '2014-08-04T23:00:00Z',\n",
       "  'submitter': 'Vincent Knight-Schrijver',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000534'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000535',\n",
       "  'lastModified': None,\n",
       "  'name': 'Dwivedi2014 - Crohns IL6 Disease model - Anti-IL6 Antibody',\n",
       "  'submissionDate': '2014-08-04T23:00:00Z',\n",
       "  'submitter': 'Vincent Knight-Schrijver',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000535'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000536',\n",
       "  'lastModified': None,\n",
       "  'name': 'Dwivedi2014 - Crohns IL6 Disease model - sgp130 activity',\n",
       "  'submissionDate': '2014-08-04T23:00:00Z',\n",
       "  'submitter': 'Vincent Knight-Schrijver',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000536'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000537',\n",
       "  'lastModified': None,\n",
       "  'name': 'Dwivedi2014 - Crohns IL6 Disease model - Anti-IL6R Antibody',\n",
       "  'submissionDate': '2014-08-04T23:00:00Z',\n",
       "  'submitter': 'Vincent Knight-Schrijver',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000537'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2505090001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Lo2016  IBD Response to Anti-TNF treatment',\n",
       "  'submissionDate': '2025-05-08T23:00:00Z',\n",
       "  'submitter': 'bastien chassagnol',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2505090001'}]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = biomodels.search_for_model(query = \"crohn's disease\",\n",
    "                                      num_results = 100)\n",
    "print(len(results))\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "eb17f723",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'format': 'SBML',\n",
       "  'id': 'BIOMD0000000955',\n",
       "  'lastModified': None,\n",
       "  'name': 'Giordano2020 - SIDARTHE model of COVID-19 spread in Italy',\n",
       "  'submissionDate': '2020-07-27T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000955'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000956',\n",
       "  'lastModified': None,\n",
       "  'name': 'Bertozzi2020 - SIR model of scenarios of COVID-19 spread in CA and NY',\n",
       "  'submissionDate': '2020-08-06T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000956'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000957',\n",
       "  'lastModified': None,\n",
       "  'name': 'Roda2020 - SIR model of COVID-19 spread in Wuhan',\n",
       "  'submissionDate': '2020-08-10T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000957'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000958',\n",
       "  'lastModified': None,\n",
       "  'name': 'Ndairou2020 - early-stage transmission dynamics of COVID-19 in Wuhan',\n",
       "  'submissionDate': '2020-08-13T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000958'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000960',\n",
       "  'lastModified': None,\n",
       "  'name': 'Paiva2020 - SEIAHRD model of transmission dynamics of COVID-19',\n",
       "  'submissionDate': '2020-08-19T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000960'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000962',\n",
       "  'lastModified': None,\n",
       "  'name': 'Zhao2020 - SUQC model of COVID-19 transmission dynamics in Wuhan, Hubei, and China',\n",
       "  'submissionDate': '2020-08-25T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000962'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000963',\n",
       "  'lastModified': None,\n",
       "  'name': 'Weitz2020 - SIR model of COVID-19 transmission with shielding',\n",
       "  'submissionDate': '2020-09-15T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000963'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000964',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mwalili2020 - SEIR model of COVID-19 transmission and environmental pathogen prevalence',\n",
       "  'submissionDate': '2020-09-20T23:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000964'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000969',\n",
       "  'lastModified': None,\n",
       "  'name': 'Cuadros2020 - SIHRD spatiotemporal model of COVID-19 transmission in Ohio',\n",
       "  'submissionDate': '2020-10-27T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000969'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000970',\n",
       "  'lastModified': None,\n",
       "  'name': 'Hou2020 - SEIR model of COVID-19 transmission in Wuhan',\n",
       "  'submissionDate': '2020-10-28T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000970'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000971',\n",
       "  'lastModified': None,\n",
       "  'name': 'Tang2020 - Estimation of transmission risk of COVID-19 and impact of public health interventions',\n",
       "  'submissionDate': '2020-11-02T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000971'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000972',\n",
       "  'lastModified': None,\n",
       "  'name': 'Tang2020 - Estimation of transmission risk of COVID-19 and impact of public health interventions - update',\n",
       "  'submissionDate': '2020-11-03T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000972'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000974',\n",
       "  'lastModified': None,\n",
       "  'name': 'Carcione2020 - Deterministic SEIR simulation of a COVID-19 outbreak',\n",
       "  'submissionDate': '2020-11-03T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000974'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000976',\n",
       "  'lastModified': None,\n",
       "  'name': 'Ghanbari2020 - forecasting the second wave of COVID-19 in Iran',\n",
       "  'submissionDate': '2020-11-23T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000976'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000977',\n",
       "  'lastModified': None,\n",
       "  'name': 'Sarkar2020 - SAIR model of COVID-19 transmission with quarantine measures in India',\n",
       "  'submissionDate': '2020-11-30T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000977'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000978',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mukandavire2020 - SEIR model of early COVID-19 transmission in South Africa',\n",
       "  'submissionDate': '2020-12-01T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000978'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000979',\n",
       "  'lastModified': None,\n",
       "  'name': 'Malkov2020 - SEIRS model of COVID-19 transmission with reinfection',\n",
       "  'submissionDate': '2020-12-04T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000979'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000980',\n",
       "  'lastModified': None,\n",
       "  'name': 'Malkov2020 - SEIRS model of COVID-19 transmission with time-varying R values and reinfection',\n",
       "  'submissionDate': '2020-12-07T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000980'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000981',\n",
       "  'lastModified': None,\n",
       "  'name': 'Wan2020 - risk estimation and prediction of the transmission of COVID-19 in maninland China excluding Hubei province',\n",
       "  'submissionDate': '2020-12-09T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000981'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000982',\n",
       "  'lastModified': None,\n",
       "  'name': 'Law2020 - SIR model of COVID-19 transmission in Malyasia with time-varying parameters',\n",
       "  'submissionDate': '2020-12-24T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000982'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000983',\n",
       "  'lastModified': None,\n",
       "  'name': 'Zongo2020 - model of COVID-19 transmission dynamics under containment measures in France',\n",
       "  'submissionDate': '2021-01-15T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000983'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000984',\n",
       "  'lastModified': None,\n",
       "  'name': 'Fang2020 - SEIR model of COVID-19 transmission considering government interventions in Wuhan',\n",
       "  'submissionDate': '2021-01-19T00:00:00Z',\n",
       "  'submitter': 'Kausthubh Ramachandran',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000984'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000988',\n",
       "  'lastModified': None,\n",
       "  'name': 'Westerhoff2020 - systems biology model of the coronavirus pandemic 2020',\n",
       "  'submissionDate': '2021-02-12T00:00:00Z',\n",
       "  'submitter': 'Paul Jonas Jost',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000988'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000991',\n",
       "  'lastModified': None,\n",
       "  'name': 'Okuonghae2020 - SEAIR model of COVID-19 transmission in Lagos, Nigeria',\n",
       "  'submissionDate': '2021-02-18T00:00:00Z',\n",
       "  'submitter': 'Paul Jonas Jost',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000991'},\n",
       " {'format': 'COMBINE archive',\n",
       "  'id': 'MODEL2003020001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Renz2020 - GEM of Human alveolar macrophage with SARS-CoV-2',\n",
       "  'submissionDate': '2020-03-02T00:00:00Z',\n",
       "  'submitter': 'Andreas Drger',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2003020001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2007210001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Bannerman2020 - Integrated model of the human airway epithelial cell and the SARS-CoV-2 virus',\n",
       "  'submissionDate': '2020-07-20T23:00:00Z',\n",
       "  'submitter': 'Bridget Bannerman',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2007210001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2109130003',\n",
       "  'lastModified': None,\n",
       "  'name': 'COVID-19 immunotherapy A mathematical model',\n",
       "  'submissionDate': '2022-06-20T23:00:00Z',\n",
       "  'submitter': 'Joo Nuno Domingues Tavares',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2109130003'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2111170001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Leander2021 - innate immune response to SARS-CoV-2 in the alveolar epithelium',\n",
       "  'submissionDate': '2021-11-17T00:00:00Z',\n",
       "  'submitter': 'Krishna Kumar Tiwari',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2111170001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2202230002',\n",
       "  'lastModified': None,\n",
       "  'name': 'COVID-19 immunotherapy A mathematical model',\n",
       "  'submissionDate': '2022-02-23T00:00:00Z',\n",
       "  'submitter': 'Joo Nuno Domingues Tavares',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2202230002'},\n",
       " {'format': 'COMBINE archive',\n",
       "  'id': 'MODEL2202240001',\n",
       "  'lastModified': None,\n",
       "  'name': 'New workflow predicts drug targets against SARS-CoV-2 via metabolic changes in infected cells',\n",
       "  'submissionDate': '2023-01-18T00:00:00Z',\n",
       "  'submitter': 'Andreas Drger',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2202240001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2203300001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Bannerman2022 Whole Genome Metabolism - Mycobacterium leprae (strain Br4923)_1',\n",
       "  'submissionDate': '2022-03-29T23:00:00Z',\n",
       "  'submitter': 'Bridget Bannerman',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2203300001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2203300002',\n",
       "  'lastModified': None,\n",
       "  'name': 'Bannerman2022 Whole Genome Metabolism - M. abscessus (strain ATCC 19977 / DSM 44196)',\n",
       "  'submissionDate': '2022-03-29T23:00:00Z',\n",
       "  'submitter': 'Bridget Bannerman',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2203300002'},\n",
       " {'format': 'MATLAB (Octave)',\n",
       "  'id': 'MODEL2205010001',\n",
       "  'lastModified': None,\n",
       "  'name': 'PraneshPadmanabhan2022 - SARS-CoV-2 virus dynamics model (human)',\n",
       "  'submissionDate': '2022-04-30T23:00:00Z',\n",
       "  'submitter': 'Rajat Desikan',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2205010001'},\n",
       " {'format': 'Other',\n",
       "  'id': 'MODEL2209020001',\n",
       "  'lastModified': None,\n",
       "  'name': 'BY-COVID Knowledge Graph: A comprehensive network integrating various data resources of BY-COVID project',\n",
       "  'submissionDate': '2022-09-01T23:00:00Z',\n",
       "  'submitter': 'Reagon Karki',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2209020001'},\n",
       " {'format': 'Other',\n",
       "  'id': 'MODEL2212160001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Gianlupi2022 - Antiviral Timing, Potency, and Heterogeneity Effects on an Epithelial Tissue Patch Infected by SARS-CoV-2',\n",
       "  'submissionDate': '2022-12-16T00:00:00Z',\n",
       "  'submitter': 'Juliano Ferrari Gianlupi',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2212160001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2212310001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Model-based prediction of SARS-CoV-2 in Ethiopia: Extended SEIR model',\n",
       "  'submissionDate': '2022-12-31T00:00:00Z',\n",
       "  'submitter': 'Simon Merkt',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2212310001'},\n",
       " {'format': 'Other',\n",
       "  'id': 'MODEL2401170001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Pradeep2024 - A Mechanical modelling framework for studying endothelial permeability',\n",
       "  'submissionDate': '2024-01-17T00:00:00Z',\n",
       "  'submitter': 'Pradeep Keshavanarayana',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2401170001'},\n",
       " {'format': 'Python',\n",
       "  'id': 'MODEL2405080005',\n",
       "  'lastModified': None,\n",
       "  'name': 'Zeng2022 - Prediction of molecular properties and drug targets using a self-supervised learning framework',\n",
       "  'submissionDate': '2024-05-07T23:00:00Z',\n",
       "  'submitter': 'Zainab Ashimiyu-Abdusalam',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2405080005'},\n",
       " {'format': 'Python',\n",
       "  'id': 'MODEL2405080006',\n",
       "  'lastModified': None,\n",
       "  'name': 'Stokes2020 - Antibiotics discovery using deep learning approach (Antiviral implementation)',\n",
       "  'submissionDate': '2024-05-07T23:00:00Z',\n",
       "  'submitter': 'Zainab Ashimiyu-Abdusalam',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2405080006'},\n",
       " {'format': 'Python',\n",
       "  'id': 'MODEL2405130001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Zeng2022 - Prediction of HIV Growth Inhibition using ImageMol',\n",
       "  'submissionDate': '2024-05-12T23:00:00Z',\n",
       "  'submitter': 'Zainab Ashimiyu-Abdusalam',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2405130001'},\n",
       " {'format': 'Python',\n",
       "  'id': 'MODEL2405130004',\n",
       "  'lastModified': None,\n",
       "  'name': 'KC2021 - A machine learning platform to estimate anti-SARS-CoV-2 activities',\n",
       "  'submissionDate': '2024-05-12T23:00:00Z',\n",
       "  'submitter': 'Zainab Ashimiyu-Abdusalam',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2405130004'},\n",
       " {'format': 'COMBINE archive',\n",
       "  'id': 'MODEL2503040001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Miroshnichenko2025 - A modular model of immune response to SARS-CoV-2 infection',\n",
       "  'submissionDate': '2025-03-04T00:00:00Z',\n",
       "  'submitter': 'Max',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2503040001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2504120001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Merkt2024 - SEIR based model of SARS-CoV-2 variant spread and cross-immunity in Ethiopia',\n",
       "  'submissionDate': '2025-04-11T23:00:00Z',\n",
       "  'submitter': 'Simon Merkt',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2504120001'}]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = biomodels.search_for_model(query = \"covid-19\",\n",
    "                                      num_results = 100)\n",
    "print(len(results))\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "15d185c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'format': 'SBML',\n",
       "  'id': 'BIOMD0000000451',\n",
       "  'lastModified': None,\n",
       "  'name': 'Carbo2013 - Cytokine driven CD4+ T Cell differentiation and phenotype plasticity',\n",
       "  'submissionDate': '2013-04-22T23:00:00Z',\n",
       "  'submitter': 'Adria Carbo',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000451'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000527',\n",
       "  'lastModified': None,\n",
       "  'name': 'Kaiser2014 - Salmonella persistence after ciprofloxacin treatment',\n",
       "  'submissionDate': '2013-12-17T00:00:00Z',\n",
       "  'submitter': 'Roland Regoes',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000527'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000583',\n",
       "  'lastModified': None,\n",
       "  'name': 'Leber2015 - Mucosal immunity and gut microbiome interaction during C. difficile infection',\n",
       "  'submissionDate': '2015-07-19T23:00:00Z',\n",
       "  'submitter': 'Andrew Leber',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000583'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000619',\n",
       "  'lastModified': None,\n",
       "  'name': 'Sluka2016  - Acetaminophen PBPK',\n",
       "  'submissionDate': '2015-09-22T23:00:00Z',\n",
       "  'submitter': 'Vijayalakshmi Chelliah',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000619'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000000947',\n",
       "  'lastModified': None,\n",
       "  'name': 'Lee2017 - Paracetamol first-pass metabolism PK model',\n",
       "  'submissionDate': '2018-03-05T00:00:00Z',\n",
       "  'submitter': 'Matthew Roberts',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000000947'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'BIOMD0000001062',\n",
       "  'lastModified': None,\n",
       "  'name': 'Kim2021 - Development of a Genome-Scale Metabolic Model and Phenome Analysis of the Probiotic Escherichia coli Strain Nissle 1917',\n",
       "  'submissionDate': '2022-04-17T23:00:00Z',\n",
       "  'submitter': 'Dohyeon Kim',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/BIOMD0000001062'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1210260004',\n",
       "  'lastModified': None,\n",
       "  'name': 'Viladomiu2012 - PPARgamma role in C.diff associated disease',\n",
       "  'submissionDate': '2012-10-25T23:00:00Z',\n",
       "  'submitter': 'Monica Viladomiu',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1210260004'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220000',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Brain medulla',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220000'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Embryonic tissue',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220002',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Cerebellum',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220002'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220003',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Colon',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220003'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220004',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Eye',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220004'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220005',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Diaphragm',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220005'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220006',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Brown fat',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220006'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220007',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Jejunum',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220007'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220008',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Kidney medulla',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220008'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220009',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Ileum',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220009'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220010',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Duodenum',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220010'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220011',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Adrenal gland',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220011'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220012',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Kidney cortex',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220012'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220013',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Brain cortex',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220013'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220014',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Heart',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220014'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220015',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Lung',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220015'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220016',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Muscle',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220016'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220017',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Liver',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220017'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220018',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Midbrain',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220018'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220019',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Thymus',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220019'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220020',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Spleeen',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220020'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220021',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Stomach',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220021'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220022',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Salivary gland',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220022'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220023',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Ovary',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220023'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220024',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Pancreas',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220024'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220025',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Olfactory bulb',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220025'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220026',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - White fat',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220026'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220027',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Tissue-specific genome-scale metabolic network - Uterus',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220027'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220028',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Generic mouse genome-scale metabolic network (MMR)',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220028'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220029',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Curated tissue-specific genome-scale metabolic network - Liver',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220029'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220030',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Curated tissue-specific genome-scale metabolic network  - Colon',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220030'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220031',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Curated tissue-specific genome-scale metabolic network - Adipose',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220031'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL1509220032',\n",
       "  'lastModified': None,\n",
       "  'name': 'Mardinoglu2015 - Curated tissue-specific genome-scale metabolic model - Small intestine',\n",
       "  'submissionDate': '2015-09-21T23:00:00Z',\n",
       "  'submitter': 'Nicolas Rodriguez',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL1509220032'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2002040002',\n",
       "  'lastModified': None,\n",
       "  'name': 'Ankrah2021 - Genome scale metabolic model of Drosophila gut microbe Acetobacter fabarum',\n",
       "  'submissionDate': '2020-02-04T00:00:00Z',\n",
       "  'submitter': 'Nana Y D Ankrah',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2002040002'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2002040003',\n",
       "  'lastModified': None,\n",
       "  'name': 'Ankrah2021 - Genome scale metabolic model of Drosophila gut microbe Acetobacter pomorum',\n",
       "  'submissionDate': '2020-02-04T00:00:00Z',\n",
       "  'submitter': 'Nana Y D Ankrah',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2002040003'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2002040004',\n",
       "  'lastModified': None,\n",
       "  'name': 'Ankrah2021 - Genome scale metabolic model of Drosophila gut microbe Acetobacter tropicalis',\n",
       "  'submissionDate': '2020-02-04T00:00:00Z',\n",
       "  'submitter': 'Nana Y D Ankrah',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2002040004'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2002040005',\n",
       "  'lastModified': None,\n",
       "  'name': 'Ankrah2021 - Genome scale metabolic model of Drosophila gut microbe Lactobacillus brevis',\n",
       "  'submissionDate': '2020-02-04T00:00:00Z',\n",
       "  'submitter': 'Nana Y D Ankrah',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2002040005'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2002040006',\n",
       "  'lastModified': None,\n",
       "  'name': 'Ankrah2021 - Genome scale metabolic model of Drosophila gut microbe Lactobacillus plantarum',\n",
       "  'submissionDate': '2020-02-04T00:00:00Z',\n",
       "  'submitter': 'Nana Y D Ankrah',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2002040006'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2002070001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Geiert2020 - Yersinia enterocolitica co-infection in mice',\n",
       "  'submissionDate': '2020-02-07T00:00:00Z',\n",
       "  'submitter': 'Andreas Drger',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2002070001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2010130002',\n",
       "  'lastModified': None,\n",
       "  'name': 'Subrmanian2015 - Energy metabolism of Leishmania infantum, constrain-based metabolic model',\n",
       "  'submissionDate': '2020-10-12T23:00:00Z',\n",
       "  'submitter': 'Abhishek Subramanian',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2010130002'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2110210002',\n",
       "  'lastModified': None,\n",
       "  'name': 'MirhakkakSchaeuble2021 - a Candida albicans genome-scale metabolic model reconstruction',\n",
       "  'submissionDate': '2021-10-20T23:00:00Z',\n",
       "  'submitter': 'Sascha Schuble',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2110210002'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Lactobacillus casei subsp. casei ATCC 393',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190001'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190002',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Lactobacillus casei BL23',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190002'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190003',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Lactobacillus casei LC5',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190003'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190004',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Lactobacillus fermentum ATCC 14931',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190004'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190006',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Lactobacillus plantarum ATCC8014',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190006'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190007',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Lactobacillus plantarum JDM1',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190007'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190008',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Lactobacillus salivarius ATCC 11741',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190008'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190009',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Lactococcus lactis subsp. cremoris NZ9000',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190009'},\n",
       " {'format': 'SBML',\n",
       "  'id': 'MODEL2210190012',\n",
       "  'lastModified': None,\n",
       "  'name': 'Koduru2022 - Leuconostoc mesenteroides subsp. mesenteroides ATCC 8293',\n",
       "  'submissionDate': '2022-10-18T23:00:00Z',\n",
       "  'submitter': 'Yi Qing Lee',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2210190012'},\n",
       " {'format': 'Other',\n",
       "  'id': 'MODEL2405300001',\n",
       "  'lastModified': None,\n",
       "  'name': 'Ioannou2024 - Synthetic Human Infant Gut Microbial Community with Human Milk Oligosaccharides',\n",
       "  'submissionDate': '2024-05-29T23:00:00Z',\n",
       "  'submitter': 'William Scott',\n",
       "  'url': 'https://www.ebi.ac.uk/biomodels/MODEL2405300001'}]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = biomodels.search_for_model(query = \"gut\",\n",
    "                                      num_results = 100)\n",
    "print(len(results))\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5f2913e",
   "metadata": {},
   "source": [
    "### Get model info tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0d540cb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "units = basico.model_info.get_model_units(model=mod_precision)\n",
    "#save the dictionary as csv\n",
    "with open(r'./tables/units_537.pkl', 'wb') as f:\n",
    "    pickle.dump(units, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f111e961",
   "metadata": {},
   "outputs": [],
   "source": [
    "units = basico.model_info.get_model_units(model=mod_epidemiology)\n",
    "with open(r'./tables/units_971.pkl', 'wb') as f:\n",
    "    pickle.dump(units, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9d31fd78",
   "metadata": {},
   "outputs": [],
   "source": [
    "basico.model_info.get_model_units(model=mod_theory)\n",
    "with open(r'./tables/units_27.pkl', 'wb') as f:\n",
    "    pickle.dump(units, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26c8c6ba",
   "metadata": {},
   "source": [
    "#### Compartments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "88318d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = basico.model_info.get_compartments(model=mod_precision)\n",
    "df.to_csv(r'./tables/compartments_537.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "3b4fd67f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = basico.model_info.get_compartments(model=mod_epidemiology)\n",
    "df.to_csv(r'./tables/compartments_971.csv', index=False)\n",
    "\n",
    "df = basico.model_info.get_compartments(model=mod_theory)\n",
    "df.to_csv(r'./tables/compartments_27.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "49a4e1e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>unit</th>\n",
       "      <th>initial_size</th>\n",
       "      <th>initial_expression</th>\n",
       "      <th>dimensionality</th>\n",
       "      <th>expression</th>\n",
       "      <th>size</th>\n",
       "      <th>rate</th>\n",
       "      <th>key</th>\n",
       "      <th>sbml_id</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>uVol</th>\n",
       "      <td>fixed</td>\n",
       "      <td>l</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Compartment_5</td>\n",
       "      <td>uVol</td>\n",
       "      <td>Compartments[uVol]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       type unit  initial_size initial_expression  dimensionality expression  \\\n",
       "name                                                                           \n",
       "uVol  fixed    l           1.0                                  3              \n",
       "\n",
       "      size  rate            key sbml_id        display_name  \n",
       "name                                                         \n",
       "uVol   1.0   0.0  Compartment_5    uVol  Compartments[uVol]  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b96a295e",
   "metadata": {},
   "source": [
    "#### Get species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7fe9d652",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = basico.model_info.get_species(model=mod_precision)\n",
    "df.reset_index(inplace=True)\n",
    "df[[\"name\" ,\"compartment\",\n",
    "    \"type\",\n",
    "    \"unit\",\n",
    "    \"initial_concentration\",\n",
    "    \"display_name\",\n",
    "                ]].to_csv(r'./tables/species_537.csv', index=False)\n",
    "\n",
    "df = basico.model_info.get_species(model=mod_epidemiology)\n",
    "df.reset_index(inplace=True)\n",
    "df[[\"name\" ,\"compartment\",\n",
    "    \"type\",\n",
    "    \"unit\",\n",
    "    \"initial_concentration\",\n",
    "    \"display_name\",\n",
    "                ]].to_csv(r'./tables/species_971.csv', index=False)\n",
    "\n",
    "df = basico.model_info.get_species(model=mod_theory)\n",
    "df.reset_index(inplace=True)\n",
    "df[[\"name\" ,\"compartment\",\n",
    "    \"type\",\n",
    "    \"unit\",\n",
    "    \"initial_concentration\",\n",
    "    \"display_name\",\n",
    "                ]].to_csv(r'./tables/species_27.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "10ebfd1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>compartment</th>\n",
       "      <th>type</th>\n",
       "      <th>unit</th>\n",
       "      <th>initial_concentration</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>uVol</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>500.0</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mp</td>\n",
       "      <td>uVol</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Mp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mpp</td>\n",
       "      <td>uVol</td>\n",
       "      <td>reactions</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Mpp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAPKK</td>\n",
       "      <td>uVol</td>\n",
       "      <td>fixed</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>50.0</td>\n",
       "      <td>MAPKK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MKP3</td>\n",
       "      <td>uVol</td>\n",
       "      <td>fixed</td>\n",
       "      <td>nmol/l</td>\n",
       "      <td>100.0</td>\n",
       "      <td>MKP3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name compartment       type    unit  initial_concentration display_name\n",
       "0      M        uVol  reactions  nmol/l                  500.0            M\n",
       "1     Mp        uVol  reactions  nmol/l                    0.0           Mp\n",
       "2    Mpp        uVol  reactions  nmol/l                    0.0          Mpp\n",
       "3  MAPKK        uVol      fixed  nmol/l                   50.0        MAPKK\n",
       "4   MKP3        uVol      fixed  nmol/l                  100.0         MKP3"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(r'./tables/species_27.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f99523c6",
   "metadata": {},
   "source": [
    "#### Get parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "e3955c23",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = basico.model_info.get_parameters(model=mod_precision)\n",
    "df.reset_index(inplace=True)\n",
    "df[[\"name\", \"type\", \"unit\", \"initial_value\", \"display_name\"]].to_csv(r'./tables/parameters_537.csv', index=False)\n",
    "\n",
    "df = basico.model_info.get_parameters(model=mod_epidemiology)\n",
    "df.reset_index(inplace=True)\n",
    "df[[\"name\", \"type\", \"unit\", \"initial_value\", \"display_name\"]].to_csv(r'./tables/parameters_971.csv', index=False)\n",
    "\n",
    "df = basico.model_info.get_parameters(model=mod_theory)\n",
    "df.reset_index(inplace=True)\n",
    "df[[\"name\", \"type\", \"unit\", \"initial_value\", \"display_name\"]].to_csv(r'./tables/parameters_27.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c54963b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>unit</th>\n",
       "      <th>initial_value</th>\n",
       "      <th>initial_expression</th>\n",
       "      <th>expression</th>\n",
       "      <th>value</th>\n",
       "      <th>rate</th>\n",
       "      <th>key</th>\n",
       "      <th>sbml_id</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k1cat</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>0.010</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_69</td>\n",
       "      <td>k1cat</td>\n",
       "      <td>Values[k1cat]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Km1</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>50.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_70</td>\n",
       "      <td>Km1</td>\n",
       "      <td>Values[Km1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k2cat</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>15.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>15.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_71</td>\n",
       "      <td>k2cat</td>\n",
       "      <td>Values[k2cat]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Km2</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>500.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>500.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_72</td>\n",
       "      <td>Km2</td>\n",
       "      <td>Values[Km2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k3cat</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>0.084</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_73</td>\n",
       "      <td>k3cat</td>\n",
       "      <td>Values[k3cat]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Km3</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>22.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_74</td>\n",
       "      <td>Km3</td>\n",
       "      <td>Values[Km3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>k4cat</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>0.060</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.060</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_75</td>\n",
       "      <td>k4cat</td>\n",
       "      <td>Values[k4cat]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Km4</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>18.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>18.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_76</td>\n",
       "      <td>Km4</td>\n",
       "      <td>Values[Km4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Km5</td>\n",
       "      <td>fixed</td>\n",
       "      <td></td>\n",
       "      <td>78.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ModelValue_77</td>\n",
       "      <td>Km5</td>\n",
       "      <td>Values[Km5]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name   type unit  initial_value initial_expression expression    value  \\\n",
       "0  k1cat  fixed               0.010                                  0.010   \n",
       "1    Km1  fixed              50.000                                 50.000   \n",
       "2  k2cat  fixed              15.000                                 15.000   \n",
       "3    Km2  fixed             500.000                                500.000   \n",
       "4  k3cat  fixed               0.084                                  0.084   \n",
       "5    Km3  fixed              22.000                                 22.000   \n",
       "6  k4cat  fixed               0.060                                  0.060   \n",
       "7    Km4  fixed              18.000                                 18.000   \n",
       "8    Km5  fixed              78.000                                 78.000   \n",
       "\n",
       "   rate            key sbml_id   display_name  \n",
       "0   0.0  ModelValue_69   k1cat  Values[k1cat]  \n",
       "1   0.0  ModelValue_70     Km1    Values[Km1]  \n",
       "2   0.0  ModelValue_71   k2cat  Values[k2cat]  \n",
       "3   0.0  ModelValue_72     Km2    Values[Km2]  \n",
       "4   0.0  ModelValue_73   k3cat  Values[k3cat]  \n",
       "5   0.0  ModelValue_74     Km3    Values[Km3]  \n",
       "6   0.0  ModelValue_75   k4cat  Values[k4cat]  \n",
       "7   0.0  ModelValue_76     Km4    Values[Km4]  \n",
       "8   0.0  ModelValue_77     Km5    Values[Km5]  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d24429f",
   "metadata": {},
   "source": [
    "### Steady state tabular"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "451a572f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stst = basico.run_steadystate(model=mod_epidemiology)\n",
    "stst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fbe30e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "stst = basico.run_steadystate(model=mod_epidemiology)\n",
    "df = basico.get_species(model=mod_epidemiology)\n",
    "df.reset_index(inplace=True)\n",
    "df_short = df[[\"display_name\", \"compartment\", \"unit\", \"initial_concentration\", 'concentration', \"rate\", \"transition_time\"]]\n",
    "#rename the columns concentration to steady_state_concentration and transition time to steady_state_transition_time\n",
    "df_short = df_short.rename(columns={\"concentration\": \"steady_state_concentration\", \"transition_time\": \"steady_state_transition_time\"})\n",
    "df_short.to_csv(r'./tables/stst_971.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "63a465c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom plotter\n",
    "stst = basico.run_steadystate(model=mod_theory)\n",
    "df = basico.get_species(model=mod_theory)\n",
    "df.reset_index(inplace=True)\n",
    "df_short = df[[\"display_name\", \"compartment\", \"unit\", \"initial_concentration\", 'concentration', \"rate\", \"transition_time\"]]\n",
    "#rename the columns concentration to steady_state_concentration and transition time to steady_state_transition_time\n",
    "df_short = df_short.rename(columns={\"concentration\": \"steady_state_concentration\", \"transition_time\": \"steady_state_transition_time\"})\n",
    "df_short.to_csv(r'./tables/stst_27.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51e9314e",
   "metadata": {},
   "source": [
    "### Custom plotter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "80b65453",
   "metadata": {},
   "outputs": [],
   "source": [
    "tc = basico.run_time_course(2016, 2016, model = mod_precision)\n",
    "tc.reset_index(inplace=True)\n",
    "tc.to_csv(r'./tables/tc_537.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "ced1c7c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "tc = basico.run_time_course(50, 50, model = mod_epidemiology)\n",
    "tc.reset_index(inplace=True)\n",
    "tc.to_csv(r'./tables/tc_971.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "eeb20c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "tc = basico.run_time_course(100, 100, model = mod_theory)\n",
    "tc.reset_index(inplace=True)\n",
    "tc.to_csv(r'./tables/tc_27.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1fc1edc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Time</th>\n",
       "      <th>sR{serum}</th>\n",
       "      <th>sgp130{serum}</th>\n",
       "      <th>R_IL6_gp130{liver}</th>\n",
       "      <th>IL6{serum}</th>\n",
       "      <th>Ab{serum}</th>\n",
       "      <th>R</th>\n",
       "      <th>sR_IL6{gut}</th>\n",
       "      <th>sR_IL6{liver}</th>\n",
       "      <th>R_IL6_gp130{gut}</th>\n",
       "      <th>...</th>\n",
       "      <th>sgp130{liver}</th>\n",
       "      <th>sR_IL6_sgp130{gut}</th>\n",
       "      <th>Ab{peripheral}</th>\n",
       "      <th>sR_IL6_sgp130{liver}</th>\n",
       "      <th>pSTAT3{gut}</th>\n",
       "      <th>STAT3{liver}</th>\n",
       "      <th>CRP Suppression (%)</th>\n",
       "      <th>CRP (% of baseline)</th>\n",
       "      <th>CRP{liver}</th>\n",
       "      <th>geneProduct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.253507</td>\n",
       "      <td>3.900000</td>\n",
       "      <td>6.599359e-05</td>\n",
       "      <td>0.000436</td>\n",
       "      <td>2.381820e-29</td>\n",
       "      <td>0.438236</td>\n",
       "      <td>1.306824e-03</td>\n",
       "      <td>9.761649e-04</td>\n",
       "      <td>8.448905e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>5.589699</td>\n",
       "      <td>0.136304</td>\n",
       "      <td>1.679209e-29</td>\n",
       "      <td>0.116344</td>\n",
       "      <td>9.389364</td>\n",
       "      <td>0.777537</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>158.325847</td>\n",
       "      <td>159.803597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>3.901765</td>\n",
       "      <td>6.420660e-05</td>\n",
       "      <td>0.000638</td>\n",
       "      <td>6.753452e+02</td>\n",
       "      <td>0.000178</td>\n",
       "      <td>1.055422e-03</td>\n",
       "      <td>9.505457e-04</td>\n",
       "      <td>7.238601e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>5.591034</td>\n",
       "      <td>0.127067</td>\n",
       "      <td>2.144021e-01</td>\n",
       "      <td>0.114914</td>\n",
       "      <td>9.389266</td>\n",
       "      <td>0.777559</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>99.999999</td>\n",
       "      <td>158.325653</td>\n",
       "      <td>159.802738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>3.905215</td>\n",
       "      <td>6.248857e-05</td>\n",
       "      <td>0.000739</td>\n",
       "      <td>6.522828e+02</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>9.592235e-04</td>\n",
       "      <td>9.156890e-04</td>\n",
       "      <td>6.787804e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>5.594230</td>\n",
       "      <td>0.115733</td>\n",
       "      <td>4.442855e-01</td>\n",
       "      <td>0.110774</td>\n",
       "      <td>9.388692</td>\n",
       "      <td>0.777686</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>99.999971</td>\n",
       "      <td>158.324515</td>\n",
       "      <td>159.797695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>3.907882</td>\n",
       "      <td>6.013415e-05</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>6.303828e+02</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>8.868956e-04</td>\n",
       "      <td>8.679425e-04</td>\n",
       "      <td>6.452335e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>5.597806</td>\n",
       "      <td>0.107157</td>\n",
       "      <td>6.659729e-01</td>\n",
       "      <td>0.105075</td>\n",
       "      <td>9.387640</td>\n",
       "      <td>0.777986</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>99.999844</td>\n",
       "      <td>158.321844</td>\n",
       "      <td>159.788467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>3.909825</td>\n",
       "      <td>5.762284e-05</td>\n",
       "      <td>0.000755</td>\n",
       "      <td>6.095828e+02</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>8.257835e-04</td>\n",
       "      <td>8.168815e-04</td>\n",
       "      <td>6.168842e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>5.600867</td>\n",
       "      <td>0.099879</td>\n",
       "      <td>8.798707e-01</td>\n",
       "      <td>0.098958</td>\n",
       "      <td>9.386196</td>\n",
       "      <td>0.778528</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>99.999505</td>\n",
       "      <td>158.317000</td>\n",
       "      <td>159.775787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2012</th>\n",
       "      <td>2012.0</td>\n",
       "      <td>0.002114</td>\n",
       "      <td>3.900000</td>\n",
       "      <td>4.346951e-08</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>5.579947e+01</td>\n",
       "      <td>0.001612</td>\n",
       "      <td>7.713982e-07</td>\n",
       "      <td>5.512026e-07</td>\n",
       "      <td>5.653722e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>5.580968</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>2.143331e+01</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.031734</td>\n",
       "      <td>9.975746</td>\n",
       "      <td>98.976568</td>\n",
       "      <td>1.023432</td>\n",
       "      <td>0.798438</td>\n",
       "      <td>1.043898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2013</th>\n",
       "      <td>2013.0</td>\n",
       "      <td>0.002119</td>\n",
       "      <td>3.900000</td>\n",
       "      <td>4.355425e-08</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>5.563380e+01</td>\n",
       "      <td>0.001617</td>\n",
       "      <td>7.729185e-07</td>\n",
       "      <td>5.522847e-07</td>\n",
       "      <td>5.664315e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>5.580968</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>2.141546e+01</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.031781</td>\n",
       "      <td>9.975708</td>\n",
       "      <td>98.975817</td>\n",
       "      <td>1.024183</td>\n",
       "      <td>0.799683</td>\n",
       "      <td>1.045452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014</th>\n",
       "      <td>2014.0</td>\n",
       "      <td>0.002123</td>\n",
       "      <td>3.900000</td>\n",
       "      <td>4.363933e-08</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>5.546843e+01</td>\n",
       "      <td>0.001622</td>\n",
       "      <td>7.744447e-07</td>\n",
       "      <td>5.533709e-07</td>\n",
       "      <td>5.674950e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>5.580968</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>2.139759e+01</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.031829</td>\n",
       "      <td>9.975670</td>\n",
       "      <td>98.975062</td>\n",
       "      <td>1.024938</td>\n",
       "      <td>0.800934</td>\n",
       "      <td>1.047014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2015</th>\n",
       "      <td>2015.0</td>\n",
       "      <td>0.002128</td>\n",
       "      <td>3.900000</td>\n",
       "      <td>4.372474e-08</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>5.530337e+01</td>\n",
       "      <td>0.001627</td>\n",
       "      <td>7.759767e-07</td>\n",
       "      <td>5.544612e-07</td>\n",
       "      <td>5.685628e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>5.580968</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>2.137969e+01</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.031877</td>\n",
       "      <td>9.975632</td>\n",
       "      <td>98.974302</td>\n",
       "      <td>1.025698</td>\n",
       "      <td>0.802191</td>\n",
       "      <td>1.048584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2016</th>\n",
       "      <td>2016.0</td>\n",
       "      <td>0.002133</td>\n",
       "      <td>3.900000</td>\n",
       "      <td>4.381050e-08</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>5.513861e+01</td>\n",
       "      <td>0.001632</td>\n",
       "      <td>7.775145e-07</td>\n",
       "      <td>5.555556e-07</td>\n",
       "      <td>5.696349e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>5.580968</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>2.136176e+01</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.031925</td>\n",
       "      <td>9.975593</td>\n",
       "      <td>98.973539</td>\n",
       "      <td>1.026461</td>\n",
       "      <td>0.803455</td>\n",
       "      <td>1.050162</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2017 rows  45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Time  sR{serum}  sgp130{serum}  R_IL6_gp130{liver}  IL6{serum}  \\\n",
       "0        0.0   4.253507       3.900000        6.599359e-05    0.000436   \n",
       "1        1.0   0.000031       3.901765        6.420660e-05    0.000638   \n",
       "2        2.0   0.000037       3.905215        6.248857e-05    0.000739   \n",
       "3        3.0   0.000043       3.907882        6.013415e-05    0.000756   \n",
       "4        4.0   0.000049       3.909825        5.762284e-05    0.000755   \n",
       "...      ...        ...            ...                 ...         ...   \n",
       "2012  2012.0   0.002114       3.900000        4.346951e-08    0.000643   \n",
       "2013  2013.0   0.002119       3.900000        4.355425e-08    0.000643   \n",
       "2014  2014.0   0.002123       3.900000        4.363933e-08    0.000643   \n",
       "2015  2015.0   0.002128       3.900000        4.372474e-08    0.000643   \n",
       "2016  2016.0   0.002133       3.900000        4.381050e-08    0.000643   \n",
       "\n",
       "         Ab{serum}         R   sR_IL6{gut}  sR_IL6{liver}  R_IL6_gp130{gut}  \\\n",
       "0     2.381820e-29  0.438236  1.306824e-03   9.761649e-04      8.448905e-05   \n",
       "1     6.753452e+02  0.000178  1.055422e-03   9.505457e-04      7.238601e-05   \n",
       "2     6.522828e+02  0.000070  9.592235e-04   9.156890e-04      6.787804e-05   \n",
       "3     6.303828e+02  0.000049  8.868956e-04   8.679425e-04      6.452335e-05   \n",
       "4     6.095828e+02  0.000039  8.257835e-04   8.168815e-04      6.168842e-05   \n",
       "...            ...       ...           ...            ...               ...   \n",
       "2012  5.579947e+01  0.001612  7.713982e-07   5.512026e-07      5.653722e-08   \n",
       "2013  5.563380e+01  0.001617  7.729185e-07   5.522847e-07      5.664315e-08   \n",
       "2014  5.546843e+01  0.001622  7.744447e-07   5.533709e-07      5.674950e-08   \n",
       "2015  5.530337e+01  0.001627  7.759767e-07   5.544612e-07      5.685628e-08   \n",
       "2016  5.513861e+01  0.001632  7.775145e-07   5.555556e-07      5.696349e-08   \n",
       "\n",
       "      ...  sgp130{liver}  sR_IL6_sgp130{gut}  Ab{peripheral}  \\\n",
       "0     ...       5.589699            0.136304    1.679209e-29   \n",
       "1     ...       5.591034            0.127067    2.144021e-01   \n",
       "2     ...       5.594230            0.115733    4.442855e-01   \n",
       "3     ...       5.597806            0.107157    6.659729e-01   \n",
       "4     ...       5.600867            0.099879    8.798707e-01   \n",
       "...   ...            ...                 ...             ...   \n",
       "2012  ...       5.580968            0.000080    2.143331e+01   \n",
       "2013  ...       5.580968            0.000080    2.141546e+01   \n",
       "2014  ...       5.580968            0.000080    2.139759e+01   \n",
       "2015  ...       5.580968            0.000080    2.137969e+01   \n",
       "2016  ...       5.580968            0.000080    2.136176e+01   \n",
       "\n",
       "      sR_IL6_sgp130{liver}  pSTAT3{gut}  STAT3{liver}  CRP Suppression (%)  \\\n",
       "0                 0.116344     9.389364      0.777537            -0.000000   \n",
       "1                 0.114914     9.389266      0.777559             0.000001   \n",
       "2                 0.110774     9.388692      0.777686             0.000029   \n",
       "3                 0.105075     9.387640      0.777986             0.000156   \n",
       "4                 0.098958     9.386196      0.778528             0.000495   \n",
       "...                    ...          ...           ...                  ...   \n",
       "2012              0.000066     0.031734      9.975746            98.976568   \n",
       "2013              0.000066     0.031781      9.975708            98.975817   \n",
       "2014              0.000067     0.031829      9.975670            98.975062   \n",
       "2015              0.000067     0.031877      9.975632            98.974302   \n",
       "2016              0.000067     0.031925      9.975593            98.973539   \n",
       "\n",
       "      CRP (% of baseline)  CRP{liver}  geneProduct  \n",
       "0              100.000000  158.325847   159.803597  \n",
       "1               99.999999  158.325653   159.802738  \n",
       "2               99.999971  158.324515   159.797695  \n",
       "3               99.999844  158.321844   159.788467  \n",
       "4               99.999505  158.317000   159.775787  \n",
       "...                   ...         ...          ...  \n",
       "2012             1.023432    0.798438     1.043898  \n",
       "2013             1.024183    0.799683     1.045452  \n",
       "2014             1.024938    0.800934     1.047014  \n",
       "2015             1.025698    0.802191     1.048584  \n",
       "2016             1.026461    0.803455     1.050162  \n",
       "\n",
       "[2017 rows x 45 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tc = basico.run_time_course(2016, 2016, model = mod_precision)\n",
    "tc.reset_index(inplace=True)\n",
    "tc.to_csv(r'./tables/tc_537.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3573a01e",
   "metadata": {},
   "source": [
    "### Get annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f58d926d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'descriptions': [{'id': 'P26696',\n",
       "   'qualifier': 'is version of',\n",
       "   'uri': 'http://identifiers.org/uniprot/P26696',\n",
       "   'resource': 'UniProt Knowledgebase'}]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basico.get_miriam_annotation(name = \"M\", model=mod_theory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "af16c5a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M\n",
      "{'descriptions': [{'id': 'P26696', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P26696', 'resource': 'UniProt Knowledgebase'}]}\n",
      "Mp\n",
      "{'descriptions': [{'id': 'P26696', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P26696', 'resource': 'UniProt Knowledgebase'}]}\n",
      "Mpp\n",
      "{'descriptions': [{'id': 'P26696', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P26696', 'resource': 'UniProt Knowledgebase'}]}\n",
      "MAPKK\n",
      "{'descriptions': [{'id': 'Q05116', 'qualifier': 'is', 'uri': 'http://identifiers.org/uniprot/Q05116', 'resource': 'UniProt Knowledgebase'}]}\n",
      "MKP3\n",
      "{'descriptions': [{'id': 'Q90W58', 'qualifier': 'is', 'uri': 'http://identifiers.org/uniprot/Q90W58', 'resource': 'UniProt Knowledgebase'}]}\n"
     ]
    }
   ],
   "source": [
    "for species in basico.get_species(model=mod_theory).index:\n",
    "    print(species)\n",
    "    print(basico.get_miriam_annotation(name = species, model=mod_theory))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8091f185",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IL6{serum}\n",
      "{'descriptions': [{'id': 'P05231', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}]}\n",
      "sgp130{serum}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "sR_IL6_sgp130{serum}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'has part', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "CRP{serum}\n",
      "{'descriptions': [{'id': 'P02741', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P02741', 'resource': 'UniProt Knowledgebase'}]}\n",
      "sR{serum}\n",
      "{'descriptions': [{'id': 'P08887', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "sR_IL6{serum}\n",
      "{'descriptions': [{'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "Ab{serum}\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}, {'id': 'CHEBI:52217', 'qualifier': 'has property', 'uri': 'http://identifiers.org/CHEBI:52217', 'resource': 'ChEBI'}]}\n",
      "Ab_sR{serum}\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'has part', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "Ab_sR_IL6{serum}\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'has part', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "CRP Suppression (%)\n",
      "{'descriptions': [{'id': 'P02741', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P02741', 'resource': 'UniProt Knowledgebase'}]}\n",
      "CRP (% of baseline)\n",
      "{'descriptions': [{'id': 'P02741', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P02741', 'resource': 'UniProt Knowledgebase'}]}\n",
      "gp130{liver}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}]}\n",
      "R_IL6_gp130{liver}\n",
      "{'descriptions': [{'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'PR:000001868', 'qualifier': 'has part', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}]}\n",
      "sR_IL6{liver}\n",
      "{'descriptions': [{'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "R\n",
      "{'descriptions': [{'id': 'P08887', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}]}\n",
      "IL6{liver}\n",
      "{'descriptions': [{'id': 'P05231', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}]}\n",
      "R_IL6\n",
      "{'descriptions': [{'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}]}\n",
      "Ractive{liver}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'has part', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0002354', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0002354', 'resource': 'PATO'}, {'id': 'SBO:0000286', 'qualifier': 'has property', 'uri': 'http://identifiers.org/SBO:0000286', 'resource': 'Systems Biology Ontology'}]}\n",
      "STAT3{liver}\n",
      "{'descriptions': [{'id': 'PR:000002089', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000002089', 'resource': 'Protein Ontology'}]}\n",
      "pSTAT3{liver}\n",
      "{'descriptions': [{'id': 'PR:000002089', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000002089', 'resource': 'Protein Ontology'}, {'id': 'SBO:0000216', 'qualifier': 'has property', 'uri': 'http://identifiers.org/SBO:0000216', 'resource': 'Systems Biology Ontology'}]}\n",
      "CRP{liver}\n",
      "{'descriptions': [{'id': 'P02741', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P02741', 'resource': 'UniProt Knowledgebase'}]}\n",
      "sR{liver}\n",
      "{'descriptions': [{'id': 'P08887', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "CRPExtracellular\n",
      "{'descriptions': [{'id': 'P02741', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P02741', 'resource': 'UniProt Knowledgebase'}]}\n",
      "sgp130{liver}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "sR_IL6_sgp130{liver}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'has part', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "Ab_sR{liver}\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'has part', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "Ab{liver}\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}]}\n",
      "Ab_R\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'has part', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}]}\n",
      "Ab_sR_IL6{liver}\n",
      "{'descriptions': [{'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'FMA:62871', 'qualifier': 'has part', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "sR_IL6{gut}\n",
      "{'descriptions': [{'id': 'P05231', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "gp130{gut}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}]}\n",
      "R_IL6_gp130{gut}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'has part', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}]}\n",
      "Ractive{gut}\n",
      "{'descriptions': [{'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'PR:000001868', 'qualifier': 'has part', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0002354', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0002354', 'resource': 'PATO'}, {'id': 'SBO:0000286', 'qualifier': 'has property', 'uri': 'http://identifiers.org/SBO:0000286', 'resource': 'Systems Biology Ontology'}]}\n",
      "STAT3{gut}\n",
      "{'descriptions': [{'id': 'PR:000002089', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000002089', 'resource': 'Protein Ontology'}]}\n",
      "pSTAT3{gut}\n",
      "{'descriptions': [{'id': 'PR:000002089', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000002089', 'resource': 'Protein Ontology'}, {'id': 'SBO:0000216', 'qualifier': 'has property', 'uri': 'http://identifiers.org/SBO:0000216', 'resource': 'Systems Biology Ontology'}]}\n",
      "geneProduct\n",
      "{'descriptions': [{'id': 'CCO:U0000003', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/CCO:U0000003', 'resource': 'Cell Cycle Ontology'}]}\n",
      "sR{gut}\n",
      "{'descriptions': [{'id': 'P08887', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "IL6{gut}\n",
      "{'descriptions': [{'id': 'P05231', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}]}\n",
      "sgp130{gut}\n",
      "{'descriptions': [{'id': 'PR:000001868', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "sR_IL6_sgp130{gut}\n",
      "{'descriptions': [{'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PR:000001868', 'qualifier': 'has part', 'uri': 'http://identifiers.org/PR:000001868', 'resource': 'Protein Ontology'}, {'id': 'P05231', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "Ab{gut}\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}]}\n",
      "Ab_sR{gut}\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'has part', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}, {'id': 'P08887', 'qualifier': 'has part', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}]}\n",
      "Ab_sR_IL6{gut}\n",
      "{'descriptions': [{'id': 'P05231', 'qualifier': 'has property', 'uri': 'http://identifiers.org/uniprot/P05231', 'resource': 'UniProt Knowledgebase'}, {'id': 'P08887', 'qualifier': 'has property', 'uri': 'http://identifiers.org/uniprot/P08887', 'resource': 'UniProt Knowledgebase'}, {'id': 'PATO:0001537', 'qualifier': 'has property', 'uri': 'http://identifiers.org/PATO:0001537', 'resource': 'PATO'}, {'id': 'FMA:62871', 'qualifier': 'has property', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}]}\n",
      "Ab{peripheral}\n",
      "{'descriptions': [{'id': 'FMA:62871', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/FMA:62871', 'resource': 'FMA'}]}\n"
     ]
    }
   ],
   "source": [
    "mod = mod_precision\n",
    "for species in basico.get_species(model =mod)[\"display_name\"]:\n",
    "    print(species)\n",
    "    print(basico.get_miriam_annotation(name = species, model=mod))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "eb11a224",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Susceptible\n",
      "{'descriptions': [{'id': '0000514', 'qualifier': 'is', 'uri': 'http://identifiers.org/ido/0000514', 'resource': 'Infectious Disease Ontology'}, {'id': 'C171133', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/ncit/C171133', 'resource': 'NCIt'}]}\n",
      "Exposed\n",
      "{'descriptions': [{'id': '0000514', 'qualifier': 'is', 'uri': 'http://identifiers.org/ido/0000514', 'resource': 'Infectious Disease Ontology'}, {'id': 'C171133', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/ncit/C171133', 'resource': 'NCIt'}, {'id': '0000597', 'qualifier': 'has property', 'uri': 'http://identifiers.org/ido/0000597', 'resource': 'Infectious Disease Ontology'}]}\n",
      "Infected\n",
      "{'created': datetime.datetime(2020, 10, 30, 3, 56, 34, tzinfo=tzutc())}\n",
      "Asymptomatic\n",
      "{'descriptions': [{'id': '0000569', 'qualifier': 'is', 'uri': 'http://identifiers.org/ido/0000569', 'resource': 'Infectious Disease Ontology'}, {'id': '0000511', 'qualifier': 'is', 'uri': 'http://identifiers.org/ido/0000511', 'resource': 'Infectious Disease Ontology'}, {'id': 'C171133', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/ncit/C171133', 'resource': 'NCIt'}]}\n",
      "Susceptible_quarantined\n",
      "{'descriptions': [{'id': 'C71902', 'qualifier': 'is', 'uri': 'http://identifiers.org/ncit/C71902', 'resource': 'NCIt'}, {'id': '0000514', 'qualifier': 'is', 'uri': 'http://identifiers.org/ido/0000514', 'resource': 'Infectious Disease Ontology'}, {'id': 'C171133', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/ncit/C171133', 'resource': 'NCIt'}]}\n",
      "Exposed_quarantined\n",
      "{'descriptions': [{'id': '0000514', 'qualifier': 'is', 'uri': 'http://identifiers.org/ido/0000514', 'resource': 'Infectious Disease Ontology'}, {'id': 'C71902', 'qualifier': 'is', 'uri': 'http://identifiers.org/ncit/C71902', 'resource': 'NCIt'}, {'id': 'C171133', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/ncit/C171133', 'resource': 'NCIt'}, {'id': '0000597', 'qualifier': 'has property', 'uri': 'http://identifiers.org/ido/0000597', 'resource': 'Infectious Disease Ontology'}]}\n",
      "Hospitalised\n",
      "{'descriptions': [{'id': '0000511', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/ido/0000511', 'resource': 'Infectious Disease Ontology'}, {'id': 'C171133', 'qualifier': 'is version of', 'uri': 'http://identifiers.org/ncit/C171133', 'resource': 'NCIt'}, {'id': 'C25179', 'qualifier': 'has property', 'uri': 'http://identifiers.org/ncit/C25179', 'resource': 'NCIt'}]}\n",
      "Recovered\n",
      "{'descriptions': [{'id': '0000621', 'qualifier': 'has property', 'uri': 'http://identifiers.org/ido/0000621', 'resource': 'Infectious Disease Ontology'}]}\n"
     ]
    }
   ],
   "source": [
    "mod = mod_epidemiology\n",
    "for species in basico.get_species(model =mod).index:\n",
    "    print(species)\n",
    "    print(basico.get_miriam_annotation(name = species, model=mod))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "67bfa244",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>compartment</th>\n",
       "      <th>type</th>\n",
       "      <th>unit</th>\n",
       "      <th>initial_concentration</th>\n",
       "      <th>initial_particle_number</th>\n",
       "      <th>initial_expression</th>\n",
       "      <th>expression</th>\n",
       "      <th>concentration</th>\n",
       "      <th>particle_number</th>\n",
       "      <th>rate</th>\n",
       "      <th>particle_number_rate</th>\n",
       "      <th>key</th>\n",
       "      <th>sbml_id</th>\n",
       "      <th>transition_time</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Susceptible</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>1.108100e+07</td>\n",
       "      <td>1.108100e+07</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.108100e+07</td>\n",
       "      <td>1.108100e+07</td>\n",
       "      <td>-2739.167828</td>\n",
       "      <td>-2739.167828</td>\n",
       "      <td>Metabolite_57</td>\n",
       "      <td>Susceptible</td>\n",
       "      <td>3968.905658</td>\n",
       "      <td>Susceptible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exposed</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>1.051000e+02</td>\n",
       "      <td>1.051000e+02</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.051000e+02</td>\n",
       "      <td>1.051000e+02</td>\n",
       "      <td>264.485659</td>\n",
       "      <td>264.485659</td>\n",
       "      <td>Metabolite_58</td>\n",
       "      <td>Exposed</td>\n",
       "      <td>0.376029</td>\n",
       "      <td>Exposed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Infected</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>2.767900e+01</td>\n",
       "      <td>2.767900e+01</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2.767900e+01</td>\n",
       "      <td>2.767900e+01</td>\n",
       "      <td>0.223005</td>\n",
       "      <td>0.223005</td>\n",
       "      <td>Metabolite_59</td>\n",
       "      <td>Infected</td>\n",
       "      <td>2.123031</td>\n",
       "      <td>Infected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Asymptomatic</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>5.353900e+01</td>\n",
       "      <td>5.353900e+01</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.353900e+01</td>\n",
       "      <td>5.353900e+01</td>\n",
       "      <td>-5.506903</td>\n",
       "      <td>-5.506903</td>\n",
       "      <td>Metabolite_60</td>\n",
       "      <td>Asymptomatic</td>\n",
       "      <td>7.154099</td>\n",
       "      <td>Asymptomatic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Susceptible_quarantined</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>7.390000e+02</td>\n",
       "      <td>7.390000e+02</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7.390000e+02</td>\n",
       "      <td>7.390000e+02</td>\n",
       "      <td>2459.667845</td>\n",
       "      <td>2459.667845</td>\n",
       "      <td>Metabolite_61</td>\n",
       "      <td>Susceptible_quarantined</td>\n",
       "      <td>0.294135</td>\n",
       "      <td>Susceptible_quarantined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exposed_quarantined</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>1.164200e+00</td>\n",
       "      <td>1.164200e+00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.164200e+00</td>\n",
       "      <td>1.164200e+00</td>\n",
       "      <td>-0.146520</td>\n",
       "      <td>-0.146520</td>\n",
       "      <td>Metabolite_62</td>\n",
       "      <td>Exposed_quarantined</td>\n",
       "      <td>7.942812</td>\n",
       "      <td>Exposed_quarantined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hospitalised</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.702211</td>\n",
       "      <td>3.702211</td>\n",
       "      <td>Metabolite_63</td>\n",
       "      <td>Hospitalised</td>\n",
       "      <td>0.261885</td>\n",
       "      <td>Hospitalised</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Recovered</th>\n",
       "      <td>Wuhan</td>\n",
       "      <td>reactions</td>\n",
       "      <td>#/ml</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>16.742018</td>\n",
       "      <td>16.742018</td>\n",
       "      <td>Metabolite_64</td>\n",
       "      <td>Recovered</td>\n",
       "      <td>0.119460</td>\n",
       "      <td>Recovered</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        compartment       type  unit  initial_concentration  \\\n",
       "name                                                                          \n",
       "Susceptible                   Wuhan  reactions  #/ml           1.108100e+07   \n",
       "Exposed                       Wuhan  reactions  #/ml           1.051000e+02   \n",
       "Infected                      Wuhan  reactions  #/ml           2.767900e+01   \n",
       "Asymptomatic                  Wuhan  reactions  #/ml           5.353900e+01   \n",
       "Susceptible_quarantined       Wuhan  reactions  #/ml           7.390000e+02   \n",
       "Exposed_quarantined           Wuhan  reactions  #/ml           1.164200e+00   \n",
       "Hospitalised                  Wuhan  reactions  #/ml           1.000000e+00   \n",
       "Recovered                     Wuhan  reactions  #/ml           2.000000e+00   \n",
       "\n",
       "                         initial_particle_number initial_expression  \\\n",
       "name                                                                  \n",
       "Susceptible                         1.108100e+07                      \n",
       "Exposed                             1.051000e+02                      \n",
       "Infected                            2.767900e+01                      \n",
       "Asymptomatic                        5.353900e+01                      \n",
       "Susceptible_quarantined             7.390000e+02                      \n",
       "Exposed_quarantined                 1.164200e+00                      \n",
       "Hospitalised                        1.000000e+00                      \n",
       "Recovered                           2.000000e+00                      \n",
       "\n",
       "                        expression  concentration  particle_number  \\\n",
       "name                                                                 \n",
       "Susceptible                          1.108100e+07     1.108100e+07   \n",
       "Exposed                              1.051000e+02     1.051000e+02   \n",
       "Infected                             2.767900e+01     2.767900e+01   \n",
       "Asymptomatic                         5.353900e+01     5.353900e+01   \n",
       "Susceptible_quarantined              7.390000e+02     7.390000e+02   \n",
       "Exposed_quarantined                  1.164200e+00     1.164200e+00   \n",
       "Hospitalised                         1.000000e+00     1.000000e+00   \n",
       "Recovered                            2.000000e+00     2.000000e+00   \n",
       "\n",
       "                                rate  particle_number_rate            key  \\\n",
       "name                                                                        \n",
       "Susceptible             -2739.167828          -2739.167828  Metabolite_57   \n",
       "Exposed                   264.485659            264.485659  Metabolite_58   \n",
       "Infected                    0.223005              0.223005  Metabolite_59   \n",
       "Asymptomatic               -5.506903             -5.506903  Metabolite_60   \n",
       "Susceptible_quarantined  2459.667845           2459.667845  Metabolite_61   \n",
       "Exposed_quarantined        -0.146520             -0.146520  Metabolite_62   \n",
       "Hospitalised                3.702211              3.702211  Metabolite_63   \n",
       "Recovered                  16.742018             16.742018  Metabolite_64   \n",
       "\n",
       "                                         sbml_id  transition_time  \\\n",
       "name                                                                \n",
       "Susceptible                          Susceptible      3968.905658   \n",
       "Exposed                                  Exposed         0.376029   \n",
       "Infected                                Infected         2.123031   \n",
       "Asymptomatic                        Asymptomatic         7.154099   \n",
       "Susceptible_quarantined  Susceptible_quarantined         0.294135   \n",
       "Exposed_quarantined          Exposed_quarantined         7.942812   \n",
       "Hospitalised                        Hospitalised         0.261885   \n",
       "Recovered                              Recovered         0.119460   \n",
       "\n",
       "                                    display_name  \n",
       "name                                              \n",
       "Susceptible                          Susceptible  \n",
       "Exposed                                  Exposed  \n",
       "Infected                                Infected  \n",
       "Asymptomatic                        Asymptomatic  \n",
       "Susceptible_quarantined  Susceptible_quarantined  \n",
       "Exposed_quarantined          Exposed_quarantined  \n",
       "Hospitalised                        Hospitalised  \n",
       "Recovered                              Recovered  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basico.get_species(model = mod_epidemiology)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0e24580a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'created': datetime.datetime(2020, 10, 30, 3, 56, 34, tzinfo=tzutc())}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basico.get_miriam_annotation(name='Infected', model = mod_epidemiology)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
